// @totvs/protheus-lib-core/esm2022/components/pro-fields/pro-branch-lookup/pro-branch-lookup.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { ReactiveFormsModule } from '@angular/forms';
import { PoFieldModule, PoTooltipModule } from '@po-ui/ng-components';
import { ProBranchLookupComponent } from './pro-branch-lookup.component';
import * as i0 from "@angular/core";
export class ProBranchLookupModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProBranchLookupModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "17.3.9", ngImport: i0, type: ProBranchLookupModule, declarations: [ProBranchLookupComponent], imports: [CommonModule, ReactiveFormsModule, PoFieldModule, PoTooltipModule], exports: [ProBranchLookupComponent] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProBranchLookupModule, imports: [CommonModule, ReactiveFormsModule, PoFieldModule, PoTooltipModule] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProBranchLookupModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, ReactiveFormsModule, PoFieldModule, PoTooltipModule],
                    declarations: [ProBranchLookupComponent],
                    exports: [ProBranchLookupComponent]
                }]
        }] });
