// @totvs/protheus-lib-core/esm2022/components/pro-fields/pro-branch-lookup/pro-branch-lookup.service.mjs
import { Injectable } from '@angular/core';
import { environment } from '../../../environments/environment';
import * as i0 from "@angular/core";
import * as i1 from "./../../../services/pro-tenant/pro-branch/services/pro-branch.service";
export class ProBranchLookupService {
    constructor(proBranchService) {
        this.proBranchService = proBranchService;
        this.useHTTP = environment.useHTTP;
        this.changeServiceChannel();
    }
    getFilteredItems(params) {
        return this.proBranchService.getListOfBranches(params.filter, params.page, params.pageSize);
    }
    getObjectByValue(value) {
        return this.proBranchService.getBranch(value);
    }
    setBranch(branch) {
        this.proBranchService.branch = branch;
    }
    setChannelAsHTTP(value) {
        this.useHTTP = value;
        this.changeServiceChannel();
    }
    isChannelHTTP() {
        return this.useHTTP;
    }
    changeServiceChannel() {
        this.proBranchService.setChannelAsHTTP(this.useHTTP);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProBranchLookupService, deps: [{ token: i1.ProBranchService }], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProBranchLookupService, providedIn: 'root' }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProBranchLookupService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root'
                }]
        }], ctorParameters: () => [{ type: i1.ProBranchService }] });
