// @totvs/protheus-lib-core/esm2022/components/pro-fields/pro-branch-lookup/pro-branch-lookup.component.mjs
import { Component, EventEmitter, Input, Output, ViewChild } from '@angular/core';
import { PoLookupComponent } from '@po-ui/ng-components';
import * as i0 from "@angular/core";
import * as i1 from "./pro-branch-lookup.service";
import * as i2 from "@angular/forms";
import * as i3 from "@po-ui/ng-components";
/**
 * @description
 *
 * Componente que disponibiliza um campo para lookup para grupo de empresa.
 * O nome do field no formulário será branch_code
 */
export class ProBranchLookupComponent {
    constructor(service) {
        this.service = service;
        this.branchSetted = new EventEmitter();
    }
    fieldFormat(value) {
        return value.Code;
    }
    /**
     * @description metodo para setar o foco no lookup de branch
     */
    setBranchFocus() {
        this.lookup.focus();
    }
    /*
    trecho para mostrar a consulta pelo atalho do F3
    mantido aqui para servir de base caso venha a ser necessário utilizar isso no futuro.
  
    // Registre o HostListener para a tecla F3
    @HostListener('document:keydown.F3', ['$event'])
    handleF3Key(event: KeyboardEvent): void {
      console.log("f3 branch out");
      if (this.lookup.inputEl.nativeElement == document.activeElement){
        console.log("f3 branch in");
        event.preventDefault(); // Isso impede que a ação padrão da tecla F3 (geralmente abrir a busca) seja executada
        this.lookup.openLookup();
      }
    } */
    setBranch(branch) {
        if (Object.keys(branch).length > 0) {
            this.parent.patchValue({ branch_description: branch.Description });
            this.service.setBranch(branch);
            this.branchSetted.emit(branch);
        }
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProBranchLookupComponent, deps: [{ token: i1.ProBranchLookupService }], target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "17.3.9", type: ProBranchLookupComponent, selector: "pro-branch-lookup", inputs: { parent: "parent", pLabel: ["p-label", "pLabel"], pHint: ["p-hint", "pHint"], columns: ["pro-columns", "columns"] }, outputs: { branchSetted: "branchSetted" }, viewQueries: [{ propertyName: "lookup", first: true, predicate: PoLookupComponent, descendants: true, static: true }], ngImport: i0, template: "<div [formGroup]=\"parent\" class=\"po-page-login-info-container\">\r\n  <po-lookup\r\n    class=\"po-md-6\"\r\n    name=\"branch_code\"\r\n    formControlName=\"branch_code\"\r\n    p-field-label=\"Code\"\r\n    p-field-value=\"Code\"\r\n    [p-label]=\"pLabel\"\r\n    [p-columns]=\"columns\"\r\n    [p-field-format]=\"fieldFormat\"\r\n    [p-filter-service]=\"service\"\r\n    (p-selected)=\"setBranch($event)\"\r\n  >\r\n  </po-lookup>\r\n  <po-input\r\n    class=\"po-md-6\"\r\n    name=\"branch_description\"\r\n    formControlName=\"branch_description\"\r\n    p-label=\"&nbsp;\"\r\n    p-disabled=\"true\"\r\n  >\r\n  </po-input>\r\n  <div class=\"po-page-login-info-icon-container tooltip-bigscreen-adjust\">\r\n    <span\r\n      class=\"po-icon po-field-icon po-icon-info\"\r\n      [p-tooltip]=\"pHint\"\r\n      p-tooltip-position=\"right\"\r\n    >\r\n    </span>\r\n  </div>\r\n</div>\r\n", styles: ["@media (min-width: 1367px){.tooltip-bigscreen-adjust{padding-left:0}}\n"], dependencies: [{ kind: "directive", type: i2.NgControlStatus, selector: "[formControlName],[ngModel],[formControl]" }, { kind: "directive", type: i2.NgControlStatusGroup, selector: "[formGroupName],[formArrayName],[ngModelGroup],[formGroup],form:not([ngNoForm]),[ngForm]" }, { kind: "directive", type: i2.FormGroupDirective, selector: "[formGroup]", inputs: ["formGroup"], outputs: ["ngSubmit"], exportAs: ["ngForm"] }, { kind: "directive", type: i2.FormControlName, selector: "[formControlName]", inputs: ["formControlName", "disabled", "ngModel"], outputs: ["ngModelChange"] }, { kind: "component", type: i3.PoInputComponent, selector: "po-input" }, { kind: "component", type: i3.PoLookupComponent, selector: "po-lookup" }, { kind: "directive", type: i3.PoTooltipDirective, selector: "[p-tooltip]" }] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProBranchLookupComponent, decorators: [{
            type: Component,
            args: [{ selector: 'pro-branch-lookup', template: "<div [formGroup]=\"parent\" class=\"po-page-login-info-container\">\r\n  <po-lookup\r\n    class=\"po-md-6\"\r\n    name=\"branch_code\"\r\n    formControlName=\"branch_code\"\r\n    p-field-label=\"Code\"\r\n    p-field-value=\"Code\"\r\n    [p-label]=\"pLabel\"\r\n    [p-columns]=\"columns\"\r\n    [p-field-format]=\"fieldFormat\"\r\n    [p-filter-service]=\"service\"\r\n    (p-selected)=\"setBranch($event)\"\r\n  >\r\n  </po-lookup>\r\n  <po-input\r\n    class=\"po-md-6\"\r\n    name=\"branch_description\"\r\n    formControlName=\"branch_description\"\r\n    p-label=\"&nbsp;\"\r\n    p-disabled=\"true\"\r\n  >\r\n  </po-input>\r\n  <div class=\"po-page-login-info-icon-container tooltip-bigscreen-adjust\">\r\n    <span\r\n      class=\"po-icon po-field-icon po-icon-info\"\r\n      [p-tooltip]=\"pHint\"\r\n      p-tooltip-position=\"right\"\r\n    >\r\n    </span>\r\n  </div>\r\n</div>\r\n", styles: ["@media (min-width: 1367px){.tooltip-bigscreen-adjust{padding-left:0}}\n"] }]
        }], ctorParameters: () => [{ type: i1.ProBranchLookupService }], propDecorators: { parent: [{
                type: Input
            }], branchSetted: [{
                type: Output
            }], pLabel: [{
                type: Input,
                args: ['p-label']
            }], pHint: [{
                type: Input,
                args: ['p-hint']
            }], columns: [{
                type: Input,
                args: ['pro-columns']
            }], lookup: [{
                type: ViewChild,
                args: [PoLookupComponent, { static: true }]
            }] } });
