// @po-ui/ng-components/esm2022/lib/components/po-menu/po-menu.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { PoBadgeModule } from '../po-badge/po-badge.module';
import { PoFieldModule } from '../po-field/po-field.module';
import { PoLoadingModule } from '../po-loading/po-loading.module';
import { PoLogoModule } from '../po-logo/po-logo.module';
import { PoIconModule } from '../po-icon/po-icon.module';
import { PoMenuFilterComponent } from './po-menu-filter/po-menu-filter.component';
import { PoMenuHeaderTemplateDirective } from './po-menu-header-template/po-menu-header-template.directive';
import { PoMenuItemComponent } from './po-menu-item/po-menu-item.component';
import { PoMenuComponent } from './po-menu.component';
import { PoTooltipModule } from '../../directives/po-tooltip/po-tooltip.module';
import * as i0 from "@angular/core";
/**
 * @description
 *
 * Módulo do componente po-menu.
 */
export class PoMenuModule {
    static ɵfac = function PoMenuModule_Factory(t) { return new (t || PoMenuModule)(); };
    static ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: PoMenuModule });
    static ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [CommonModule,
            RouterModule,
            PoBadgeModule,
            PoFieldModule,
            PoLoadingModule,
            PoLogoModule,
            PoIconModule,
            PoTooltipModule] });
}
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(PoMenuModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    RouterModule,
                    PoBadgeModule,
                    PoFieldModule,
                    PoLoadingModule,
                    PoLogoModule,
                    PoIconModule,
                    PoTooltipModule
                ],
                declarations: [PoMenuComponent, PoMenuFilterComponent, PoMenuHeaderTemplateDirective, PoMenuItemComponent],
                exports: [PoMenuComponent, PoMenuHeaderTemplateDirective]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(PoMenuModule, { declarations: [PoMenuComponent, PoMenuFilterComponent, PoMenuHeaderTemplateDirective, PoMenuItemComponent], imports: [CommonModule,
        RouterModule,
        PoBadgeModule,
        PoFieldModule,
        PoLoadingModule,
        PoLogoModule,
        PoIconModule,
        PoTooltipModule], exports: [PoMenuComponent, PoMenuHeaderTemplateDirective] }); })();
