// @totvs/protheus-lib-core/esm2022/services/pro-theme/services/pro-theme.service.mjs
import { Injectable } from '@angular/core';
import { PoThemeTypeEnum } from '@po-ui/ng-components';
import { ProTheme } from '../models/pro-theme';
import { ProThemeEnum } from './../models/pro-theme';
import * as i0 from "@angular/core";
import * as i1 from "@po-ui/ng-components";
export class ProThemeService {
    constructor(poTheme) {
        this.poTheme = poTheme;
        this.proTheme = new ProTheme();
    }
    /**
     * @description Método para configurar o tema do aplicativo com base
     * na configuração da localStorage - ProTheme
     */
    setTheme() {
        const sessionProtheusTheme = this.getProtheusConfig();
        switch (sessionProtheusTheme?.Configuration) {
            case 0: //totvs light theme
                this.proTheme.setTheme(ProThemeEnum.proThemeTotvsClassic);
                this.poTheme.setTheme(this.proTheme, PoThemeTypeEnum.light);
                break;
            case 1: //totvs dark theme
                this.proTheme.setTheme(ProThemeEnum.proThemeTotvsClassic);
                this.poTheme.setTheme(this.proTheme, PoThemeTypeEnum.dark);
                break;
            case 2: //sunset light theme
                this.proTheme.setTheme(ProThemeEnum.proThemeTotvsSunset);
                this.poTheme.setTheme(this.proTheme, PoThemeTypeEnum.light);
                break;
            default:
                this.proTheme.setTheme(ProThemeEnum.proThemeTotvsClassic);
                this.poTheme.setTheme(this.proTheme, PoThemeTypeEnum.light);
        }
    }
    getProtheusConfig() {
        return JSON.parse(localStorage.getItem('ProTheme'));
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProThemeService, deps: [{ token: i1.PoThemeService }], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProThemeService, providedIn: 'root' }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProThemeService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root',
                }]
        }], ctorParameters: () => [{ type: i1.PoThemeService }] });
