// @po-ui/ng-components/esm2022/lib/components/po-listbox/po-search-list/po-search-list.component.mjs
import { ChangeDetectionStrategy, Component, ElementRef, EventEmitter, Input, Output, ViewChild } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "../../po-icon/po-icon.component";
const _c0 = ["inputElement"];
/**
 * @docsPrivate
 *
 * @description
 *
 * Componente de pesquisa que será criado dentro do search do `po-listbox`.
 */
export class PoSearchListComponent {
    cd;
    inputElement;
    /** Propriedade que recebe as literais definidas no `po-listbox`. */
    literals;
    fieldValue;
    /** Evento que será disparado a cada tecla digitada no campo de busca. */
    change = new EventEmitter();
    _placeholder;
    constructor(cd) {
        this.cd = cd;
    }
    /**
     * @optional
     *
     * @description
     *
     * Placeholder do campo de pesquisa.
     *
     * > Caso o mesmo não seja informado, o valor padrão será traduzido com base no idioma do navegador (pt, es e en).
     *
     * @default `Buscar`
     */
    set placeholder(placeholder) {
        this._placeholder =
            placeholder && this.isTypeof(placeholder, 'string') ? placeholder : this.literals.placeholderSearch;
    }
    get placeholder() {
        return this._placeholder || this.literals.placeholderSearch;
    }
    get inputValue() {
        return this.inputElement.nativeElement.value;
    }
    onChange(event) {
        this.change.emit({ event: event, [this.fieldValue]: this.inputElement.nativeElement.value });
    }
    setFocus() {
        this.inputElement.nativeElement.focus();
    }
    clean() {
        this.inputElement.nativeElement.value = '';
        this.cd.markForCheck();
    }
    isTypeof(object, type) {
        return typeof object === type;
    }
    static ɵfac = function PoSearchListComponent_Factory(t) { return new (t || PoSearchListComponent)(i0.ɵɵdirectiveInject(i0.ChangeDetectorRef)); };
    static ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: PoSearchListComponent, selectors: [["po-search-list"]], viewQuery: function PoSearchListComponent_Query(rf, ctx) { if (rf & 1) {
            i0.ɵɵviewQuery(_c0, 7, ElementRef);
        } if (rf & 2) {
            let _t;
            i0.ɵɵqueryRefresh(_t = i0.ɵɵloadQuery()) && (ctx.inputElement = _t.first);
        } }, inputs: { literals: [i0.ɵɵInputFlags.None, "p-literals", "literals"], fieldValue: [i0.ɵɵInputFlags.None, "p-field-value", "fieldValue"], placeholder: [i0.ɵɵInputFlags.None, "p-placeholder", "placeholder"] }, outputs: { change: "p-change" }, decls: 5, vars: 1, consts: [[1, "po-field-container-content", "po-listbox-container-search"], [1, "po-field-icon-container-left"], ["p-icon", "ICON_SEARCH", 1, "po-field-icon", "po-icon-input"], ["type", "text", 1, "po-input", "po-input-icon-left", "po-listbox-search-input", 3, "placeholder", "keyup"], ["inputElement", ""]], template: function PoSearchListComponent_Template(rf, ctx) { if (rf & 1) {
            i0.ɵɵelementStart(0, "div", 0)(1, "div", 1);
            i0.ɵɵelement(2, "po-icon", 2);
            i0.ɵɵelementEnd();
            i0.ɵɵelementStart(3, "input", 3, 4);
            i0.ɵɵlistener("keyup", function PoSearchListComponent_Template_input_keyup_3_listener($event) { return ctx.onChange($event); });
            i0.ɵɵelementEnd()();
        } if (rf & 2) {
            i0.ɵɵadvance(3);
            i0.ɵɵproperty("placeholder", ctx.placeholder);
        } }, dependencies: [i1.PoIconComponent], encapsulation: 2, changeDetection: 0 });
}
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(PoSearchListComponent, [{
        type: Component,
        args: [{ selector: 'po-search-list', changeDetection: ChangeDetectionStrategy.OnPush, template: "<div class=\"po-field-container-content po-listbox-container-search\">\n  <div class=\"po-field-icon-container-left\">\n    <po-icon p-icon=\"ICON_SEARCH\" class=\"po-field-icon po-icon-input\"></po-icon>\n  </div>\n\n  <input\n    #inputElement\n    class=\"po-input po-input-icon-left po-listbox-search-input\"\n    type=\"text\"\n    [placeholder]=\"placeholder\"\n    (keyup)=\"onChange($event)\"\n  />\n</div>\n" }]
    }], () => [{ type: i0.ChangeDetectorRef }], { inputElement: [{
            type: ViewChild,
            args: ['inputElement', { read: ElementRef, static: true }]
        }], literals: [{
            type: Input,
            args: ['p-literals']
        }], fieldValue: [{
            type: Input,
            args: ['p-field-value']
        }], change: [{
            type: Output,
            args: ['p-change']
        }], placeholder: [{
            type: Input,
            args: ['p-placeholder']
        }] }); })();
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassDebugInfo(PoSearchListComponent, { className: "PoSearchListComponent", filePath: "lib/components/po-listbox/po-search-list/po-search-list.component.ts", lineNumber: 25 }); })();
