// @po-ui/ng-templates/esm2022/lib/components/po-page-job-scheduler/po-page-job-scheduler-parameters/po-job-scheduler-parameters-template/po-job-scheduler-parameters-template.directive.mjs
import { Directive, Input } from '@angular/core';
import * as i0 from "@angular/core";
/**
 * @usedBy PoPageJobScheduler
 *
 * @description
 *
 * Esta diretiva permite personalizar o conteúdo da uma ou várias etapa(s) de parametrização do componente de PoPageJobScheduler.
 *
 * Para repassar as alterações realizadas no componente customizado ao model do PoPageJobScheduler, deve
 * ser atualizado os valores através da propriedade p-execution-parameter. Dessa forma as alterações serão adicionadas ao
 * atributo executionParameter do objeto de envio a Api.
 * É possível também controlar a permissão de avançar, fazendo uso da propriedade p-disable-advance.
 *
 * ```
 * ...
 * <po-page-job-scheduler [p-service-api]="serviceApi">
 *     <ng-template p-job-scheduler-parameters-template
 *      [p-execution-parameter]="executionParameter"
 *      [p-disable-advance]="disableAdvance"
 *      p-label-step="titleStep"
 *     >
 *       ...
 *     </ng-template>
 *     <ng-template p-job-scheduler-parameters-template
 *      [p-execution-parameter]="executionParameter"
 *      [p-disable-advance]="disableAdvance"
 *      p-label-step="titleStep"
 *     >
 *       ...
 *     </ng-template>
 * </po-page-job-scheduler>
 * ...
 * ```
 *
 */
export class PoJobSchedulerParametersTemplateDirective {
    templateRef;
    /**
     * @optional
     *
     * @description
     *
     * Objeto que deve conter as alterações feitas pelo componente de template que serão repassadas dentro do atributo
     * `executionParameter` para envio na api.
     *
     * > O componente deve manter essa propriedade atualizada. É chamada após o avançar da etapa de parametrização.
     */
    executionParameter;
    /**
     * @optional
     *
     * @default false
     *
     * @description
     *
     * Determina se deve desabilitar o botão de avançar para a próxima etapa
     *
     * > Pode ser utilizado para validar campos antes de avançar.
     */
    disabledAdvance = false;
    /**
     * @optional
     *
     * @description
     *
     * Determina o label do step
     */
    title;
    // Necessário manter templateRef para o funcionamento do row template.
    constructor(templateRef) {
        this.templateRef = templateRef;
    }
    static ɵfac = function PoJobSchedulerParametersTemplateDirective_Factory(t) { return new (t || PoJobSchedulerParametersTemplateDirective)(i0.ɵɵdirectiveInject(i0.TemplateRef)); };
    static ɵdir = /*@__PURE__*/ i0.ɵɵdefineDirective({ type: PoJobSchedulerParametersTemplateDirective, selectors: [["", "p-job-scheduler-parameters-template", ""]], inputs: { executionParameter: [i0.ɵɵInputFlags.None, "p-execution-parameter", "executionParameter"], disabledAdvance: [i0.ɵɵInputFlags.None, "p-disable-advance", "disabledAdvance"], title: [i0.ɵɵInputFlags.None, "p-label-step", "title"] } });
}
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(PoJobSchedulerParametersTemplateDirective, [{
        type: Directive,
        args: [{
                selector: '[p-job-scheduler-parameters-template]'
            }]
    }], () => [{ type: i0.TemplateRef }], { executionParameter: [{
            type: Input,
            args: ['p-execution-parameter']
        }], disabledAdvance: [{
            type: Input,
            args: ['p-disable-advance']
        }], title: [{
            type: Input,
            args: ['p-label-step']
        }] }); })();
