// @po-ui/ng-components/esm2022/lib/components/po-popover/po-popover.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { PoPopoverComponent } from './po-popover.component';
import * as i0 from "@angular/core";
/**
 * @description
 * Módulo do componente po-popover.
 */
export class PoPopoverModule {
    static ɵfac = function PoPopoverModule_Factory(t) { return new (t || PoPopoverModule)(); };
    static ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: PoPopoverModule });
    static ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [CommonModule] });
}
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(PoPopoverModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule],
                declarations: [PoPopoverComponent],
                exports: [PoPopoverComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(PoPopoverModule, { declarations: [PoPopoverComponent], imports: [CommonModule], exports: [PoPopoverComponent] }); })();
