// @po-ui/ng-components/esm2022/lib/components/po-toolbar/po-toolbar-actions/po-toolbar-actions.component.mjs
import { Component, Input, TemplateRef } from '@angular/core';
import { isTypeof } from '../../../utils/util';
import * as i0 from "@angular/core";
import * as i1 from "../../po-popup/po-popup.component";
import * as i2 from "../../po-icon/po-icon.component";
const poToolbarActionsIconDefault = 'ICON_MORE';
/**
 * @docsPrivate
 *
 * @usedBy PoToolbarComponent
 *
 * @description
 *
 * O componente `po-toolbar-actions` tem como objetivo receber uma lista de ações e um ícone que podem ser personalizados.
 */
export class PoToolbarActionsComponent {
    /** Define uma lista de ações. */
    actions;
    _actionsIcon = poToolbarActionsIconDefault;
    /** Define o ícone das ações. */
    set actionsIcon(icon) {
        this._actionsIcon = isTypeof(icon, 'string') || icon instanceof TemplateRef ? icon : poToolbarActionsIconDefault;
    }
    get actionsIcon() {
        return this._actionsIcon;
    }
    static ɵfac = function PoToolbarActionsComponent_Factory(t) { return new (t || PoToolbarActionsComponent)(); };
    static ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: PoToolbarActionsComponent, selectors: [["po-toolbar-actions"]], inputs: { actions: [i0.ɵɵInputFlags.None, "p-actions", "actions"], actionsIcon: [i0.ɵɵInputFlags.None, "p-actions-icon", "actionsIcon"] }, decls: 6, vars: 3, consts: [[1, "po-toolbar-actions", 3, "click"], ["toolbarActions", ""], [1, "po-toolbar-icon", "po-clickable", 3, "p-icon"], [3, "p-actions", "p-target"], ["popup", ""]], template: function PoToolbarActionsComponent_Template(rf, ctx) { if (rf & 1) {
            const _r2 = i0.ɵɵgetCurrentView();
            i0.ɵɵelementStart(0, "div", 0);
            i0.ɵɵlistener("click", function PoToolbarActionsComponent_Template_div_click_0_listener() { i0.ɵɵrestoreView(_r2); const _r1 = i0.ɵɵreference(5); return i0.ɵɵresetView(_r1.toggle()); });
            i0.ɵɵelementStart(1, "span", null, 1);
            i0.ɵɵelement(3, "po-icon", 2);
            i0.ɵɵelementEnd()();
            i0.ɵɵelement(4, "po-popup", 3, 4);
        } if (rf & 2) {
            const _r0 = i0.ɵɵreference(2);
            i0.ɵɵadvance(3);
            i0.ɵɵproperty("p-icon", ctx.actionsIcon);
            i0.ɵɵadvance();
            i0.ɵɵproperty("p-actions", ctx.actions)("p-target", _r0);
        } }, dependencies: [i1.PoPopupComponent, i2.PoIconComponent], encapsulation: 2 });
}
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(PoToolbarActionsComponent, [{
        type: Component,
        args: [{ selector: 'po-toolbar-actions', template: "<div class=\"po-toolbar-actions\" (click)=\"popup.toggle()\">\n  <span #toolbarActions>\n    <po-icon class=\"po-toolbar-icon po-clickable\" [p-icon]=\"actionsIcon\"></po-icon>\n  </span>\n</div>\n\n<po-popup #popup [p-actions]=\"actions\" [p-target]=\"toolbarActions\"> </po-popup>\n" }]
    }], null, { actions: [{
            type: Input,
            args: ['p-actions']
        }], actionsIcon: [{
            type: Input,
            args: ['p-actions-icon']
        }] }); })();
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassDebugInfo(PoToolbarActionsComponent, { className: "PoToolbarActionsComponent", filePath: "lib/components/po-toolbar/po-toolbar-actions/po-toolbar-actions.component.ts", lineNumber: 22 }); })();
