// @totvs/protheus-lib-core/esm2022/services/pro-authorization/services/pro-user-info/services/pro-user-access/pro-user-access.service.mjs
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common/http";
export class ProUserAccessService {
    constructor(http) {
        this.http = http;
        this.URL_ALIAS = '/api/framework/v1/accessPrivilegesServices/aliasAccess/';
        this.URL_FUNCTION = '/api/framework/v1/accessPrivilegesServices/functionAccess/';
    }
    /**
     * efetua um get na api referente a função MpUserHasAccess do protheus
     * a api utiliza o usuário logado para efetuar a busca
     * @param cRotina Nome da rotina a ser pesquisada
     * @param nOpc parametro opcional, numero da rotina no menudef a ser pesquisada
     * caso não seja passada a pesquisa será com base apenas na função
     * @returns Observable com o retorno da api
     */
    userHasAccess(cRotina, nOpc = 0) {
        if (nOpc) {
            return this.http.get(`${this.URL_FUNCTION}${cRotina}/${nOpc}`);
        }
        else {
            return this.http.get(`${this.URL_FUNCTION}${cRotina}`);
        }
    }
    /**
     * efetua um get na api referente a função FWChkTblAccess do protheus
     * a api utiliza o usuário logado para efetuar a busca
     * @param cTabela Tabela a ser pesquisada
     * @param nOpc parametro opcional, numero da rotina no menudef a ser pesquisada
     * caso não seja passada a pesquisa será com base apenas na função
     * @returns Observable com o retorno da api
     */
    aliasHasAccess(cTabela, nOpc = 0) {
        if (nOpc) {
            return this.http.get(`${this.URL_ALIAS}${cTabela}/${nOpc}`);
        }
        else {
            return this.http.get(`${this.URL_ALIAS}${cTabela}`);
        }
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProUserAccessService, deps: [{ token: i1.HttpClient }], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProUserAccessService, providedIn: 'root' }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProUserAccessService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root'
                }]
        }], ctorParameters: () => [{ type: i1.HttpClient }] });
