// @totvs/protheus-lib-core/esm2022/services/pro-system-info/pro-system-modules/interceptors/pro-system-modules.inteceptor.mjs
import { Injectable } from '@angular/core';
import { ProAppConfigService } from '../../../pro-app-config';
import * as i0 from "@angular/core";
const CACHE_KEY = 'ProModule';
/**
 * @description
 *
 * Interceptor que pega a requisição html e insere no header o módulo.
 */
export class ProSystemModulesInteceptor {
    constructor(injector) {
        this.injector = injector;
    }
    /**
     * @description Intercept http para tratamento do módulo do Protheus nas requisições
     * @param original_request Requisição http
     * @param next Handler da requisição http
     * @returns Observable da requisição http com o header do módulo caso o ERP seja Protheus
     */
    intercept(original_request, next) {
        let requestResult = original_request;
        const proAppConfigService = this.injector.get(ProAppConfigService);
        if (!(original_request.url.search(/assets/) >= 0)) {
            const isProtheus = proAppConfigService.isProtheusRender;
            if (isProtheus) {
                requestResult = this.appendModuleToRequest(original_request);
            }
        }
        return next.handle(requestResult);
    }
    appendModuleToRequest(request) {
        const module = this.getModule();
        if (module) {
            return request.clone({
                headers: request.headers.set('x-erp-module', module)
            });
        }
        else {
            return request.clone();
        }
    }
    getModule() {
        return sessionStorage[CACHE_KEY];
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProSystemModulesInteceptor, deps: [{ token: i0.Injector }], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProSystemModulesInteceptor, providedIn: 'root' }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProSystemModulesInteceptor, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root'
                }]
        }], ctorParameters: () => [{ type: i0.Injector }] });
