// @po-ui/ng-components/esm2022/lib/components/po-menu/po-menu-filter/po-menu-filter.component.mjs
import { Component, ElementRef, EventEmitter, Input, Output, ViewChild } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "../../../services/po-language/po-language.service";
import * as i2 from "@angular/common";
import * as i3 from "../../po-field/po-clean/po-clean.component";
import * as i4 from "../../po-loading/po-loading-icon/po-loading-icon.component";
import * as i5 from "../../po-icon/po-icon.component";
const _c0 = ["inputFilter"];
function PoMenuFilterComponent_po_icon_4_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelement(0, "po-icon", 8);
} }
function PoMenuFilterComponent_po_loading_icon_5_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelement(0, "po-loading-icon", 9);
} }
export const poMenuFilterLiteralsDefault = {
    en: { search: 'Search' },
    es: { search: 'Buscar' },
    pt: { search: 'Pesquisar' },
    ru: { search: 'Поиск' }
};
/**
 * @docsPrivate
 *
 * @description
 *
 * Componente que apresenta o campo de pesquisa no po-menu.
 */
export class PoMenuFilterComponent {
    languageService;
    loading;
    enableCollapse = false;
    // utilizado para repassar ao po-clean
    inputFilterElement;
    filter = new EventEmitter();
    literals;
    constructor(languageService) {
        this.languageService = languageService;
        this.literals = {
            ...poMenuFilterLiteralsDefault[this.languageService?.getLanguageDefault()],
            ...poMenuFilterLiteralsDefault[this.languageService?.getShortLanguage()]
        };
    }
    filterItems(search) {
        this.filter.emit(search);
    }
    static ɵfac = function PoMenuFilterComponent_Factory(t) { return new (t || PoMenuFilterComponent)(i0.ɵɵdirectiveInject(i1.PoLanguageService)); };
    static ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: PoMenuFilterComponent, selectors: [["po-menu-filter"]], viewQuery: function PoMenuFilterComponent_Query(rf, ctx) { if (rf & 1) {
            i0.ɵɵviewQuery(_c0, 7, ElementRef);
        } if (rf & 2) {
            let _t;
            i0.ɵɵqueryRefresh(_t = i0.ɵɵloadQuery()) && (ctx.inputFilterElement = _t.first);
        } }, inputs: { loading: [i0.ɵɵInputFlags.None, "p-loading", "loading"], enableCollapse: [i0.ɵɵInputFlags.None, "p-enable-collapse", "enableCollapse"] }, outputs: { filter: "p-filter" }, decls: 8, vars: 6, consts: [[1, "po-menu-filter-container"], ["type", "text", 1, "po-menu-filter", 3, "placeholder", "keyup"], ["inputFilter", ""], [1, "po-menu-filter-search-icon-container"], ["p-icon", "ICON_SEARCH po-field-icon", "class", "po-menu-filter-icon", 4, "ngIf"], ["p-size", "sm", 4, "ngIf"], [1, "po-menu-filter-close-icon-container"], [3, "p-element-ref", "p-change-event"], ["p-icon", "ICON_SEARCH po-field-icon", 1, "po-menu-filter-icon"], ["p-size", "sm"]], template: function PoMenuFilterComponent_Template(rf, ctx) { if (rf & 1) {
            const _r3 = i0.ɵɵgetCurrentView();
            i0.ɵɵelementStart(0, "div", 0)(1, "input", 1, 2);
            i0.ɵɵlistener("keyup", function PoMenuFilterComponent_Template_input_keyup_1_listener() { i0.ɵɵrestoreView(_r3); const _r0 = i0.ɵɵreference(2); return i0.ɵɵresetView(ctx.filterItems(_r0.value)); });
            i0.ɵɵelementEnd();
            i0.ɵɵelementStart(3, "div", 3);
            i0.ɵɵtemplate(4, PoMenuFilterComponent_po_icon_4_Template, 1, 0, "po-icon", 4)(5, PoMenuFilterComponent_po_loading_icon_5_Template, 1, 0, "po-loading-icon", 5);
            i0.ɵɵelementEnd();
            i0.ɵɵelementStart(6, "div", 6)(7, "po-clean", 7);
            i0.ɵɵlistener("p-change-event", function PoMenuFilterComponent_Template_po_clean_p_change_event_7_listener() { i0.ɵɵrestoreView(_r3); const _r0 = i0.ɵɵreference(2); return i0.ɵɵresetView(ctx.filterItems(_r0.value)); });
            i0.ɵɵelementEnd()()();
        } if (rf & 2) {
            i0.ɵɵclassProp("po-menu-filter-container-hidden", ctx.enableCollapse);
            i0.ɵɵadvance();
            i0.ɵɵproperty("placeholder", ctx.literals.search);
            i0.ɵɵadvance(3);
            i0.ɵɵproperty("ngIf", !ctx.loading);
            i0.ɵɵadvance();
            i0.ɵɵproperty("ngIf", ctx.loading);
            i0.ɵɵadvance(2);
            i0.ɵɵproperty("p-element-ref", ctx.inputFilterElement);
        } }, dependencies: [i2.NgIf, i3.PoCleanComponent, i4.PoLoadingIconComponent, i5.PoIconComponent], encapsulation: 2 });
}
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(PoMenuFilterComponent, [{
        type: Component,
        args: [{ selector: 'po-menu-filter', template: "<div class=\"po-menu-filter-container\" [class.po-menu-filter-container-hidden]=\"enableCollapse\">\n  <input\n    #inputFilter\n    type=\"text\"\n    class=\"po-menu-filter\"\n    [placeholder]=\"literals.search\"\n    (keyup)=\"filterItems(inputFilter.value)\"\n  />\n\n  <div class=\"po-menu-filter-search-icon-container\">\n    <po-icon p-icon=\"ICON_SEARCH po-field-icon\" *ngIf=\"!loading\" class=\"po-menu-filter-icon\"></po-icon>\n    <po-loading-icon *ngIf=\"loading\" p-size=\"sm\"></po-loading-icon>\n  </div>\n  <div class=\"po-menu-filter-close-icon-container\">\n    <po-clean [p-element-ref]=\"inputFilterElement\" (p-change-event)=\"filterItems(inputFilter.value)\"></po-clean>\n  </div>\n</div>\n" }]
    }], () => [{ type: i1.PoLanguageService }], { loading: [{
            type: Input,
            args: ['p-loading']
        }], enableCollapse: [{
            type: Input,
            args: ['p-enable-collapse']
        }], inputFilterElement: [{
            type: ViewChild,
            args: ['inputFilter', { read: ElementRef, static: true }]
        }], filter: [{
            type: Output,
            args: ['p-filter']
        }] }); })();
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassDebugInfo(PoMenuFilterComponent, { className: "PoMenuFilterComponent", filePath: "lib/components/po-menu/po-menu-filter/po-menu-filter.component.ts", lineNumber: 23 }); })();
