// @totvs/protheus-lib-core/esm2022/components/pro-pages/pro-session-settings/guards/pro-session-settings-defaults/pro-session-settings-defaults.resolver.mjs
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "../../services/pro-session-settings.service";
export class ProSessionSettingsDefaultsResolver {
    constructor(sessionSettingsService) {
        this.sessionSettingsService = sessionSettingsService;
    }
    resolve(_route, _state) {
        return this.sessionSettingsService.requestSettingsDefaults();
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProSessionSettingsDefaultsResolver, deps: [{ token: i1.ProSessionSettingsService }], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProSessionSettingsDefaultsResolver, providedIn: 'root' }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProSessionSettingsDefaultsResolver, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root'
                }]
        }], ctorParameters: () => [{ type: i1.ProSessionSettingsService }] });
