// @po-ui/ng-components/esm2022/lib/components/po-page/po-page-slide/po-page-slide.component.mjs
import { animate, animateChild, group, query, style, transition, trigger } from '@angular/animations';
import { Component, ElementRef, ViewChild } from '@angular/core';
import { ReplaySubject } from 'rxjs';
import { delay, take } from 'rxjs/operators';
import { getFocusableElements, uuid } from '../../../utils/util';
import { PoPageSlideBaseComponent } from './po-page-slide-base.component';
import * as i0 from "@angular/core";
import * as i1 from "../../../services/po-active-overlay/po-active-overlay.service";
import * as i2 from "@angular/common";
import * as i3 from "../../po-button/po-button.component";
const _c0 = ["pageContent"];
function PoPageSlideComponent_div_0_div_9_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "div", 11);
    i0.ɵɵtext(1);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r2 = i0.ɵɵnextContext(2);
    i0.ɵɵadvance();
    i0.ɵɵtextInterpolate(ctx_r2.subtitle);
} }
function PoPageSlideComponent_div_0_po_button_10_Template(rf, ctx) { if (rf & 1) {
    const _r5 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "po-button", 12);
    i0.ɵɵlistener("p-click", function PoPageSlideComponent_div_0_po_button_10_Template_po_button_p_click_0_listener() { i0.ɵɵrestoreView(_r5); const ctx_r4 = i0.ɵɵnextContext(2); return i0.ɵɵresetView(ctx_r4.close()); });
    i0.ɵɵelementEnd();
} }
const _c1 = a0 => ({ "width": a0 });
function PoPageSlideComponent_div_0_Template(rf, ctx) { if (rf & 1) {
    const _r7 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "div", 1)(1, "div", 2);
    i0.ɵɵlistener("mousedown", function PoPageSlideComponent_div_0_Template_div_mousedown_1_listener($event) { i0.ɵɵrestoreView(_r7); const ctx_r6 = i0.ɵɵnextContext(); return i0.ɵɵresetView(ctx_r6.onClickOut($event)); });
    i0.ɵɵelementEnd();
    i0.ɵɵelementStart(2, "div", 3)(3, "div", 4, 5)(5, "div", 6)(6, "div", 7)(7, "span");
    i0.ɵɵtext(8);
    i0.ɵɵelementEnd();
    i0.ɵɵtemplate(9, PoPageSlideComponent_div_0_div_9_Template, 2, 1, "div", 8);
    i0.ɵɵelementEnd();
    i0.ɵɵtemplate(10, PoPageSlideComponent_div_0_po_button_10_Template, 1, 0, "po-button", 9);
    i0.ɵɵelementEnd();
    i0.ɵɵelementStart(11, "div", 10);
    i0.ɵɵprojection(12);
    i0.ɵɵelementEnd()()()();
} if (rf & 2) {
    const ctx_r0 = i0.ɵɵnextContext();
    i0.ɵɵproperty("@fade", undefined);
    i0.ɵɵadvance(2);
    i0.ɵɵclassMapInterpolate1("po-page-slide-container po-page-slide-right po-page-slide-", ctx_r0.size, "");
    i0.ɵɵproperty("@slide", undefined)("ngStyle", i0.ɵɵpureFunction1(9, _c1, ctx_r0.flexibleWidth ? "" : ctx_r0.size === "auto" ? "auto" : ""));
    i0.ɵɵadvance(6);
    i0.ɵɵtextInterpolate(ctx_r0.title);
    i0.ɵɵadvance();
    i0.ɵɵproperty("ngIf", ctx_r0.subtitle);
    i0.ɵɵadvance();
    i0.ɵɵproperty("ngIf", !ctx_r0.hideClose);
} }
const _c2 = ["*"];
/**
 * @docsExtends PoPageSlideBaseComponent
 *
 * @example
 *
 * <example name="po-page-slide-basic" title="PO Page Slide Basic">
 *  <file name="sample-po-page-slide-basic/sample-po-page-slide-basic.component.html"> </file>
 *  <file name="sample-po-page-slide-basic/sample-po-page-slide-basic.component.ts"> </file>
 * </example>
 *
 * <example name="po-page-slide-labs" title="PO Page Slide Labs">
 *  <file name="sample-po-page-slide-labs/sample-po-page-slide-labs.component.html"> </file>
 *  <file name="sample-po-page-slide-labs/sample-po-page-slide-labs.component.ts"> </file>
 * </example>
 *
 * <example name="po-page-slide-configuration" title="PO Page Slide - Configuration">
 *  <file name="sample-po-page-slide-configuration/sample-po-page-slide-configuration.component.html"> </file>
 *  <file name="sample-po-page-slide-configuration/sample-po-page-slide-configuration.component.ts"> </file>
 * </example>
 */
export class PoPageSlideComponent extends PoPageSlideBaseComponent {
    poActiveOverlayService;
    _pageContent;
    firstElement;
    id = uuid();
    loadingCompleted = new ReplaySubject();
    sourceElement;
    focusEvent;
    set pageContent(pageContent) {
        if (pageContent) {
            this._pageContent = pageContent;
            this.loadingCompleted.next();
        }
    }
    get pageContent() {
        return this._pageContent;
    }
    constructor(poActiveOverlayService) {
        super();
        this.poActiveOverlayService = poActiveOverlayService;
    }
    open() {
        this.sourceElement = document.activeElement;
        super.open();
        this.loadingCompleted.pipe(take(1)).pipe(delay(0)).subscribe(this.handleFocus.bind(this));
    }
    close() {
        this.poActiveOverlayService.activeOverlay.pop();
        super.close();
        this.removeEventListeners();
        this.sourceElement.focus();
    }
    onClickOut(event) {
        if (this.clickOut && !this.pageContent.nativeElement.contains(event.target)) {
            this.close();
        }
    }
    handleFocus() {
        this.poActiveOverlayService.activeOverlay.push(this.id);
        this.loadFirstElement();
        this.initFocus();
        document.addEventListener('focus', this.focusEvent, true);
    }
    initFocus() {
        // O foco não pode sair da página.
        this.focusEvent = (event) => {
            if (!this.pageContent.nativeElement.contains(event.target) &&
                this.poActiveOverlayService.activeOverlay[this.poActiveOverlayService.activeOverlay.length - 1] === this.id) {
                event.stopPropagation();
                this.firstElement.focus();
            }
        };
        if (this.hideClose) {
            this.firstElement.focus();
        }
        else {
            const elements = getFocusableElements(this.pageContent.nativeElement);
            const element = elements[0] || this.pageContent.nativeElement;
            element.focus();
        }
    }
    loadFirstElement() {
        this.firstElement = getFocusableElements(this.pageContent.nativeElement)[0] || this.pageContent.nativeElement;
    }
    removeEventListeners() {
        document.removeEventListener('focus', this.focusEvent, true);
        this.loadingCompleted.complete();
    }
    static ɵfac = function PoPageSlideComponent_Factory(t) { return new (t || PoPageSlideComponent)(i0.ɵɵdirectiveInject(i1.PoActiveOverlayService)); };
    static ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: PoPageSlideComponent, selectors: [["po-page-slide"]], viewQuery: function PoPageSlideComponent_Query(rf, ctx) { if (rf & 1) {
            i0.ɵɵviewQuery(_c0, 5, ElementRef);
        } if (rf & 2) {
            let _t;
            i0.ɵɵqueryRefresh(_t = i0.ɵɵloadQuery()) && (ctx.pageContent = _t.first);
        } }, features: [i0.ɵɵProvidersFeature([]), i0.ɵɵInheritDefinitionFeature], ngContentSelectors: _c2, decls: 1, vars: 1, consts: [["class", "po-page-slide", "tabindex", "0", 4, "ngIf"], ["tabindex", "0", 1, "po-page-slide"], [1, "po-page-slide-overlay", 3, "mousedown"], [3, "ngStyle"], ["tabindex", "-1", 1, "po-page-slide-content"], ["pageContent", ""], [1, "po-page-slide-header"], [1, "po-page-slide-title"], ["class", "po-page-slide-subtitle", 4, "ngIf"], ["p-icon", "ICON_CLOSE", "p-kind", "tertiary", 3, "p-click", 4, "ngIf"], [1, "po-page-slide-body"], [1, "po-page-slide-subtitle"], ["p-icon", "ICON_CLOSE", "p-kind", "tertiary", 3, "p-click"]], template: function PoPageSlideComponent_Template(rf, ctx) { if (rf & 1) {
            i0.ɵɵprojectionDef();
            i0.ɵɵtemplate(0, PoPageSlideComponent_div_0_Template, 13, 11, "div", 0);
        } if (rf & 2) {
            i0.ɵɵproperty("ngIf", !ctx.hidden);
        } }, dependencies: [i2.NgIf, i2.NgStyle, i3.PoButtonComponent], encapsulation: 2, data: { animation: [
                trigger('fade', [
                    transition(':enter', [
                        style({ opacity: 0 }),
                        group([animate('150ms', style({ opacity: 1 })), query('@slide', animateChild())])
                    ]),
                    transition(':leave', group([query('@slide', animateChild()), animate('150ms', style({ opacity: 0 }))]))
                ]),
                trigger('slide', [
                    transition(':enter', [
                        style({ transform: 'translateX(50px)' }),
                        animate('691ms ease-in-out', style({ transform: 'none' }))
                    ]),
                    transition(':leave', [animate('150ms', style({ transform: 'translateX(50px)' }))])
                ])
            ] } });
}
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(PoPageSlideComponent, [{
        type: Component,
        args: [{ selector: 'po-page-slide', providers: [], animations: [
                    trigger('fade', [
                        transition(':enter', [
                            style({ opacity: 0 }),
                            group([animate('150ms', style({ opacity: 1 })), query('@slide', animateChild())])
                        ]),
                        transition(':leave', group([query('@slide', animateChild()), animate('150ms', style({ opacity: 0 }))]))
                    ]),
                    trigger('slide', [
                        transition(':enter', [
                            style({ transform: 'translateX(50px)' }),
                            animate('691ms ease-in-out', style({ transform: 'none' }))
                        ]),
                        transition(':leave', [animate('150ms', style({ transform: 'translateX(50px)' }))])
                    ])
                ], template: "<div class=\"po-page-slide\" tabindex=\"0\" *ngIf=\"!hidden\" [@fade]>\n  <div class=\"po-page-slide-overlay\" (mousedown)=\"onClickOut($event)\"></div>\n  <div\n    class=\"po-page-slide-container po-page-slide-right po-page-slide-{{ size }}\"\n    [@slide]\n    [ngStyle]=\"{ 'width': flexibleWidth ? '' : size === 'auto' ? 'auto' : '' }\"\n  >\n    <div class=\"po-page-slide-content\" tabindex=\"-1\" #pageContent>\n      <div class=\"po-page-slide-header\">\n        <div class=\"po-page-slide-title\">\n          <span>{{ title }}</span>\n          <div class=\"po-page-slide-subtitle\" *ngIf=\"subtitle\">{{ subtitle }}</div>\n        </div>\n        <po-button *ngIf=\"!hideClose\" p-icon=\"ICON_CLOSE\" p-kind=\"tertiary\" (p-click)=\"close()\"> </po-button>\n      </div>\n      <div class=\"po-page-slide-body\">\n        <ng-content></ng-content>\n      </div>\n    </div>\n  </div>\n</div>\n" }]
    }], () => [{ type: i1.PoActiveOverlayService }], { pageContent: [{
            type: ViewChild,
            args: ['pageContent', { read: ElementRef }]
        }] }); })();
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassDebugInfo(PoPageSlideComponent, { className: "PoPageSlideComponent", filePath: "lib/components/po-page/po-page-slide/po-page-slide.component.ts", lineNumber: 52 }); })();
