// @totvs/protheus-lib-core/esm2022/services/pro-language/interceptors/pro-system-idiom.inteceptor.mjs
import { Injectable } from '@angular/core';
import { ProAppConfigService } from '../../pro-app-config';
import * as i0 from "@angular/core";
const CACHE_KEY = 'ProIdiom';
/**
 * @description
 *
 * Interceptor que pega a requisição html e insere no header do idioma.
 */
export class ProSystemIdiomInteceptor {
    constructor(injector) {
        this.injector = injector;
    }
    /**
     * @description Intercept http para tratamento do módulo do Protheus nas requisições
     * @param original_request Requisição http
     * @param next Handler da requisição http
     * @returns Observable da requisição http com o header do módulo caso o ERP seja Protheus
     */
    intercept(original_request, next) {
        let requestResult = original_request;
        const proAppConfigService = this.injector.get(ProAppConfigService);
        if (!(original_request.url.search(/assets/) >= 0)) {
            const isProtheus = proAppConfigService.isProtheusRender;
            if (isProtheus) {
                requestResult = this.appendLanguageToRequest(original_request);
            }
        }
        return next.handle(requestResult);
    }
    /**
     * @description Efetua a validação e adição do header de idioma
     * @param request Requisição HTTP Original
     * @returns HttpRequest, podendo ser alterada com a inclusão do idioma conforme o Protheus
     */
    appendLanguageToRequest(request) {
        const language = this.getLanguage();
        if (language) {
            return request.clone({
                headers: request.headers.set('Accept-Language', language)
            });
        }
        else {
            return request.clone();
        }
    }
    /**
     * @description Retorna o idioma presente no storage do navegador
     * @returns string contendo o valor de idioma do Protheus
     */
    getLanguage() {
        return sessionStorage[CACHE_KEY];
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProSystemIdiomInteceptor, deps: [{ token: i0.Injector }], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProSystemIdiomInteceptor, providedIn: 'root' }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProSystemIdiomInteceptor, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root'
                }]
        }], ctorParameters: () => [{ type: i0.Injector }] });
