// @po-ui/ng-components/schematics/ng-update/v15/changes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.poButtonReplacesDanger = exports.poButtonReplacesTertiary = exports.poButtonReplacesSecondary = exports.poButtonReplaces = exports.updateDepedenciesVersion = exports.regexPoDanger = exports.regexPoTertiary = exports.regexPoSecondary = exports.regexPoButton = void 0;
//regex para alterar o po-button de `p-type` para p-kind
exports.regexPoButton = new RegExp('(<po-button\\b[^<>]* p-)type(="[^"]*"[^<>]*>)', 'gmi');
//regex para alterar o po-button de `p-type=default` para `p-kind=secondary`
exports.regexPoSecondary = new RegExp('(<po-button\\b[^<>]* p-)type(="default")', 'gmi');
//regex para alterar o po-button de `p-type=link` para `p-kind=tertiary`
exports.regexPoTertiary = new RegExp('(<po-button\\b[^<>]* p-)type(="link")', 'gmi');
//regex para alterar o po-button de `p-type=danger` para `p-danger=true`
exports.regexPoDanger = new RegExp('(<po-button\\b[^<>]*)p-type="danger"', 'gmi');
exports.updateDepedenciesVersion = {
    dependencies: [
        '@po-ui/ng-components',
        '@po-ui/ng-code-editor',
        '@po-ui/ng-templates',
        '@po-ui/ng-storage',
        '@po-ui/ng-sync',
        '@totvs/po-theme',
        '@po-ui/style'
    ]
};
//altera a propriedade p-type do  button para p-kind utilizando o regex regexPoButton
exports.poButtonReplaces = [{ replace: exports.regexPoButton, replaceWith: `$1kind$2` }];
//altera para o secondary utilizando o regex da constante regexPoSecondary
exports.poButtonReplacesSecondary = [
    { replace: exports.regexPoSecondary, replaceWith: `$1kind="secondary"` }
];
//altera para o tertiary utilizando o regex da constante regexPoTertiary
exports.poButtonReplacesTertiary = [
    { replace: exports.regexPoTertiary, replaceWith: `$1kind="tertiary"` }
];
//altera para [p-danger]=true utilizando o regex da constante regexPoDanger
exports.poButtonReplacesDanger = [
    { replace: exports.regexPoDanger, replaceWith: `$1[p-danger]="true"` }
];
