// @po-ui/ng-components/esm2022/lib/components/po-table/po-table-list-manager/po-table-list-manager.component.mjs
import { Component, EventEmitter, Input, Output, forwardRef } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { poLocaleDefault } from '../../../services/po-language/po-language.constant';
import { convertToBoolean } from '../../../utils/util';
import { PoCheckboxGroupComponent } from '../../po-field/po-checkbox-group/po-checkbox-group.component';
import * as i0 from "@angular/core";
import * as i1 from "../../../services/po-language/po-language.service";
import * as i2 from "@angular/common";
import * as i3 from "../../po-button/po-button.component";
import * as i4 from "../../../directives/po-tooltip/po-tooltip.directive";
import * as i5 from "../../po-field/po-switch/po-switch.component";
function PoTableListManagerComponent_section_0_ng_container_3_ng_container_1_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainer(0);
} }
const _c0 = a0 => ({ $implicit: a0 });
function PoTableListManagerComponent_section_0_ng_container_3_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵelementStart(1, "li", 8);
    i0.ɵɵtemplate(2, PoTableListManagerComponent_section_0_ng_container_3_ng_container_1_ng_container_2_Template, 1, 0, "ng-container", 9);
    i0.ɵɵelementEnd();
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const option_r9 = i0.ɵɵnextContext().$implicit;
    const ctx_r10 = i0.ɵɵnextContext(2);
    const _r3 = i0.ɵɵreference(6);
    i0.ɵɵadvance();
    i0.ɵɵclassProp("po-checkbox-group-item-disabled", option_r9.disabled || ctx_r10.disabled);
    i0.ɵɵadvance();
    i0.ɵɵproperty("ngTemplateOutlet", _r3)("ngTemplateOutletContext", i0.ɵɵpureFunction1(4, _c0, option_r9));
} }
function PoTableListManagerComponent_section_0_ng_container_3_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtemplate(1, PoTableListManagerComponent_section_0_ng_container_3_ng_container_1_Template, 3, 6, "ng-container", 7);
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const option_r9 = ctx.$implicit;
    const ctx_r8 = i0.ɵɵnextContext(2);
    i0.ɵɵadvance();
    i0.ɵɵproperty("ngIf", ctx_r8.isFixed(option_r9));
} }
function PoTableListManagerComponent_section_0_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "section", 1)(1, "p", 2);
    i0.ɵɵtext(2);
    i0.ɵɵelementEnd();
    i0.ɵɵtemplate(3, PoTableListManagerComponent_section_0_ng_container_3_Template, 2, 1, "ng-container", 3);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = i0.ɵɵnextContext();
    i0.ɵɵadvance(2);
    i0.ɵɵtextInterpolate(ctx_r0.literals.fixedColumns);
    i0.ɵɵadvance();
    i0.ɵɵproperty("ngForOf", ctx_r0.checkboxGroupOptionsView)("ngForTrackBy", ctx_r0.trackByFn);
} }
function PoTableListManagerComponent_ng_container_4_ng_container_1_ng_container_2_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainer(0);
} }
function PoTableListManagerComponent_ng_container_4_ng_container_1_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtemplate(1, PoTableListManagerComponent_ng_container_4_ng_container_1_ng_container_2_ng_container_1_Template, 1, 0, "ng-container", 9);
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const option_r13 = i0.ɵɵnextContext(2).$implicit;
    i0.ɵɵnextContext();
    const _r3 = i0.ɵɵreference(6);
    i0.ɵɵadvance();
    i0.ɵɵproperty("ngTemplateOutlet", _r3)("ngTemplateOutletContext", i0.ɵɵpureFunction1(2, _c0, option_r13));
} }
function PoTableListManagerComponent_ng_container_4_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵelementStart(1, "li", 8);
    i0.ɵɵtemplate(2, PoTableListManagerComponent_ng_container_4_ng_container_1_ng_container_2_Template, 2, 4, "ng-container", 7);
    i0.ɵɵelementEnd();
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const option_r13 = i0.ɵɵnextContext().$implicit;
    const ctx_r14 = i0.ɵɵnextContext();
    i0.ɵɵadvance();
    i0.ɵɵclassProp("po-checkbox-group-item-disabled", option_r13.disabled || ctx_r14.disabled);
    i0.ɵɵadvance();
    i0.ɵɵproperty("ngIf", !ctx_r14.isFixed(option_r13));
} }
function PoTableListManagerComponent_ng_container_4_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtemplate(1, PoTableListManagerComponent_ng_container_4_ng_container_1_Template, 3, 3, "ng-container", 7);
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const option_r13 = ctx.$implicit;
    const ctx_r1 = i0.ɵɵnextContext();
    i0.ɵɵadvance();
    i0.ɵɵproperty("ngIf", !ctx_r1.isFixed(option_r13));
} }
function PoTableListManagerComponent_ng_template_5_po_button_6_Template(rf, ctx) { if (rf & 1) {
    const _r23 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "po-button", 18);
    i0.ɵɵlistener("p-click", function PoTableListManagerComponent_ng_template_5_po_button_6_Template_po_button_p_click_0_listener() { i0.ɵɵrestoreView(_r23); const option_r19 = i0.ɵɵnextContext().$implicit; const ctx_r21 = i0.ɵɵnextContext(); return i0.ɵɵresetView(ctx_r21.emitFixed(option_r19)); });
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const option_r19 = i0.ɵɵnextContext().$implicit;
    const ctx_r20 = i0.ɵɵnextContext();
    const _r7 = i0.ɵɵreference(10);
    const _r5 = i0.ɵɵreference(8);
    i0.ɵɵproperty("p-icon", ctx_r20.isFixed(option_r19) ? _r7 : _r5)("p-disabled", !option_r19.visible || ctx_r20.checksIfHasFiveFixed(option_r19));
} }
function PoTableListManagerComponent_ng_template_5_Template(rf, ctx) { if (rf & 1) {
    const _r26 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "div", 10)(1, "div", 11)(2, "po-switch", 12);
    i0.ɵɵlistener("p-change", function PoTableListManagerComponent_ng_template_5_Template_po_switch_p_change_2_listener() { const restoredCtx = i0.ɵɵrestoreView(_r26); const option_r19 = restoredCtx.$implicit; const ctx_r25 = i0.ɵɵnextContext(); return i0.ɵɵresetView(ctx_r25.clickSwitch(option_r19)); })("keydown", function PoTableListManagerComponent_ng_template_5_Template_po_switch_keydown_2_listener($event) { const restoredCtx = i0.ɵɵrestoreView(_r26); const option_r19 = restoredCtx.$implicit; const ctx_r27 = i0.ɵɵnextContext(); return i0.ɵɵresetView(ctx_r27.onKeyDown($event, option_r19)); });
    i0.ɵɵelementEnd();
    i0.ɵɵelementStart(3, "span", 13);
    i0.ɵɵtext(4);
    i0.ɵɵelementEnd()();
    i0.ɵɵelementStart(5, "div", 14);
    i0.ɵɵtemplate(6, PoTableListManagerComponent_ng_template_5_po_button_6_Template, 1, 2, "po-button", 15);
    i0.ɵɵelementStart(7, "po-button", 16);
    i0.ɵɵlistener("p-click", function PoTableListManagerComponent_ng_template_5_Template_po_button_p_click_7_listener() { const restoredCtx = i0.ɵɵrestoreView(_r26); const option_r19 = restoredCtx.$implicit; const ctx_r28 = i0.ɵɵnextContext(); return i0.ɵɵresetView(ctx_r28.emitChangePosition(option_r19, "up")); });
    i0.ɵɵelementEnd();
    i0.ɵɵelementStart(8, "po-button", 17);
    i0.ɵɵlistener("p-click", function PoTableListManagerComponent_ng_template_5_Template_po_button_p_click_8_listener() { const restoredCtx = i0.ɵɵrestoreView(_r26); const option_r19 = restoredCtx.$implicit; const ctx_r29 = i0.ɵɵnextContext(); return i0.ɵɵresetView(ctx_r29.emitChangePosition(option_r19, "down")); });
    i0.ɵɵelementEnd()()();
} if (rf & 2) {
    const option_r19 = ctx.$implicit;
    const ctx_r2 = i0.ɵɵnextContext();
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("p-disabled", option_r19.disabled || ctx_r2.disabled || ctx_r2.isFixed(option_r19))("p-value", ctx_r2.checkedOptions[option_r19.value]);
    i0.ɵɵadvance(2);
    i0.ɵɵtextInterpolate1(" ", option_r19.label, " ");
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", !ctx_r2.hideActionFixedColumns);
    i0.ɵɵadvance();
    i0.ɵɵproperty("p-tooltip", ctx_r2.literals.up)("p-disabled", ctx_r2.verifyArrowDisabled(option_r19, "up") || ctx_r2.isFixed(option_r19));
    i0.ɵɵadvance();
    i0.ɵɵproperty("p-tooltip", ctx_r2.literals.down)("p-disabled", ctx_r2.verifyArrowDisabled(option_r19, "down") || ctx_r2.isFixed(option_r19));
} }
function PoTableListManagerComponent_ng_template_7_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵnamespaceSVG();
    i0.ɵɵelementStart(0, "svg", 19)(1, "g", 20);
    i0.ɵɵelement(2, "path", 21);
    i0.ɵɵelementEnd();
    i0.ɵɵelementStart(3, "defs")(4, "clipPath", 22);
    i0.ɵɵelement(5, "rect", 23);
    i0.ɵɵelementEnd()()();
} }
function PoTableListManagerComponent_ng_template_9_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵnamespaceSVG();
    i0.ɵɵelementStart(0, "svg", 19)(1, "g", 24);
    i0.ɵɵelement(2, "path", 25);
    i0.ɵɵelementEnd();
    i0.ɵɵelementStart(3, "defs")(4, "clipPath", 26);
    i0.ɵɵelement(5, "rect", 27);
    i0.ɵɵelementEnd()()();
} }
export const poTableListManagerLiterals = {
    en: {
        up: 'up',
        down: 'down',
        otherColumns: 'Other columns',
        fixedColumns: 'Fixed'
    },
    es: {
        up: 'arriba',
        down: 'abajo',
        otherColumns: 'Otras columnas',
        fixedColumns: 'Fijado'
    },
    pt: {
        up: 'acima',
        down: 'abaixo',
        otherColumns: 'Outras colunas',
        fixedColumns: 'Fixo'
    },
    ru: {
        up: 'вверх',
        down: 'вниз',
        otherColumns: 'Другие столбцы',
        fixedColumns: 'зафиксированный'
    }
};
export class PoTableListManagerComponent extends PoCheckboxGroupComponent {
    changePosition = new EventEmitter();
    changeColumnFixed = new EventEmitter();
    columnsManager;
    hideActionFixedColumns = false;
    literals;
    constructor(languageService, changeDetector) {
        super(changeDetector);
        const language = languageService.getShortLanguage();
        this.literals = {
            ...poTableListManagerLiterals[poLocaleDefault],
            ...poTableListManagerLiterals[language]
        };
    }
    emitChangePosition(option, direction) {
        if (!this.isFixed(option)) {
            const infoPosition = { option, direction };
            const hasDisabled = this.verifyArrowDisabled(option, direction);
            if (!hasDisabled) {
                this.changePosition.emit(infoPosition);
            }
        }
    }
    verifyArrowDisabled(option, direction) {
        const index = this.columnsManager.findIndex(el => el.property === option.value);
        const existsDetail = this.columnsManager.some(function (el) {
            return el.property === 'detail';
        });
        const valueSubtraction = existsDetail ? 2 : 1;
        if (index === 0 && direction === 'up') {
            return true;
        }
        if (index !== 0 && direction === 'up' && this.columnsManager[index - 1].fixed) {
            return true;
        }
        if (index === this.columnsManager.length - valueSubtraction && direction === 'down') {
            return true;
        }
        return false;
    }
    emitFixed(option) {
        if (option.visible) {
            const index = this.columnsManager.findIndex(el => el.property === option.value);
            if (this.columnsManager[index].fixed === null ||
                this.columnsManager[index].fixed === undefined ||
                this.columnsManager[index].fixed === false) {
                this.columnsManager[index].fixed = true;
                option.fixed = true;
            }
            else {
                this.columnsManager[index].fixed = false;
                option.fixed = false;
            }
            this.changeColumnFixed.emit(option);
        }
    }
    isFixed(option) {
        const index = this.columnsManager.findIndex(el => el.property === option.value);
        if (this.columnsManager[index].fixed === true) {
            return true;
        }
        return false;
    }
    existedFixedItem() {
        return this.columnsManager.some(option => option['fixed'] === true);
    }
    checksIfHasFiveFixed(option) {
        const isMoreThanFive = this.columnsManager.filter(item => item.fixed === true).length > 4;
        const isNotFixed = !this.isFixed(option);
        return isMoreThanFive && isNotFixed;
    }
    clickSwitch(option) {
        this.checkOption(option);
    }
    static ɵfac = function PoTableListManagerComponent_Factory(t) { return new (t || PoTableListManagerComponent)(i0.ɵɵdirectiveInject(i1.PoLanguageService), i0.ɵɵdirectiveInject(i0.ChangeDetectorRef)); };
    static ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: PoTableListManagerComponent, selectors: [["po-table-list-manager"]], inputs: { columnsManager: [i0.ɵɵInputFlags.None, "p-columns-manager", "columnsManager"], hideActionFixedColumns: [i0.ɵɵInputFlags.HasDecoratorInputTransform, "p-hide-action-fixed-columns", "hideActionFixedColumns", convertToBoolean] }, outputs: { changePosition: "p-change-position", changeColumnFixed: "p-change-fixed" }, features: [i0.ɵɵProvidersFeature([
                {
                    provide: NG_VALUE_ACCESSOR,
                    useExisting: forwardRef(() => PoTableListManagerComponent),
                    multi: true
                }
            ]), i0.ɵɵInputTransformsFeature, i0.ɵɵInheritDefinitionFeature], decls: 11, vars: 4, consts: [["class", "po-table-list-manager-section", 4, "ngIf"], [1, "po-table-list-manager-section"], [1, "po-table-list-manager-title-group-items"], [4, "ngFor", "ngForOf", "ngForTrackBy"], ["listItemsManagerDefault", ""], ["templateIconNotFixed", ""], ["templateIconFixed", ""], [4, "ngIf"], [1, "po-table-list-manager-container"], [4, "ngTemplateOutlet", "ngTemplateOutletContext"], [1, "po-table-list-manager-item"], [1, "po-table-list-manager-item-switch"], ["name", "switch", "p-label-off", " ", "p-label-on", " ", 3, "p-disabled", "p-value", "p-change", "keydown"], [1, "po-table-list-manager-item-text"], [1, "po-table-list-manager-item-buttons"], [3, "p-icon", "p-disabled", "p-click", 4, "ngIf"], ["p-tooltip-position", "left", "p-icon", "ICON_ARROW_UP", 3, "p-tooltip", "p-disabled", "p-click"], ["p-tooltip-position", "top", "p-icon", "ICON_ARROW_DOWN", 3, "p-tooltip", "p-disabled", "p-click"], [3, "p-icon", "p-disabled", "p-click"], ["xmlns", "http://www.w3.org/2000/svg", "width", "24", "height", "24", "viewBox", "0 0 24 24", "fill", "none"], ["clip-path", "url(#clip0_12562_15948)"], ["d", "M13.7143 5.71429V10C13.7143 10.96 14.0314 11.8514 14.5714 12.5714H9.42857C9.98571 11.8343 10.2857 10.9429 10.2857 10V5.71429H13.7143ZM16.2857 4H7.71429C7.24286 4 6.85714 4.38571 6.85714 4.85714C6.85714 5.32857 7.24286 5.71429 7.71429 5.71429H8.57143V10C8.57143 11.4229 7.42286 12.5714 6 12.5714V14.2857H11.1171V20.2857L11.9743 21.1429L12.8314 20.2857V14.2857H18V12.5714C16.5771 12.5714 15.4286 11.4229 15.4286 10V5.71429H16.2857C16.7571 5.71429 17.1429 5.32857 17.1429 4.85714C17.1429 4.38571 16.7571 4 16.2857 4Z", "fill", "black"], ["id", "clip0_12562_15948"], ["width", "24", "height", "24", "fill", "white"], ["clip-path", "url(#clip0_12562_17487)"], ["d", "M16.2857 4H7.71429C7.24286 4 6.85714 4.38571 6.85714 4.85714C6.85714 5.32857 7.24286 5.71429 7.71429 5.71429H8.57143V10C8.57143 11.4229 7.42286 12.5714 6 12.5714V14.2857H11.1171V20.2857L11.9743 21.1429L12.8314 20.2857V14.2857H18V12.5714C16.5771 12.5714 15.4286 11.4229 15.4286 10V5.71429H16.2857C16.7571 5.71429 17.1429 5.32857 17.1429 4.85714C17.1429 4.38571 16.7571 4 16.2857 4Z", "fill", "#B4B4C0"], ["id", "clip0_12562_17487"], ["width", "24", "height", "24", "fill", "black"]], template: function PoTableListManagerComponent_Template(rf, ctx) { if (rf & 1) {
            i0.ɵɵtemplate(0, PoTableListManagerComponent_section_0_Template, 4, 3, "section", 0);
            i0.ɵɵelementStart(1, "section", 1)(2, "p", 2);
            i0.ɵɵtext(3);
            i0.ɵɵelementEnd();
            i0.ɵɵtemplate(4, PoTableListManagerComponent_ng_container_4_Template, 2, 1, "ng-container", 3);
            i0.ɵɵelementEnd();
            i0.ɵɵtemplate(5, PoTableListManagerComponent_ng_template_5_Template, 9, 8, "ng-template", null, 4, i0.ɵɵtemplateRefExtractor)(7, PoTableListManagerComponent_ng_template_7_Template, 6, 0, "ng-template", null, 5, i0.ɵɵtemplateRefExtractor)(9, PoTableListManagerComponent_ng_template_9_Template, 6, 0, "ng-template", null, 6, i0.ɵɵtemplateRefExtractor);
        } if (rf & 2) {
            i0.ɵɵproperty("ngIf", ctx.existedFixedItem());
            i0.ɵɵadvance(3);
            i0.ɵɵtextInterpolate(ctx.literals.otherColumns);
            i0.ɵɵadvance();
            i0.ɵɵproperty("ngForOf", ctx.checkboxGroupOptionsView)("ngForTrackBy", ctx.trackByFn);
        } }, dependencies: [i2.NgForOf, i2.NgIf, i2.NgTemplateOutlet, i3.PoButtonComponent, i4.PoTooltipDirective, i5.PoSwitchComponent], encapsulation: 2 });
}
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(PoTableListManagerComponent, [{
        type: Component,
        args: [{ selector: 'po-table-list-manager', providers: [
                    {
                        provide: NG_VALUE_ACCESSOR,
                        useExisting: forwardRef(() => PoTableListManagerComponent),
                        multi: true
                    }
                ], template: "<section class=\"po-table-list-manager-section\" *ngIf=\"existedFixedItem()\">\n  <p class=\"po-table-list-manager-title-group-items\">{{ literals.fixedColumns }}</p>\n  <ng-container *ngFor=\"let option of checkboxGroupOptionsView; trackBy: trackByFn\">\n    <ng-container *ngIf=\"isFixed(option)\">\n      <li class=\"po-table-list-manager-container\" [class.po-checkbox-group-item-disabled]=\"option.disabled || disabled\">\n        <ng-container *ngTemplateOutlet=\"listItemsManagerDefault; context: { $implicit: option }\"></ng-container>\n      </li>\n    </ng-container>\n  </ng-container>\n</section>\n\n<section class=\"po-table-list-manager-section\">\n  <p class=\"po-table-list-manager-title-group-items\">{{ literals.otherColumns }}</p>\n  <ng-container *ngFor=\"let option of checkboxGroupOptionsView; trackBy: trackByFn\">\n    <ng-container *ngIf=\"!isFixed(option)\">\n      <li class=\"po-table-list-manager-container\" [class.po-checkbox-group-item-disabled]=\"option.disabled || disabled\">\n        <ng-container *ngIf=\"!isFixed(option)\">\n          <ng-container *ngTemplateOutlet=\"listItemsManagerDefault; context: { $implicit: option }\"></ng-container>\n        </ng-container>\n      </li>\n    </ng-container>\n  </ng-container>\n</section>\n\n<ng-template #listItemsManagerDefault let-option>\n  <div class=\"po-table-list-manager-item\">\n    <div class=\"po-table-list-manager-item-switch\">\n      <po-switch\n        name=\"switch\"\n        (p-change)=\"clickSwitch(option)\"\n        (keydown)=\"onKeyDown($event, option)\"\n        [p-disabled]=\"option.disabled || disabled || isFixed(option)\"\n        [p-value]=\"checkedOptions[option.value]\"\n        p-label-off=\" \"\n        p-label-on=\" \"\n      >\n      </po-switch>\n      <span class=\"po-table-list-manager-item-text\">\n        {{ option.label }}\n      </span>\n    </div>\n\n    <div class=\"po-table-list-manager-item-buttons\">\n      <po-button\n        *ngIf=\"!hideActionFixedColumns\"\n        [p-icon]=\"isFixed(option) ? templateIconFixed : templateIconNotFixed\"\n        (p-click)=\"emitFixed(option)\"\n        [p-disabled]=\"!option.visible || checksIfHasFiveFixed(option)\"\n      >\n      </po-button>\n\n      <po-button\n        [p-tooltip]=\"literals.up\"\n        p-tooltip-position=\"left\"\n        p-icon=\"ICON_ARROW_UP\"\n        [p-disabled]=\"verifyArrowDisabled(option, 'up') || isFixed(option)\"\n        (p-click)=\"emitChangePosition(option, 'up')\"\n      >\n      </po-button>\n\n      <po-button\n        [p-tooltip]=\"literals.down\"\n        p-tooltip-position=\"top\"\n        p-icon=\"ICON_ARROW_DOWN\"\n        [p-disabled]=\"verifyArrowDisabled(option, 'down') || isFixed(option)\"\n        (p-click)=\"emitChangePosition(option, 'down')\"\n      >\n      </po-button>\n    </div>\n  </div>\n</ng-template>\n\n<ng-template #templateIconNotFixed>\n  <svg xmlns=\"http://www.w3.org/2000/svg\" width=\"24\" height=\"24\" viewBox=\"0 0 24 24\" fill=\"none\">\n    <g clip-path=\"url(#clip0_12562_15948)\">\n      <path\n        d=\"M13.7143 5.71429V10C13.7143 10.96 14.0314 11.8514 14.5714 12.5714H9.42857C9.98571 11.8343 10.2857 10.9429 10.2857 10V5.71429H13.7143ZM16.2857 4H7.71429C7.24286 4 6.85714 4.38571 6.85714 4.85714C6.85714 5.32857 7.24286 5.71429 7.71429 5.71429H8.57143V10C8.57143 11.4229 7.42286 12.5714 6 12.5714V14.2857H11.1171V20.2857L11.9743 21.1429L12.8314 20.2857V14.2857H18V12.5714C16.5771 12.5714 15.4286 11.4229 15.4286 10V5.71429H16.2857C16.7571 5.71429 17.1429 5.32857 17.1429 4.85714C17.1429 4.38571 16.7571 4 16.2857 4Z\"\n        fill=\"black\"\n      />\n    </g>\n    <defs>\n      <clipPath id=\"clip0_12562_15948\">\n        <rect width=\"24\" height=\"24\" fill=\"white\" />\n      </clipPath>\n    </defs>\n  </svg>\n</ng-template>\n\n<ng-template #templateIconFixed>\n  <svg xmlns=\"http://www.w3.org/2000/svg\" width=\"24\" height=\"24\" viewBox=\"0 0 24 24\" fill=\"none\">\n    <g clip-path=\"url(#clip0_12562_17487)\">\n      <path\n        d=\"M16.2857 4H7.71429C7.24286 4 6.85714 4.38571 6.85714 4.85714C6.85714 5.32857 7.24286 5.71429 7.71429 5.71429H8.57143V10C8.57143 11.4229 7.42286 12.5714 6 12.5714V14.2857H11.1171V20.2857L11.9743 21.1429L12.8314 20.2857V14.2857H18V12.5714C16.5771 12.5714 15.4286 11.4229 15.4286 10V5.71429H16.2857C16.7571 5.71429 17.1429 5.32857 17.1429 4.85714C17.1429 4.38571 16.7571 4 16.2857 4Z\"\n        fill=\"#B4B4C0\"\n      />\n    </g>\n    <defs>\n      <clipPath id=\"clip0_12562_17487\">\n        <rect width=\"24\" height=\"24\" fill=\"black\" />\n      </clipPath>\n    </defs>\n  </svg>\n</ng-template>\n" }]
    }], () => [{ type: i1.PoLanguageService }, { type: i0.ChangeDetectorRef }], { changePosition: [{
            type: Output,
            args: ['p-change-position']
        }], changeColumnFixed: [{
            type: Output,
            args: ['p-change-fixed']
        }], columnsManager: [{
            type: Input,
            args: ['p-columns-manager']
        }], hideActionFixedColumns: [{
            type: Input,
            args: [{ alias: 'p-hide-action-fixed-columns', transform: convertToBoolean }]
        }] }); })();
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassDebugInfo(PoTableListManagerComponent, { className: "PoTableListManagerComponent", filePath: "lib/components/po-table/po-table-list-manager/po-table-list-manager.component.ts", lineNumber: 50 }); })();
