// @totvs/protheus-lib-core/esm2022/services/pro-authorization/pro-authorization.module.mjs
import { CommonModule } from '@angular/common';
import { HTTP_INTERCEPTORS } from '@angular/common/http';
import { NgModule } from '@angular/core';
import { ProAuthGuard } from './guards/pro-auth.guard';
import { ProAuthInteceptor } from './interceptors/pro-auth.inteceptor';
import { ProAuthService } from './services/pro-auth/pro-auth.service';
import { ProUserInfoService } from './services/pro-user-info/pro-user-info.service';
import * as i0 from "@angular/core";
export class ProAuthorizationModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProAuthorizationModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "17.3.9", ngImport: i0, type: ProAuthorizationModule, imports: [CommonModule] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProAuthorizationModule, providers: [
            ProAuthService,
            ProAuthGuard,
            ProUserInfoService,
            { provide: HTTP_INTERCEPTORS, useClass: ProAuthInteceptor, multi: true }
        ], imports: [CommonModule] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProAuthorizationModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule],
                    declarations: [],
                    providers: [
                        ProAuthService,
                        ProAuthGuard,
                        ProUserInfoService,
                        { provide: HTTP_INTERCEPTORS, useClass: ProAuthInteceptor, multi: true }
                    ]
                }]
        }] });
