// @totvs/protheus-lib-core/esm2022/services/pro-thread-info/services/pro-thread-info.service.mjs
import { map } from 'rxjs';
import { Injectable } from "@angular/core";
import * as i0 from "@angular/core";
import * as i1 from "../../pro-authorization";
import * as i2 from "./../../pro-authorization/services/pro-user-info/pro-user-info.service";
/***
 * @description Classe com métodos para recuperar informações da thread do Protheus, thread a qual abriu o app via FWCallApp
 */
export class ProThreadInfoService {
    constructor(proAuthService, proUserInfoService) {
        this.proAuthService = proAuthService;
        this.proUserInfoService = proUserInfoService;
    }
    /**
     * @description Retorna um objeto contendo informações do usuário logado no app e na thread do Protheus
     * @returns ProThreadInfo = {userId, userName}
     * userId = ID do usuário
     * userName = Nome (login) do usuário
     */
    get proThreadInfo() {
        const token = this.proAuthService.getTokenPayload();
        const proThreadInfo = {};
        proThreadInfo.userId = token.userid;
        proThreadInfo.userName = token.sub;
        return proThreadInfo;
    }
    /**
     * @description Recupera o ID do usuário logado no app e no Protheus
     * @returns ID do usuário
     */
    get userId() {
        return this.proAuthService.userId;
    }
    /**
     * @description Recupera o nome (login) do usuário logado no app e no Protheus
     * @returns Nome do usuário (login)
     */
    get userName() {
        return this.proThreadInfo.userName;
    }
    /**
     * @description Retorna informações do usuário logado (id, nome de usuário, nome e emails)
     * @returns ProUserInfo = {id, userName, displayName, emails}
     */
    getUserInfoThread() {
        return this.proUserInfoService.get(this.userId, true).pipe(map((user) => ({
            id: user.id,
            userName: user.userName,
            displayName: user.displayName,
            emails: user.emails
        })));
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProThreadInfoService, deps: [{ token: i1.ProAuthService }, { token: i2.ProUserInfoService }], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProThreadInfoService, providedIn: 'root' }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProThreadInfoService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root'
                }]
        }], ctorParameters: () => [{ type: i1.ProAuthService }, { type: i2.ProUserInfoService }] });
