// @po-ui/ng-components/esm2022/lib/components/po-field/po-multiselect/po-multiselect-option-template/po-multiselect-option-template.directive.mjs
import { Directive } from '@angular/core';
import * as i0 from "@angular/core";
/**
 * @usedBy PoMultiselectComponent
 *
 * @description
 *
 * Esta diretiva permite personalizar o conteúdo dos itens exibidos na lista de opções do componente.
 *
 * Para personalizar o conteúdo de cada item da lista deve-se utilizar a diretiva `p-multiselect-option-template` com `ng-template`
 * dentro da *tag* `po-multiselect`, o seu array de `p-options` repassado para com componente deve conter a propriedade `value` e `label`.
 *
 * Para obter a referência do item atual utilize `let-option`, com isso você terá acesso aos valores e poderá personalizar sua exibição.
 *
 * Exemplo de uso:
 *
 * ```
 * ...
 * <po-multiselect
 *   name="multiselect"
 *   [(ngModel)]="multiselect"
 *   [p-options]="options"
 *   [p-hide-select-all]="true">
 *     <ng-template p-multiselect-option-template let-option>
 *        <!-- template customizado -->
 *       <div class="po-font-text-large-bold" [innerHtml]="option.label"></div>
 *       <!-- template customizado -->
 *     </ng-template>
 * </po-multiselect>
 * ...
 * ```
 *
 */
export class PoMultiselectOptionTemplateDirective {
    templateRef;
    // Necessário manter templateRef para o funcionamento do row template.
    constructor(templateRef) {
        this.templateRef = templateRef;
    }
    static ɵfac = function PoMultiselectOptionTemplateDirective_Factory(t) { return new (t || PoMultiselectOptionTemplateDirective)(i0.ɵɵdirectiveInject(i0.TemplateRef)); };
    static ɵdir = /*@__PURE__*/ i0.ɵɵdefineDirective({ type: PoMultiselectOptionTemplateDirective, selectors: [["", "p-multiselect-option-template", ""]] });
}
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(PoMultiselectOptionTemplateDirective, [{
        type: Directive,
        args: [{
                selector: '[p-multiselect-option-template]'
            }]
    }], () => [{ type: i0.TemplateRef }], null); })();
