// @totvs/protheus-lib-core/esm2022/components/pro-fields/pro-system-module-lookup/pro-system-module-lookup.service.mjs
import { Injectable } from '@angular/core';
import { environment } from '../../../environments/environment';
import * as i0 from "@angular/core";
import * as i1 from "./../../../services/pro-system-info/pro-system-modules/services/pro-system-modules.service";
export class ProSystemModuleLookupService {
    constructor(proSystemModuleService) {
        this.proSystemModuleService = proSystemModuleService;
        this.useHTTP = environment.useHTTP;
        this.changeServiceChannel();
    }
    getFilteredItems(params) {
        return this.proSystemModuleService.getListOfSystemModules(params.filter, params.page, params.pageSize);
    }
    getObjectByValue(moduleId) {
        this.setSystemModule({ id: moduleId });
        return this.proSystemModuleService.getSystemModule(moduleId);
    }
    setSystemModule(systemModule) {
        this.proSystemModuleService.systemModule = systemModule;
    }
    setChannelAsHTTP(value) {
        this.useHTTP = value;
        this.changeServiceChannel();
    }
    isChannelHTTP() {
        return this.useHTTP;
    }
    changeServiceChannel() {
        this.proSystemModuleService.setChannelAsHTTP(this.useHTTP);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProSystemModuleLookupService, deps: [{ token: i1.ProSystemModuleService }], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProSystemModuleLookupService, providedIn: 'root' }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProSystemModuleLookupService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root'
                }]
        }], ctorParameters: () => [{ type: i1.ProSystemModuleService }] });
