// @po-ui/ng-components/esm2022/lib/components/po-field/po-upload/po-upload-drag-drop/po-upload-drag-drop-area/po-upload-drag-drop-area.component.mjs
import { ChangeDetectionStrategy, Component, EventEmitter, Input, Output, ViewChild } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
import * as i2 from "../../../../po-icon/po-icon.component";
const _c0 = ["selectFilesLink"];
function PoUploadDragDropAreaComponent_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainer(0);
} }
function PoUploadDragDropAreaComponent_ng_template_3_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelement(0, "po-icon", 5);
    i0.ɵɵelementStart(1, "div", 6);
    i0.ɵɵtext(2);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r1 = i0.ɵɵnextContext();
    i0.ɵɵadvance(2);
    i0.ɵɵtextInterpolate1(" ", ctx_r1.directoryCompatible ? ctx_r1.literals == null ? null : ctx_r1.literals.dropFoldersHere : ctx_r1.literals == null ? null : ctx_r1.literals.dropFilesHere, " ");
} }
const _c1 = a0 => ({ "po-clickable": a0 });
function PoUploadDragDropAreaComponent_ng_template_5_Template(rf, ctx) { if (rf & 1) {
    const _r7 = i0.ɵɵgetCurrentView();
    i0.ɵɵelement(0, "po-icon", 7);
    i0.ɵɵelementStart(1, "div", 8);
    i0.ɵɵtext(2);
    i0.ɵɵelementEnd();
    i0.ɵɵelementStart(3, "button", 9, 10);
    i0.ɵɵlistener("click", function PoUploadDragDropAreaComponent_ng_template_5_Template_button_click_3_listener() { i0.ɵɵrestoreView(_r7); const ctx_r6 = i0.ɵɵnextContext(); return i0.ɵɵresetView(ctx_r6.selectFiles.emit()); });
    i0.ɵɵtext(5);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r3 = i0.ɵɵnextContext();
    i0.ɵɵadvance(2);
    i0.ɵɵtextInterpolate1(" ", ctx_r3.directoryCompatible ? ctx_r3.literals == null ? null : ctx_r3.literals.dragFoldersHere : ctx_r3.literals == null ? null : ctx_r3.literals.dragFilesHere, " ");
    i0.ɵɵadvance();
    i0.ɵɵproperty("disabled", ctx_r3.disabled)("ngClass", i0.ɵɵpureFunction1(4, _c1, !ctx_r3.disabled));
    i0.ɵɵadvance(2);
    i0.ɵɵtextInterpolate1(" ", ctx_r3.directoryCompatible ? ctx_r3.literals == null ? null : ctx_r3.literals.selectFolderOnComputer : ctx_r3.literals == null ? null : ctx_r3.literals.selectFilesOnComputer, " ");
} }
const _c2 = a0 => ({ "po-upload-drag-drop-area-disabled": a0 });
export class PoUploadDragDropAreaComponent {
    elementRef;
    directoryCompatible;
    disabled;
    height;
    literals;
    overlay;
    selectFiles = new EventEmitter();
    selectFilesLinkElement;
    constructor(elementRef) {
        this.elementRef = elementRef;
    }
    focus() {
        this.selectFilesLinkElement.nativeElement.focus();
    }
    static ɵfac = function PoUploadDragDropAreaComponent_Factory(t) { return new (t || PoUploadDragDropAreaComponent)(i0.ɵɵdirectiveInject(i0.ElementRef)); };
    static ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: PoUploadDragDropAreaComponent, selectors: [["po-upload-drag-drop-area"]], viewQuery: function PoUploadDragDropAreaComponent_Query(rf, ctx) { if (rf & 1) {
            i0.ɵɵviewQuery(_c0, 5);
        } if (rf & 2) {
            let _t;
            i0.ɵɵqueryRefresh(_t = i0.ɵɵloadQuery()) && (ctx.selectFilesLinkElement = _t.first);
        } }, inputs: { directoryCompatible: [i0.ɵɵInputFlags.None, "p-directory-compatible", "directoryCompatible"], disabled: [i0.ɵɵInputFlags.None, "p-disabled", "disabled"], height: [i0.ɵɵInputFlags.None, "p-height", "height"], literals: [i0.ɵɵInputFlags.None, "p-literals", "literals"], overlay: [i0.ɵɵInputFlags.None, "p-overlay", "overlay"] }, outputs: { selectFiles: "p-select-files" }, decls: 7, vars: 8, consts: [[1, "po-upload-drag-drop-area", 3, "ngClass"], [1, "po-upload-drag-drop-area-container"], [4, "ngIf", "ngIfThen", "ngIfElse"], ["overlayTemplate", ""], ["defaultTemplate", ""], ["p-icon", "ICON_UPLOAD_CLOUD", 1, "po-upload-drag-drop-area-overlay-icon"], [1, "po-upload-drag-drop-area-overlay-label"], ["p-icon", "ICON_UPLOAD_CLOUD", 1, "po-upload-drag-drop-area-icon"], [1, "po-upload-drag-drop-area-label"], [1, "po-upload-drag-drop-area-select-files", 3, "disabled", "ngClass", "click"], ["selectFilesLink", ""]], template: function PoUploadDragDropAreaComponent_Template(rf, ctx) { if (rf & 1) {
            i0.ɵɵelementStart(0, "div", 0)(1, "div", 1);
            i0.ɵɵtemplate(2, PoUploadDragDropAreaComponent_ng_container_2_Template, 1, 0, "ng-container", 2);
            i0.ɵɵelementEnd()();
            i0.ɵɵtemplate(3, PoUploadDragDropAreaComponent_ng_template_3_Template, 3, 1, "ng-template", null, 3, i0.ɵɵtemplateRefExtractor)(5, PoUploadDragDropAreaComponent_ng_template_5_Template, 6, 6, "ng-template", null, 4, i0.ɵɵtemplateRefExtractor);
        } if (rf & 2) {
            const _r2 = i0.ɵɵreference(4);
            const _r4 = i0.ɵɵreference(6);
            i0.ɵɵstyleProp("height", ctx.height, "px");
            i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction1(6, _c2, ctx.disabled));
            i0.ɵɵadvance(2);
            i0.ɵɵproperty("ngIf", ctx.overlay && !ctx.disabled)("ngIfThen", _r2)("ngIfElse", _r4);
        } }, dependencies: [i1.NgClass, i1.NgIf, i2.PoIconComponent], encapsulation: 2, changeDetection: 0 });
}
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(PoUploadDragDropAreaComponent, [{
        type: Component,
        args: [{ selector: 'po-upload-drag-drop-area', changeDetection: ChangeDetectionStrategy.OnPush, template: "<div\n  class=\"po-upload-drag-drop-area\"\n  [ngClass]=\"{ 'po-upload-drag-drop-area-disabled': disabled }\"\n  [style.height.px]=\"height\"\n>\n  <div class=\"po-upload-drag-drop-area-container\">\n    <ng-container *ngIf=\"overlay && !disabled; then overlayTemplate; else defaultTemplate\"> </ng-container>\n  </div>\n</div>\n\n<ng-template #overlayTemplate>\n  <po-icon p-icon=\"ICON_UPLOAD_CLOUD\" class=\"po-upload-drag-drop-area-overlay-icon\"></po-icon>\n  <div class=\"po-upload-drag-drop-area-overlay-label\">\n    {{ directoryCompatible ? literals?.dropFoldersHere : literals?.dropFilesHere }}\n  </div>\n</ng-template>\n\n<ng-template #defaultTemplate>\n  <po-icon p-icon=\"ICON_UPLOAD_CLOUD\" class=\"po-upload-drag-drop-area-icon\"></po-icon>\n  <div class=\"po-upload-drag-drop-area-label\">\n    {{ directoryCompatible ? literals?.dragFoldersHere : literals?.dragFilesHere }}\n  </div>\n\n  <button\n    #selectFilesLink\n    class=\"po-upload-drag-drop-area-select-files\"\n    [disabled]=\"disabled\"\n    [ngClass]=\"{ 'po-clickable': !disabled }\"\n    (click)=\"selectFiles.emit()\"\n  >\n    {{ directoryCompatible ? literals?.selectFolderOnComputer : literals?.selectFilesOnComputer }}\n  </button>\n</ng-template>\n" }]
    }], () => [{ type: i0.ElementRef }], { directoryCompatible: [{
            type: Input,
            args: ['p-directory-compatible']
        }], disabled: [{
            type: Input,
            args: ['p-disabled']
        }], height: [{
            type: Input,
            args: ['p-height']
        }], literals: [{
            type: Input,
            args: ['p-literals']
        }], overlay: [{
            type: Input,
            args: ['p-overlay']
        }], selectFiles: [{
            type: Output,
            args: ['p-select-files']
        }], selectFilesLinkElement: [{
            type: ViewChild,
            args: ['selectFilesLink']
        }] }); })();
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassDebugInfo(PoUploadDragDropAreaComponent, { className: "PoUploadDragDropAreaComponent", filePath: "lib/components/po-field/po-upload/po-upload-drag-drop/po-upload-drag-drop-area/po-upload-drag-drop-area.component.ts", lineNumber: 10 }); })();
