// @po-ui/ng-components/esm2022/lib/components/po-field/po-switch/po-switch.module.mjs
import { FormsModule } from '@angular/forms';
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { PoSwitchComponent } from './po-switch.component';
import { PoIconModule } from '../../po-icon';
import { PoLabelModule } from '../../po-label';
import { PoFieldContainerModule } from './../po-field-container/po-field-container.module';
import * as i0 from "@angular/core";
export class PoSwitchModule {
    static ɵfac = function PoSwitchModule_Factory(t) { return new (t || PoSwitchModule)(); };
    static ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: PoSwitchModule });
    static ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [CommonModule, FormsModule, PoFieldContainerModule, PoLabelModule, PoIconModule] });
}
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(PoSwitchModule, [{
        type: NgModule,
        args: [{
                declarations: [PoSwitchComponent],
                exports: [PoSwitchComponent],
                imports: [CommonModule, FormsModule, PoFieldContainerModule, PoLabelModule, PoIconModule]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(PoSwitchModule, { declarations: [PoSwitchComponent], imports: [CommonModule, FormsModule, PoFieldContainerModule, PoLabelModule, PoIconModule], exports: [PoSwitchComponent] }); })();
