// @po-ui/ng-components/esm2022/lib/components/po-field/po-checkbox/po-checkbox.module.mjs
import { FormsModule } from '@angular/forms';
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { PoCheckboxComponent } from './po-checkbox.component';
import { PoLabelModule } from '../../po-label/po-label.module';
import * as i0 from "@angular/core";
export class PoCheckboxModule {
    static ɵfac = function PoCheckboxModule_Factory(t) { return new (t || PoCheckboxModule)(); };
    static ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: PoCheckboxModule });
    static ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [CommonModule, FormsModule, PoLabelModule] });
}
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(PoCheckboxModule, [{
        type: NgModule,
        args: [{
                declarations: [PoCheckboxComponent],
                exports: [PoCheckboxComponent],
                imports: [CommonModule, FormsModule, PoLabelModule]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(PoCheckboxModule, { declarations: [PoCheckboxComponent], imports: [CommonModule, FormsModule, PoLabelModule], exports: [PoCheckboxComponent] }); })();
