// @po-ui/ng-components/esm2022/lib/components/po-button/po-button.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { PoIconModule } from './../po-icon/index';
import { PoLoadingModule } from './../po-loading/index';
import { PoButtonComponent } from './po-button.component';
import * as i0 from "@angular/core";
/**
 * @description
 *
 * Módulo do componente po-button.
 */
export class PoButtonModule {
    static ɵfac = function PoButtonModule_Factory(t) { return new (t || PoButtonModule)(); };
    static ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: PoButtonModule });
    static ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [CommonModule, PoLoadingModule, PoIconModule] });
}
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(PoButtonModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, PoLoadingModule, PoIconModule],
                declarations: [PoButtonComponent],
                exports: [PoButtonComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(PoButtonModule, { declarations: [PoButtonComponent], imports: [CommonModule, PoLoadingModule, PoIconModule], exports: [PoButtonComponent] }); })();
