// @totvs/protheus-lib-core/esm2022/services/pro-theme/models/pro-theme.mjs
import { poThemeDefaultActions, poThemeDefaultDarkValues, poThemeDefaultFeedback, poThemeDefaultLightValues, poThemeDefaultNeutrals, PoThemeTypeEnum, } from '@po-ui/ng-components';
import { ProThemeSunsetLight, ProThemeTotvsDark, ProThemeTotvsLight, } from '../constants/pro-theme.constants';
/**
 * @description Enum que indica quais são os temas disponíveis já criados pelo
 * protheus lib service
 */
export var ProThemeEnum;
(function (ProThemeEnum) {
    ProThemeEnum["proThemeTotvsClassic"] = "proThemeTotvsClassic";
    ProThemeEnum["proThemeTotvsSunset"] = "proThemeTotvsSunset";
})(ProThemeEnum || (ProThemeEnum = {}));
/**
 * @description Classe de tema contendo estrutura necessária para o serviço PoThemeService
 */
export class ProTheme {
    constructor() {
        this._name = '';
        this._type = this.getTypeStructure();
        this._active = PoThemeTypeEnum.light;
    }
    get name() {
        return this._name;
    }
    set name(name) {
        this._name = name;
    }
    get active() {
        return this._active;
    }
    set active(active) {
        this._active = active;
    }
    get type() {
        return this._type;
    }
    set type(type) {
        this._type = type;
    }
    /**
     * @description Método para consultar quais são os temas disponíveis para configuração
     * automática, sendo uma lista de opções do ProThemeEnum
     */
    availableThemes() {
        const themes = Object.values(ProThemeEnum);
        return themes;
    }
    /**
     * @description Método para configurar o tema atual do aplicativo
     *
     */
    setTheme(theme) {
        let _theme;
        switch (theme) {
            case ProThemeEnum.proThemeTotvsClassic:
                _theme = new ProThemeTotvsClassic();
                break;
            case ProThemeEnum.proThemeTotvsSunset:
                _theme = new ProThemeTotvsSunset();
                break;
        }
        this.name = _theme.name;
        this.active = _theme.active;
        this.type = _theme.type;
    }
    /**
     * @description Método para devolver a estrutura solicitada do PoTheme para configuração
     * do aplicativo
     */
    getTypeStructure() {
        return {
            light: {
                color: {
                    brand: {
                        '01': {
                            lightest: '',
                            lighter: '',
                            light: '',
                            base: '',
                            dark: '',
                            darker: '',
                            darkest: '',
                        },
                        '02': {
                            base: '',
                        },
                        '03': {
                            base: '',
                        },
                    },
                    action: {
                        ...poThemeDefaultActions,
                    },
                    feedback: {
                        ...poThemeDefaultFeedback,
                        info: {
                            ...poThemeDefaultFeedback.info,
                            base: '',
                        },
                    },
                    neutral: {
                        ...poThemeDefaultNeutrals,
                    },
                },
            },
            dark: {
                color: {
                    brand: {
                        '01': {
                            lightest: '',
                            lighter: '',
                            light: '',
                            base: '',
                            dark: '',
                            darker: '',
                            darkest: '',
                        },
                        '02': {
                            base: '',
                        },
                        '03': {
                            base: '',
                        },
                    },
                    action: {
                        default: '',
                        hover: '',
                        pressed: '',
                        disabled: '',
                        focus: '',
                    },
                    feedback: {
                        ...poThemeDefaultFeedback,
                        info: {
                            ...poThemeDefaultFeedback.info,
                            base: '',
                        },
                    },
                    neutral: {
                        light: {
                            '00': '',
                            '05': '',
                            '10': '',
                            '20': '',
                            '30': '',
                        },
                        mid: {
                            '40': '',
                            '60': '',
                        },
                        dark: {
                            '70': '',
                            '80': '',
                            '90': '',
                            '95': '',
                        },
                    },
                },
            },
        };
    }
}
/**
 * @description Classe filha que extende a ProTheme, com os tokens de cores do
 * protheus clássico
 */
export class ProThemeTotvsClassic extends ProTheme {
    constructor() {
        super();
        this.name = 'proThemeTotvsClassic';
        this.type = this.getProThemeTotvsClassic();
    }
    getProThemeTotvsClassic() {
        return {
            light: this.getProThemeTotvsClassicTokensDark(),
            dark: this.getProThemeTotvsClassicTokensLight(),
        };
    }
    getProThemeTotvsClassicTokensLight() {
        return {
            color: {
                brand: {
                    '01': {
                        lightest: ProThemeTotvsDark['--color-brand-01-lightest'],
                        lighter: ProThemeTotvsDark['--color-brand-01-lighter'],
                        light: ProThemeTotvsDark['--color-brand-01-light'],
                        base: ProThemeTotvsDark['--color-brand-01-base'],
                        dark: ProThemeTotvsDark['--color-brand-01-dark'],
                        darker: ProThemeTotvsDark['--color-brand-01-darker'],
                        darkest: ProThemeTotvsDark['--color-brand-01-darkest'],
                    },
                    '02': {
                        base: ProThemeTotvsDark['--color-brand-02-base'],
                    },
                    '03': {
                        base: ProThemeTotvsDark['--color-brand-03-base'],
                    },
                },
                action: {
                    default: ProThemeTotvsDark['--color-action-default'],
                    hover: ProThemeTotvsDark['--color-action-hover'],
                    pressed: ProThemeTotvsDark['--color-action-pressed'],
                    disabled: ProThemeTotvsDark['--color-action-disabled'],
                    focus: ProThemeTotvsDark['--color-action-focus'],
                },
                neutral: {
                    light: {
                        '10': ProThemeTotvsDark['--color-neutral-light-10'],
                        '20': ProThemeTotvsDark['--color-neutral-light-20'],
                        '30': ProThemeTotvsDark['--color-neutral-light-30'],
                        '00': ProThemeTotvsDark['--color-neutral-light-00'],
                        '05': ProThemeTotvsDark['--color-neutral-light-05'],
                    },
                    mid: {
                        '40': ProThemeTotvsDark['--color-neutral-mid-40'],
                        '60': ProThemeTotvsDark['--color-neutral-mid-60'],
                    },
                    dark: {
                        '70': ProThemeTotvsDark['--color-neutral-dark-70'],
                        '80': ProThemeTotvsDark['--color-neutral-dark-80'],
                        '90': ProThemeTotvsDark['--color-neutral-dark-90'],
                        '95': ProThemeTotvsDark['--color-neutral-dark-95'],
                    },
                },
                feedback: {
                    info: {
                        base: ProThemeTotvsDark['--color-feedback-info-base'],
                        lightest: ProThemeTotvsDark['--color-feedback-info-lightest'],
                        lighter: ProThemeTotvsDark['--color-brand-info-lighter'],
                        light: ProThemeTotvsDark['--color-feedback-info-light'],
                        dark: ProThemeTotvsDark['--color-feedback-info-dark'],
                        darker: ProThemeTotvsDark['--color-feedback-info-darker'],
                        darkest: ProThemeTotvsDark['--color-feedback-info-darkerst'],
                    },
                    negative: {
                        base: ProThemeTotvsDark['--color-feedback-negative-base'],
                        lightest: ProThemeTotvsDark['--color-feedback-negative-lightest'],
                        lighter: ProThemeTotvsDark['--color-brand-negative-lighter'],
                        light: ProThemeTotvsDark['--color-feedback-negative-light'],
                        dark: ProThemeTotvsDark['--color-feedback-negative-dark'],
                        darker: ProThemeTotvsDark['--color-feedback-negative-darker'],
                        darkest: ProThemeTotvsDark['--color-feedback-negative-darkerst'],
                    },
                    positive: {
                        base: ProThemeTotvsDark['--color-feedback-positive-base'],
                        lightest: ProThemeTotvsDark['--color-feedback-positive-lightest'],
                        lighter: ProThemeTotvsDark['--color-brand-positive-lighter'],
                        light: ProThemeTotvsDark['--color-feedback-positive-light'],
                        dark: ProThemeTotvsDark['--color-feedback-positive-dark'],
                        darker: ProThemeTotvsDark['--color-feedback-positive-darker'],
                        darkest: ProThemeTotvsDark['--color-feedback-positive-darkerst'],
                    },
                    warning: {
                        base: ProThemeTotvsDark['--color-feedback-warning-base'],
                        lightest: ProThemeTotvsDark['--color-feedback-warning-lightest'],
                        lighter: ProThemeTotvsDark['--color-brand-warning-lighter'],
                        light: ProThemeTotvsDark['--color-feedback-warning-light'],
                        dark: ProThemeTotvsDark['--color-feedback-warning-dark'],
                        darker: ProThemeTotvsDark['--color-feedback-warning-darker'],
                        darkest: ProThemeTotvsDark['--color-feedback-warning-darkerst'],
                    },
                },
            },
            onRoot: {
                ...poThemeDefaultDarkValues.onRoot,
            },
            perComponent: {
                ...poThemeDefaultDarkValues.perComponent,
            },
        };
    }
    getProThemeTotvsClassicTokensDark() {
        return {
            color: {
                brand: {
                    '01': {
                        lightest: ProThemeTotvsLight['--color-brand-01-lightest'],
                        lighter: ProThemeTotvsLight['--color-brand-01-lighter'],
                        light: ProThemeTotvsLight['--color-brand-01-light'],
                        base: ProThemeTotvsLight['--color-brand-01-base'],
                        dark: ProThemeTotvsLight['--color-brand-01-dark'],
                        darker: ProThemeTotvsLight['--color-brand-01-darker'],
                        darkest: ProThemeTotvsLight['--color-brand-01-darkest'],
                    },
                    '02': {
                        base: ProThemeTotvsLight['--color-brand-02-base'],
                    },
                    '03': {
                        base: ProThemeTotvsLight['--color-brand-03-base'],
                    },
                },
                action: {
                    default: ProThemeTotvsLight['--color-action-default'],
                    hover: ProThemeTotvsLight['--color-action-hover'],
                    pressed: ProThemeTotvsLight['--color-action-pressed'],
                    disabled: ProThemeTotvsLight['--color-action-disabled'],
                    focus: ProThemeTotvsLight['--color-action-focus'],
                },
                neutral: {
                    light: {
                        '10': ProThemeTotvsLight['--color-neutral-light-10'],
                        '20': ProThemeTotvsLight['--color-neutral-light-20'],
                        '30': ProThemeTotvsLight['--color-neutral-light-30'],
                        '00': ProThemeTotvsLight['--color-neutral-light-00'],
                        '05': ProThemeTotvsLight['--color-neutral-light-05'],
                    },
                    mid: {
                        '40': ProThemeTotvsLight['--color-neutral-mid-40'],
                        '60': ProThemeTotvsLight['--color-neutral-mid-60'],
                    },
                    dark: {
                        '70': ProThemeTotvsLight['--color-neutral-dark-70'],
                        '80': ProThemeTotvsLight['--color-neutral-dark-80'],
                        '90': ProThemeTotvsLight['--color-neutral-dark-90'],
                        '95': ProThemeTotvsLight['--color-neutral-dark-95'],
                    },
                },
                feedback: {
                    info: {
                        base: ProThemeTotvsLight['--color-feedback-info-base'],
                        lightest: ProThemeTotvsLight['--color-feedback-info-lightest'],
                        lighter: ProThemeTotvsLight['--color-brand-info-lighter'],
                        light: ProThemeTotvsLight['--color-feedback-info-light'],
                        dark: ProThemeTotvsLight['--color-feedback-info-dark'],
                        darker: ProThemeTotvsLight['--color-feedback-info-darker'],
                        darkest: ProThemeTotvsLight['--color-feedback-info-darkerst'],
                    },
                    negative: {
                        base: ProThemeTotvsLight['--color-feedback-negative-base'],
                        lightest: ProThemeTotvsLight['--color-feedback-negative-lightest'],
                        lighter: ProThemeTotvsLight['--color-brand-negative-lighter'],
                        light: ProThemeTotvsLight['--color-feedback-negative-light'],
                        dark: ProThemeTotvsLight['--color-feedback-negative-dark'],
                        darker: ProThemeTotvsLight['--color-feedback-negative-darker'],
                        darkest: ProThemeTotvsLight['--color-feedback-negative-darkerst'],
                    },
                    positive: {
                        base: ProThemeTotvsLight['--color-feedback-positive-base'],
                        lightest: ProThemeTotvsLight['--color-feedback-positive-lightest'],
                        lighter: ProThemeTotvsLight['--color-brand-positive-lighter'],
                        light: ProThemeTotvsLight['--color-feedback-positive-light'],
                        dark: ProThemeTotvsLight['--color-feedback-positive-dark'],
                        darker: ProThemeTotvsLight['--color-feedback-positive-darker'],
                        darkest: ProThemeTotvsLight['--color-feedback-positive-darkerst'],
                    },
                    warning: {
                        base: ProThemeTotvsLight['--color-feedback-warning-base'],
                        lightest: ProThemeTotvsLight['--color-feedback-warning-lightest'],
                        lighter: ProThemeTotvsLight['--color-brand-warning-lighter'],
                        light: ProThemeTotvsLight['--color-feedback-warning-light'],
                        dark: ProThemeTotvsLight['--color-feedback-warning-dark'],
                        darker: ProThemeTotvsLight['--color-feedback-warning-darker'],
                        darkest: ProThemeTotvsLight['--color-feedback-warning-darkerst'],
                    },
                },
            },
            onRoot: {
                ...poThemeDefaultLightValues.onRoot,
            },
            perComponent: {
                ...poThemeDefaultLightValues.perComponent,
            },
        };
    }
}
/**
 * @description Classe filha que extende a ProTheme, com os tokens de cores do
 * protheus sunset
 */
export class ProThemeTotvsSunset extends ProTheme {
    constructor() {
        super();
        this.name = 'proThemeTotvsSunset';
        this.type = this.getProThemeTotvsSunset();
    }
    getProThemeTotvsSunset() {
        return {
            light: this.getProThemeTotvsSunsetTokensLight(),
        };
    }
    getProThemeTotvsSunsetTokensLight() {
        return {
            color: {
                brand: {
                    '01': {
                        lightest: ProThemeSunsetLight['--color-brand-01-lightest'],
                        lighter: ProThemeSunsetLight['--color-brand-01-lighter'],
                        light: ProThemeSunsetLight['--color-brand-01-light'],
                        base: ProThemeSunsetLight['--color-brand-01-base'],
                        dark: ProThemeSunsetLight['--color-brand-01-dark'],
                        darker: ProThemeSunsetLight['--color-brand-01-darker'],
                        darkest: ProThemeSunsetLight['--color-brand-01-darkest'],
                    },
                    '02': {
                        base: ProThemeSunsetLight['--color-brand-02-base'],
                    },
                    '03': {
                        base: ProThemeSunsetLight['--color-brand-03-base'],
                    },
                },
                action: {
                    default: ProThemeSunsetLight['--color-action-default'],
                    hover: ProThemeSunsetLight['--color-action-hover'],
                    pressed: ProThemeSunsetLight['--color-action-pressed'],
                    disabled: ProThemeSunsetLight['--color-action-disabled'],
                    focus: ProThemeSunsetLight['--color-action-focus'],
                },
                neutral: {
                    light: {
                        '10': ProThemeSunsetLight['--color-neutral-light-10'],
                        '20': ProThemeSunsetLight['--color-neutral-light-20'],
                        '30': ProThemeSunsetLight['--color-neutral-light-30'],
                        '00': ProThemeSunsetLight['--color-neutral-light-00'],
                        '05': ProThemeSunsetLight['--color-neutral-light-05'],
                    },
                    mid: {
                        '40': ProThemeSunsetLight['--color-neutral-mid-40'],
                        '60': ProThemeSunsetLight['--color-neutral-mid-60'],
                    },
                    dark: {
                        '70': ProThemeSunsetLight['--color-neutral-dark-70'],
                        '80': ProThemeSunsetLight['--color-neutral-dark-80'],
                        '90': ProThemeSunsetLight['--color-neutral-dark-90'],
                        '95': ProThemeSunsetLight['--color-neutral-dark-95'],
                    },
                },
                feedback: {
                    info: {
                        base: ProThemeSunsetLight['--color-feedback-info-base'],
                        lightest: ProThemeSunsetLight['--color-feedback-info-lightest'],
                        lighter: ProThemeSunsetLight['--color-brand-info-lighter'],
                        light: ProThemeSunsetLight['--color-feedback-info-light'],
                        dark: ProThemeSunsetLight['--color-feedback-info-dark'],
                        darker: ProThemeSunsetLight['--color-feedback-info-darker'],
                        darkest: ProThemeSunsetLight['--color-feedback-info-darkerst'],
                    },
                    negative: {
                        base: ProThemeSunsetLight['--color-feedback-negative-base'],
                        lightest: ProThemeSunsetLight['--color-feedback-negative-lightest'],
                        lighter: ProThemeSunsetLight['--color-brand-negative-lighter'],
                        light: ProThemeSunsetLight['--color-feedback-negative-light'],
                        dark: ProThemeSunsetLight['--color-feedback-negative-dark'],
                        darker: ProThemeSunsetLight['--color-feedback-negative-darker'],
                        darkest: ProThemeSunsetLight['--color-feedback-negative-darkerst'],
                    },
                    positive: {
                        base: ProThemeSunsetLight['--color-feedback-positive-base'],
                        lightest: ProThemeSunsetLight['--color-feedback-positive-lightest'],
                        lighter: ProThemeSunsetLight['--color-brand-positive-lighter'],
                        light: ProThemeSunsetLight['--color-feedback-positive-light'],
                        dark: ProThemeSunsetLight['--color-feedback-positive-dark'],
                        darker: ProThemeSunsetLight['--color-feedback-positive-darker'],
                        darkest: ProThemeSunsetLight['--color-feedback-positive-darkerst'],
                    },
                    warning: {
                        base: ProThemeSunsetLight['--color-feedback-warning-base'],
                        lightest: ProThemeSunsetLight['--color-feedback-warning-lightest'],
                        lighter: ProThemeSunsetLight['--color-brand-warning-lighter'],
                        light: ProThemeSunsetLight['--color-feedback-warning-light'],
                        dark: ProThemeSunsetLight['--color-feedback-warning-dark'],
                        darker: ProThemeSunsetLight['--color-feedback-warning-darker'],
                        darkest: ProThemeSunsetLight['--color-feedback-warning-darkerst'],
                    },
                },
            },
            onRoot: {
                ...poThemeDefaultLightValues.onRoot,
            },
            perComponent: {
                ...poThemeDefaultLightValues.perComponent,
            },
        };
    }
}
