// @po-ui/ng-components/esm2022/lib/components/po-table/po-table.module.mjs
import { CommonModule, DecimalPipe } from '@angular/common';
import { FormsModule } from '@angular/forms';
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { ScrollingModule } from '@angular/cdk/scrolling';
import { DragDropModule } from '@angular/cdk/drag-drop';
import { PoButtonModule } from './../po-button/po-button.module';
import { PoCheckboxGroupModule } from '../po-field/po-checkbox-group/po-checkbox-group.module';
import { PoContainerModule } from '../po-container/po-container.module';
import { PoLoadingModule } from '../po-loading/po-loading.module';
import { PoModalModule } from '../po-modal/po-modal.module';
import { PoPopoverModule } from '../po-popover/po-popover.module';
import { PoPopupModule } from './../po-popup/po-popup.module';
import { PoTagModule } from '../po-tag/po-tag.module';
import { PoTimeModule } from '../../pipes/po-time/index';
import { PoTooltipModule } from '../../directives/po-tooltip/index';
import { PoIconModule } from './../po-icon/po-icon.module';
import { PoCheckboxModule } from './../po-field/po-checkbox/po-checkbox.module';
import { PoRadioModule } from './../po-field/po-radio/po-radio.module';
import { PoLinkModule } from '../po-link/po-link.module';
import { PoPageSlideModule } from './../po-page/po-page-slide/po-page-slide.module';
import { PoTableColumnIconComponent } from './po-table-column-icon/po-table-column-icon.component';
import { PoTableColumnLabelComponent } from './po-table-column-label/po-table-column-label.component';
import { PoTableColumnLinkComponent } from './po-table-column-link/po-table-column-link.component';
import { PoTableColumnManagerComponent } from './po-table-column-manager/po-table-column-manager.component';
import { PoTableComponent } from './po-table.component';
import { PoTableDetailComponent } from './po-table-detail/po-table-detail.component';
import { PoTableIconComponent } from './po-table-icon/po-table-icon.component';
import { PoTableRowTemplateDirective } from './po-table-row-template/po-table-row-template.directive';
import { PoTableShowSubtitleComponent } from './po-table-show-subtitle/po-table-show-subtitle.component';
import { PoTableSubtitleCircleComponent } from './po-table-subtitle-circle/po-table-subtitle-circle.component';
import { PoTableSubtitleFooterComponent } from './po-table-subtitle-footer/po-table-subtitle-footer.component';
import { PoTableCellTemplateDirective } from './po-table-cell-template/po-table-cell-template.directive';
import { PoTableColumnTemplateDirective } from './po-table-column-template/po-table-column-template.directive';
import { PoTableListManagerComponent } from './po-table-list-manager/po-table-list-manager.component';
import { PoSwitchModule } from './../po-field/po-switch/po-switch.module';
import { PoTableColumnFrozenDirective } from './po-table-column-frozen/po-table-column-frozen.directive';
import { PoSearchModule } from '../po-search/po-search.module';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
import * as i2 from "@angular/cdk/scrolling";
import * as i3 from "@angular/cdk/drag-drop";
import * as i4 from "../po-button/po-button.component";
import * as i5 from "../po-container/po-container.component";
import * as i6 from "../po-loading/po-loading-overlay/po-loading-overlay.component";
import * as i7 from "../po-modal/po-modal.component";
import * as i8 from "../po-popup/po-popup.component";
import * as i9 from "../../directives/po-tooltip/po-tooltip.directive";
import * as i10 from "../po-icon/po-icon.component";
import * as i11 from "../po-field/po-checkbox/po-checkbox.component";
import * as i12 from "../po-field/po-radio/po-radio.component";
import * as i13 from "../po-search/po-search.component";
import * as i14 from "../../pipes/po-time/po-time.pipe";
/**
 * @description
 * Módulo do componente po-table
 */
export class PoTableModule {
    static ɵfac = function PoTableModule_Factory(t) { return new (t || PoTableModule)(); };
    static ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: PoTableModule });
    static ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ providers: [DecimalPipe], imports: [CommonModule,
            FormsModule,
            ScrollingModule,
            DragDropModule,
            RouterModule,
            PoButtonModule,
            PoCheckboxGroupModule,
            PoContainerModule,
            PoLoadingModule,
            PoModalModule,
            PoPopoverModule,
            PoPopupModule,
            PoTagModule,
            PoTimeModule,
            PoTooltipModule,
            PoIconModule,
            PoCheckboxModule,
            PoRadioModule,
            PoLinkModule,
            PoPageSlideModule,
            PoSwitchModule,
            PoSearchModule] });
}
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(PoTableModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    FormsModule,
                    ScrollingModule,
                    DragDropModule,
                    RouterModule,
                    PoButtonModule,
                    PoCheckboxGroupModule,
                    PoContainerModule,
                    PoLoadingModule,
                    PoModalModule,
                    PoPopoverModule,
                    PoPopupModule,
                    PoTagModule,
                    PoTimeModule,
                    PoTooltipModule,
                    PoIconModule,
                    PoCheckboxModule,
                    PoRadioModule,
                    PoLinkModule,
                    PoPageSlideModule,
                    PoSwitchModule,
                    PoSearchModule
                ],
                declarations: [
                    PoTableComponent,
                    PoTableColumnIconComponent,
                    PoTableColumnLabelComponent,
                    PoTableColumnLinkComponent,
                    PoTableColumnManagerComponent,
                    PoTableListManagerComponent,
                    PoTableDetailComponent,
                    PoTableIconComponent,
                    PoTableRowTemplateDirective,
                    PoTableShowSubtitleComponent,
                    PoTableSubtitleCircleComponent,
                    PoTableSubtitleFooterComponent,
                    PoTableCellTemplateDirective,
                    PoTableColumnTemplateDirective,
                    PoTableColumnFrozenDirective
                ],
                exports: [
                    PoTableComponent,
                    PoTableRowTemplateDirective,
                    PoTableCellTemplateDirective,
                    PoTableColumnTemplateDirective
                ],
                providers: [DecimalPipe]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(PoTableModule, { declarations: [PoTableComponent,
        PoTableColumnIconComponent,
        PoTableColumnLabelComponent,
        PoTableColumnLinkComponent,
        PoTableColumnManagerComponent,
        PoTableListManagerComponent,
        PoTableDetailComponent,
        PoTableIconComponent,
        PoTableRowTemplateDirective,
        PoTableShowSubtitleComponent,
        PoTableSubtitleCircleComponent,
        PoTableSubtitleFooterComponent,
        PoTableCellTemplateDirective,
        PoTableColumnTemplateDirective,
        PoTableColumnFrozenDirective], imports: [CommonModule,
        FormsModule,
        ScrollingModule,
        DragDropModule,
        RouterModule,
        PoButtonModule,
        PoCheckboxGroupModule,
        PoContainerModule,
        PoLoadingModule,
        PoModalModule,
        PoPopoverModule,
        PoPopupModule,
        PoTagModule,
        PoTimeModule,
        PoTooltipModule,
        PoIconModule,
        PoCheckboxModule,
        PoRadioModule,
        PoLinkModule,
        PoPageSlideModule,
        PoSwitchModule,
        PoSearchModule], exports: [PoTableComponent,
        PoTableRowTemplateDirective,
        PoTableCellTemplateDirective,
        PoTableColumnTemplateDirective] }); })();
i0.ɵɵsetComponentScope(PoTableComponent, [i1.NgClass, i1.NgForOf, i1.NgIf, i1.NgTemplateOutlet, i1.NgStyle, i1.NgSwitch, i1.NgSwitchCase, i1.NgSwitchDefault, i1.NgPlural, i1.NgPluralCase, i2.CdkFixedSizeVirtualScroll, i2.CdkVirtualForOf, i2.CdkVirtualScrollViewport, i3.CdkDropList, i3.CdkDrag, i3.CdkDragHandle, i4.PoButtonComponent, i5.PoContainerComponent, i6.PoLoadingOverlayComponent, i7.PoModalComponent, i8.PoPopupComponent, i9.PoTooltipDirective, i10.PoIconComponent, i11.PoCheckboxComponent, i12.PoRadioComponent, i13.PoSearchComponent, PoTableColumnIconComponent,
    PoTableColumnLabelComponent,
    PoTableColumnLinkComponent,
    PoTableColumnManagerComponent,
    PoTableDetailComponent,
    PoTableSubtitleCircleComponent,
    PoTableSubtitleFooterComponent,
    PoTableColumnFrozenDirective], [i1.LowerCasePipe, i1.TitleCasePipe, i1.CurrencyPipe, i1.DatePipe, i14.PoTimePipe]);
