// @po-ui/ng-components/esm2022/lib/components/po-tabs/po-tab/po-tab.component.mjs
import { Component } from '@angular/core';
import { PoTabBaseComponent } from './po-tab-base.component';
import * as i0 from "@angular/core";
import * as i1 from "../po-tabs.service";
const _c0 = ["*"];
/**
 * @docsExtends PoTabBaseComponent
 */
export class PoTabComponent extends PoTabBaseComponent {
    elementRef;
    tabsService;
    constructor(elementRef, tabsService) {
        super();
        this.elementRef = elementRef;
        this.tabsService = tabsService;
    }
    ngAfterContentInit() {
        this.setDisplayOnActive();
    }
    ngOnChanges(changes) {
        setTimeout(() => {
            this.tabsService.triggerOnChanges();
            if (changes?.active?.currentValue) {
                this.tabsService.triggerActiveOnChanges(this);
            }
        }, 100);
    }
    setDisplayOnActive() {
        this.elementRef.nativeElement.style.display = this.active ? '' : 'none';
    }
    static ɵfac = function PoTabComponent_Factory(t) { return new (t || PoTabComponent)(i0.ɵɵdirectiveInject(i0.ElementRef), i0.ɵɵdirectiveInject(i1.PoTabsService)); };
    static ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: PoTabComponent, selectors: [["po-tab"]], features: [i0.ɵɵInheritDefinitionFeature, i0.ɵɵNgOnChangesFeature], ngContentSelectors: _c0, decls: 1, vars: 0, template: function PoTabComponent_Template(rf, ctx) { if (rf & 1) {
            i0.ɵɵprojectionDef();
            i0.ɵɵprojection(0);
        } }, encapsulation: 2 });
}
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(PoTabComponent, [{
        type: Component,
        args: [{ selector: 'po-tab', template: "<ng-content></ng-content>\n" }]
    }], () => [{ type: i0.ElementRef }, { type: i1.PoTabsService }], null); })();
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassDebugInfo(PoTabComponent, { className: "PoTabComponent", filePath: "lib/components/po-tabs/po-tab/po-tab.component.ts", lineNumber: 13 }); })();
