// @totvs/protheus-lib-core/esm2022/components/pro-pages/pro-login/guards/pro-login-defaults/pro-login-defaults.resolver.mjs
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "../../../../../services/pro-authorization/services/pro-auth/pro-auth.service";
export class ProLoginDefaultsResolver {
    constructor(proAuthService) {
        this.proAuthService = proAuthService;
    }
    resolve(_route, _state) {
        return this.proAuthService.requestLoginDefaults();
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProLoginDefaultsResolver, deps: [{ token: i1.ProAuthService }], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProLoginDefaultsResolver, providedIn: 'root' }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProLoginDefaultsResolver, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root'
                }]
        }], ctorParameters: () => [{ type: i1.ProAuthService }] });
