// @po-ui/ng-components/esm2022/lib/components/po-field/po-datepicker/po-datepicker.module.mjs
import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';
import { NgModule } from '@angular/core';
import { PoCalendarModule } from '../../po-calendar/po-calendar.module';
import { PoFieldContainerModule } from '../po-field-container/po-field-container.module';
import { PoCleanModule } from '../po-clean/po-clean.module';
import { PoButtonModule } from '../../po-button';
import { PoDatepickerComponent } from './po-datepicker.component';
import * as i0 from "@angular/core";
/**
 * @description
 *
 * Módulo do componente `po-datepicker`.
 */
export class PoDatepickerModule {
    static ɵfac = function PoDatepickerModule_Factory(t) { return new (t || PoDatepickerModule)(); };
    static ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: PoDatepickerModule });
    static ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [CommonModule, FormsModule, PoFieldContainerModule, PoCleanModule, PoCalendarModule, PoButtonModule] });
}
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(PoDatepickerModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, FormsModule, PoFieldContainerModule, PoCleanModule, PoCalendarModule, PoButtonModule],
                exports: [PoDatepickerComponent],
                declarations: [PoDatepickerComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(PoDatepickerModule, { declarations: [PoDatepickerComponent], imports: [CommonModule, FormsModule, PoFieldContainerModule, PoCleanModule, PoCalendarModule, PoButtonModule], exports: [PoDatepickerComponent] }); })();
