// @totvs/protheus-lib-core/esm2022/services/pro-system-info/pro-system-database/interceptors/pro-system-database.interceptor.mjs
import { Injectable } from '@angular/core';
import { ProAppConfigService } from '../../../pro-app-config';
import * as i0 from "@angular/core";
const CACHE_KEY = 'ProDatabase';
/**
 * @description
 *
 * Interceptor que pega a requisição html e insere no header o database.
 */
export class ProSystemDatabaseInterceptor {
    constructor(injector) {
        this.injector = injector;
    }
    /**
     * @description Intercept http para tratamento do database do Protheus nas requisições
     * @param original_request Requisição http
     * @param next Handler da requisição http
     * @returns Observable da requisição http com o header do database caso o ERP seja Protheus
     */
    intercept(original_request, next) {
        let requestResult = original_request;
        const proAppConfigService = this.injector.get(ProAppConfigService);
        if (!(original_request.url.search(/assets/) >= 0)) {
            const isProtheus = proAppConfigService.isProtheusRender;
            if (isProtheus) {
                requestResult = this.appendDataBaseToRequest(original_request);
            }
        }
        return next.handle(requestResult);
    }
    /**
     * @description Adiciona o header da database na requisição caso exista
     * @param request Objeto HttpRequest da requisição
     * @returns HttpRequest da requisição com a database ou apenas um cópia
     */
    appendDataBaseToRequest(request) {
        const database = this.getDatabase();
        if (database) {
            return request.clone({
                headers: request.headers.set('x-erp-database', database)
            });
        }
        else {
            return request.clone();
        }
    }
    /**
     * @description Verifica o valor da database no SessionStorage
     * @returns Retorna o valor da database no SessionStorage, podendo não ter valor algum
     */
    getDatabase() {
        return sessionStorage[CACHE_KEY];
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProSystemDatabaseInterceptor, deps: [{ token: i0.Injector }], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProSystemDatabaseInterceptor, providedIn: 'root' }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProSystemDatabaseInterceptor, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root'
                }]
        }], ctorParameters: () => [{ type: i0.Injector }] });
