// @po-ui/ng-components/esm2022/lib/components/po-calendar/po-calendar-header/po-calendar-header.component.mjs
import { ChangeDetectionStrategy, Component, EventEmitter, Input, Output } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
import * as i2 from "../../po-icon/po-icon.component";
function PoCalendarHeaderComponent_po_icon_1_Template(rf, ctx) { if (rf & 1) {
    const _r3 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "po-icon", 4);
    i0.ɵɵlistener("click", function PoCalendarHeaderComponent_po_icon_1_Template_po_icon_click_0_listener() { i0.ɵɵrestoreView(_r3); const ctx_r2 = i0.ɵɵnextContext(); return i0.ɵɵresetView(ctx_r2.previous.emit()); });
    i0.ɵɵelementEnd();
} }
function PoCalendarHeaderComponent_po_icon_4_Template(rf, ctx) { if (rf & 1) {
    const _r5 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "po-icon", 5);
    i0.ɵɵlistener("click", function PoCalendarHeaderComponent_po_icon_4_Template_po_icon_click_0_listener() { i0.ɵɵrestoreView(_r5); const ctx_r4 = i0.ɵɵnextContext(); return i0.ɵɵresetView(ctx_r4.next.emit()); });
    i0.ɵɵelementEnd();
} }
const _c0 = ["*"];
export class PoCalendarHeaderComponent {
    hidePrevious = false;
    hideNext = false;
    previous = new EventEmitter();
    next = new EventEmitter();
    constructor() { }
    static ɵfac = function PoCalendarHeaderComponent_Factory(t) { return new (t || PoCalendarHeaderComponent)(); };
    static ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: PoCalendarHeaderComponent, selectors: [["po-calendar-header"]], inputs: { hidePrevious: [i0.ɵɵInputFlags.None, "p-hide-previous", "hidePrevious"], hideNext: [i0.ɵɵInputFlags.None, "p-hide-next", "hideNext"] }, outputs: { previous: "p-previous", next: "p-next" }, ngContentSelectors: _c0, decls: 5, vars: 2, consts: [[1, "po-calendar-header"], ["p-icon", "ICON_ARROW_LEFT", "class", "po-calendar-header-left", 3, "click", 4, "ngIf"], ["attr-calendar", "", 1, "po-calendar-header-title"], ["p-icon", "ICON_ARROW_RIGHT", "class", "po-calendar-header-right", 3, "click", 4, "ngIf"], ["p-icon", "ICON_ARROW_LEFT", 1, "po-calendar-header-left", 3, "click"], ["p-icon", "ICON_ARROW_RIGHT", 1, "po-calendar-header-right", 3, "click"]], template: function PoCalendarHeaderComponent_Template(rf, ctx) { if (rf & 1) {
            i0.ɵɵprojectionDef();
            i0.ɵɵelementStart(0, "div", 0);
            i0.ɵɵtemplate(1, PoCalendarHeaderComponent_po_icon_1_Template, 1, 0, "po-icon", 1);
            i0.ɵɵelementStart(2, "div", 2);
            i0.ɵɵprojection(3);
            i0.ɵɵelementEnd();
            i0.ɵɵtemplate(4, PoCalendarHeaderComponent_po_icon_4_Template, 1, 0, "po-icon", 3);
            i0.ɵɵelementEnd();
        } if (rf & 2) {
            i0.ɵɵadvance();
            i0.ɵɵproperty("ngIf", (ctx.previous == null ? null : ctx.previous.observers) && !ctx.hidePrevious);
            i0.ɵɵadvance(3);
            i0.ɵɵproperty("ngIf", (ctx.next == null ? null : ctx.next.observers.length) && !ctx.hideNext);
        } }, dependencies: [i1.NgIf, i2.PoIconComponent], encapsulation: 2, changeDetection: 0 });
}
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(PoCalendarHeaderComponent, [{
        type: Component,
        args: [{ selector: 'po-calendar-header', changeDetection: ChangeDetectionStrategy.OnPush, template: "<div class=\"po-calendar-header\">\n  <po-icon\n    p-icon=\"ICON_ARROW_LEFT\"\n    *ngIf=\"previous?.observers && !hidePrevious\"\n    class=\"po-calendar-header-left\"\n    (click)=\"previous.emit()\"\n  >\n  </po-icon>\n  <div class=\"po-calendar-header-title\" attr-calendar>\n    <ng-content></ng-content>\n  </div>\n\n  <po-icon\n    p-icon=\"ICON_ARROW_RIGHT\"\n    *ngIf=\"next?.observers.length && !hideNext\"\n    class=\"po-calendar-header-right\"\n    (click)=\"next.emit()\"\n  >\n  </po-icon>\n</div>\n" }]
    }], () => [], { hidePrevious: [{
            type: Input,
            args: ['p-hide-previous']
        }], hideNext: [{
            type: Input,
            args: ['p-hide-next']
        }], previous: [{
            type: Output,
            args: ['p-previous']
        }], next: [{
            type: Output,
            args: ['p-next']
        }] }); })();
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassDebugInfo(PoCalendarHeaderComponent, { className: "PoCalendarHeaderComponent", filePath: "lib/components/po-calendar/po-calendar-header/po-calendar-header.component.ts", lineNumber: 8 }); })();
