// @po-ui/ng-components/esm2022/lib/components/po-dynamic/po-dynamic-container/po-dynamic-container.component.mjs
import { Component, Input } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
import * as i2 from "../../po-container/po-container.component";
function PoDynamicContainerComponent_ng_template_1_div_1_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "div", 5);
    i0.ɵɵtext(1);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const fieldsInContainer_r1 = i0.ɵɵnextContext().$implicit;
    i0.ɵɵadvance();
    i0.ɵɵtextInterpolate1(" ", fieldsInContainer_r1[0].container, " ");
} }
function PoDynamicContainerComponent_ng_template_1_ng_container_4_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainer(0);
} }
const _c0 = a0 => ({ $implicit: a0 });
function PoDynamicContainerComponent_ng_template_1_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "div");
    i0.ɵɵtemplate(1, PoDynamicContainerComponent_ng_template_1_div_1_Template, 2, 1, "div", 2);
    i0.ɵɵelementStart(2, "po-container")(3, "div", 3);
    i0.ɵɵtemplate(4, PoDynamicContainerComponent_ng_template_1_ng_container_4_Template, 1, 0, "ng-container", 4);
    i0.ɵɵelementEnd()()();
} if (rf & 2) {
    const fieldsInContainer_r1 = ctx.$implicit;
    const ctx_r0 = i0.ɵɵnextContext();
    i0.ɵɵadvance();
    i0.ɵɵproperty("ngIf", fieldsInContainer_r1[0].container);
    i0.ɵɵadvance(3);
    i0.ɵɵproperty("ngTemplateOutlet", ctx_r0.content)("ngTemplateOutletContext", i0.ɵɵpureFunction1(3, _c0, fieldsInContainer_r1));
} }
export class PoDynamicContainerComponent {
    content;
    fields;
    static ɵfac = function PoDynamicContainerComponent_Factory(t) { return new (t || PoDynamicContainerComponent)(); };
    static ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: PoDynamicContainerComponent, selectors: [["po-dynamic-container"]], inputs: { content: [i0.ɵɵInputFlags.None, "p-content", "content"], fields: [i0.ɵɵInputFlags.None, "p-fields", "fields"] }, decls: 2, vars: 1, consts: [[1, "po-dynamic-container"], ["ngFor", "", 3, "ngForOf"], ["class", "po-dynamic-container-title", 4, "ngIf"], [1, "po-row"], [4, "ngTemplateOutlet", "ngTemplateOutletContext"], [1, "po-dynamic-container-title"]], template: function PoDynamicContainerComponent_Template(rf, ctx) { if (rf & 1) {
            i0.ɵɵelementStart(0, "div", 0);
            i0.ɵɵtemplate(1, PoDynamicContainerComponent_ng_template_1_Template, 5, 5, "ng-template", 1);
            i0.ɵɵelementEnd();
        } if (rf & 2) {
            i0.ɵɵadvance();
            i0.ɵɵproperty("ngForOf", ctx.fields);
        } }, dependencies: [i1.NgForOf, i1.NgIf, i1.NgTemplateOutlet, i2.PoContainerComponent], encapsulation: 2 });
}
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(PoDynamicContainerComponent, [{
        type: Component,
        args: [{ selector: 'po-dynamic-container', template: "<div class=\"po-dynamic-container\">\n  <ng-template ngFor let-fieldsInContainer [ngForOf]=\"fields\">\n    <div>\n      <div class=\"po-dynamic-container-title\" *ngIf=\"fieldsInContainer[0].container\">\n        {{ fieldsInContainer[0].container }}\n      </div>\n      <po-container>\n        <div class=\"po-row\">\n          <ng-container *ngTemplateOutlet=\"content; context: { $implicit: fieldsInContainer }\"> </ng-container>\n        </div>\n      </po-container>\n    </div>\n  </ng-template>\n</div>\n" }]
    }], null, { content: [{
            type: Input,
            args: ['p-content']
        }], fields: [{
            type: Input,
            args: ['p-fields']
        }] }); })();
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassDebugInfo(PoDynamicContainerComponent, { className: "PoDynamicContainerComponent", filePath: "lib/components/po-dynamic/po-dynamic-container/po-dynamic-container.component.ts", lineNumber: 7 }); })();
