// @totvs/protheus-lib-core/esm2022/components/pro-pages/pro-page-background/pro-page-background.component.mjs
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { environment } from '../../../environments/environment';
import { convertToBoolean, isTypeof } from '../../../util';
import * as i0 from "@angular/core";
import * as i1 from "@po-ui/ng-components";
import * as i2 from "../../../services/pro-language/services/pro-language.service";
import * as i3 from "@angular/common";
import * as i4 from "@angular/forms";
/**
 * @docsPrivate
 *
 * @description
 *
 * Componente para definição de cor de fundo e dos logotipos primário e secundário para os templates
 * de `po-page-login` e demais templates de login.
 */
export class ProPageBackgroundComponent {
    /** Caminho para a logomarca localizada na parte superior. */
    set logo(value) {
        this._logo = isTypeof(value, 'string') && value.trim() ? value : undefined;
    }
    get logo() {
        return this._logo;
    }
    /**
     * @optional
     *
     * @description
     *
     * Caminho para a logomarca localizada no rodapé.
     */
    set secondaryLogo(value) {
        this._secondaryLogo =
            isTypeof(value, 'string') && value.trim() ? value : undefined;
    }
    get secondaryLogo() {
        return this._secondaryLogo;
    }
    /** Define se o seletor de idiomas deve ser exibido. */
    set showSelectLanguage(showSelectLanguage) {
        this._showSelectLanguage = convertToBoolean(showSelectLanguage);
    }
    get showSelectLanguage() {
        return this._showSelectLanguage;
    }
    /**
     * @description Construtor do componente
     * @param poLanguageService Serviço de idioma do PO UI
     * @param proLanguageService Serviço de idiomas do Protheus
     */
    constructor(poLanguageService, proLanguageService) {
        this.poLanguageService = poLanguageService;
        this.proLanguageService = proLanguageService;
        this.selectLanguageOptions = [];
        /**
         * Evento disparado ao selecionar alguma opção no seletor de idiomas.
         * Para este evento será passado como parâmetro o valor de idioma selecionado.
         */
        // tslint:disable-next-line: no-output-rename
        this.selectedLanguage = new EventEmitter();
        this.updateLangs();
    }
    /**
     * @description Atualiza os idiomas disponíveis
     */
    updateLangs() {
        this.proLanguageService.setChannelAsHTTP(environment.useHTTP);
        this.proLanguageService.getListOfLanguages().subscribe({
            next: (langs) => {
                this.selectLanguageOptions = langs.map((language) => {
                    return {
                        label: language.description,
                        value: language.language,
                    };
                });
            },
            error: (error) => {
                console.error(error);
            },
        });
    }
    ngOnInit() {
        this.selectedLanguageOption = this.poLanguageService.getShortLanguage();
    }
    onChangeLanguage() {
        this.selectedLanguage.emit(this.selectedLanguageOption);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProPageBackgroundComponent, deps: [{ token: i1.PoLanguageService }, { token: i2.ProLanguageService }], target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "17.3.9", type: ProPageBackgroundComponent, selector: "pro-page-background", inputs: { background: ["p-background", "background"], hideLogo: ["p-hide-logo", "hideLogo"], highlightInfo: ["p-highlight-info", "highlightInfo"], logo: ["p-logo", "logo"], secondaryLogo: ["p-secondary-logo", "secondaryLogo"], showSelectLanguage: ["p-show-select-language", "showSelectLanguage"] }, outputs: { selectedLanguage: "p-selected-language" }, ngImport: i0, template: "<div class=\"po-page-login-container\">\r\n  <div\r\n    class=\"po-page-login-panel\">\r\n    <img\r\n      *ngIf=\"logo\"\r\n      class=\"po-page-background-main-logo-image\"\r\n      [class.po-page-background-hide-logo-image]=\"hideLogo\"\r\n      alt=\"main-logo\"\r\n      [src]=\"logo\"\r\n    />\r\n\r\n    <div class=\"po-page-login-body\">\r\n      <div class=\"po-page-login-panel-content pro-bg-component\">\r\n        <ng-content></ng-content>\r\n      </div>\r\n    </div>\r\n    <div class=\"po-page-background-footer po-sm-12\">\r\n      <po-divider class=\"po-page-background-footer-mobile-only\"></po-divider>\r\n\r\n      <div class=\"po-page-background-footer-content\">\r\n        <div\r\n          *ngIf=\"showSelectLanguage\"\r\n          class=\"po-page-background-footer-select\"\r\n        >\r\n          <po-select\r\n            *ngIf=\"selectLanguageOptions\"\r\n            name=\"selectedLanguageOption\"\r\n            [(ngModel)]=\"selectedLanguageOption\"\r\n            [p-options]=\"selectLanguageOptions\"\r\n            (p-change)=\"onChangeLanguage()\"\r\n          >\r\n          </po-select>\r\n        </div>\r\n\r\n        <div\r\n          class=\"po-page-background-secondary-logo\"\r\n          [ngClass]=\"\r\n            showSelectLanguage\r\n              ? 'po-page-background-secondary-logo-right'\r\n              : 'po-page-background-secondary-logo-centered'\r\n          \"\r\n        >\r\n          <img\r\n            *ngIf=\"secondaryLogo\"\r\n            class=\"po-page-background-secondary-logo-image\"\r\n            alt=\"secondary-logo\"\r\n            [src]=\"secondaryLogo\"\r\n          />\r\n        </div>\r\n      </div>\r\n    </div>\r\n  </div>\r\n\r\n  <div\r\n    *ngIf=\"background\"\r\n    [ngClass]=\"\r\n      background\r\n        ? 'po-page-login-highlight-image'\r\n        : 'po-page-login-highlight-image-off'\r\n    \"\r\n    [style.background-image]=\"'url(' + background + ')'\"\r\n  >\r\n    <div class=\"po-page-login-highlight-text\">\r\n      <div class=\"po-font-display\">{{ highlightInfo }}</div>\r\n    </div>\r\n  </div>\r\n</div>\r\n", styles: [".pro-bg-component{margin-bottom:2%;padding:8%}\n"], dependencies: [{ kind: "directive", type: i3.NgClass, selector: "[ngClass]", inputs: ["class", "ngClass"] }, { kind: "directive", type: i3.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }, { kind: "directive", type: i4.NgControlStatus, selector: "[formControlName],[ngModel],[formControl]" }, { kind: "directive", type: i4.NgModel, selector: "[ngModel]:not([formControlName]):not([formControl])", inputs: ["name", "disabled", "ngModel", "ngModelOptions"], outputs: ["ngModelChange"], exportAs: ["ngModel"] }, { kind: "component", type: i1.PoDividerComponent, selector: "po-divider" }, { kind: "component", type: i1.PoSelectComponent, selector: "po-select", inputs: ["p-readonly", "p-placeholder", "p-options", "p-field-label", "p-field-value"], outputs: ["ngModelChange"] }] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProPageBackgroundComponent, decorators: [{
            type: Component,
            args: [{ selector: 'pro-page-background', template: "<div class=\"po-page-login-container\">\r\n  <div\r\n    class=\"po-page-login-panel\">\r\n    <img\r\n      *ngIf=\"logo\"\r\n      class=\"po-page-background-main-logo-image\"\r\n      [class.po-page-background-hide-logo-image]=\"hideLogo\"\r\n      alt=\"main-logo\"\r\n      [src]=\"logo\"\r\n    />\r\n\r\n    <div class=\"po-page-login-body\">\r\n      <div class=\"po-page-login-panel-content pro-bg-component\">\r\n        <ng-content></ng-content>\r\n      </div>\r\n    </div>\r\n    <div class=\"po-page-background-footer po-sm-12\">\r\n      <po-divider class=\"po-page-background-footer-mobile-only\"></po-divider>\r\n\r\n      <div class=\"po-page-background-footer-content\">\r\n        <div\r\n          *ngIf=\"showSelectLanguage\"\r\n          class=\"po-page-background-footer-select\"\r\n        >\r\n          <po-select\r\n            *ngIf=\"selectLanguageOptions\"\r\n            name=\"selectedLanguageOption\"\r\n            [(ngModel)]=\"selectedLanguageOption\"\r\n            [p-options]=\"selectLanguageOptions\"\r\n            (p-change)=\"onChangeLanguage()\"\r\n          >\r\n          </po-select>\r\n        </div>\r\n\r\n        <div\r\n          class=\"po-page-background-secondary-logo\"\r\n          [ngClass]=\"\r\n            showSelectLanguage\r\n              ? 'po-page-background-secondary-logo-right'\r\n              : 'po-page-background-secondary-logo-centered'\r\n          \"\r\n        >\r\n          <img\r\n            *ngIf=\"secondaryLogo\"\r\n            class=\"po-page-background-secondary-logo-image\"\r\n            alt=\"secondary-logo\"\r\n            [src]=\"secondaryLogo\"\r\n          />\r\n        </div>\r\n      </div>\r\n    </div>\r\n  </div>\r\n\r\n  <div\r\n    *ngIf=\"background\"\r\n    [ngClass]=\"\r\n      background\r\n        ? 'po-page-login-highlight-image'\r\n        : 'po-page-login-highlight-image-off'\r\n    \"\r\n    [style.background-image]=\"'url(' + background + ')'\"\r\n  >\r\n    <div class=\"po-page-login-highlight-text\">\r\n      <div class=\"po-font-display\">{{ highlightInfo }}</div>\r\n    </div>\r\n  </div>\r\n</div>\r\n", styles: [".pro-bg-component{margin-bottom:2%;padding:8%}\n"] }]
        }], ctorParameters: () => [{ type: i1.PoLanguageService }, { type: i2.ProLanguageService }], propDecorators: { background: [{
                type: Input,
                args: ['p-background']
            }], hideLogo: [{
                type: Input,
                args: ['p-hide-logo']
            }], highlightInfo: [{
                type: Input,
                args: ['p-highlight-info']
            }], logo: [{
                type: Input,
                args: ['p-logo']
            }], secondaryLogo: [{
                type: Input,
                args: ['p-secondary-logo']
            }], showSelectLanguage: [{
                type: Input,
                args: ['p-show-select-language']
            }], selectedLanguage: [{
                type: Output,
                args: ['p-selected-language']
            }] } });
