// @totvs/protheus-lib-core/esm2022/components/pro-pages/pro-page-background/pro-page-background.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { RouterModule } from '@angular/router';
import { PoDividerModule, PoFieldModule } from '@po-ui/ng-components';
import { ProPageBackgroundComponent } from './pro-page-background.component';
import * as i0 from "@angular/core";
/**
 * @description
 *
 * Módulo do template do pro-page-background.
 */
export class ProPageBackgroundModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProPageBackgroundModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "17.3.9", ngImport: i0, type: ProPageBackgroundModule, declarations: [ProPageBackgroundComponent], imports: [CommonModule,
            FormsModule,
            RouterModule,
            PoDividerModule,
            PoFieldModule], exports: [ProPageBackgroundComponent] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProPageBackgroundModule, imports: [CommonModule,
            FormsModule,
            RouterModule,
            PoDividerModule,
            PoFieldModule] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProPageBackgroundModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        FormsModule,
                        RouterModule,
                        PoDividerModule,
                        PoFieldModule
                    ],
                    declarations: [ProPageBackgroundComponent],
                    exports: [ProPageBackgroundComponent]
                }]
        }] });
