// @totvs/protheus-lib-core/esm2022/services/pro-brand/services/pro-brand.service.mjs
import { Injectable } from '@angular/core';
import { Observable, of } from 'rxjs';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common/http";
import * as i2 from "../../pro-js-to-advpl";
import * as i3 from "../../pro-app-config/services/pro-app-config.service";
export class ProBrandService {
    constructor(http, advplService, proAppConfigService) {
        this.http = http;
        this.advplService = advplService;
        this.proAppConfigService = proAppConfigService;
        this.url = '/api/brand/v1/erp-brand';
        this.useHTTP = true;
        this.EVENT_GET_BRANDCONTEXT = 'getBrandContext';
        this.EVENT_SET_BRANDCONTEXT = 'setBrandContext';
    }
    getERPBrand() {
        if (this.proAppConfigService.isProtheusRender) {
            if (this.useHTTP) {
                return this.http.get(this.url, { responseType: 'text' });
            }
            return this.getAdvplBrandContext();
        }
        else {
            return of('TOTVS');
        }
    }
    /**
     * @description Cria um observable para a comunicação com o ADVPL
     * @returns Observable de string sobre a marca do ambiente
     */
    getAdvplBrandContext() {
        if (!this.advplService.protheusConnected()) {
            return this.advplNotPrepared();
        }
        return this.advplService.buildObservable(({ protheusResponse, subscriber }) => {
            if (subscriber) {
                if (protheusResponse.length === 0) {
                    subscriber.error({
                        status: 400,
                        description: 'Brand context not found'
                    });
                }
                else {
                    subscriber.next(protheusResponse);
                }
                subscriber.complete();
            }
        }, {
            sendInfo: {
                type: this.EVENT_GET_BRANDCONTEXT
            },
            autoDestruct: true,
            receiveId: this.EVENT_SET_BRANDCONTEXT
        });
    }
    advplNotPrepared() {
        return new Observable(subscriber => {
            subscriber.error({
                status: 400,
                description: 'advplService not prepared in ProBrandService'
            });
            subscriber.complete();
        });
    }
    isChannelHTTP() {
        return this.useHTTP;
    }
    setChannelAsHTTP(value) {
        this.useHTTP = value;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProBrandService, deps: [{ token: i1.HttpClient }, { token: i2.ProJsToAdvplService }, { token: i3.ProAppConfigService }], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProBrandService, providedIn: 'root' }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProBrandService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root'
                }]
        }], ctorParameters: () => [{ type: i1.HttpClient }, { type: i2.ProJsToAdvplService }, { type: i3.ProAppConfigService }] });
