// @po-ui/ng-components/esm2022/lib/components/po-disclaimer-group/po-disclaimer-group.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { PoDisclaimerModule } from '../po-disclaimer/po-disclaimer.module';
import { PoTagModule } from '../po-tag/po-tag.module';
import { PoDisclaimerGroupComponent } from './po-disclaimer-group.component';
import { PoDisclaimerRemoveComponent } from './po-disclaimer-remove/po-disclaimer-remove.component';
import * as i0 from "@angular/core";
/**
 * @description
 *
 * Módulo do componente po-disclaimer-group.
 */
export class PoDisclaimerGroupModule {
    static ɵfac = function PoDisclaimerGroupModule_Factory(t) { return new (t || PoDisclaimerGroupModule)(); };
    static ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: PoDisclaimerGroupModule });
    static ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [CommonModule, PoDisclaimerModule, PoTagModule] });
}
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(PoDisclaimerGroupModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, PoDisclaimerModule, PoTagModule],
                declarations: [PoDisclaimerGroupComponent, PoDisclaimerRemoveComponent],
                exports: [PoDisclaimerGroupComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(PoDisclaimerGroupModule, { declarations: [PoDisclaimerGroupComponent, PoDisclaimerRemoveComponent], imports: [CommonModule, PoDisclaimerModule, PoTagModule], exports: [PoDisclaimerGroupComponent] }); })();
