// @po-ui/ng-components/esm2022/lib/components/po-label/po-label.component.mjs
import { Component, Input } from '@angular/core';
import { convertToBoolean } from '../../utils/util';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
function PoLabelComponent_label_0_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "label", 2);
    i0.ɵɵtext(1);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = i0.ɵɵnextContext();
    i0.ɵɵclassProp("po-label-disabled", ctx_r0.disabled);
    i0.ɵɵattribute("for", ctx_r0.field && ctx_r0.for ? ctx_r0.for : undefined)("p-field", ctx_r0.field);
    i0.ɵɵadvance();
    i0.ɵɵtextInterpolate(ctx_r0.label);
} }
function PoLabelComponent_span_1_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "span", 3);
    i0.ɵɵtext(1);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r1 = i0.ɵɵnextContext();
    i0.ɵɵadvance();
    i0.ɵɵtextInterpolate(ctx_r1.requirement);
} }
/**
 * @docsPrivate
 *
 * Componente de uso interno.
 */
export class PoLabelComponent {
    _disabled = false;
    _field = false;
    /**
     * @optional
     *
     * @description
     *
     * Indica se o label está desativado.
     *
     * @default `false`
     */
    set disabled(value) {
        this._disabled = value === '' ? true : convertToBoolean(value);
    }
    get disabled() {
        return this._disabled;
    }
    /**
     * @optional
     *
     * @description
     *
     * Indica se o label será de um field.
     *
     * @default `false`
     */
    set field(value) {
        this._field = value === '' ? true : convertToBoolean(value);
    }
    get field() {
        return this._field;
    }
    /** Indica o campo vinculado ao label */
    for;
    /** Label do campo. */
    label;
    /** Indica o tipo do campo vinculado ao label */
    requirement;
    static ɵfac = function PoLabelComponent_Factory(t) { return new (t || PoLabelComponent)(); };
    static ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: PoLabelComponent, selectors: [["po-label"]], inputs: { disabled: [i0.ɵɵInputFlags.None, "p-disabled", "disabled"], field: [i0.ɵɵInputFlags.None, "p-field", "field"], for: [i0.ɵɵInputFlags.None, "p-for", "for"], label: [i0.ɵɵInputFlags.None, "p-label", "label"], requirement: [i0.ɵɵInputFlags.None, "p-requirement", "requirement"] }, decls: 2, vars: 2, consts: [["class", "po-label", 3, "po-label-disabled", 4, "ngIf"], ["class", "po-label-requirement", 4, "ngIf"], [1, "po-label"], [1, "po-label-requirement"]], template: function PoLabelComponent_Template(rf, ctx) { if (rf & 1) {
            i0.ɵɵtemplate(0, PoLabelComponent_label_0_Template, 2, 5, "label", 0)(1, PoLabelComponent_span_1_Template, 2, 1, "span", 1);
        } if (rf & 2) {
            i0.ɵɵproperty("ngIf", ctx.label);
            i0.ɵɵadvance();
            i0.ɵɵproperty("ngIf", ctx.field && ctx.requirement);
        } }, dependencies: [i1.NgIf], encapsulation: 2 });
}
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(PoLabelComponent, [{
        type: Component,
        args: [{ selector: 'po-label', template: "<label\n  *ngIf=\"label\"\n  class=\"po-label\"\n  [attr.for]=\"field && for ? for : undefined\"\n  [attr.p-field]=\"field\"\n  [class.po-label-disabled]=\"disabled\"\n  >{{ label }}</label\n>\n<span *ngIf=\"field && requirement\" class=\"po-label-requirement\">{{ requirement }}</span>\n" }]
    }], null, { disabled: [{
            type: Input,
            args: ['p-disabled']
        }], field: [{
            type: Input,
            args: ['p-field']
        }], for: [{
            type: Input,
            args: ['p-for']
        }], label: [{
            type: Input,
            args: ['p-label']
        }], requirement: [{
            type: Input,
            args: ['p-requirement']
        }] }); })();
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassDebugInfo(PoLabelComponent, { className: "PoLabelComponent", filePath: "lib/components/po-label/po-label.component.ts", lineNumber: 14 }); })();
