// @po-ui/ng-components/esm2022/lib/components/po-dynamic/po-dynamic-view/services/po-dynamic-view.service.mjs
import { HttpHeaders } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { map } from 'rxjs';
import { isTypeof } from '../../../../utils/util';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common/http";
/**
 * @docsPrivate
 *
 * @description
 *
 * Serviço padrão utilizado para filtrar os dados dos campos que utilizam a propriedade `searchService`.
 */
export class PoDynamicViewService {
    httpClient;
    headers = new HttpHeaders({
        'X-PO-No-Message': 'true'
    });
    url;
    constructor(httpClient) {
        this.httpClient = httpClient;
    }
    getObjectByValue(value, filterParams) {
        const validatedFilterParams = this.validateParams(filterParams);
        const encodedValue = encodeURIComponent(value);
        const newURL = `${this.url}/${encodedValue}`;
        return this.httpClient
            .get(newURL, { headers: this.headers, params: validatedFilterParams })
            .pipe(map((response) => ('items' in response ? response.items : response)));
    }
    onLoad(url, value) {
        return this.httpClient.post(url, value).toPromise();
    }
    setConfig(url) {
        this.url = url;
    }
    validateParams(params) {
        return isTypeof(params, 'object') && !Array.isArray(params) ? params : undefined;
    }
    static ɵfac = function PoDynamicViewService_Factory(t) { return new (t || PoDynamicViewService)(i0.ɵɵinject(i1.HttpClient)); };
    static ɵprov = /*@__PURE__*/ i0.ɵɵdefineInjectable({ token: PoDynamicViewService, factory: PoDynamicViewService.ɵfac });
}
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(PoDynamicViewService, [{
        type: Injectable
    }], () => [{ type: i1.HttpClient }], null); })();
