// @totvs/protheus-lib-core/esm2022/services/pro-language/services/pro-language.service.mjs
import { Injectable } from '@angular/core';
import { Observable, of } from 'rxjs';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common/http";
import * as i2 from "../../pro-js-to-advpl";
import * as i3 from "../../pro-app-config/services/pro-app-config.service";
export class ProLanguageService {
    constructor(http, advplService, proAppConfigService) {
        this.http = http;
        this.advplService = advplService;
        this.proAppConfigService = proAppConfigService;
        this.url = '/api/language/v1/protheus-suported-languages';
        this.useHTTP = true;
        this.EVENT_GET_LANGUAGES = 'getLanguages';
        this.EVENT_SET_LANGUAGES = 'setLanguages';
    }
    getListOfLanguages() {
        if (this.proAppConfigService.isProtheusRender) {
            if (this.useHTTP) {
                return this.http.get(this.url);
            }
            return this.getAdvplLanguages();
        }
        else {
            return of([]);
        }
    }
    getAdvplLanguages() {
        if (!this.advplService.protheusConnected()) {
            return this.advplNotPrepared();
        }
        return this.advplService.buildObservable(({ protheusResponse, subscriber }) => {
            if (protheusResponse.length === 0) {
                subscriber.error({
                    status: 400,
                    description: 'No language can be found'
                });
            }
            else {
                const laguages = JSON.parse(protheusResponse);
                subscriber.next(laguages);
            }
            subscriber.complete();
        }, {
            sendInfo: {
                type: this.EVENT_GET_LANGUAGES
            },
            autoDestruct: true,
            receiveId: this.EVENT_SET_LANGUAGES
        });
    }
    advplNotPrepared() {
        return new Observable(subscriber => {
            subscriber.error({
                status: 400,
                description: 'advplService not prepared in ProLanguageService'
            });
            subscriber.complete();
        });
    }
    isChannelHTTP() {
        return this.useHTTP;
    }
    setChannelAsHTTP(value) {
        this.useHTTP = value;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProLanguageService, deps: [{ token: i1.HttpClient }, { token: i2.ProJsToAdvplService }, { token: i3.ProAppConfigService }], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProLanguageService, providedIn: 'root' }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProLanguageService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root'
                }]
        }], ctorParameters: () => [{ type: i1.HttpClient }, { type: i2.ProJsToAdvplService }, { type: i3.ProAppConfigService }] });
