// @po-ui/ng-components/esm2022/lib/components/po-avatar/po-avatar.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { PoIconModule } from '../po-icon';
import { PoAvatarComponent } from './po-avatar.component';
import * as i0 from "@angular/core";
/**
 * @description
 *
 * Módulo do componente po-avatar.
 */
export class PoAvatarModule {
    static ɵfac = function PoAvatarModule_Factory(t) { return new (t || PoAvatarModule)(); };
    static ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: PoAvatarModule });
    static ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [CommonModule, PoIconModule] });
}
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(PoAvatarModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, PoIconModule],
                declarations: [PoAvatarComponent],
                exports: [PoAvatarComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(PoAvatarModule, { declarations: [PoAvatarComponent], imports: [CommonModule, PoIconModule], exports: [PoAvatarComponent] }); })();
