// @po-ui/ng-templates/esm2022/lib/services/po-page-dynamic/po-page-dynamic.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { PoPageDynamicService } from './po-page-dynamic.service';
import * as i0 from "@angular/core";
/**
 * @description
 *
 * Módulo do serviço do po-page-dynamic-service.
 */
export class PoPageDynamicModule {
    static ɵfac = function PoPageDynamicModule_Factory(t) { return new (t || PoPageDynamicModule)(); };
    static ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: PoPageDynamicModule });
    static ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ providers: [PoPageDynamicService], imports: [CommonModule] });
}
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(PoPageDynamicModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule],
                providers: [PoPageDynamicService]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(PoPageDynamicModule, { imports: [CommonModule] }); })();
