// @po-ui/ng-templates/esm2022/lib/components/po-page-login/po-page-login.module.mjs
import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { PoModule } from '@po-ui/ng-components';
import { PoPageBackgroundModule } from '../po-page-background/index';
import { PoPageLoginComponent } from './po-page-login.component';
import { PoPageLoginPopoverComponent } from './po-page-login-popover/po-page-login-popover.component';
import { PoPageLoginService } from './po-page-login.service';
import * as i0 from "@angular/core";
/**
 * @description
 *
 * Módulo do template do po-page-login.
 */
export class PoPageLoginModule {
    static ɵfac = function PoPageLoginModule_Factory(t) { return new (t || PoPageLoginModule)(); };
    static ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: PoPageLoginModule });
    static ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ providers: [PoPageLoginService], imports: [CommonModule, FormsModule, RouterModule, PoPageBackgroundModule, PoModule] });
}
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(PoPageLoginModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, FormsModule, RouterModule, PoPageBackgroundModule, PoModule],
                declarations: [PoPageLoginComponent, PoPageLoginPopoverComponent],
                exports: [PoPageLoginComponent],
                providers: [PoPageLoginService]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(PoPageLoginModule, { declarations: [PoPageLoginComponent, PoPageLoginPopoverComponent], imports: [CommonModule, FormsModule, RouterModule, PoPageBackgroundModule, PoModule], exports: [PoPageLoginComponent] }); })();
