// @totvs/protheus-lib-core/esm2022/components/pro-fields/pro-role-lookup/pro-role-lookup.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { ReactiveFormsModule } from '@angular/forms';
import { PoFieldModule, PoTooltipModule } from '@po-ui/ng-components';
import { ProRoleLookupComponent } from './pro-role-lookup.component';
import * as i0 from "@angular/core";
export class ProRoleLookupModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProRoleLookupModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "17.3.9", ngImport: i0, type: ProRoleLookupModule, declarations: [ProRoleLookupComponent], imports: [CommonModule, ReactiveFormsModule, PoFieldModule, PoTooltipModule], exports: [ProRoleLookupComponent] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProRoleLookupModule, imports: [CommonModule, ReactiveFormsModule, PoFieldModule, PoTooltipModule] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProRoleLookupModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, ReactiveFormsModule, PoFieldModule, PoTooltipModule],
                    declarations: [ProRoleLookupComponent],
                    exports: [ProRoleLookupComponent]
                }]
        }] });
