// @totvs/protheus-lib-core/esm2022/components/pro-fields/pro-role-lookup/pro-role-lookup.service.mjs
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "./../../../services/pro-role/pro-role-modules/services/pro-role.service";
export class ProRoleLookupService {
    constructor(proRoleService) {
        this.proRoleService = proRoleService;
    }
    getFilteredItems(params) {
        return this.proRoleService.getListofRoles(params.filter, params.page, params.pageSize);
    }
    getObjectByValue(code) {
        return this.proRoleService.getRoleByCode(code);
    }
    setRole(role) {
        this.proRoleService.role = role;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProRoleLookupService, deps: [{ token: i1.ProRoleService }], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProRoleLookupService, providedIn: 'root' }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProRoleLookupService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root'
                }]
        }], ctorParameters: () => [{ type: i1.ProRoleService }] });
