// @po-ui/ng-components/esm2022/lib/components/po-menu/services/po-menu-global.service.mjs
import { Injectable } from '@angular/core';
import { Subject } from 'rxjs';
import * as i0 from "@angular/core";
export class PoMenuGlobalService {
    applicationMenu = new Subject();
    menus = new Subject();
    removedApplicationMenu = new Subject();
    // eslint-disable-next-line @typescript-eslint/member-ordering
    receiveApplicationMenu$ = this.applicationMenu.asObservable();
    // eslint-disable-next-line @typescript-eslint/member-ordering
    receiveMenus$ = this.menus.asObservable();
    // eslint-disable-next-line @typescript-eslint/member-ordering
    receiveRemovedApplicationMenu$ = this.removedApplicationMenu.asObservable();
    sendApplicationMenu(menu) {
        this.applicationMenu.next(menu);
    }
    sendMenus(menus) {
        this.menus.next(menus);
    }
    sendRemovedApplicationMenu(id) {
        this.removedApplicationMenu.next(id);
    }
    static ɵfac = function PoMenuGlobalService_Factory(t) { return new (t || PoMenuGlobalService)(); };
    static ɵprov = /*@__PURE__*/ i0.ɵɵdefineInjectable({ token: PoMenuGlobalService, factory: PoMenuGlobalService.ɵfac, providedIn: 'root' });
}
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(PoMenuGlobalService, [{
        type: Injectable,
        args: [{
                providedIn: 'root'
            }]
    }], null, null); })();
