// @po-ui/ng-components/esm2022/lib/components/po-menu/services/po-menu.service.mjs
import { Injectable } from '@angular/core';
import { map } from 'rxjs/operators';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common/http";
/**
 * @docsPrivate
 *
 * @description
 *
 * Serviço que implementa a interface `PoMenuFilter`, utilizado para fazer requisições ao serviço informado pelo usuário,
 * caso for uma URL, no componente `po-menu`.
 */
export class PoMenuService {
    http;
    _url;
    get url() {
        return this._url;
    }
    constructor(http) {
        this.http = http;
    }
    configProperties(url) {
        this._url = url;
    }
    getFilteredData(search, params) {
        const filterParams = {
            search,
            ...params
        };
        return this.http
            .get(this.url, { params: filterParams })
            .pipe(map((response) => response && response.items));
    }
    static ɵfac = function PoMenuService_Factory(t) { return new (t || PoMenuService)(i0.ɵɵinject(i1.HttpClient)); };
    static ɵprov = /*@__PURE__*/ i0.ɵɵdefineInjectable({ token: PoMenuService, factory: PoMenuService.ɵfac });
}
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(PoMenuService, [{
        type: Injectable
    }], () => [{ type: i1.HttpClient }], null); })();
