// @po-ui/ng-components/esm2022/lib/components/po-field/po-field-container/po-field-container-bottom/po-field-container-bottom.component.mjs
import { ChangeDetectionStrategy, Component, Input } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
import * as i2 from "../../../po-icon/po-icon.component";
function PoFieldContainerBottomComponent_div_1_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "div", 3);
    i0.ɵɵelement(1, "po-icon", 4);
    i0.ɵɵelementStart(2, "span");
    i0.ɵɵtext(3);
    i0.ɵɵelementEnd()();
} if (rf & 2) {
    const ctx_r0 = i0.ɵɵnextContext();
    i0.ɵɵadvance(3);
    i0.ɵɵtextInterpolate1(" ", ctx_r0.errorPattern, "");
} }
function PoFieldContainerBottomComponent_span_2_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "span", 5);
    i0.ɵɵtext(1);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r1 = i0.ɵɵnextContext();
    i0.ɵɵadvance();
    i0.ɵɵtextInterpolate1(" ", ctx_r1.help, " ");
} }
/**
 * @docsPrivate
 *
 * @description
 *
 * Componente de uso interno, responsável por gerar uma margem inferior nos componentes que utilizam o po-field-container.
 * Essa margem inferior pode conter uma mensagem de erro.
 */
export class PoFieldContainerBottomComponent {
    /**
     * Mensagem que será apresentada quando o pattern ou a máscara não for satisfeita.
     * Obs: Esta mensagem não é apresentada quando o campo estiver vazio, mesmo que ele seja requerido.
     */
    errorPattern = '';
    disabled = false;
    help;
    static ɵfac = function PoFieldContainerBottomComponent_Factory(t) { return new (t || PoFieldContainerBottomComponent)(); };
    static ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: PoFieldContainerBottomComponent, selectors: [["po-field-container-bottom"]], inputs: { errorPattern: [i0.ɵɵInputFlags.None, "p-error-pattern", "errorPattern"], disabled: [i0.ɵɵInputFlags.None, "p-disabled", "disabled"], help: [i0.ɵɵInputFlags.None, "p-help", "help"] }, decls: 3, vars: 4, consts: [[1, "po-field-container-bottom"], ["class", "po-field-container-bottom-text-error", 4, "ngIf"], ["class", "po-field-container-bottom-help-text", 4, "ngIf"], [1, "po-field-container-bottom-text-error"], ["p-icon", "ICON_WARNING"], [1, "po-field-container-bottom-help-text"]], template: function PoFieldContainerBottomComponent_Template(rf, ctx) { if (rf & 1) {
            i0.ɵɵelementStart(0, "div", 0);
            i0.ɵɵtemplate(1, PoFieldContainerBottomComponent_div_1_Template, 4, 1, "div", 1)(2, PoFieldContainerBottomComponent_span_2_Template, 2, 1, "span", 2);
            i0.ɵɵelementEnd();
        } if (rf & 2) {
            i0.ɵɵclassProp("po-field-container-bottom-disabled", ctx.disabled);
            i0.ɵɵadvance();
            i0.ɵɵproperty("ngIf", ctx.errorPattern);
            i0.ɵɵadvance();
            i0.ɵɵproperty("ngIf", ctx.help && !ctx.errorPattern);
        } }, dependencies: [i1.NgIf, i2.PoIconComponent], encapsulation: 2, changeDetection: 0 });
}
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(PoFieldContainerBottomComponent, [{
        type: Component,
        args: [{ selector: 'po-field-container-bottom', changeDetection: ChangeDetectionStrategy.OnPush, template: "<div class=\"po-field-container-bottom\" [class.po-field-container-bottom-disabled]=\"disabled\">\n  <div *ngIf=\"errorPattern\" class=\"po-field-container-bottom-text-error\">\n    <po-icon p-icon=\"ICON_WARNING\"></po-icon>\n    <span> {{ errorPattern }}</span>\n  </div>\n  <span *ngIf=\"help && !errorPattern\" class=\"po-field-container-bottom-help-text\">\n    {{ help }}\n  </span>\n</div>\n" }]
    }], null, { errorPattern: [{
            type: Input,
            args: ['p-error-pattern']
        }], disabled: [{
            type: Input,
            args: ['p-disabled']
        }], help: [{
            type: Input,
            args: ['p-help']
        }] }); })();
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassDebugInfo(PoFieldContainerBottomComponent, { className: "PoFieldContainerBottomComponent", filePath: "lib/components/po-field/po-field-container/po-field-container-bottom/po-field-container-bottom.component.ts", lineNumber: 16 }); })();
