// @totvs/protheus-lib-core/esm2022/services/pro-system-info/pro-system-modules/services/pro-system-modules.service.mjs
import { HttpHeaders, HttpParams } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common/http";
import * as i2 from "../../../pro-js-to-advpl";
const CACHE_KEY = 'ProSystemModule';
export class ProSystemModuleService {
    constructor(http, advplService) {
        this.http = http;
        this.advplService = advplService;
        this.url = '/api/framework/v1/systemModules';
        this.useHTTP = true;
        this.EVENT_SET_LIST = 'setModulesList';
        this.EVENT_GET_LIST = 'getModulesList';
        this.EVENT_SET_ONE = 'setModuleInfo';
        this.EVENT_GET_ONE = 'getModuleInfo';
    }
    getListOfSystemModules(description = '', page = 1, _pageSize = 10) {
        const pageSize = 100;
        if (this.useHTTP) {
            return this.getListOfSystemModulesFromApi(description, page, pageSize);
        }
        return this.getListOfSystemModulesFromAdvpl(description, page, pageSize);
    }
    getListOfSystemModulesFromApi(description, page, pageSize) {
        let params = new HttpParams()
            .append('page', page.toString())
            .append('pageSize', pageSize.toString()); // Alterado o tamanho da página para melhorar a experiência do usuário.
        if (description !== '') {
            params = params.append('description', description);
        }
        const headers = new HttpHeaders().append('Accept', 'application/json; charset=utf-8');
        return this.http.get(this.url, { headers, params });
    }
    getListOfSystemModulesFromAdvpl(description, page, pageSize) {
        if (!this.advplService.protheusConnected()) {
            return this.advplNotPrepared();
        }
        const stringContent = JSON.stringify({
            description,
            page,
            pageSize,
        });
        return this.advplService.buildObservable(({ protheusResponse, subscriber }) => {
            if (protheusResponse.length === 0) {
                subscriber.error({
                    status: 400,
                    description: `module ${description} could not be found`
                });
            }
            else {
                const modulesData = JSON.parse(protheusResponse);
                subscriber.next(modulesData);
            }
            subscriber.complete();
        }, {
            sendInfo: {
                type: this.EVENT_GET_LIST,
                content: stringContent
            },
            autoDestruct: true,
            receiveId: this.EVENT_SET_LIST
        });
    }
    getSystemModule(systemModuleId) {
        if (this.useHTTP) {
            return this.getSystemModuleFromAPI(systemModuleId);
        }
        return this.getSystemModuleFromAdvpl(systemModuleId);
    }
    getSystemModuleFromAPI(systemModuleId) {
        const headers = new HttpHeaders().append('Accept', 'application/json; charset=utf-8');
        return this.http.get(`${this.url}/${systemModuleId}`, {
            headers
        });
    }
    getSystemModuleFromAdvpl(systemModuleId) {
        if (!this.advplService.protheusConnected()) {
            return this.advplNotPrepared();
        }
        return this.advplService.buildObservable(({ protheusResponse, subscriber }) => {
            if (protheusResponse.length === 0) {
                subscriber.error({
                    status: 400,
                    description: `module ${systemModuleId} could not be found`
                });
            }
            else {
                const moduleData = JSON.parse(protheusResponse);
                subscriber.next(moduleData);
            }
            subscriber.complete();
        }, {
            sendInfo: {
                type: this.EVENT_GET_ONE,
                content: systemModuleId
            },
            autoDestruct: true,
            receiveId: this.EVENT_SET_ONE
        });
    }
    get systemModule() {
        if (sessionStorage[CACHE_KEY]) {
            return JSON.parse(sessionStorage[CACHE_KEY]);
        }
        else {
            return { id: '', name: '', description: '' };
        }
    }
    set systemModule(systemModule) {
        sessionStorage[CACHE_KEY] = JSON.stringify(systemModule);
    }
    advplNotPrepared() {
        return new Observable(subscriber => {
            subscriber.error({
                status: 400,
                description: 'advplService not prepared in ProSystemModuleService'
            });
            subscriber.complete();
        });
    }
    isChannelHTTP() {
        return this.useHTTP;
    }
    setChannelAsHTTP(value) {
        this.useHTTP = value;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProSystemModuleService, deps: [{ token: i1.HttpClient }, { token: i2.ProJsToAdvplService }], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProSystemModuleService, providedIn: 'root' }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProSystemModuleService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root'
                }]
        }], ctorParameters: () => [{ type: i1.HttpClient }, { type: i2.ProJsToAdvplService }] });
