// @po-ui/ng-components/esm2022/lib/components/po-accordion/po-accordion-item-body/po-accordion-item-body.component.mjs
import { Component, Input } from '@angular/core';
import { animate, style, transition, trigger } from '@angular/animations';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
function PoAccordionItemBodyComponent_div_0_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "div", 1)(1, "div", 2);
    i0.ɵɵprojection(2);
    i0.ɵɵelementEnd()();
} if (rf & 2) {
    i0.ɵɵproperty("@toggleBody", undefined);
} }
const _c0 = ["*"];
export class PoAccordionItemBodyComponent {
    expanded = false;
    static ɵfac = function PoAccordionItemBodyComponent_Factory(t) { return new (t || PoAccordionItemBodyComponent)(); };
    static ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: PoAccordionItemBodyComponent, selectors: [["po-accordion-item-body"]], inputs: { expanded: [i0.ɵɵInputFlags.None, "p-expanded", "expanded"] }, ngContentSelectors: _c0, decls: 1, vars: 1, consts: [["class", "po-accordion-item-body", 4, "ngIf"], [1, "po-accordion-item-body"], [1, "po-accordion-item-body-content"]], template: function PoAccordionItemBodyComponent_Template(rf, ctx) { if (rf & 1) {
            i0.ɵɵprojectionDef();
            i0.ɵɵtemplate(0, PoAccordionItemBodyComponent_div_0_Template, 3, 1, "div", 0);
        } if (rf & 2) {
            i0.ɵɵproperty("ngIf", ctx.expanded);
        } }, dependencies: [i1.NgIf], encapsulation: 2, data: { animation: [
                trigger('toggleBody', [
                    transition('* => void', [style({ height: '*' }), animate(200, style({ height: 0 }))]),
                    transition('void => *', [style({ height: '0' }), animate(200, style({ height: '*' }))])
                ])
            ] } });
}
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(PoAccordionItemBodyComponent, [{
        type: Component,
        args: [{ selector: 'po-accordion-item-body', animations: [
                    trigger('toggleBody', [
                        transition('* => void', [style({ height: '*' }), animate(200, style({ height: 0 }))]),
                        transition('void => *', [style({ height: '0' }), animate(200, style({ height: '*' }))])
                    ])
                ], template: "<div class=\"po-accordion-item-body\" *ngIf=\"expanded\" @toggleBody>\n  <div class=\"po-accordion-item-body-content\">\n    <ng-content></ng-content>\n  </div>\n</div>\n" }]
    }], null, { expanded: [{
            type: Input,
            args: ['p-expanded']
        }] }); })();
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassDebugInfo(PoAccordionItemBodyComponent, { className: "PoAccordionItemBodyComponent", filePath: "lib/components/po-accordion/po-accordion-item-body/po-accordion-item-body.component.ts", lineNumber: 14 }); })();
