// @po-ui/ng-components/esm2022/lib/components/po-badge/po-badge.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { PoBadgeComponent } from './po-badge.component';
import { PoIconModule } from '../po-icon';
import * as i0 from "@angular/core";
/**
 * @description
 *
 * Módulo do componente po-badge.
 */
export class PoBadgeModule {
    static ɵfac = function PoBadgeModule_Factory(t) { return new (t || PoBadgeModule)(); };
    static ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: PoBadgeModule });
    static ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [CommonModule, PoIconModule] });
}
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(PoBadgeModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, PoIconModule],
                declarations: [PoBadgeComponent],
                exports: [PoBadgeComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(PoBadgeModule, { declarations: [PoBadgeComponent], imports: [CommonModule, PoIconModule], exports: [PoBadgeComponent] }); })();
