// @po-ui/ng-components/esm2022/lib/components/po-tag/po-tag.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { PoIconModule } from '../po-icon/po-icon.module';
import { PoTooltipModule } from '../../directives/po-tooltip/index';
import { PoTagComponent } from './po-tag.component';
import * as i0 from "@angular/core";
/**
 * @description
 *
 * Módulo do componente `po-tag`.
 */
export class PoTagModule {
    static ɵfac = function PoTagModule_Factory(t) { return new (t || PoTagModule)(); };
    static ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: PoTagModule });
    static ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [CommonModule, PoIconModule, PoTooltipModule] });
}
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(PoTagModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, PoIconModule, PoTooltipModule],
                declarations: [PoTagComponent],
                exports: [PoTagComponent],
                providers: [],
                schemas: []
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(PoTagModule, { declarations: [PoTagComponent], imports: [CommonModule, PoIconModule, PoTooltipModule], exports: [PoTagComponent] }); })();
