// @po-ui/ng-components/esm2022/lib/components/po-progress/po-progress-bar/po-progress-bar.component.mjs
import { ChangeDetectionStrategy, Component, Input } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
function PoProgressBarComponent_div_0_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "div", 2);
    i0.ɵɵelement(1, "div", 3)(2, "div", 4);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = i0.ɵɵnextContext();
    i0.ɵɵattribute("aria-valuenow", ctx_r0.value);
    i0.ɵɵadvance(2);
    i0.ɵɵstyleProp("left", "-" + (100 - ctx_r0.value) + "%");
} }
function PoProgressBarComponent_div_1_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "div", 5);
    i0.ɵɵelement(1, "div", 6);
    i0.ɵɵelementEnd();
} }
export class PoProgressBarComponent {
    indeterminate;
    value;
    get valueScale() {
        return `${this.value / 100}`;
    }
    static ɵfac = function PoProgressBarComponent_Factory(t) { return new (t || PoProgressBarComponent)(); };
    static ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: PoProgressBarComponent, selectors: [["po-progress-bar"]], inputs: { indeterminate: [i0.ɵɵInputFlags.None, "p-indeterminate", "indeterminate"], value: [i0.ɵɵInputFlags.None, "p-value", "value"] }, decls: 2, vars: 2, consts: [["role", "progressbar", "aria-valuemin", "0", "aria-valuemax", "100", "aria-live", "polite", "class", "po-progress-bar-default", 4, "ngIf"], ["class", "po-progress-bar-indeterminate-track", 4, "ngIf"], ["role", "progressbar", "aria-valuemin", "0", "aria-valuemax", "100", "aria-live", "polite", 1, "po-progress-bar-default"], [1, "po-progress-bar-element", "po-progress-bar-primary"], [1, "po-progress-bar-element", "po-progress-bar-secondary"], [1, "po-progress-bar-indeterminate-track"], [1, "po-progress-bar-indeterminate-track-bar"]], template: function PoProgressBarComponent_Template(rf, ctx) { if (rf & 1) {
            i0.ɵɵtemplate(0, PoProgressBarComponent_div_0_Template, 3, 3, "div", 0)(1, PoProgressBarComponent_div_1_Template, 2, 0, "div", 1);
        } if (rf & 2) {
            i0.ɵɵproperty("ngIf", !ctx.indeterminate);
            i0.ɵɵadvance();
            i0.ɵɵproperty("ngIf", ctx.indeterminate);
        } }, dependencies: [i1.NgIf], encapsulation: 2, changeDetection: 0 });
}
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(PoProgressBarComponent, [{
        type: Component,
        args: [{ selector: 'po-progress-bar', changeDetection: ChangeDetectionStrategy.OnPush, template: "<div\n  *ngIf=\"!indeterminate\"\n  role=\"progressbar\"\n  [attr.aria-valuenow]=\"value\"\n  aria-valuemin=\"0\"\n  aria-valuemax=\"100\"\n  aria-live=\"polite\"\n  class=\"po-progress-bar-default\"\n>\n  <div class=\"po-progress-bar-element po-progress-bar-primary\"></div>\n  <div class=\"po-progress-bar-element po-progress-bar-secondary\" [style.left]=\"'-' + (100 - value) + '%'\"></div>\n</div>\n\n<div *ngIf=\"indeterminate\" class=\"po-progress-bar-indeterminate-track\">\n  <div class=\"po-progress-bar-indeterminate-track-bar\"></div>\n</div>\n" }]
    }], null, { indeterminate: [{
            type: Input,
            args: ['p-indeterminate']
        }], value: [{
            type: Input,
            args: ['p-value']
        }] }); })();
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassDebugInfo(PoProgressBarComponent, { className: "PoProgressBarComponent", filePath: "lib/components/po-progress/po-progress-bar/po-progress-bar.component.ts", lineNumber: 8 }); })();
