// @totvs/protheus-lib-core/esm2022/components/pro-pages/pro-session-settings/guards/pro-session-settings-user-info/pro-session-settings-user-info.resolver.mjs
import { Injectable } from '@angular/core';
import { environment } from '../../../../../environments/environment';
import * as i0 from "@angular/core";
import * as i1 from "./../../../../../services/pro-authorization/services/pro-user-info/pro-user-info.service";
import * as i2 from "./../../../../../services/pro-authorization/services/pro-auth/pro-auth.service";
export class ProSessionSettingsUserInfoResolver {
    constructor(proUserInfoService, proAuthService) {
        this.proUserInfoService = proUserInfoService;
        this.proAuthService = proAuthService;
        this.proUserInfoService.setChannelAsHTTP(environment.useHTTP);
    }
    resolve(_route, _state) {
        const userId = this.proAuthService.userId;
        return this.proUserInfoService.get(userId);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProSessionSettingsUserInfoResolver, deps: [{ token: i1.ProUserInfoService }, { token: i2.ProAuthService }], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProSessionSettingsUserInfoResolver, providedIn: 'root' }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProSessionSettingsUserInfoResolver, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root'
                }]
        }], ctorParameters: () => [{ type: i1.ProUserInfoService }, { type: i2.ProAuthService }] });
