// @po-ui/ng-components/esm2022/lib/components/po-chart/po-chart-container/po-chart-line/po-chart-path/po-chart-path.component.mjs
import { Component, Input, ViewChild } from '@angular/core';
import { convertToBoolean, isIE } from '../../../../../utils/util';
import * as i0 from "@angular/core";
const _c0 = ["chartPath"];
const _c1 = ["po-chart-path", ""];
const pathDashoffsetDefaultWidth = 0;
export class PoChartPathComponent {
    changeDetector;
    isActive = true;
    chartLine = false;
    coordinates;
    chartPath;
    pathWidth;
    _animate;
    _color;
    set color(value) {
        this._color = value.includes('po-color') && this.chartLine ? value.replace('po-color', 'po-border-color') : value;
    }
    get color() {
        return this._color;
    }
    set animate(value) {
        this._animate = value;
        if (!this.animate) {
            this.pathWidth = pathDashoffsetDefaultWidth;
        }
    }
    get animate() {
        return this._animate;
    }
    constructor(changeDetector) {
        this.changeDetector = changeDetector;
    }
    ngAfterViewInit() {
        if (!isIE()) {
            this.pathAnimation();
        }
    }
    pathAnimation() {
        this.pathWidth = this.chartPath.nativeElement.getTotalLength();
        this.changeDetector.detectChanges();
        setTimeout(this.preventsAnimatingAgain.bind(this), 700);
    }
    // Redefine pathWidth para zero após o período da animação para que a linha se expanda normalmente on resize.
    preventsAnimatingAgain() {
        this.pathWidth = pathDashoffsetDefaultWidth;
    }
    static ɵfac = function PoChartPathComponent_Factory(t) { return new (t || PoChartPathComponent)(i0.ɵɵdirectiveInject(i0.ChangeDetectorRef)); };
    static ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: PoChartPathComponent, selectors: [["", "po-chart-path", ""]], viewQuery: function PoChartPathComponent_Query(rf, ctx) { if (rf & 1) {
            i0.ɵɵviewQuery(_c0, 7);
        } if (rf & 2) {
            let _t;
            i0.ɵɵqueryRefresh(_t = i0.ɵɵloadQuery()) && (ctx.chartPath = _t.first);
        } }, inputs: { isActive: [i0.ɵɵInputFlags.HasDecoratorInputTransform, "p-is-active", "isActive", convertToBoolean], chartLine: [i0.ɵɵInputFlags.HasDecoratorInputTransform, "p-chart-line", "chartLine", convertToBoolean], coordinates: [i0.ɵɵInputFlags.None, "p-coordinates", "coordinates"], color: [i0.ɵɵInputFlags.None, "p-color", "color"], animate: [i0.ɵɵInputFlags.None, "p-animate", "animate"] }, features: [i0.ɵɵInputTransformsFeature], attrs: _c1, decls: 2, vars: 11, consts: [[1, "po-chart-path"], ["chartPath", ""]], template: function PoChartPathComponent_Template(rf, ctx) { if (rf & 1) {
            i0.ɵɵnamespaceSVG();
            i0.ɵɵelement(0, "path", 0, 1);
        } if (rf & 2) {
            i0.ɵɵclassMap((ctx.color == null ? null : ctx.color.includes("po-border-color")) || (ctx.color == null ? null : ctx.color.includes("po-color")) ? ctx.color : "");
            i0.ɵɵclassProp("po-chart-active-path", ctx.isActive)("po-chart-line-path", ctx.chartLine);
            i0.ɵɵattribute("fill", ctx.chartLine ? "none" : ctx.color)("stroke-dasharray", ctx.pathWidth)("stroke-dashoffset", ctx.pathWidth)("stroke", ctx.chartLine ? ctx.color : "")("d", ctx.coordinates);
        } }, encapsulation: 2 });
}
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(PoChartPathComponent, [{
        type: Component,
        args: [{ selector: '[po-chart-path]', template: "<svg:path #chartPath\n  class=\"po-chart-path\"\n  [class.po-chart-active-path]=\"isActive\"\n  [class.po-chart-line-path]=\"chartLine\"\n  [class]=\"color?.includes('po-border-color') || color?.includes('po-color') ? color : ''\"\n  [attr.fill]=\"chartLine ? 'none' : color\"\n  [attr.stroke-dasharray]=\"pathWidth\"\n  [attr.stroke-dashoffset]=\"pathWidth\"\n  [attr.stroke]=\"chartLine ? color : ''\"\n  [attr.d]=\"coordinates\">\n</svg:path> " }]
    }], () => [{ type: i0.ChangeDetectorRef }], { isActive: [{
            type: Input,
            args: [{ alias: 'p-is-active', transform: convertToBoolean }]
        }], chartLine: [{
            type: Input,
            args: [{ alias: 'p-chart-line', transform: convertToBoolean }]
        }], coordinates: [{
            type: Input,
            args: ['p-coordinates']
        }], chartPath: [{
            type: ViewChild,
            args: ['chartPath', { static: true }]
        }], color: [{
            type: Input,
            args: ['p-color']
        }], animate: [{
            type: Input,
            args: ['p-animate']
        }] }); })();
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassDebugInfo(PoChartPathComponent, { className: "PoChartPathComponent", filePath: "lib/components/po-chart/po-chart-container/po-chart-line/po-chart-path/po-chart-path.component.ts", lineNumber: 11 }); })();
