// @po-ui/ng-templates/esm2022/lib/services/services.module.mjs
import { NgModule } from '@angular/core';
import { PoPageCustomizationModule } from './po-page-customization/po-page-customization.module';
import { PoPageDynamicModule } from './po-page-dynamic/po-page-dynamic.module';
import * as i0 from "@angular/core";
export class PoServicesModule {
    static ɵfac = function PoServicesModule_Factory(t) { return new (t || PoServicesModule)(); };
    static ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: PoServicesModule });
    static ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [PoPageCustomizationModule, PoPageDynamicModule, PoPageCustomizationModule, PoPageDynamicModule] });
}
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(PoServicesModule, [{
        type: NgModule,
        args: [{
                imports: [PoPageCustomizationModule, PoPageDynamicModule],
                exports: [PoPageCustomizationModule, PoPageDynamicModule]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(PoServicesModule, { imports: [PoPageCustomizationModule, PoPageDynamicModule], exports: [PoPageCustomizationModule, PoPageDynamicModule] }); })();
