// @po-ui/ng-components/esm2022/lib/components/po-stepper/po-stepper-circle/po-stepper-circle.component.mjs
import { Component, Input } from '@angular/core';
import { PoStepperStatus } from '../enums/po-stepper-status.enum';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
import * as i2 from "../../po-icon/po-icon.component";
function PoStepperCircleComponent_po_icon_1_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "po-icon", 3);
    i0.ɵɵtext(1);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = i0.ɵɵnextContext();
    i0.ɵɵclassProp("po-icon", ctx_r0.icons || ctx_r0.isDone)("po-stepper-circle-content-lg", ctx_r0.isLargeStep)("po-stepper-circle-content-md", ctx_r0.isMediumStep);
    i0.ɵɵproperty("p-icon", ctx_r0.icons && ctx_r0.isError ? "ICON_EXCLAMATION" : ctx_r0.icons && (ctx_r0.isActive || ctx_r0.isDefault || ctx_r0.isDisabled) ? "ICON_INFO" : ctx_r0.isDone ? "ICON_OK" : "");
    i0.ɵɵadvance();
    i0.ɵɵtextInterpolate1(" ", !ctx_r0.icons && !ctx_r0.isDone ? ctx_r0.content : "", " ");
} }
function PoStepperCircleComponent_div_2_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelement(0, "div", 4);
} }
const poMediumStepSize = 32;
const poLargeStepSize = 48;
/**
 * @docsPrivate
 *
 * @description
 *
 * Componente responsável por exibir os círculos dos *steps*.
 */
export class PoStepperCircleComponent {
    // Conteúdo que irá aparecer no círculo do *step*.
    content;
    // Define se serão exibidos ícones no lugar de números nos steps.
    icons;
    // Tamanho do *step-circle*.
    size;
    // Status do *step*.
    status;
    get isActive() {
        return this.status === PoStepperStatus.Active;
    }
    get isDefault() {
        return this.status === PoStepperStatus.Default;
    }
    get isDisabled() {
        return this.status === PoStepperStatus.Disabled;
    }
    get isDone() {
        return this.status === PoStepperStatus.Done;
    }
    get isError() {
        return this.status === PoStepperStatus.Error;
    }
    get isLargeStep() {
        return this.size >= poLargeStepSize;
    }
    get isMediumStep() {
        return this.size >= poMediumStepSize && !this.isLargeStep;
    }
    static ɵfac = function PoStepperCircleComponent_Factory(t) { return new (t || PoStepperCircleComponent)(); };
    static ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: PoStepperCircleComponent, selectors: [["po-stepper-circle"]], inputs: { content: [i0.ɵɵInputFlags.None, "p-content", "content"], icons: [i0.ɵɵInputFlags.None, "p-icons", "icons"], size: [i0.ɵɵInputFlags.None, "p-size", "size"], status: [i0.ɵɵInputFlags.None, "p-status", "status"] }, decls: 3, vars: 7, consts: [[1, "po-stepper-circle", 3, "tabindex"], ["class", "po-stepper-circle-content", 3, "po-icon", "p-icon", "po-stepper-circle-content-lg", "po-stepper-circle-content-md", 4, "ngIf"], ["class", "po-stepper-circle-active", 4, "ngIf"], [1, "po-stepper-circle-content", 3, "p-icon"], [1, "po-stepper-circle-active"]], template: function PoStepperCircleComponent_Template(rf, ctx) { if (rf & 1) {
            i0.ɵɵelementStart(0, "div", 0);
            i0.ɵɵtemplate(1, PoStepperCircleComponent_po_icon_1_Template, 2, 8, "po-icon", 1)(2, PoStepperCircleComponent_div_2_Template, 1, 0, "div", 2);
            i0.ɵɵelementEnd();
        } if (rf & 2) {
            i0.ɵɵstyleProp("height", ctx.size, "px")("width", ctx.size, "px");
            i0.ɵɵproperty("tabindex", ctx.isDisabled ? -1 : 0);
            i0.ɵɵadvance();
            i0.ɵɵproperty("ngIf", !ctx.isActive);
            i0.ɵɵadvance();
            i0.ɵɵproperty("ngIf", ctx.isActive || ctx.isError);
        } }, dependencies: [i1.NgIf, i2.PoIconComponent], encapsulation: 2 });
}
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(PoStepperCircleComponent, [{
        type: Component,
        args: [{ selector: 'po-stepper-circle', template: "<div class=\"po-stepper-circle\" [style.height.px]=\"size\" [style.width.px]=\"size\" [tabindex]=\"isDisabled ? -1 : 0\">\n  <po-icon\n    *ngIf=\"!isActive\"\n    class=\"po-stepper-circle-content\"\n    [class.po-icon]=\"icons || isDone\"\n    [p-icon]=\"\n      icons && isError\n        ? 'ICON_EXCLAMATION'\n        : icons && (isActive || isDefault || isDisabled)\n          ? 'ICON_INFO'\n          : isDone\n            ? 'ICON_OK'\n            : ''\n    \"\n    [class.po-stepper-circle-content-lg]=\"isLargeStep\"\n    [class.po-stepper-circle-content-md]=\"isMediumStep\"\n  >\n    {{ !icons && !isDone ? content : '' }}\n  </po-icon>\n\n  <div *ngIf=\"isActive || isError\" class=\"po-stepper-circle-active\"></div>\n</div>\n" }]
    }], null, { content: [{
            type: Input,
            args: ['p-content']
        }], icons: [{
            type: Input,
            args: ['p-icons']
        }], size: [{
            type: Input,
            args: ['p-size']
        }], status: [{
            type: Input,
            args: ['p-status']
        }] }); })();
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassDebugInfo(PoStepperCircleComponent, { className: "PoStepperCircleComponent", filePath: "lib/components/po-stepper/po-stepper-circle/po-stepper-circle.component.ts", lineNumber: 19 }); })();
