// @po-ui/ng-templates/esm2022/lib/services/po-page-customization/po-page-customization.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { PoPageCustomizationService } from './po-page-customization.service';
import * as i0 from "@angular/core";
/**
 * @description
 *
 * Módulo do serviço do po-page-customization-service.
 */
export class PoPageCustomizationModule {
    static ɵfac = function PoPageCustomizationModule_Factory(t) { return new (t || PoPageCustomizationModule)(); };
    static ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: PoPageCustomizationModule });
    static ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ providers: [PoPageCustomizationService], imports: [CommonModule] });
}
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(PoPageCustomizationModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule],
                providers: [PoPageCustomizationService]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(PoPageCustomizationModule, { imports: [CommonModule] }); })();
