// @po-ui/ng-components/esm2022/lib/components/po-dynamic/po-dynamic-field-type.enum.mjs
/**
 * @usedBy PoDynamicFormComponent
 *
 * @description
 *
 * Enum para definição do tipo de campo que será criado dinamicamente.
 */
export var PoDynamicFieldType;
(function (PoDynamicFieldType) {
    /** Valor booleano. */
    PoDynamicFieldType["Boolean"] = "boolean";
    /** Valor numérico que contém casas decimais e milhar. */
    PoDynamicFieldType["Currency"] = "currency";
    /** Valor numérico que contém casas decimais e milhar. */
    PoDynamicFieldType["Decimal"] = "decimal";
    /** Valor para data. */
    PoDynamicFieldType["Date"] = "date";
    /** Valor para data e hora. */
    PoDynamicFieldType["DateTime"] = "datetime";
    /** Utilizado para informar/exibir hora. */
    PoDynamicFieldType["Time"] = "time";
    /** Valor numérico. */
    PoDynamicFieldType["Number"] = "number";
    /** Texto. */
    PoDynamicFieldType["String"] = "string";
    /** Utilizado para fazer uploads de arquivos. */
    PoDynamicFieldType["Upload"] = "upload";
})(PoDynamicFieldType || (PoDynamicFieldType = {}));
