// @totvs/protheus-lib-core/esm2022/components/pro-fields/pro-company-lookup/pro-company-lookup.service.mjs
import { Injectable } from '@angular/core';
import { environment } from '../../../environments/environment';
import * as i0 from "@angular/core";
import * as i1 from "./../../../services/pro-tenant/pro-company/services/pro-company.service";
export class ProCompanyLookupService {
    constructor(proCompanyService) {
        this.proCompanyService = proCompanyService;
        this.useHTTP = environment.useHTTP;
        this.changeServiceChannel();
    }
    getFilteredItems(params) {
        return this.proCompanyService.getListOfCompanies(params.filter, params.page, params.pageSize);
    }
    getObjectByValue(value) {
        return this.proCompanyService.getCompany(value);
    }
    setCompany(company) {
        this.proCompanyService.company = company;
    }
    setChannelAsHTTP(value) {
        this.useHTTP = value;
        this.changeServiceChannel();
    }
    isChannelHTTP() {
        return this.useHTTP;
    }
    changeServiceChannel() {
        this.proCompanyService.setChannelAsHTTP(this.useHTTP);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProCompanyLookupService, deps: [{ token: i1.ProCompanyService }], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProCompanyLookupService, providedIn: 'root' }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProCompanyLookupService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root'
                }]
        }], ctorParameters: () => [{ type: i1.ProCompanyService }] });
