// @totvs/protheus-lib-core/esm2022/components/pro-pages/pro-login/pro-login.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { ReactiveFormsModule } from '@angular/forms';
import { PoComponentsModule } from '@po-ui/ng-templates';
import { ProLoginComponent } from './pro-login.component';
import * as i0 from "@angular/core";
export class ProLoginModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProLoginModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "17.3.9", ngImport: i0, type: ProLoginModule, declarations: [ProLoginComponent], imports: [CommonModule, PoComponentsModule, ReactiveFormsModule], exports: [ProLoginComponent] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProLoginModule, imports: [CommonModule, PoComponentsModule, ReactiveFormsModule] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProLoginModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, PoComponentsModule, ReactiveFormsModule],
                    declarations: [ProLoginComponent],
                    exports: [ProLoginComponent]
                }]
        }] });
