// @po-ui/ng-components/esm2022/lib/components/po-loading/po-loading.component.mjs
import { Component } from '@angular/core';
import { PoLoadingBaseComponent } from './po-loading-base.component';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
import * as i2 from "./po-loading-icon/po-loading-icon.component";
function PoLoadingComponent_span_2_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "span", 2);
    i0.ɵɵtext(1);
    i0.ɵɵelementStart(2, "div", 3);
    i0.ɵɵtext(3, ".");
    i0.ɵɵelementEnd();
    i0.ɵɵelementStart(4, "div", 3);
    i0.ɵɵtext(5, ".");
    i0.ɵɵelementEnd();
    i0.ɵɵelementStart(6, "div", 3);
    i0.ɵɵtext(7, ".");
    i0.ɵɵelementEnd()();
} if (rf & 2) {
    const ctx_r0 = i0.ɵɵnextContext();
    i0.ɵɵadvance();
    i0.ɵɵtextInterpolate1("", ctx_r0.text, " ");
} }
/**
 * @docsPrivate
 *
 * @docsExtends PoLoadingBaseComponent
 */
export class PoLoadingComponent extends PoLoadingBaseComponent {
    static ɵfac = /*@__PURE__*/ (() => { let ɵPoLoadingComponent_BaseFactory; return function PoLoadingComponent_Factory(t) { return (ɵPoLoadingComponent_BaseFactory || (ɵPoLoadingComponent_BaseFactory = i0.ɵɵgetInheritedFactory(PoLoadingComponent)))(t || PoLoadingComponent); }; })();
    static ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: PoLoadingComponent, selectors: [["po-loading"]], features: [i0.ɵɵInheritDefinitionFeature], decls: 3, vars: 5, consts: [[3, "p-size"], ["class", "po-loading-label po-text-ellipsis", 4, "ngIf"], [1, "po-loading-label", "po-text-ellipsis"], ["aria-hidden", "true", 1, "po-loading-dot"]], template: function PoLoadingComponent_Template(rf, ctx) { if (rf & 1) {
            i0.ɵɵelementStart(0, "div");
            i0.ɵɵelement(1, "po-loading-icon", 0);
            i0.ɵɵtemplate(2, PoLoadingComponent_span_2_Template, 8, 1, "span", 1);
            i0.ɵɵelementEnd();
        } if (rf & 2) {
            i0.ɵɵclassMapInterpolate1("po-loading po-loading-", ctx.size, "");
            i0.ɵɵadvance();
            i0.ɵɵproperty("p-size", ctx.size);
            i0.ɵɵadvance();
            i0.ɵɵproperty("ngIf", ctx.text);
        } }, dependencies: [i1.NgIf, i2.PoLoadingIconComponent], encapsulation: 2 });
}
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(PoLoadingComponent, [{
        type: Component,
        args: [{ selector: 'po-loading', template: "<div class=\"po-loading po-loading-{{ size }}\">\n  <po-loading-icon [p-size]=\"size\"></po-loading-icon>\n\n  <span *ngIf=\"text\" class=\"po-loading-label po-text-ellipsis\"\n    >{{ text }}\n    <div class=\"po-loading-dot\" aria-hidden=\"true\">.</div>\n    <div class=\"po-loading-dot\" aria-hidden=\"true\">.</div>\n    <div class=\"po-loading-dot\" aria-hidden=\"true\">.</div>\n  </span>\n</div>\n" }]
    }], null, null); })();
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassDebugInfo(PoLoadingComponent, { className: "PoLoadingComponent", filePath: "lib/components/po-loading/po-loading.component.ts", lineNumber: 14 }); })();
