// @po-ui/ng-components/esm2022/lib/components/po-loading/po-loading-base.component.mjs
import { Input, Directive } from '@angular/core';
import { PoLoadingIconSize } from './po-loading-icon/po-loading-icon-size-enum';
import * as i0 from "@angular/core";
import * as i1 from "../../services/po-language/po-language.service";
export const poLoadingLiteralsDefault = {
    en: {
        loading: 'Loading'
    },
    es: {
        loading: 'Cargando'
    },
    pt: {
        loading: 'Carregando'
    },
    ru: {
        loading: 'Загрузка'
    }
};
/**
 * @docsPrivate
 *
 * @description
 *
 * Este componente tem o objetivo de mostrar visualmente aos usuários que a aplicação está processando
 * ou aguardando a resposta de alguma requisição.
 */
export class PoLoadingBaseComponent {
    languageService;
    _text;
    _size;
    /**
     * Texto a ser exibido no componente.
     */
    set text(value) {
        this._text = value || (value === '' ? '' : this.getTextDefault());
    }
    get text() {
        return this._text;
    }
    /**
     * Define o tamanho do ícone.
     *
     * @default `lg`
     *
     * Valores válidos:
     *  - `xs`: tamanho `extra small`
     *  - `sm`: tamanho `small`
     *  - `md`: tamanho `medium`
     *  - `lg`: tamanho `large`
     */
    set size(value) {
        this._size = PoLoadingIconSize[value] ? PoLoadingIconSize[value] : PoLoadingIconSize.lg;
    }
    get size() {
        return this._size;
    }
    constructor(languageService) {
        this.languageService = languageService;
        this.text = this.getTextDefault();
    }
    getTextDefault() {
        const language = this.languageService.getShortLanguage();
        return poLoadingLiteralsDefault[language].loading;
    }
    static ɵfac = function PoLoadingBaseComponent_Factory(t) { return new (t || PoLoadingBaseComponent)(i0.ɵɵdirectiveInject(i1.PoLanguageService)); };
    static ɵdir = /*@__PURE__*/ i0.ɵɵdefineDirective({ type: PoLoadingBaseComponent, inputs: { text: [i0.ɵɵInputFlags.None, "p-text", "text"], size: [i0.ɵɵInputFlags.None, "p-size", "size"] } });
}
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(PoLoadingBaseComponent, [{
        type: Directive
    }], () => [{ type: i1.PoLanguageService }], { text: [{
            type: Input,
            args: ['p-text']
        }], size: [{
            type: Input,
            args: ['p-size']
        }] }); })();
