// @po-ui/ng-components/esm2022/lib/components/po-image/po-image.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule, NgOptimizedImage } from '@angular/common';
import { PoImageComponent } from './po-image.component';
import { PoContainerModule } from '../po-container';
import * as i0 from "@angular/core";
/**
 * @description
 *
 * Módulo do componente po-image
 */
export class PoImageModule {
    static ɵfac = function PoImageModule_Factory(t) { return new (t || PoImageModule)(); };
    static ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: PoImageModule });
    static ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [CommonModule, PoContainerModule] });
}
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(PoImageModule, [{
        type: NgModule,
        args: [{
                declarations: [PoImageComponent],
                imports: [CommonModule, PoContainerModule, NgOptimizedImage],
                exports: [PoImageComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(PoImageModule, { declarations: [PoImageComponent], imports: [CommonModule, PoContainerModule, NgOptimizedImage], exports: [PoImageComponent] }); })();
