// @po-ui/ng-components/esm2022/lib/components/po-field/po-radio/po-radio.module.mjs
import { FormsModule } from '@angular/forms';
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { PoRadioComponent } from './po-radio.component';
import { PoLabelModule } from '../../po-label/po-label.module';
import * as i0 from "@angular/core";
export class PoRadioModule {
    static ɵfac = function PoRadioModule_Factory(t) { return new (t || PoRadioModule)(); };
    static ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: PoRadioModule });
    static ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [CommonModule, FormsModule, PoLabelModule] });
}
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(PoRadioModule, [{
        type: NgModule,
        args: [{
                declarations: [PoRadioComponent],
                exports: [PoRadioComponent],
                imports: [CommonModule, FormsModule, PoLabelModule]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(PoRadioModule, { declarations: [PoRadioComponent], imports: [CommonModule, FormsModule, PoLabelModule], exports: [PoRadioComponent] }); })();
