// @totvs/protheus-lib-core/esm2022/services/pro-generic-adapter/pro-generic-adapter.service.mjs
import { HttpParams } from '@angular/common/http';
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common/http";
export class ProGenericAdapterService {
    constructor(http) {
        this.http = http;
        this.URL_LIST = "/api/framework/v1/genericList";
        this.URL_QUERY = "/api/framework/v1/genericQuery";
    }
    /**
     * Metodo para devolver uma listagem de registro baseado na api genericList
     * @param searchParams parametros da interface ProAdapterListInterface a ser enviado
     * para api genericList
     * @returns Observable de ProAdapterBaseV2 a ser utilizado em um po-table
     */
    list(searchParams) {
        const params = this.convertSearchParamsToHttpParams(searchParams);
        return this.http.get(this.URL_LIST, { params });
    }
    /**
     * Metodo para devolver uma listagem de registro baseado na api genericQuery
     * @param searchParams parametros da interface ProAdapterQueryInterface a ser enviado
     * para api genericQuery
     * @returns Observable de ProAdapterBaseV2 a ser utilizado em um po-table
     */
    query(searchParams) {
        const params = this.convertSearchParamsToHttpParams(searchParams);
        return this.http.get(this.URL_QUERY, { params });
    }
    /**
     * converte uma interface para httpParams
     * cada propriedade vira um parametro http
     * @param searchParams parametro a ser convetido
     * @returns httpParams a ser utilizado em uma requisição
     */
    convertSearchParamsToHttpParams(searchParams) {
        let params = new HttpParams();
        Object.keys(searchParams).forEach(key => {
            params = params.set(key, searchParams[key]);
        });
        return params;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProGenericAdapterService, deps: [{ token: i1.HttpClient }], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProGenericAdapterService, providedIn: 'root' }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProGenericAdapterService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root'
                }]
        }], ctorParameters: () => [{ type: i1.HttpClient }] });
