// @totvs/protheus-lib-core/esm2022/services/pro-role/pro-role-modules/services/pro-role.service.mjs
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import * as i0 from "@angular/core";
import * as i1 from "./../../../pro-js-to-advpl/services/pro-js-to-advpl.service";
const CACHE_KEY = 'ProRole';
export class ProRoleService {
    constructor(advplService) {
        this.advplService = advplService;
    }
    getListofRoles(description = '', page = 1, pageSize = 10) {
        pageSize = 100; // Alterado o tamanho da página para melhorar a experiência do usuário.
        if (this.advplService.protheusConnected()) {
            return this.advplService.buildObservable(({ protheusResponse, subscriber }) => {
                if (protheusResponse.length === 0) {
                    subscriber.next({
                        hasNext: false,
                        items: []
                    });
                }
                else {
                    const roleList = JSON.parse(protheusResponse)
                        .map((role) => {
                        return {
                            Code: role[0],
                            Description: role[1]
                        };
                    })
                        .filter((role) => {
                        return role.Description.includes(description);
                    });
                    subscriber.next({
                        hasNext: roleList.length > page * pageSize,
                        items: roleList.slice((page - 1) * pageSize, page * pageSize)
                    });
                }
                subscriber.complete();
            }, {
                autoDestruct: true,
                receiveId: 'setRoleList',
                sendInfo: { type: 'getRoleList' }
            });
        }
        else {
            return new Observable(subscriber => {
                subscriber.next({
                    hasNext: false,
                    items: []
                });
                subscriber.complete();
            });
        }
    }
    getRoleByCode(roleCode) {
        if (this.advplService.protheusConnected()) {
            return this.advplService.buildObservable(({ protheusResponse, subscriber }) => {
                const role = JSON.parse(protheusResponse);
                const response = { Code: role[0], Description: role[1] };
                subscriber.next(response);
                subscriber.complete();
            }, {
                autoDestruct: true,
                receiveId: 'setRoleItem',
                sendInfo: { type: 'getRoleItem', content: roleCode }
            });
        }
        else {
            return new Observable(subscriber => {
                subscriber.next({ Code: '', Description: '' });
                subscriber.complete();
            });
        }
    }
    get role() {
        if (sessionStorage[CACHE_KEY]) {
            return JSON.parse(sessionStorage[CACHE_KEY]);
        }
        else {
            return { Code: '', Description: '' };
        }
    }
    set role(role) {
        sessionStorage[CACHE_KEY] = JSON.stringify(role);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProRoleService, deps: [{ token: i1.ProJsToAdvplService }], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProRoleService, providedIn: 'root' }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProRoleService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root'
                }]
        }], ctorParameters: () => [{ type: i1.ProJsToAdvplService }] });
