// @po-ui/ng-components/esm2022/lib/components/po-progress/po-progress.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { PoProgressBarComponent } from './po-progress-bar/po-progress-bar.component';
import { PoProgressComponent } from './po-progress.component';
import { PoButtonModule } from '../po-button/po-button.module';
import { PoIconModule } from '../po-icon/po-icon.module';
import { PoLabelModule } from '../po-label/po-label.module';
import * as i0 from "@angular/core";
/**
 * @description
 *
 * Módulo do componente `po-progress`.
 */
export class PoProgressModule {
    static ɵfac = function PoProgressModule_Factory(t) { return new (t || PoProgressModule)(); };
    static ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: PoProgressModule });
    static ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [CommonModule, PoButtonModule, PoIconModule, PoLabelModule] });
}
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(PoProgressModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, PoButtonModule, PoIconModule, PoLabelModule],
                exports: [PoProgressComponent],
                declarations: [PoProgressBarComponent, PoProgressComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(PoProgressModule, { declarations: [PoProgressBarComponent, PoProgressComponent], imports: [CommonModule, PoButtonModule, PoIconModule, PoLabelModule], exports: [PoProgressComponent] }); })();
