// @po-ui/ng-templates/esm2022/lib/components/po-page-job-scheduler/po-page-job-scheduler-lookup.service.mjs
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "./po-page-job-scheduler.service";
export class PoPageJobSchedulerLookupService {
    poPageJobSchedulerService;
    constructor(poPageJobSchedulerService) {
        this.poPageJobSchedulerService = poPageJobSchedulerService;
    }
    getFilteredItems({ filter, page, pageSize }) {
        const params = { page, pageSize, search: filter };
        return this.poPageJobSchedulerService.getProcesses(params);
    }
    getObjectByValue(processId) {
        return this.poPageJobSchedulerService.getProcess(processId);
    }
    static ɵfac = function PoPageJobSchedulerLookupService_Factory(t) { return new (t || PoPageJobSchedulerLookupService)(i0.ɵɵinject(i1.PoPageJobSchedulerService)); };
    static ɵprov = /*@__PURE__*/ i0.ɵɵdefineInjectable({ token: PoPageJobSchedulerLookupService, factory: PoPageJobSchedulerLookupService.ɵfac });
}
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(PoPageJobSchedulerLookupService, [{
        type: Injectable
    }], () => [{ type: i1.PoPageJobSchedulerService }], null); })();
