// @po-ui/ng-components/esm2022/lib/components/po-page/po-page-header/po-page-header.component.mjs
import { Component } from '@angular/core';
import { PoPageHeaderBaseComponent } from './po-page-header-base.component';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
import * as i2 from "../../po-breadcrumb/po-breadcrumb.component";
function PoPageHeaderComponent_div_1_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "div", 4);
    i0.ɵɵelement(1, "po-breadcrumb", 5);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = i0.ɵɵnextContext();
    i0.ɵɵadvance();
    i0.ɵɵproperty("p-favorite-service", ctx_r0.breadcrumb.favorite)("p-items", ctx_r0.breadcrumb.items)("p-params-service", ctx_r0.breadcrumb.params);
} }
function PoPageHeaderComponent_div_3_h1_1_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "h1", 9);
    i0.ɵɵtext(1);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r2 = i0.ɵɵnextContext(2);
    i0.ɵɵadvance();
    i0.ɵɵtextInterpolate1(" ", ctx_r2.title, " ");
} }
function PoPageHeaderComponent_div_3_div_2_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "div", 10);
    i0.ɵɵtext(1);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r3 = i0.ɵɵnextContext(2);
    i0.ɵɵadvance();
    i0.ɵɵtextInterpolate1(" ", ctx_r3.subtitle, " ");
} }
function PoPageHeaderComponent_div_3_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "div", 6);
    i0.ɵɵtemplate(1, PoPageHeaderComponent_div_3_h1_1_Template, 2, 1, "h1", 7)(2, PoPageHeaderComponent_div_3_div_2_Template, 2, 1, "div", 8);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r1 = i0.ɵɵnextContext();
    i0.ɵɵadvance();
    i0.ɵɵproperty("ngIf", ctx_r1.title);
    i0.ɵɵadvance();
    i0.ɵɵproperty("ngIf", ctx_r1.subtitle);
} }
const _c0 = ["*"];
/**
 * @docsPrivate
 * @docsExtends PoPageHeaderBaseComponent
 */
export class PoPageHeaderComponent extends PoPageHeaderBaseComponent {
    static ɵfac = /*@__PURE__*/ (() => { let ɵPoPageHeaderComponent_BaseFactory; return function PoPageHeaderComponent_Factory(t) { return (ɵPoPageHeaderComponent_BaseFactory || (ɵPoPageHeaderComponent_BaseFactory = i0.ɵɵgetInheritedFactory(PoPageHeaderComponent)))(t || PoPageHeaderComponent); }; })();
    static ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: PoPageHeaderComponent, selectors: [["po-page-header"]], features: [i0.ɵɵInheritDefinitionFeature], ngContentSelectors: _c0, decls: 5, vars: 2, consts: [[1, "po-page-header"], ["class", "po-page-header-breadcrumb", 4, "ngIf"], [1, "po-page-header-container"], ["class", "po-page-header-title-container", 4, "ngIf"], [1, "po-page-header-breadcrumb"], [3, "p-favorite-service", "p-items", "p-params-service"], [1, "po-page-header-title-container"], ["class", "po-page-header-title", 4, "ngIf"], ["class", "po-page-header-subtitle", 4, "ngIf"], [1, "po-page-header-title"], [1, "po-page-header-subtitle"]], template: function PoPageHeaderComponent_Template(rf, ctx) { if (rf & 1) {
            i0.ɵɵprojectionDef();
            i0.ɵɵelementStart(0, "div", 0);
            i0.ɵɵtemplate(1, PoPageHeaderComponent_div_1_Template, 2, 3, "div", 1);
            i0.ɵɵelementStart(2, "div", 2);
            i0.ɵɵtemplate(3, PoPageHeaderComponent_div_3_Template, 3, 2, "div", 3);
            i0.ɵɵprojection(4);
            i0.ɵɵelementEnd()();
        } if (rf & 2) {
            i0.ɵɵadvance();
            i0.ɵɵproperty("ngIf", ctx.breadcrumb && ctx.breadcrumb.items.length);
            i0.ɵɵadvance(2);
            i0.ɵɵproperty("ngIf", ctx.title || ctx.subtitle);
        } }, dependencies: [i1.NgIf, i2.PoBreadcrumbComponent], encapsulation: 2 });
}
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(PoPageHeaderComponent, [{
        type: Component,
        args: [{ selector: 'po-page-header', template: "<div class=\"po-page-header\">\n  <!-- Breadcrumb -->\n  <div class=\"po-page-header-breadcrumb\" *ngIf=\"breadcrumb && breadcrumb.items.length\">\n    <po-breadcrumb\n      [p-favorite-service]=\"breadcrumb.favorite\"\n      [p-items]=\"breadcrumb.items\"\n      [p-params-service]=\"breadcrumb.params\"\n    >\n    </po-breadcrumb>\n  </div>\n\n  <div class=\"po-page-header-container\">\n    <!-- Titulo e SubTitulo-->\n    <div class=\"po-page-header-title-container\" *ngIf=\"title || subtitle\">\n      <h1 class=\"po-page-header-title\" *ngIf=\"title\">\n        {{ title }}\n      </h1>\n      <div class=\"po-page-header-subtitle\" *ngIf=\"subtitle\">\n        {{ subtitle }}\n      </div>\n    </div>\n\n    <!-- Opera\u00E7\u00F5es -->\n    <ng-content></ng-content>\n  </div>\n</div>\n" }]
    }], null, null); })();
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassDebugInfo(PoPageHeaderComponent, { className: "PoPageHeaderComponent", filePath: "lib/components/po-page/po-page-header/po-page-header.component.ts", lineNumber: 13 }); })();
