// @po-ui/ng-templates/esm2022/lib/components/po-page-blocked-user/po-page-blocked-user.component.mjs
import { Component } from '@angular/core';
import { poLocaleDefault } from '@po-ui/ng-components';
import { isExternalLink } from '../../utils/util';
import { PoPageBlockedUserBaseComponent } from './po-page-blocked-user-base.component';
import * as i0 from "@angular/core";
import * as i1 from "@angular/router";
import * as i2 from "@po-ui/ng-components";
import * as i3 from "../po-page-background/po-page-background.component";
import * as i4 from "./po-page-blocked-user-contacts/po-page-blocked-user-contacts.component";
import * as i5 from "./po-page-blocked-user-reason/po-page-blocked-user-reason.component";
export const poPageBlockedUserButtonLiterals = {
    en: {
        primaryButton: 'Back to home screen'
    },
    es: {
        primaryButton: 'Volver al inicio'
    },
    pt: {
        primaryButton: 'Voltar para o início'
    },
    ru: {
        primaryButton: 'Перейти к началу страницы'
    }
};
/**
 * @docsExtends PoPageBlockedUserBaseComponent
 *
 * @example
 *
 * <example name="po-page-blocked-user-basic" title="PO Page Blocked User Basic">
 *  <file name="sample-po-page-blocked-user-basic/sample-po-page-blocked-user-basic.component.html"> </file>
 *  <file name="sample-po-page-blocked-user-basic/sample-po-page-blocked-user-basic.component.ts"> </file>
 * </example>
 *
 * <example name="po-page-blocked-user-labs" title="PO Page Blocked User Labs">
 *  <file name="sample-po-page-blocked-user-labs/sample-po-page-blocked-user-labs.component.html"> </file>
 *  <file name="sample-po-page-blocked-user-labs/sample-po-page-blocked-user-labs.component.ts"> </file>
 * </example>
 *
 * <example name="po-page-blocked-user-exceeded-attempts" title="PO Page Blocked User Exceeded Attempts">
 *  <file name="sample-po-page-blocked-user-exceeded-attempts/sample-po-page-blocked-user-exceeded-attempts.component.html"> </file>
 *  <file name="sample-po-page-blocked-user-exceeded-attempts/sample-po-page-blocked-user-exceeded-attempts.component.ts"> </file>
 * </example>
 *
 * <example name="po-page-blocked-user-expired-password" title="PO Page Blocked User Expired Password">
 *  <file name="sample-po-page-blocked-user-expired-password/sample-po-page-blocked-user-expired-password.component.html"> </file>
 *  <file name="sample-po-page-blocked-user-expired-password/sample-po-page-blocked-user-expired-password.component.ts"> </file>
 * </example>
 */
export class PoPageBlockedUserComponent extends PoPageBlockedUserBaseComponent {
    activatedRoute;
    router;
    literals;
    constructor(activatedRoute, router, languageService) {
        super();
        this.activatedRoute = activatedRoute;
        this.router = router;
        const language = languageService.getShortLanguage();
        this.literals = {
            ...poPageBlockedUserButtonLiterals[poLocaleDefault],
            ...poPageBlockedUserButtonLiterals[language]
        };
    }
    ngOnInit() {
        this.checkingForRouteMetadata(this.activatedRoute.snapshot.data);
    }
    navigateTo(url) {
        isExternalLink(url) ? window.open(url) : this.router.navigate([url || '/']);
    }
    checkingForMetadataProperty(object, property) {
        if (Object.prototype.hasOwnProperty.call(object, property)) {
            return object[property];
        }
    }
    checkingForRouteMetadata(data) {
        if (Object.keys(data).length !== 0) {
            this.contactEmail = this.checkingForMetadataProperty(data, 'contactEmail') || this.contactEmail;
            this.contactPhone = this.checkingForMetadataProperty(data, 'contactPhone') || this.contactPhone;
            this.reason = this.checkingForMetadataProperty(data, 'reason') || this.reason;
            this.urlBack = this.checkingForMetadataProperty(data, 'urlBack') || this.urlBack;
        }
    }
    static ɵfac = function PoPageBlockedUserComponent_Factory(t) { return new (t || PoPageBlockedUserComponent)(i0.ɵɵdirectiveInject(i1.ActivatedRoute), i0.ɵɵdirectiveInject(i1.Router), i0.ɵɵdirectiveInject(i2.PoLanguageService)); };
    static ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: PoPageBlockedUserComponent, selectors: [["po-page-blocked-user"]], features: [i0.ɵɵInheritDefinitionFeature], decls: 8, vars: 8, consts: [[1, "po-page-blocked-user-wrapper"], [1, "po-page-background-wrap"], [3, "p-hide-logo", "p-logo", "p-secondary-logo"], [3, "p-params", "p-reason"], [1, "po-page-blocked-user-mobile-bottom-alignment"], [3, "p-email", "p-phone"], [1, "po-row"], ["p-kind", "primary", 1, "po-lg-12", "po-sm-12", "po-page-blocked-user-button", 3, "p-label", "p-click"]], template: function PoPageBlockedUserComponent_Template(rf, ctx) { if (rf & 1) {
            i0.ɵɵelementStart(0, "div", 0)(1, "div", 1)(2, "po-page-background", 2);
            i0.ɵɵelement(3, "po-page-blocked-user-reason", 3);
            i0.ɵɵelementStart(4, "div", 4);
            i0.ɵɵelement(5, "po-page-blocked-user-contacts", 5);
            i0.ɵɵelementStart(6, "div", 6)(7, "po-button", 7);
            i0.ɵɵlistener("p-click", function PoPageBlockedUserComponent_Template_po_button_p_click_7_listener() { return ctx.navigateTo(ctx.urlBack); });
            i0.ɵɵelementEnd()()()()()();
        } if (rf & 2) {
            i0.ɵɵadvance(2);
            i0.ɵɵproperty("p-hide-logo", true)("p-logo", ctx.logo)("p-secondary-logo", ctx.secondaryLogo);
            i0.ɵɵadvance();
            i0.ɵɵproperty("p-params", ctx.params)("p-reason", ctx.reason);
            i0.ɵɵadvance(2);
            i0.ɵɵproperty("p-email", ctx.contactEmail)("p-phone", ctx.contactPhone);
            i0.ɵɵadvance(2);
            i0.ɵɵproperty("p-label", ctx.literals == null ? null : ctx.literals.primaryButton);
        } }, dependencies: [i2.PoButtonComponent, i3.PoPageBackgroundComponent, i4.PoPageBlockedUserContactsComponent, i5.PoPageBlockedUserReasonComponent], encapsulation: 2 });
}
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(PoPageBlockedUserComponent, [{
        type: Component,
        args: [{ selector: 'po-page-blocked-user', template: "<div class=\"po-page-blocked-user-wrapper\">\n  <div class=\"po-page-background-wrap\">\n    <po-page-background [p-hide-logo]=\"true\" [p-logo]=\"logo\" [p-secondary-logo]=\"secondaryLogo\">\n      <po-page-blocked-user-reason [p-params]=\"params\" [p-reason]=\"reason\"> </po-page-blocked-user-reason>\n\n      <div class=\"po-page-blocked-user-mobile-bottom-alignment\">\n        <po-page-blocked-user-contacts [p-email]=\"contactEmail\" [p-phone]=\"contactPhone\">\n        </po-page-blocked-user-contacts>\n\n        <div class=\"po-row\">\n          <po-button\n            class=\"po-lg-12 po-sm-12 po-page-blocked-user-button\"\n            p-kind=\"primary\"\n            [p-label]=\"literals?.primaryButton\"\n            (p-click)=\"navigateTo(urlBack)\"\n          >\n          </po-button>\n        </div>\n      </div>\n    </po-page-background>\n  </div>\n</div>\n" }]
    }], () => [{ type: i1.ActivatedRoute }, { type: i1.Router }, { type: i2.PoLanguageService }], null); })();
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassDebugInfo(PoPageBlockedUserComponent, { className: "PoPageBlockedUserComponent", filePath: "lib/components/po-page-blocked-user/po-page-blocked-user.component.ts", lineNumber: 55 }); })();
