// @po-ui/ng-components/esm2022/lib/components/po-field/po-clean/po-clean.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { PoCleanComponent } from './po-clean.component';
import { PoIconModule } from '../../po-icon';
import * as i0 from "@angular/core";
/**
 * @description
 *
 * Módulo do componente `po-clean`.
 */
export class PoCleanModule {
    static ɵfac = function PoCleanModule_Factory(t) { return new (t || PoCleanModule)(); };
    static ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: PoCleanModule });
    static ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [CommonModule, PoIconModule] });
}
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(PoCleanModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, PoIconModule],
                exports: [PoCleanComponent],
                declarations: [PoCleanComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(PoCleanModule, { declarations: [PoCleanComponent], imports: [CommonModule, PoIconModule], exports: [PoCleanComponent] }); })();
