// @po-ui/ng-components/esm2022/lib/components/po-field/po-clean/po-clean.component.mjs
import { Component } from '@angular/core';
import { PoCleanBaseComponent } from './po-clean-base.component';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
import * as i2 from "../../po-icon/po-icon.component";
function PoCleanComponent_po_icon_0_Template(rf, ctx) { if (rf & 1) {
    const _r2 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "po-icon", 1);
    i0.ɵɵlistener("click", function PoCleanComponent_po_icon_0_Template_po_icon_click_0_listener() { i0.ɵɵrestoreView(_r2); const ctx_r1 = i0.ɵɵnextContext(); return i0.ɵɵresetView(ctx_r1.clear()); });
    i0.ɵɵelementEnd();
} }
/**
 * @docsPrivate
 *
 * @docsExtends PoCleanBaseComponent
 *
 * @examplePrivate
 *
 * <example-private name="po-clean-labs" title="PO Clean Labs">
 *   <file name="sample-po-clean-labs.component.html"> </file>
 *   <file name="sample-po-clean-labs.component.ts"> </file>
 * </example-private>
 */
export class PoCleanComponent extends PoCleanBaseComponent {
    setInputValue(value) {
        if (this.inputRef && this.inputRef.nativeElement) {
            this.inputRef.nativeElement.value = value;
        }
    }
    getInputValue() {
        if (this.inputRef && this.inputRef.nativeElement) {
            return this.inputRef.nativeElement.value;
        }
    }
    static ɵfac = /*@__PURE__*/ (() => { let ɵPoCleanComponent_BaseFactory; return function PoCleanComponent_Factory(t) { return (ɵPoCleanComponent_BaseFactory || (ɵPoCleanComponent_BaseFactory = i0.ɵɵgetInheritedFactory(PoCleanComponent)))(t || PoCleanComponent); }; })();
    static ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: PoCleanComponent, selectors: [["po-clean"]], features: [i0.ɵɵInheritDefinitionFeature], decls: 1, vars: 1, consts: [["p-icon", "ICON_CLEAR_CONTENT po-field-icon", 3, "click", 4, "ngIf"], ["p-icon", "ICON_CLEAR_CONTENT po-field-icon", 3, "click"]], template: function PoCleanComponent_Template(rf, ctx) { if (rf & 1) {
            i0.ɵɵtemplate(0, PoCleanComponent_po_icon_0_Template, 1, 0, "po-icon", 0);
        } if (rf & 2) {
            i0.ɵɵproperty("ngIf", ctx.showIcon());
        } }, dependencies: [i1.NgIf, i2.PoIconComponent], encapsulation: 2 });
}
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(PoCleanComponent, [{
        type: Component,
        args: [{ selector: 'po-clean', template: "<po-icon p-icon=\"ICON_CLEAR_CONTENT po-field-icon\" *ngIf=\"showIcon()\" (click)=\"clear()\"> </po-icon>\n" }]
    }], null, null); })();
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassDebugInfo(PoCleanComponent, { className: "PoCleanComponent", filePath: "lib/components/po-field/po-clean/po-clean.component.ts", lineNumber: 21 }); })();
