// @po-ui/ng-components/esm2022/lib/services/po-theme/po-theme.module.mjs
import { NgModule } from '@angular/core';
import { PoThemeService } from './po-theme.service';
import * as i0 from "@angular/core";
/**
 * Módulo responsável por fornecer serviços relacionados ao tema PO.
 */
export class PoThemeModule {
    static ɵfac = function PoThemeModule_Factory(t) { return new (t || PoThemeModule)(); };
    static ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: PoThemeModule });
    static ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ providers: [
            PoThemeService,
            // Fornecer o objeto 'Window' para uso no serviço
            { provide: 'Window', useValue: window }
        ] });
}
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(PoThemeModule, [{
        type: NgModule,
        args: [{
                providers: [
                    PoThemeService,
                    // Fornecer o objeto 'Window' para uso no serviço
                    { provide: 'Window', useValue: window }
                ],
                bootstrap: []
            }]
    }], null, null); })();
