// @po-ui/ng-components/esm2022/lib/components/po-widget/po-widget.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { PoIconModule } from '../po-icon';
import { PoContainerModule } from '../po-container/index';
import { PoWidgetComponent } from './po-widget.component';
import * as i0 from "@angular/core";
/**
 * @description
 *
 * Módulo do componente po-widget
 */
export class PoWidgetModule {
    static ɵfac = function PoWidgetModule_Factory(t) { return new (t || PoWidgetModule)(); };
    static ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: PoWidgetModule });
    static ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [CommonModule, PoContainerModule, PoIconModule] });
}
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(PoWidgetModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, PoContainerModule, PoIconModule],
                exports: [PoWidgetComponent],
                declarations: [PoWidgetComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(PoWidgetModule, { declarations: [PoWidgetComponent], imports: [CommonModule, PoContainerModule, PoIconModule], exports: [PoWidgetComponent] }); })();
