// @totvs/protheus-lib-core/esm2022/services/pro-authorization/services/pro-user-info/pro-user-info.service.mjs
import { HttpHeaders } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { map, tap } from 'rxjs/operators';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common/http";
import * as i2 from "../../../pro-js-to-advpl";
const CACHE_KEY_USER = 'ProUser';
const USER_ENDPOINT = '/api/framework/v1/users';
/**
 * Essa classe corresponde ao serviço de busca de informações do usuário
 */
export class ProUserInfoService {
    constructor(http, advplService) {
        this.http = http;
        this.advplService = advplService;
        this.useHTTP = true;
        this.EVENT_GET_ID = 'getUserInfo';
        this.EVENT_SET_ID = 'setUserInfo';
        this.URL_PSWRET = '/api/framework/v1/basicProtheusServices/pswret/';
    }
    /**
     * @description Retorna informações do usuário
     * @param userId string, id do usuário
     * @param isUseHttp boolean, indica se utiliza requisição http
     * @returns Observable de ProUserInfo, dados do usuário
     */
    get(userId, isUseHttp = false) {
        if (this.useHTTP || isUseHttp) {
            return this.getFromApi(userId);
        }
        return this.getFromAdvpl(userId);
    }
    getFromApi(userId) {
        const headers = new HttpHeaders()
            .append('X-Totvs-No-Error', 'true')
            .append('Accept', 'application/json; charset=utf-8');
        return this.http.get(`${USER_ENDPOINT}/${userId}`, { headers }).pipe(map((userInfo) => {
            userInfo.complete_name = userInfo.userName;
            return userInfo;
        }), tap((userInfo) => {
            sessionStorage[CACHE_KEY_USER] = JSON.stringify(userInfo);
        }));
    }
    getFromAdvpl(userId) {
        if (!this.advplService.protheusConnected()) {
            return this.advplNotPrepared();
        }
        const randomEventId = this.advplService.generateEventId();
        const observableParams = {
            sendInfo: {
                type: `${this.EVENT_GET_ID}#id#${randomEventId}`,
                content: userId
            },
            autoDestruct: true,
            receiveId: `${this.EVENT_SET_ID}#id#${randomEventId}`
        };
        const observableCallback = ({ protheusResponse, subscriber }) => {
            if (protheusResponse.length === 0) {
                subscriber.error({
                    status: 400,
                    description: `user ${userId} could not be found`
                });
            }
            else {
                const userInfo = JSON.parse(protheusResponse);
                userInfo.complete_name = userInfo.userName;
                sessionStorage[CACHE_KEY_USER] = JSON.stringify(protheusResponse);
                subscriber.next(userInfo);
            }
            subscriber.complete();
        };
        return this.advplService.buildObservable(observableCallback, observableParams);
    }
    removeFromStorage() {
        sessionStorage.removeItem(CACHE_KEY_USER);
    }
    advplNotPrepared() {
        return new Observable(subscriber => {
            subscriber.error({
                status: 400,
                description: 'advplService not prepared in ProUserInfoService'
                /**
                 * Esse erro está sendo emitido ao utilizar o app com a otimização ligada e a opção de SSO.
                 * O que foi percebido é que em determinada situação a conexão AdvPL
                 * não fica disponível quando o serviço é solicitado. Foram feitos testes adicionando
                 * múltiplas tentativas mas sem sucesso. Utilizando a opção de canal HTTP funciona corretamente,
                 * porém interfere na criação de dicionários do lado AdvPL, o que torna a opção inviável.
                 */
            });
            subscriber.complete();
        });
    }
    isChannelHTTP() {
        return this.useHTTP;
    }
    setChannelAsHTTP(value) {
        this.useHTTP = value;
    }
    /**
     * efetua um get na api referente a função PswRet do protheus
     * a api utiliza o usuário logado para efetuar a busca
     * @returns Observable com o retorno da api
     */
    pswRet() {
        return this.http.get(this.URL_PSWRET);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProUserInfoService, deps: [{ token: i1.HttpClient }, { token: i2.ProJsToAdvplService }], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProUserInfoService, providedIn: 'root' }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProUserInfoService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root'
                }]
        }], ctorParameters: () => [{ type: i1.HttpClient }, { type: i2.ProJsToAdvplService }] });
