// @po-ui/ng-components/esm2022/lib/components/po-field/po-upload/po-upload-file.mjs
import { PoUploadStatus } from './po-upload-status.enum';
// Classe responsável pelo arquivo do PO Upload
export class PoUploadFile {
    // Nome do arquivo.
    name;
    // Arquivo bruto.
    rawFile;
    // Status de envio do arquivo.
    status;
    // Identificador do arquivo.
    uid;
    // Extensão do arquivo.
    extension;
    // Tamanho do arquivo em bytes;
    size;
    // propriedade para auxiliar a exibição do texto no componente progress
    displayName;
    // porcentagem utilizada para repassar ao componente progress
    percent;
    constructor(file) {
        if (file) {
            this.name = file.name;
            this.displayName = `${file.name} - ${this.getFileSize(file.size)}`;
            this.extension = this.getExtension(file.name);
            this.size = file.size;
            this.rawFile = file;
            this.uid = this.generateUUID();
            this.status = PoUploadStatus.None;
        }
    }
    getExtension(value) {
        if (value) {
            const extension = value.substr(value.lastIndexOf('.'));
            return extension.toLowerCase();
        }
        return '';
    }
    generateUUID() {
        function s4() {
            return Math.floor((1 + Math.random()) * 0x10000)
                .toString(16)
                .substring(1);
        }
        return s4() + s4() + '-' + s4() + '-' + s4() + '-' + s4() + '-' + s4() + s4() + s4();
    }
    // Retorna o tamanho do arquivo em KBytes.
    getFileSize(size) {
        let kbSize = 0;
        if (size) {
            kbSize = Math.ceil(size / 1024);
        }
        return `${kbSize} KB`;
    }
}
