// @po-ui/ng-templates/esm2022/lib/components/po-page-dynamic-edit/po-page-dynamic-edit.module.mjs
import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { PoButtonModule, PoDividerModule, PoDynamicModule, PoGridModule, PoIconModule, PoPageModule, PoWidgetModule } from '@po-ui/ng-components';
import { PoPageDynamicEditComponent } from './po-page-dynamic-edit.component';
import { PoPageDynamicModule } from '../../services/po-page-dynamic/po-page-dynamic.module';
import * as i0 from "@angular/core";
/**
 * @description
 *
 * Módulo do template do po-page-dynamic-detail.
 */
export class PoPageDynamicEditModule {
    static ɵfac = function PoPageDynamicEditModule_Factory(t) { return new (t || PoPageDynamicEditModule)(); };
    static ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: PoPageDynamicEditModule });
    static ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [CommonModule,
            FormsModule,
            RouterModule,
            PoButtonModule,
            PoDividerModule,
            PoDynamicModule,
            PoGridModule,
            PoIconModule,
            PoPageModule,
            PoWidgetModule,
            PoPageDynamicModule] });
}
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(PoPageDynamicEditModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    FormsModule,
                    RouterModule,
                    PoButtonModule,
                    PoDividerModule,
                    PoDynamicModule,
                    PoGridModule,
                    PoIconModule,
                    PoPageModule,
                    PoWidgetModule,
                    PoPageDynamicModule
                ],
                declarations: [PoPageDynamicEditComponent],
                exports: [PoPageDynamicEditComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(PoPageDynamicEditModule, { declarations: [PoPageDynamicEditComponent], imports: [CommonModule,
        FormsModule,
        RouterModule,
        PoButtonModule,
        PoDividerModule,
        PoDynamicModule,
        PoGridModule,
        PoIconModule,
        PoPageModule,
        PoWidgetModule,
        PoPageDynamicModule], exports: [PoPageDynamicEditComponent] }); })();
