// @po-ui/ng-components/esm2022/lib/components/po-search/po-search.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { PoCleanModule } from '../po-field';
import { PoIconModule } from '../po-icon';
import { PoLoadingModule } from '../po-loading';
import { PoSearchComponent } from './po-search.component';
import { FormsModule } from '@angular/forms';
import { PoAccordionModule } from '../po-accordion/po-accordion.module';
import { PoListBoxModule } from '../po-listbox';
import { PoDropdownModule } from '../po-dropdown';
import * as i0 from "@angular/core";
/**
 * @description
 *
 * Módulo do componente po-search.
 */
export class PoSearchModule {
    static ɵfac = function PoSearchModule_Factory(t) { return new (t || PoSearchModule)(); };
    static ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: PoSearchModule });
    static ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [CommonModule,
            PoCleanModule,
            PoIconModule,
            PoLoadingModule,
            PoAccordionModule,
            FormsModule,
            PoListBoxModule,
            PoDropdownModule] });
}
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(PoSearchModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    PoCleanModule,
                    PoIconModule,
                    PoLoadingModule,
                    PoAccordionModule,
                    FormsModule,
                    PoListBoxModule,
                    PoDropdownModule
                ],
                declarations: [PoSearchComponent],
                exports: [PoSearchComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(PoSearchModule, { declarations: [PoSearchComponent], imports: [CommonModule,
        PoCleanModule,
        PoIconModule,
        PoLoadingModule,
        PoAccordionModule,
        FormsModule,
        PoListBoxModule,
        PoDropdownModule], exports: [PoSearchComponent] }); })();
