// @po-ui/ng-components/esm2022/lib/components/po-tabs/po-tab-dropdown/po-tab-dropdown.component.mjs
import { Component, EventEmitter, Input, Output, ViewChild } from '@angular/core';
import { PoButtonComponent } from '../../po-button/po-button.component';
import * as i0 from "@angular/core";
import * as i1 from "../../po-popover/po-popover.component";
import * as i2 from "../../po-button/po-button.component";
import * as i3 from "../../po-listbox/po-listbox.component";
const _c0 = ["popover"];
/**
 * @docsPrivate
 *
 * @description
 *
 * Componente para agrupamento de componentes `po-tab-button` que os rendereiza em uma um caixa de diálogo
 * no formato de lista.
 */
export class PoTabDropdownComponent {
    popover;
    button;
    // Rótulo do `po-tab-button`
    label;
    // Diminui o tamanho do botão
    small;
    // Lista de abas
    tabs = [];
    // Evento que será emitido ao ativar uma aba
    activated = new EventEmitter();
    // Evento que será emitido a aba for desabilitada ou ocultada
    changeState = new EventEmitter();
    // Evento de click
    click = new EventEmitter();
    closeAndReturnToButtom() {
        this.popover.close();
        this.button.focus();
    }
    get buttonElement() {
        return this.button.buttonElement;
    }
    static ɵfac = function PoTabDropdownComponent_Factory(t) { return new (t || PoTabDropdownComponent)(); };
    static ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: PoTabDropdownComponent, selectors: [["po-tab-dropdown"]], viewQuery: function PoTabDropdownComponent_Query(rf, ctx) { if (rf & 1) {
            i0.ɵɵviewQuery(_c0, 7);
            i0.ɵɵviewQuery(PoButtonComponent, 7);
        } if (rf & 2) {
            let _t;
            i0.ɵɵqueryRefresh(_t = i0.ɵɵloadQuery()) && (ctx.popover = _t.first);
            i0.ɵɵqueryRefresh(_t = i0.ɵɵloadQuery()) && (ctx.button = _t.first);
        } }, inputs: { label: [i0.ɵɵInputFlags.None, "p-label", "label"], small: [i0.ɵɵInputFlags.None, "p-small", "small"], tabs: [i0.ɵɵInputFlags.None, "p-tabs", "tabs"] }, outputs: { activated: "p-activated", changeState: "p-change-state", click: "p-click" }, decls: 9, vars: 4, consts: [[1, "po-tab-dropdown-content", "po-tab-button-default"], ["p-kind", "tertiary", "p-icon", "ICON_ARROW_DOWN", 3, "p-aria-label"], ["button", ""], ["p-hide-arrow", "", "p-position", "bottom", 3, "p-target"], ["popover", ""], [1, "po-tab-dropdown-container"], ["p-type", "action", 3, "p-items", "p-is-tabs", "p-activated-tabs", "p-change-state-tabs", "p-click-tabs", "p-close"], ["listbox", "", "poListBoxRef", ""]], template: function PoTabDropdownComponent_Template(rf, ctx) { if (rf & 1) {
            const _r4 = i0.ɵɵgetCurrentView();
            i0.ɵɵelementStart(0, "div", 0);
            i0.ɵɵelement(1, "po-button", 1, 2);
            i0.ɵɵelementEnd();
            i0.ɵɵelementStart(3, "po-popover", 3, 4)(5, "div", 5)(6, "po-listbox", 6, 7);
            i0.ɵɵlistener("p-activated-tabs", function PoTabDropdownComponent_Template_po_listbox_p_activated_tabs_6_listener($event) { return ctx.activated.emit($event); })("p-change-state-tabs", function PoTabDropdownComponent_Template_po_listbox_p_change_state_tabs_6_listener($event) { return ctx.changeState.emit($event); })("p-click-tabs", function PoTabDropdownComponent_Template_po_listbox_p_click_tabs_6_listener($event) { i0.ɵɵrestoreView(_r4); const _r1 = i0.ɵɵreference(4); _r1.close(); return i0.ɵɵresetView(ctx.click.emit($event)); })("p-close", function PoTabDropdownComponent_Template_po_listbox_p_close_6_listener() { return ctx.closeAndReturnToButtom(); });
            i0.ɵɵelementEnd()()();
        } if (rf & 2) {
            i0.ɵɵadvance();
            i0.ɵɵproperty("p-aria-label", ctx.label);
            i0.ɵɵadvance(2);
            i0.ɵɵproperty("p-target", ctx.buttonElement);
            i0.ɵɵadvance(3);
            i0.ɵɵproperty("p-items", ctx.tabs)("p-is-tabs", true);
        } }, dependencies: [i1.PoPopoverComponent, i2.PoButtonComponent, i3.PoListBoxComponent], encapsulation: 2 });
}
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(PoTabDropdownComponent, [{
        type: Component,
        args: [{ selector: 'po-tab-dropdown', template: "<div class=\"po-tab-dropdown-content po-tab-button-default\">\n  <po-button #button p-kind=\"tertiary\" [p-aria-label]=\"label\" p-icon=\"ICON_ARROW_DOWN\"> </po-button>\n</div>\n\n<po-popover #popover p-hide-arrow p-position=\"bottom\" [p-target]=\"buttonElement\">\n  <div class=\"po-tab-dropdown-container\">\n    <po-listbox\n      #listbox\n      #poListBoxRef\n      p-type=\"action\"\n      [p-items]=\"tabs\"\n      [p-is-tabs]=\"true\"\n      (p-activated-tabs)=\"activated.emit($event)\"\n      (p-change-state-tabs)=\"changeState.emit($event)\"\n      (p-click-tabs)=\"popover.close(); click.emit($event)\"\n      (p-close)=\"closeAndReturnToButtom()\"\n    >\n    </po-listbox>\n  </div>\n</po-popover>\n" }]
    }], null, { popover: [{
            type: ViewChild,
            args: ['popover', { static: true }]
        }], button: [{
            type: ViewChild,
            args: [PoButtonComponent, { static: true }]
        }], label: [{
            type: Input,
            args: ['p-label']
        }], small: [{
            type: Input,
            args: ['p-small']
        }], tabs: [{
            type: Input,
            args: ['p-tabs']
        }], activated: [{
            type: Output,
            args: ['p-activated']
        }], changeState: [{
            type: Output,
            args: ['p-change-state']
        }], click: [{
            type: Output,
            args: ['p-click']
        }] }); })();
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassDebugInfo(PoTabDropdownComponent, { className: "PoTabDropdownComponent", filePath: "lib/components/po-tabs/po-tab-dropdown/po-tab-dropdown.component.ts", lineNumber: 19 }); })();
