// @po-ui/ng-components/esm2022/lib/components/po-field/po-field-validate.model.mjs
import { Directive, Input } from '@angular/core';
import { convertToBoolean } from '../..//utils/util';
import { PoFieldModel } from './po-field.model';
import { requiredFailed } from './validators';
import * as i0 from "@angular/core";
/**
 * @docsExtends PoFieldModel
 */
export class PoFieldValidateModel extends PoFieldModel {
    /**
     * @optional
     *
     * @description
     *
     * Define se a indicação de campo opcional será exibida.
     *
     * > Não será exibida a indicação se:
     * - O campo conter `p-required`;
     * - Não possuir `p-help` e/ou `p-label`.
     *
     * @default `false`
     */
    optional = false;
    /**
     * @optional
     *
     * @description
     *
     * Define que o campo será obrigatório.
     *
     * @default `false`
     */
    required = false;
    /**
     *  Define se a indicação de campo obrigatório será exibida.
     *
     * > Não será exibida a indicação se:
     * - Não possuir `p-help` e/ou `p-label`.
     */
    showRequired = false;
    onValidatorChange;
    validate(abstractControl) {
        if (requiredFailed(this.required, this.disabled, abstractControl.value)) {
            return {
                required: {
                    valid: false
                }
            };
        }
        return this.extraValidation(abstractControl);
    }
    registerOnValidatorChange(fn) {
        this.onValidatorChange = fn;
    }
    validateModel() {
        if (this.onValidatorChange) {
            this.onValidatorChange();
        }
    }
    static ɵfac = /*@__PURE__*/ (() => { let ɵPoFieldValidateModel_BaseFactory; return function PoFieldValidateModel_Factory(t) { return (ɵPoFieldValidateModel_BaseFactory || (ɵPoFieldValidateModel_BaseFactory = i0.ɵɵgetInheritedFactory(PoFieldValidateModel)))(t || PoFieldValidateModel); }; })();
    static ɵdir = /*@__PURE__*/ i0.ɵɵdefineDirective({ type: PoFieldValidateModel, inputs: { optional: [i0.ɵɵInputFlags.HasDecoratorInputTransform, "p-optional", "optional", convertToBoolean], required: [i0.ɵɵInputFlags.HasDecoratorInputTransform, "p-required", "required", convertToBoolean], showRequired: [i0.ɵɵInputFlags.None, "p-show-required", "showRequired"] }, features: [i0.ɵɵInputTransformsFeature, i0.ɵɵInheritDefinitionFeature] });
}
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(PoFieldValidateModel, [{
        type: Directive
    }], null, { optional: [{
            type: Input,
            args: [{ alias: 'p-optional', transform: convertToBoolean }]
        }], required: [{
            type: Input,
            args: [{ alias: 'p-required', transform: convertToBoolean }]
        }], showRequired: [{
            type: Input,
            args: ['p-show-required']
        }] }); })();
