// @po-ui/ng-components/esm2022/lib/components/po-field/po-field.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';
import { OverlayModule } from '@angular/cdk/overlay';
import { PoButtonGroupModule } from '../po-button-group/index';
import { PoButtonModule } from '../po-button/index';
import { PoCheckboxGroupModule } from './po-checkbox-group/po-checkbox-group.module';
import { PoRadioGroupModule } from './po-radio-group/po-radio-group.module';
import { PoRadioModule } from './po-radio/po-radio.module';
import { PoContainerModule } from '../po-container/index';
import { PoCalendarModule } from '../po-calendar/po-calendar.module';
import { PoCleanModule } from './po-clean/po-clean.module';
import { PoDatepickerModule } from './po-datepicker/po-datepicker.module';
import { PoDisclaimerGroupModule } from './../po-disclaimer-group/po-disclaimer-group.module';
import { PoDisclaimerModule } from './../po-disclaimer/po-disclaimer.module';
import { PoFieldContainerModule } from './po-field-container/po-field-container.module';
import { PoLoadingModule } from '../po-loading/index';
import { PoModalModule } from '../po-modal/po-modal.module';
import { PoProgressModule } from './../po-progress/po-progress.module';
import { PoServicesModule } from '../../services/services.module';
import { PoTableModule } from '../po-table/po-table.module';
import { PoTooltipModule } from './../../directives/po-tooltip/po-tooltip.module';
import { PoIconModule } from '../po-icon/po-icon.module';
import { PoListBoxModule } from '../po-listbox/po-listbox.module';
import { PoComboComponent } from './po-combo/po-combo.component';
import { PoComboOptionTemplateDirective } from './po-combo/po-combo-option-template/po-combo-option-template.directive';
import { PoMultiselectOptionTemplateDirective } from './po-multiselect/po-multiselect-option-template/po-multiselect-option-template.directive';
import { PoDatepickerRangeComponent } from './po-datepicker-range/po-datepicker-range.component';
import { PoDecimalComponent } from './po-decimal/po-decimal.component';
import { PoEmailComponent } from './po-email/po-email.component';
import { PoLoginComponent } from './po-login/po-login.component';
import { PoLookupComponent } from './po-lookup/po-lookup.component';
import { PoLookupModalComponent } from './po-lookup/po-lookup-modal/po-lookup-modal.component';
import { PoMultiselectDropdownComponent } from './po-multiselect/po-multiselect-dropdown/po-multiselect-dropdown.component';
import { PoMultiselectComponent } from './po-multiselect/po-multiselect.component';
import { PoRichTextBodyComponent } from './po-rich-text/po-rich-text-body/po-rich-text-body.component';
import { PoRichTextComponent } from './po-rich-text/po-rich-text.component';
import { PoRichTextImageModalComponent } from './po-rich-text/po-rich-text-image-modal/po-rich-text-image-modal.component';
import { PoRichTextLinkModalComponent } from './po-rich-text/po-rich-text-link-modal/po-rich-text-link-modal.component';
import { PoRichTextToolbarComponent } from './po-rich-text/po-rich-text-toolbar/po-rich-text-toolbar.component';
import { PoInputComponent } from './po-input/po-input.component';
import { PoNumberComponent } from './po-number/po-number.component';
import { PoPasswordComponent } from './po-password/po-password.component';
import { PoSelectComponent } from './po-select/po-select.component';
import { PoTextareaComponent } from './po-textarea/po-textarea.component';
import { PoUploadComponent } from './po-upload/po-upload.component';
import { PoUploadDragDropComponent } from './po-upload/po-upload-drag-drop/po-upload-drag-drop.component';
import { PoUploadDragDropDirective } from './po-upload/po-upload-drag-drop/po-upload-drag-drop.directive';
import { PoUploadDragDropAreaOverlayComponent } from './po-upload/po-upload-drag-drop/po-upload-drag-drop-area-overlay/po-upload-drag-drop-area-overlay.component';
import { PoUploadDragDropAreaComponent } from './po-upload/po-upload-drag-drop/po-upload-drag-drop-area/po-upload-drag-drop-area.component';
import { PoUploadFileRestrictionsComponent } from './po-upload/po-upload-file-restrictions/po-upload-file-restrictions.component';
import { PoUrlComponent } from './po-url/po-url.component';
import { PoCheckboxModule } from './po-checkbox/po-checkbox.module';
import { PoSwitchModule } from './po-switch/po-switch.module';
import { PoLabelModule } from '../po-label';
import { PoTagModule } from '../po-tag';
import * as i0 from "@angular/core";
import * as i1 from "@angular/forms";
import * as i2 from "@angular/common";
import * as i3 from "../po-disclaimer-group/po-disclaimer-group.component";
import * as i4 from "./po-field-container/po-field-container.component";
import * as i5 from "../po-modal/po-modal.component";
import * as i6 from "../po-table/po-table.component";
import * as i7 from "../po-icon/po-icon.component";
import * as i8 from "@angular/cdk/overlay";
import * as i9 from "./po-field-container/po-field-container-bottom/po-field-container-bottom.component";
import * as i10 from "../po-tag/po-tag.component";
/**
 * @description
 *
 * Módulo dos componentes: po-checkbox, po-checkbox-group, po-combo, po-datepicker, po-datepicker-range, po-email, po-input,
 * po-lookup, po-number, po-multiselect, po-password, po-radio-group, po-select, po-switch, po-textarea, po-upload
 * e po-url.
 *
 * > Não esqueça de importar o módulo `FormsModule` para usar os componentes de formulários e caso esteja trabalhando com
 * > formulários reativos, importe o módulo `ReactiveFormsModule`, ambos nativos do Angular.
 */
export class PoFieldModule {
    static ɵfac = function PoFieldModule_Factory(t) { return new (t || PoFieldModule)(); };
    static ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: PoFieldModule });
    static ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [CommonModule,
            FormsModule,
            FormsModule.withConfig({
                callSetDisabledState: 'whenDisabledForLegacyCode'
            }),
            OverlayModule,
            PoButtonGroupModule,
            PoButtonModule,
            PoCleanModule,
            PoCalendarModule,
            PoCheckboxGroupModule,
            PoRadioGroupModule,
            PoContainerModule,
            PoDatepickerModule,
            PoDisclaimerGroupModule,
            PoDisclaimerModule,
            PoFieldContainerModule,
            PoLoadingModule,
            PoModalModule,
            PoProgressModule,
            PoServicesModule,
            PoTableModule,
            PoTooltipModule,
            PoIconModule,
            PoCheckboxModule,
            PoRadioModule,
            PoLabelModule,
            PoListBoxModule,
            PoSwitchModule,
            PoTagModule, PoCheckboxGroupModule,
            PoRadioGroupModule,
            PoCleanModule,
            PoDatepickerModule,
            PoFieldContainerModule,
            PoCheckboxModule,
            PoRadioModule,
            PoLabelModule,
            PoSwitchModule,
            PoTagModule] });
}
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(PoFieldModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    FormsModule,
                    FormsModule.withConfig({
                        callSetDisabledState: 'whenDisabledForLegacyCode'
                    }),
                    OverlayModule,
                    PoButtonGroupModule,
                    PoButtonModule,
                    PoCleanModule,
                    PoCalendarModule,
                    PoCheckboxGroupModule,
                    PoRadioGroupModule,
                    PoContainerModule,
                    PoDatepickerModule,
                    PoDisclaimerGroupModule,
                    PoDisclaimerModule,
                    PoFieldContainerModule,
                    PoLoadingModule,
                    PoModalModule,
                    PoProgressModule,
                    PoServicesModule,
                    PoTableModule,
                    PoTooltipModule,
                    PoIconModule,
                    PoCheckboxModule,
                    PoRadioModule,
                    PoLabelModule,
                    PoListBoxModule,
                    PoSwitchModule,
                    PoTagModule
                ],
                exports: [
                    PoCheckboxGroupModule,
                    PoRadioGroupModule,
                    PoCleanModule,
                    PoDatepickerModule,
                    PoComboComponent,
                    PoComboOptionTemplateDirective,
                    PoMultiselectOptionTemplateDirective,
                    PoDecimalComponent,
                    PoDatepickerRangeComponent,
                    PoEmailComponent,
                    PoFieldContainerModule,
                    PoInputComponent,
                    PoLoginComponent,
                    PoLookupComponent,
                    PoLookupModalComponent,
                    PoMultiselectComponent,
                    PoNumberComponent,
                    PoPasswordComponent,
                    PoRichTextComponent,
                    PoSelectComponent,
                    PoTextareaComponent,
                    PoUploadComponent,
                    PoUrlComponent,
                    PoCheckboxModule,
                    PoRadioModule,
                    PoLabelModule,
                    PoSwitchModule,
                    PoTagModule
                ],
                declarations: [
                    PoComboComponent,
                    PoComboOptionTemplateDirective,
                    PoMultiselectOptionTemplateDirective,
                    PoDecimalComponent,
                    PoDatepickerRangeComponent,
                    PoEmailComponent,
                    PoInputComponent,
                    PoLoginComponent,
                    PoLookupComponent,
                    PoLookupModalComponent,
                    PoMultiselectComponent,
                    PoMultiselectDropdownComponent,
                    PoNumberComponent,
                    PoPasswordComponent,
                    PoRichTextBodyComponent,
                    PoRichTextComponent,
                    PoRichTextImageModalComponent,
                    PoRichTextLinkModalComponent,
                    PoRichTextToolbarComponent,
                    PoSelectComponent,
                    PoTextareaComponent,
                    PoUploadComponent,
                    PoUploadDragDropComponent,
                    PoUploadDragDropDirective,
                    PoUploadDragDropAreaOverlayComponent,
                    PoUploadDragDropAreaComponent,
                    PoUploadFileRestrictionsComponent,
                    PoUrlComponent
                ],
                providers: []
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(PoFieldModule, { declarations: [PoComboComponent,
        PoComboOptionTemplateDirective,
        PoMultiselectOptionTemplateDirective,
        PoDecimalComponent,
        PoDatepickerRangeComponent,
        PoEmailComponent,
        PoInputComponent,
        PoLoginComponent,
        PoLookupComponent,
        PoLookupModalComponent,
        PoMultiselectComponent,
        PoMultiselectDropdownComponent,
        PoNumberComponent,
        PoPasswordComponent,
        PoRichTextBodyComponent,
        PoRichTextComponent,
        PoRichTextImageModalComponent,
        PoRichTextLinkModalComponent,
        PoRichTextToolbarComponent,
        PoSelectComponent,
        PoTextareaComponent,
        PoUploadComponent,
        PoUploadDragDropComponent,
        PoUploadDragDropDirective,
        PoUploadDragDropAreaOverlayComponent,
        PoUploadDragDropAreaComponent,
        PoUploadFileRestrictionsComponent,
        PoUrlComponent], imports: [CommonModule,
        FormsModule, i1.FormsModule, OverlayModule,
        PoButtonGroupModule,
        PoButtonModule,
        PoCleanModule,
        PoCalendarModule,
        PoCheckboxGroupModule,
        PoRadioGroupModule,
        PoContainerModule,
        PoDatepickerModule,
        PoDisclaimerGroupModule,
        PoDisclaimerModule,
        PoFieldContainerModule,
        PoLoadingModule,
        PoModalModule,
        PoProgressModule,
        PoServicesModule,
        PoTableModule,
        PoTooltipModule,
        PoIconModule,
        PoCheckboxModule,
        PoRadioModule,
        PoLabelModule,
        PoListBoxModule,
        PoSwitchModule,
        PoTagModule], exports: [PoCheckboxGroupModule,
        PoRadioGroupModule,
        PoCleanModule,
        PoDatepickerModule,
        PoComboComponent,
        PoComboOptionTemplateDirective,
        PoMultiselectOptionTemplateDirective,
        PoDecimalComponent,
        PoDatepickerRangeComponent,
        PoEmailComponent,
        PoFieldContainerModule,
        PoInputComponent,
        PoLoginComponent,
        PoLookupComponent,
        PoLookupModalComponent,
        PoMultiselectComponent,
        PoNumberComponent,
        PoPasswordComponent,
        PoRichTextComponent,
        PoSelectComponent,
        PoTextareaComponent,
        PoUploadComponent,
        PoUrlComponent,
        PoCheckboxModule,
        PoRadioModule,
        PoLabelModule,
        PoSwitchModule,
        PoTagModule] }); })();
i0.ɵɵsetComponentScope(PoLookupModalComponent, function () { return [i2.NgIf, i1.DefaultValueAccessor, i1.NgControlStatus, i1.NgModel, i3.PoDisclaimerGroupComponent, i4.PoFieldContainerComponent, i5.PoModalComponent, i6.PoTableComponent, i7.PoIconComponent]; }, []);
i0.ɵɵsetComponentScope(PoMultiselectComponent, function () { return [i2.NgClass, i2.NgForOf, i2.NgIf, i2.NgTemplateOutlet, i8.CdkConnectedOverlay, i8.CdkOverlayOrigin, i9.PoFieldContainerBottomComponent, i4.PoFieldContainerComponent, i7.PoIconComponent, i10.PoTagComponent, PoMultiselectDropdownComponent]; }, []);
