// @po-ui/ng-components/esm2022/lib/components/po-breadcrumb/po-breadcrumb.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { PoBreadcrumbComponent } from './po-breadcrumb.component';
import { PoBreadcrumbFavoriteComponent } from './po-breadcrumb-favorite/po-breadcrumb-favorite.component';
import { PoLinkModule } from '../po-link/po-link.module';
import { PoPopupModule } from '../po-popup/po-popup.module';
import { PoIconModule } from '../po-icon/po-icon.module';
import * as i0 from "@angular/core";
/**
 * @description
 *
 * Módulo do componente po-breadcrumb.
 *
 */
export class PoBreadcrumbModule {
    static ɵfac = function PoBreadcrumbModule_Factory(t) { return new (t || PoBreadcrumbModule)(); };
    static ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: PoBreadcrumbModule });
    static ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [CommonModule, RouterModule, PoLinkModule, PoPopupModule, PoIconModule] });
}
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(PoBreadcrumbModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, RouterModule, PoLinkModule, PoPopupModule, PoIconModule],
                declarations: [PoBreadcrumbComponent, PoBreadcrumbFavoriteComponent],
                exports: [PoBreadcrumbComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(PoBreadcrumbModule, { declarations: [PoBreadcrumbComponent, PoBreadcrumbFavoriteComponent], imports: [CommonModule, RouterModule, PoLinkModule, PoPopupModule, PoIconModule], exports: [PoBreadcrumbComponent] }); })();
