// @po-ui/ng-components/esm2022/lib/components/po-disclaimer/po-disclaimer-base.component.mjs
import { EventEmitter, Input, Output, Directive } from '@angular/core';
import { convertToBoolean } from '../../utils/util';
import { PoDisclaimerLiterals } from './po-disclaimer.literals';
import * as i0 from "@angular/core";
import * as i1 from "./../../services/po-language/po-language.service";
const PO_DISCLAIMER_TYPES = ['default', 'danger'];
const PO_DISCLAIMER_DEFAULT_TYPE = 'default';
/**
 * @docsPrivate
 *
 * @description
 *
 * O componente po-disclaimer é responsável por representar tags.
 * Seu uso é recomendado em buscas e em campos onde é necessário representar objetos selecionados,
 * como por exemplo, no po-multi-select.
 *
 */
export class PoDisclaimerBaseComponent {
    languageService;
    /**
     * Label que aparecerá dentro do po-disclaimer.
     * Quando não for definido um label será apresentada a propriedade p-value.
     */
    label;
    /** Valor do po-disclaimer. */
    value;
    /** Nome da propriedade vinculada à este po-disclaimer. */
    property;
    disclaimerCustomWidth = 201;
    lastDisclaimer = false;
    tooltipPosition = 'bottom';
    /**
     * @optional
     *
     * @description
     *
     * Evento disparado ao fechar o disclaimer.
     * Para este evento será passado como parâmetro um objeto com value, label e property.
     */
    closeAction = new EventEmitter();
    clickNumber = new EventEmitter();
    literals;
    showDisclaimer = true;
    _type = 'default';
    _hideClose = false;
    /**
     * @description
     *
     * Esta propriedade esconde o botão para fechamento do po-disclaimer, ao utilizar esta propriedade
     * sem passar valor a mesma é setada como false, onde o botão de fechamento está visível.
     *
     * @default false
     */
    set hideClose(value) {
        this._hideClose = value === '' ? true : convertToBoolean(value);
    }
    get hideClose() {
        return this._hideClose;
    }
    /**
     * @description
     *
     * Tipo do po-disclaimer. Pode ser 'default' ou 'danger'.
     *
     * @default default
     * @optional
     */
    set type(type) {
        this._type = PO_DISCLAIMER_TYPES.includes(type) ? type : PO_DISCLAIMER_DEFAULT_TYPE;
    }
    get type() {
        return this._type;
    }
    constructor(languageService) {
        this.languageService = languageService;
        const language = this.languageService?.getShortLanguage();
        this.literals = {
            ...PoDisclaimerLiterals[language]
        };
    }
    close() {
        this.showDisclaimer = false;
        this.closeAction.emit({ value: this.value, label: this.label, property: this.property });
    }
    getLabel() {
        return this.label ? this.label : this.value;
    }
    setAriaLabel() {
        return this.label ? this.label + ' ' + this.literals.remove : this.value + ' ' + this.literals.remove;
    }
    static ɵfac = function PoDisclaimerBaseComponent_Factory(t) { return new (t || PoDisclaimerBaseComponent)(i0.ɵɵdirectiveInject(i1.PoLanguageService)); };
    static ɵdir = /*@__PURE__*/ i0.ɵɵdefineDirective({ type: PoDisclaimerBaseComponent, inputs: { label: [i0.ɵɵInputFlags.None, "p-label", "label"], value: [i0.ɵɵInputFlags.None, "p-value", "value"], property: [i0.ɵɵInputFlags.None, "p-property", "property"], disclaimerCustomWidth: [i0.ɵɵInputFlags.None, "p-disclaimer-custom-width", "disclaimerCustomWidth"], lastDisclaimer: [i0.ɵɵInputFlags.None, "p-last-disclaimer", "lastDisclaimer"], tooltipPosition: [i0.ɵɵInputFlags.None, "p-tooltip-position", "tooltipPosition"], hideClose: [i0.ɵɵInputFlags.None, "p-hide-close", "hideClose"], type: [i0.ɵɵInputFlags.None, "p-type", "type"] }, outputs: { closeAction: "p-close-action", clickNumber: "p-click-number" } });
}
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(PoDisclaimerBaseComponent, [{
        type: Directive
    }], () => [{ type: i1.PoLanguageService }], { label: [{
            type: Input,
            args: ['p-label']
        }], value: [{
            type: Input,
            args: ['p-value']
        }], property: [{
            type: Input,
            args: ['p-property']
        }], disclaimerCustomWidth: [{
            type: Input,
            args: ['p-disclaimer-custom-width']
        }], lastDisclaimer: [{
            type: Input,
            args: ['p-last-disclaimer']
        }], tooltipPosition: [{
            type: Input,
            args: ['p-tooltip-position']
        }], closeAction: [{
            type: Output,
            args: ['p-close-action']
        }], clickNumber: [{
            type: Output,
            args: ['p-click-number']
        }], hideClose: [{
            type: Input,
            args: ['p-hide-close']
        }], type: [{
            type: Input,
            args: ['p-type']
        }] }); })();
