// @po-ui/ng-components/esm2022/lib/components/po-disclaimer/po-disclaimer.component.mjs
import { Component, ViewChild } from '@angular/core';
import { isKeyCodeEnter } from '../../utils/util';
import { PoDisclaimerBaseComponent } from './po-disclaimer-base.component';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
import * as i2 from "../po-icon/po-icon.component";
import * as i3 from "../../directives/po-tooltip/po-tooltip.directive";
const _c0 = ["disclaimerContainer"];
function PoDisclaimerComponent_div_2_po_icon_4_Template(rf, ctx) { if (rf & 1) {
    const _r4 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "po-icon", 7);
    i0.ɵɵlistener("click", function PoDisclaimerComponent_div_2_po_icon_4_Template_po_icon_click_0_listener() { i0.ɵɵrestoreView(_r4); const ctx_r3 = i0.ɵɵnextContext(2); return i0.ɵɵresetView(ctx_r3.close()); })("keydown", function PoDisclaimerComponent_div_2_po_icon_4_Template_po_icon_keydown_0_listener($event) { i0.ɵɵrestoreView(_r4); const ctx_r5 = i0.ɵɵnextContext(2); return i0.ɵɵresetView(ctx_r5.onKeyPress($event)); });
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r2 = i0.ɵɵnextContext(2);
    i0.ɵɵclassProp("po-disclaimer-remove-danger", ctx_r2.type === "danger");
    i0.ɵɵattribute("aria-label", ctx_r2.setAriaLabel())("role", !ctx_r2.hideClose ? "button" : "");
} }
function PoDisclaimerComponent_div_2_Template(rf, ctx) { if (rf & 1) {
    const _r7 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "div", 3);
    i0.ɵɵlistener("click", function PoDisclaimerComponent_div_2_Template_div_click_0_listener() { i0.ɵɵrestoreView(_r7); const ctx_r6 = i0.ɵɵnextContext(); return i0.ɵɵresetView(ctx_r6.emitLastDisclaimer(ctx_r6.lastDisclaimer)); })("keydown.enter", function PoDisclaimerComponent_div_2_Template_div_keydown_enter_0_listener() { i0.ɵɵrestoreView(_r7); const ctx_r8 = i0.ɵɵnextContext(); return i0.ɵɵresetView(ctx_r8.emitLastDisclaimer(ctx_r8.lastDisclaimer)); });
    i0.ɵɵelementStart(1, "div", 4)(2, "span", 5);
    i0.ɵɵtext(3);
    i0.ɵɵelementEnd()();
    i0.ɵɵtemplate(4, PoDisclaimerComponent_div_2_po_icon_4_Template, 1, 4, "po-icon", 6);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r1 = i0.ɵɵnextContext();
    i0.ɵɵclassProp("po-disclaimer-remove-all", ctx_r1.hideClose)("po-disclaimer-danger", ctx_r1.type === "danger");
    i0.ɵɵproperty("p-tooltip", ctx_r1.getWidthDisclaimer() ? ctx_r1.getLabel() : "")("p-tooltip-position", ctx_r1.tooltipPosition)("tabindex", ctx_r1.hideClose ? "0" : "-1");
    i0.ɵɵadvance();
    i0.ɵɵclassProp("show-close", !ctx_r1.hideClose);
    i0.ɵɵadvance(2);
    i0.ɵɵtextInterpolate1(" ", ctx_r1.getLabel(), " ");
    i0.ɵɵadvance();
    i0.ɵɵproperty("ngIf", ctx_r1.hideClose === false);
} }
/**
 * @docsPrivate
 *
 * @docsExtends PoDisclaimerBaseComponent
 *
 * @examplePrivate
 *
 * <example-private name="po-disclaimer" title="PO Disclaimer">
 *   <file name="sample-po-disclaimer.component.html"> </file>
 *   <file name="sample-po-disclaimer.component.ts"> </file>
 * </example-private>
 */
export class PoDisclaimerComponent extends PoDisclaimerBaseComponent {
    disclaimerContainer;
    onKeyPress(event) {
        if (isKeyCodeEnter(event)) {
            this.close();
        }
    }
    emitLastDisclaimer(isLast) {
        if (isLast) {
            this.clickNumber.emit();
        }
    }
    getWidthDisclaimer() {
        return this.disclaimerContainer.nativeElement.offsetWidth > this.disclaimerCustomWidth;
    }
    static ɵfac = /*@__PURE__*/ (() => { let ɵPoDisclaimerComponent_BaseFactory; return function PoDisclaimerComponent_Factory(t) { return (ɵPoDisclaimerComponent_BaseFactory || (ɵPoDisclaimerComponent_BaseFactory = i0.ɵɵgetInheritedFactory(PoDisclaimerComponent)))(t || PoDisclaimerComponent); }; })();
    static ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: PoDisclaimerComponent, selectors: [["po-disclaimer"]], viewQuery: function PoDisclaimerComponent_Query(rf, ctx) { if (rf & 1) {
            i0.ɵɵviewQuery(_c0, 7);
        } if (rf & 2) {
            let _t;
            i0.ɵɵqueryRefresh(_t = i0.ɵɵloadQuery()) && (ctx.disclaimerContainer = _t.first);
        } }, features: [i0.ɵɵInheritDefinitionFeature], decls: 3, vars: 1, consts: [[1, "disclaimer-container"], ["disclaimerContainer", ""], ["class", "po-disclaimer", 3, "po-disclaimer-remove-all", "po-disclaimer-danger", "p-tooltip", "p-tooltip-position", "tabindex", "click", "keydown.enter", 4, "ngIf"], [1, "po-disclaimer", 3, "p-tooltip", "p-tooltip-position", "tabindex", "click", "keydown.enter"], [1, "po-disclaimer-label"], [1, "label"], ["p-icon", "ICON_CLOSE", "class", "po-disclaimer-remove po-clickable", "tabindex", "0", 3, "po-disclaimer-remove-danger", "click", "keydown", 4, "ngIf"], ["p-icon", "ICON_CLOSE", "tabindex", "0", 1, "po-disclaimer-remove", "po-clickable", 3, "click", "keydown"]], template: function PoDisclaimerComponent_Template(rf, ctx) { if (rf & 1) {
            i0.ɵɵelementStart(0, "div", 0, 1);
            i0.ɵɵtemplate(2, PoDisclaimerComponent_div_2_Template, 5, 11, "div", 2);
            i0.ɵɵelementEnd();
        } if (rf & 2) {
            i0.ɵɵadvance(2);
            i0.ɵɵproperty("ngIf", ctx.showDisclaimer === true);
        } }, dependencies: [i1.NgIf, i2.PoIconComponent, i3.PoTooltipDirective], encapsulation: 2 });
}
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(PoDisclaimerComponent, [{
        type: Component,
        args: [{ selector: 'po-disclaimer', template: "<div #disclaimerContainer class=\"disclaimer-container\">\n  <div\n    *ngIf=\"showDisclaimer === true\"\n    class=\"po-disclaimer\"\n    [class.po-disclaimer-remove-all]=\"hideClose\"\n    [class.po-disclaimer-danger]=\"type === 'danger'\"\n    [p-tooltip]=\"getWidthDisclaimer() ? getLabel() : ''\"\n    [p-tooltip-position]=\"tooltipPosition\"\n    [tabindex]=\"hideClose ? '0' : '-1'\"\n    (click)=\"emitLastDisclaimer(lastDisclaimer)\"\n    (keydown.enter)=\"emitLastDisclaimer(lastDisclaimer)\"\n  >\n    <div class=\"po-disclaimer-label\" [class.show-close]=\"!hideClose\">\n      <span class=\"label\">\n        {{ getLabel() }}\n      </span>\n    </div>\n    <po-icon\n      p-icon=\"ICON_CLOSE\"\n      *ngIf=\"hideClose === false\"\n      [attr.aria-label]=\"setAriaLabel()\"\n      [attr.role]=\"!hideClose ? 'button' : ''\"\n      class=\"po-disclaimer-remove po-clickable\"\n      tabindex=\"0\"\n      [class.po-disclaimer-remove-danger]=\"type === 'danger'\"\n      (click)=\"close()\"\n      (keydown)=\"onKeyPress($event)\"\n    ></po-icon>\n  </div>\n</div>\n" }]
    }], null, { disclaimerContainer: [{
            type: ViewChild,
            args: ['disclaimerContainer', { static: true }]
        }] }); })();
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassDebugInfo(PoDisclaimerComponent, { className: "PoDisclaimerComponent", filePath: "lib/components/po-disclaimer/po-disclaimer.component.ts", lineNumber: 23 }); })();
