// @po-ui/ng-components/esm2022/lib/components/po-disclaimer/po-disclaimer.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { PoIconModule } from '../po-icon';
import { PoTooltipModule } from '../../directives/po-tooltip/index';
import { PoDisclaimerComponent } from './po-disclaimer.component';
import * as i0 from "@angular/core";
/**
 * @description
 * Módulo do componente po-disclaimer.
 */
export class PoDisclaimerModule {
    static ɵfac = function PoDisclaimerModule_Factory(t) { return new (t || PoDisclaimerModule)(); };
    static ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: PoDisclaimerModule });
    static ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [CommonModule, PoIconModule, PoTooltipModule] });
}
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(PoDisclaimerModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, PoIconModule, PoTooltipModule],
                declarations: [PoDisclaimerComponent],
                exports: [PoDisclaimerComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(PoDisclaimerModule, { declarations: [PoDisclaimerComponent], imports: [CommonModule, PoIconModule, PoTooltipModule], exports: [PoDisclaimerComponent] }); })();
