// @po-ui/ng-components/esm2022/lib/components/po-toolbar/po-toolbar-base.component.mjs
import { Input, Directive } from '@angular/core';
import { convertToInt } from '../../utils/util';
import * as i0 from "@angular/core";
/**
 * @description
 *
 * O componente `po-toolbar` é um cabeçalho para o título da aplicação e informações de usuário e notificações quando houver necessidade.
 */
export class PoToolbarBaseComponent {
    /**
     * @optional
     *
     * @description
     *
     * Define uma lista de ações que serão exibidas ao clicar no ícone declarado em `p-actions-icon`.
     */
    actions;
    /**
     * @optional
     *
     * @description
     *
     * Define um [ícone](/guides/icons) para a propriedade `p-actions`.
     *
     * É possível usar qualquer um dos ícones da [Biblioteca de ícones](/guides/icons). conforme exemplo abaixo:
     * ```
     * <po-toolbar p-actions-icon="po-icon-user" [p-actions]="actions"></po-toolbar>
     * ```
     * Também é possível utilizar outras fontes de ícones, por exemplo a biblioteca *Font Awesome*, da seguinte forma:
     * ```
     * <po-toolbar p-actions-icon="far fa-comment-alt" [p-actions]="actions"></po-toolbar>
     * ```
     * Outra opção seria a customização do ícone através do `TemplateRef`, conforme exemplo abaixo:
     * ```
     * <po-toolbar [p-actions-icon]="template" [p-actions]="actions"></po-toolbar>
     *
     * <ng-template #template>
     *  <ion-icon style="font-size: inherit" name="heart"></ion-icon>
     * </ng-template>
     * ```
     * > Para o ícone enquadrar corretamente, deve-se utilizar `font-size: inherit` caso o ícone utilizado não aplique-o.
     *
     * > Caso não haja ações definidas em `p-actions`, o ícone não será exibido.
     *
     * @default `po-icon-more`
     */
    actionsIcon;
    /** Define o objeto que será o cabeçalho da lista de ações com as informações do perfil. */
    profile;
    /** Define uma lista de ações que serão exibidas ao clicar no ícone do perfil. */
    profileActions;
    /** Se falso, oculta o ícone de notificações. */
    showNotification = true;
    /** Título do *toolbar* e aplicação. */
    title;
    /**
     * @optional
     *
     * @description
     *
     * Lista de ações da notificação.
     */
    notificationActions;
    _notificationNumber;
    /**
     * @optional
     *
     * @description
     *
     * Número de notificações.
     */
    set notificationNumber(value) {
        this._notificationNumber = convertToInt(value, 0);
    }
    get notificationNumber() {
        return this._notificationNumber;
    }
    get isShowProfile() {
        return !!(this.profile || this.profileActions);
    }
    static ɵfac = function PoToolbarBaseComponent_Factory(t) { return new (t || PoToolbarBaseComponent)(); };
    static ɵdir = /*@__PURE__*/ i0.ɵɵdefineDirective({ type: PoToolbarBaseComponent, inputs: { actions: [i0.ɵɵInputFlags.None, "p-actions", "actions"], actionsIcon: [i0.ɵɵInputFlags.None, "p-actions-icon", "actionsIcon"], profile: [i0.ɵɵInputFlags.None, "p-profile", "profile"], profileActions: [i0.ɵɵInputFlags.None, "p-profile-actions", "profileActions"], showNotification: [i0.ɵɵInputFlags.None, "p-show-notification", "showNotification"], title: [i0.ɵɵInputFlags.None, "p-title", "title"], notificationActions: [i0.ɵɵInputFlags.None, "p-notification-actions", "notificationActions"], notificationNumber: [i0.ɵɵInputFlags.None, "p-notification-number", "notificationNumber"] } });
}
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(PoToolbarBaseComponent, [{
        type: Directive
    }], null, { actions: [{
            type: Input,
            args: ['p-actions']
        }], actionsIcon: [{
            type: Input,
            args: ['p-actions-icon']
        }], profile: [{
            type: Input,
            args: ['p-profile']
        }], profileActions: [{
            type: Input,
            args: ['p-profile-actions']
        }], showNotification: [{
            type: Input,
            args: ['p-show-notification']
        }], title: [{
            type: Input,
            args: ['p-title']
        }], notificationActions: [{
            type: Input,
            args: ['p-notification-actions']
        }], notificationNumber: [{
            type: Input,
            args: ['p-notification-number']
        }] }); })();
