// @po-ui/ng-components/esm2022/lib/components/po-accordion/po-accordion-item-header/po-accordion-item-header.component.mjs
import { ChangeDetectionStrategy, Component, ElementRef, EventEmitter, Input, Output, ViewChild } from '@angular/core';
import { poLocaleDefault } from '../../../services/po-language/po-language.constant';
import * as i0 from "@angular/core";
import * as i1 from "../../../services/po-language";
import * as i2 from "@angular/common";
import * as i3 from "../../po-tag/po-tag.component";
import * as i4 from "../../po-icon/po-icon.component";
const _c0 = ["accordionElement"];
function PoAccordionItemHeaderComponent_po_tag_6_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelement(0, "po-tag", 7);
} if (rf & 2) {
    const ctx_r1 = i0.ɵɵnextContext();
    i0.ɵɵproperty("p-value", ctx_r1.labelTag)("p-type", ctx_r1.typeTag);
} }
export class PoAccordionItemHeaderComponent {
    language = poLocaleDefault;
    accordionElement;
    expanded = false;
    label;
    labelTag;
    typeTag;
    disabledItem;
    toggle = new EventEmitter();
    constructor(languageService) {
        this.language = languageService.getShortLanguage();
    }
    onClick() {
        this.expanded = !this.expanded;
        this.toggle.emit(this.expanded);
    }
    static ɵfac = function PoAccordionItemHeaderComponent_Factory(t) { return new (t || PoAccordionItemHeaderComponent)(i0.ɵɵdirectiveInject(i1.PoLanguageService)); };
    static ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: PoAccordionItemHeaderComponent, selectors: [["po-accordion-item-header"]], viewQuery: function PoAccordionItemHeaderComponent_Query(rf, ctx) { if (rf & 1) {
            i0.ɵɵviewQuery(_c0, 7, ElementRef);
        } if (rf & 2) {
            let _t;
            i0.ɵɵqueryRefresh(_t = i0.ɵɵloadQuery()) && (ctx.accordionElement = _t.first);
        } }, inputs: { expanded: [i0.ɵɵInputFlags.None, "p-expanded", "expanded"], label: [i0.ɵɵInputFlags.None, "p-label", "label"], labelTag: [i0.ɵɵInputFlags.None, "p-label-tag", "labelTag"], typeTag: [i0.ɵɵInputFlags.None, "p-type-tag", "typeTag"], disabledItem: [i0.ɵɵInputFlags.None, "p-disabled", "disabledItem"] }, outputs: { toggle: "p-toggle" }, decls: 8, vars: 5, consts: [[1, "po-accordion-item-header"], [1, "po-accordion-item-header-button", "po-clickable", 3, "disabled", "click"], ["accordionElement", ""], [1, "po-accordion-item-header-button-content"], [1, "po-text-ellipsis", "po-accordion-item-header-title"], ["class", "po-accordion-item-header-tag", 3, "p-value", "p-type", 4, "ngIf"], ["p-icon", "ICON_ARROW_DOWN", 1, "po-icon", "po-accordion-item-header-icon"], [1, "po-accordion-item-header-tag", 3, "p-value", "p-type"]], template: function PoAccordionItemHeaderComponent_Template(rf, ctx) { if (rf & 1) {
            i0.ɵɵelementStart(0, "div", 0)(1, "button", 1, 2);
            i0.ɵɵlistener("click", function PoAccordionItemHeaderComponent_Template_button_click_1_listener() { return ctx.onClick(); });
            i0.ɵɵelementStart(3, "div", 3)(4, "div", 4);
            i0.ɵɵtext(5);
            i0.ɵɵelementEnd();
            i0.ɵɵtemplate(6, PoAccordionItemHeaderComponent_po_tag_6_Template, 1, 2, "po-tag", 5);
            i0.ɵɵelementEnd();
            i0.ɵɵelement(7, "po-icon", 6);
            i0.ɵɵelementEnd()();
        } if (rf & 2) {
            i0.ɵɵadvance();
            i0.ɵɵproperty("disabled", ctx.disabledItem);
            i0.ɵɵattribute("aria-label", ctx.label)("aria-expanded", ctx.expanded || false);
            i0.ɵɵadvance(4);
            i0.ɵɵtextInterpolate(ctx.label);
            i0.ɵɵadvance();
            i0.ɵɵproperty("ngIf", ctx.labelTag);
        } }, dependencies: [i2.NgIf, i3.PoTagComponent, i4.PoIconComponent], encapsulation: 2, changeDetection: 0 });
}
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(PoAccordionItemHeaderComponent, [{
        type: Component,
        args: [{ selector: 'po-accordion-item-header', changeDetection: ChangeDetectionStrategy.OnPush, template: "<div class=\"po-accordion-item-header\">\n  <button\n    #accordionElement\n    [disabled]=\"disabledItem\"\n    [attr.aria-label]=\"label\"\n    [attr.aria-expanded]=\"expanded || false\"\n    class=\"po-accordion-item-header-button po-clickable\"\n    (click)=\"onClick()\"\n  >\n    <div class=\"po-accordion-item-header-button-content\">\n      <div class=\"po-text-ellipsis po-accordion-item-header-title\">{{ label }}</div>\n      <po-tag *ngIf=\"labelTag\" class=\"po-accordion-item-header-tag\" [p-value]=\"labelTag\" [p-type]=\"typeTag\"> </po-tag>\n    </div>\n    <po-icon p-icon=\"ICON_ARROW_DOWN\" class=\"po-icon po-accordion-item-header-icon\"></po-icon>\n  </button>\n</div>\n" }]
    }], () => [{ type: i1.PoLanguageService }], { accordionElement: [{
            type: ViewChild,
            args: ['accordionElement', { read: ElementRef, static: true }]
        }], expanded: [{
            type: Input,
            args: ['p-expanded']
        }], label: [{
            type: Input,
            args: ['p-label']
        }], labelTag: [{
            type: Input,
            args: ['p-label-tag']
        }], typeTag: [{
            type: Input,
            args: ['p-type-tag']
        }], disabledItem: [{
            type: Input,
            args: ['p-disabled']
        }], toggle: [{
            type: Output,
            args: ['p-toggle']
        }] }); })();
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassDebugInfo(PoAccordionItemHeaderComponent, { className: "PoAccordionItemHeaderComponent", filePath: "lib/components/po-accordion/po-accordion-item-header/po-accordion-item-header.component.ts", lineNumber: 10 }); })();
