// @po-ui/ng-components/esm2022/lib/components/po-dropdown/po-dropdown.component.mjs
import { ChangeDetectionStrategy, Component, ElementRef, ViewChild } from '@angular/core';
import { isKeyCodeEnter } from './../../utils/util';
import { PoDropdownBaseComponent } from './po-dropdown-base.component';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
import * as i2 from "../po-popup/po-popup.component";
import * as i3 from "../po-icon/po-icon.component";
const _c0 = ["dropdownRef"];
const _c1 = ["popupRef"];
const _c2 = (a0, a1) => ({ "po-dropdown-button-disabled": a0, "po-dropdown-button-open": a1 });
const _c3 = () => ["bottom-left", "top-left"];
/**
 * @docsExtends PoDropdownBaseComponent
 *
 * @example
 *
 * <example name="po-dropdown-basic" title="PO Dropdown Basic" >
 *  <file name="sample-po-dropdown-basic/sample-po-dropdown-basic.component.html"> </file>
 *  <file name="sample-po-dropdown-basic/sample-po-dropdown-basic.component.ts"> </file>
 *  <file name="sample-po-dropdown-basic/sample-po-dropdown-basic.component.e2e-spec.ts"> </file>
 *  <file name="sample-po-dropdown-basic/sample-po-dropdown-basic.component.po.ts"> </file>
 * </example>
 *
 * <example name="po-dropdown-labs" title="PO Dropdown Labs" >
 *  <file name="sample-po-dropdown-labs/sample-po-dropdown-labs.component.html"> </file>
 *  <file name="sample-po-dropdown-labs/sample-po-dropdown-labs.component.ts"> </file>
 * </example>
 *
 * <example name="po-dropdown-social-network" title="PO Dropdown - Social Network" >
 *  <file name="sample-po-dropdown-social-network/sample-po-dropdown-social-network.component.html"> </file>
 *  <file name="sample-po-dropdown-social-network/sample-po-dropdown-social-network.component.ts"> </file>
 * </example>
 */
export class PoDropdownComponent extends PoDropdownBaseComponent {
    renderer;
    changeDetector;
    dropdownRef;
    popupRef;
    clickoutListener;
    resizeListener;
    constructor(renderer, changeDetector) {
        super();
        this.renderer = renderer;
        this.changeDetector = changeDetector;
    }
    onKeyDown(event) {
        if (isKeyCodeEnter(event)) {
            this.toggleDropdown();
        }
    }
    toggleDropdown() {
        this.dropdownRef && !this.open && !this.disabled ? this.showDropdown() : this.hideDropdown();
    }
    checkClickArea(event) {
        return this.dropdownRef && this.dropdownRef.nativeElement.contains(event.target);
    }
    hideDropdown() {
        this.icon = 'ICON_ARROW_DOWN';
        this.removeListeners();
        this.popupRef.close();
        this.open = false;
        this.changeDetector.detectChanges();
    }
    initializeListeners() {
        this.clickoutListener = this.renderer.listen('document', 'click', (event) => {
            this.wasClickedOnDropdown(event);
        });
        this.resizeListener = this.renderer.listen('window', 'resize', () => {
            this.hideDropdown();
        });
        window.addEventListener('scroll', this.onScroll, true);
    }
    onScroll = ({ target }) => {
        if (this.open && target.className !== 'po-popup-container' && !this.isDropdownClosed()) {
            this.hideDropdown();
        }
    };
    isDropdownClosed() {
        const dropdownRect = this.dropdownRef.nativeElement.getBoundingClientRect();
        return dropdownRect.top >= 0 && dropdownRect.bottom <= window.innerHeight;
    }
    removeListeners() {
        if (this.clickoutListener) {
            this.clickoutListener();
        }
        if (this.resizeListener) {
            this.resizeListener();
        }
        window.removeEventListener('scroll', this.onScroll, true);
    }
    showDropdown() {
        this.icon = 'ICON_ARROW_UP';
        this.initializeListeners();
        this.popupRef.open();
        this.open = true;
        this.changeDetector.detectChanges();
    }
    wasClickedOnDropdown(event) {
        const clickedOnDropdown = this.checkClickArea(event);
        if (!clickedOnDropdown) {
            this.hideDropdown();
        }
    }
    static ɵfac = function PoDropdownComponent_Factory(t) { return new (t || PoDropdownComponent)(i0.ɵɵdirectiveInject(i0.Renderer2), i0.ɵɵdirectiveInject(i0.ChangeDetectorRef)); };
    static ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: PoDropdownComponent, selectors: [["po-dropdown"]], viewQuery: function PoDropdownComponent_Query(rf, ctx) { if (rf & 1) {
            i0.ɵɵviewQuery(_c0, 7, ElementRef);
            i0.ɵɵviewQuery(_c1, 5);
        } if (rf & 2) {
            let _t;
            i0.ɵɵqueryRefresh(_t = i0.ɵɵloadQuery()) && (ctx.dropdownRef = _t.first);
            i0.ɵɵqueryRefresh(_t = i0.ɵɵloadQuery()) && (ctx.popupRef = _t.first);
        } }, features: [i0.ɵɵInheritDefinitionFeature], decls: 8, vars: 12, consts: [[1, "po-dropdown", 3, "tabindex", "click", "keydown"], ["dropdownRef", ""], [1, "po-dropdown-button", 3, "ngClass"], [1, "po-dropdown-icon", 3, "p-icon"], ["p-hide-arrow", "", "p-is-corner-align", "", "p-position", "bottom-left", 3, "p-actions", "p-custom-positions", "p-target"], ["popupRef", ""]], template: function PoDropdownComponent_Template(rf, ctx) { if (rf & 1) {
            i0.ɵɵelementStart(0, "div", 0, 1);
            i0.ɵɵlistener("click", function PoDropdownComponent_Template_div_click_0_listener() { return ctx.toggleDropdown(); })("keydown", function PoDropdownComponent_Template_div_keydown_0_listener($event) { return ctx.onKeyDown($event); });
            i0.ɵɵelementStart(2, "div", 2)(3, "span");
            i0.ɵɵtext(4);
            i0.ɵɵelementEnd();
            i0.ɵɵelement(5, "po-icon", 3);
            i0.ɵɵelementEnd()();
            i0.ɵɵelement(6, "po-popup", 4, 5);
        } if (rf & 2) {
            const _r0 = i0.ɵɵreference(1);
            i0.ɵɵpropertyInterpolate("tabindex", ctx.disabled ? -1 : 0);
            i0.ɵɵattribute("disabled", ctx.disabled);
            i0.ɵɵadvance(2);
            i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction2(8, _c2, ctx.disabled, ctx.open));
            i0.ɵɵadvance(2);
            i0.ɵɵtextInterpolate(ctx.label);
            i0.ɵɵadvance();
            i0.ɵɵproperty("p-icon", ctx.icon);
            i0.ɵɵadvance();
            i0.ɵɵproperty("p-actions", ctx.actions)("p-custom-positions", i0.ɵɵpureFunction0(11, _c3))("p-target", _r0);
        } }, dependencies: [i1.NgClass, i2.PoPopupComponent, i3.PoIconComponent], encapsulation: 2, changeDetection: 0 });
}
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(PoDropdownComponent, [{
        type: Component,
        args: [{ selector: 'po-dropdown', changeDetection: ChangeDetectionStrategy.OnPush, template: "<div\n  #dropdownRef\n  class=\"po-dropdown\"\n  [attr.disabled]=\"disabled\"\n  tabindex=\"{{ disabled ? -1 : 0 }}\"\n  (click)=\"toggleDropdown()\"\n  (keydown)=\"onKeyDown($event)\"\n>\n  <div\n    class=\"po-dropdown-button\"\n    [ngClass]=\"{ 'po-dropdown-button-disabled': disabled, 'po-dropdown-button-open': open }\"\n  >\n    <span>{{ label }}</span>\n    <po-icon class=\"po-dropdown-icon\" [p-icon]=\"icon\"></po-icon>\n  </div>\n</div>\n\n<po-popup\n  #popupRef\n  p-hide-arrow\n  p-is-corner-align\n  p-position=\"bottom-left\"\n  [p-actions]=\"actions\"\n  [p-custom-positions]=\"['bottom-left', 'top-left']\"\n  [p-target]=\"dropdownRef\"\n>\n</po-popup>\n" }]
    }], () => [{ type: i0.Renderer2 }, { type: i0.ChangeDetectorRef }], { dropdownRef: [{
            type: ViewChild,
            args: ['dropdownRef', { read: ElementRef, static: true }]
        }], popupRef: [{
            type: ViewChild,
            args: ['popupRef']
        }] }); })();
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassDebugInfo(PoDropdownComponent, { className: "PoDropdownComponent", filePath: "lib/components/po-dropdown/po-dropdown.component.ts", lineNumber: 34 }); })();
