// @totvs/protheus-lib-core/esm2022/services/pro-tenant/interceptors/pro-tenant.inteceptor.mjs
import { Injectable } from '@angular/core';
import { ProCompanyService } from '../pro-company/services/pro-company.service';
import { ProBranchService } from './../pro-branch/services/pro-branch.service';
import { ProAppConfigService } from '../../pro-app-config';
import * as i0 from "@angular/core";
/**
 * @description
 *
 * Interceptor que pega a requisição html e insere no header o token.
 */
export class ProTenantInteceptor {
    constructor(injector) {
        this.injector = injector;
    }
    intercept(original_request, next) {
        let requestResult = original_request;
        const proAppConfigService = this.injector.get(ProAppConfigService);
        if (!(original_request.url.search(/assets/) >= 0)) {
            const isProtheus = proAppConfigService.isProtheusRender;
            if (isProtheus) {
                requestResult = this.appendTenantToRequest(original_request);
            }
        }
        return next.handle(requestResult);
    }
    appendTenantToRequest(request) {
        const tenantid = this.getTenantId();
        if (tenantid) {
            return request.clone({
                headers: request.headers.set('tenantid', tenantid)
            });
        }
        else {
            return request.clone();
        }
    }
    getTenantId() {
        const proBranchService = this.injector.get(ProBranchService);
        const proCompanyService = this.injector.get(ProCompanyService);
        const companyCode = proCompanyService.company.Code;
        const branchCode = proBranchService.branch.Code;
        let tenantid = '';
        if (companyCode !== '') {
            tenantid += companyCode;
            if (branchCode !== '') {
                // adicionada a virgula a direita para preservar o espaço a direita
                // no codigo da filial quando houver
                tenantid += `,${branchCode},`;
            }
            return tenantid;
        }
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProTenantInteceptor, deps: [{ token: i0.Injector }], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProTenantInteceptor, providedIn: 'root' }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProTenantInteceptor, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root'
                }]
        }], ctorParameters: () => [{ type: i0.Injector }] });
