// @po-ui/ng-components/esm2022/lib/components/po-chart/po-chart-legend/po-chart-legend.component.mjs
import { ChangeDetectionStrategy, Component, Input } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
function PoChartLegendComponent_div_2_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "div", 3);
    i0.ɵɵelement(1, "div", 4);
    i0.ɵɵelementStart(2, "span", 5);
    i0.ɵɵtext(3);
    i0.ɵɵelementEnd()();
} if (rf & 2) {
    const serie_r1 = ctx.$implicit;
    i0.ɵɵadvance();
    i0.ɵɵclassMap((serie_r1.color == null ? null : serie_r1.color.includes("po-color")) ? serie_r1.color : "");
    i0.ɵɵstyleProp("background", serie_r1.color);
    i0.ɵɵadvance(2);
    i0.ɵɵtextInterpolate(serie_r1.label || serie_r1.category);
} }
export class PoChartLegendComponent {
    type;
    _series;
    set series(value) {
        this._series = value;
    }
    get series() {
        return this._series;
    }
    static ɵfac = function PoChartLegendComponent_Factory(t) { return new (t || PoChartLegendComponent)(); };
    static ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: PoChartLegendComponent, selectors: [["po-chart-legend"]], inputs: { type: [i0.ɵɵInputFlags.None, "p-type", "type"], series: [i0.ɵɵInputFlags.None, "p-series", "series"] }, decls: 3, vars: 1, consts: [[1, "po-chart-legend"], [1, "po-chart-legend-container"], ["class", "po-chart-legend-item", 4, "ngFor", "ngForOf"], [1, "po-chart-legend-item"], [1, "po-chart-legend-square"], [1, "po-chart-legend-text"]], template: function PoChartLegendComponent_Template(rf, ctx) { if (rf & 1) {
            i0.ɵɵelementStart(0, "div", 0)(1, "div", 1);
            i0.ɵɵtemplate(2, PoChartLegendComponent_div_2_Template, 4, 5, "div", 2);
            i0.ɵɵelementEnd()();
        } if (rf & 2) {
            i0.ɵɵadvance(2);
            i0.ɵɵproperty("ngForOf", ctx.series);
        } }, dependencies: [i1.NgForOf], encapsulation: 2, changeDetection: 0 });
}
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(PoChartLegendComponent, [{
        type: Component,
        args: [{ selector: 'po-chart-legend', changeDetection: ChangeDetectionStrategy.OnPush, template: "<div class=\"po-chart-legend\">\n  <div class=\"po-chart-legend-container\">\n    <div class=\"po-chart-legend-item\" *ngFor=\"let serie of series; let i = index\">\n      <div\n        class=\"po-chart-legend-square\"\n        [class]=\"serie.color?.includes('po-color') ? serie.color : ''\"\n        [style.background]=\"serie.color\"\n      ></div>\n      <span class=\"po-chart-legend-text\">{{ serie.label || serie.category }}</span>\n    </div>\n  </div>\n</div>\n" }]
    }], null, { type: [{
            type: Input,
            args: ['p-type']
        }], series: [{
            type: Input,
            args: ['p-series']
        }] }); })();
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassDebugInfo(PoChartLegendComponent, { className: "PoChartLegendComponent", filePath: "lib/components/po-chart/po-chart-legend/po-chart-legend.component.ts", lineNumber: 10 }); })();
