// @po-ui/ng-components/esm2022/lib/components/po-link/po-link.component.mjs
import { Component } from '@angular/core';
import { PoLinkBaseComponent } from './po-link-base.component';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
import * as i2 from "@angular/router";
function PoLinkComponent_button_1_Template(rf, ctx) { if (rf & 1) {
    const _r5 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "button", 5);
    i0.ɵɵlistener("click", function PoLinkComponent_button_1_Template_button_click_0_listener() { i0.ɵɵrestoreView(_r5); const ctx_r4 = i0.ɵɵnextContext(); return i0.ɵɵresetView(ctx_r4.onClick()); });
    i0.ɵɵtext(1);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = i0.ɵɵnextContext();
    i0.ɵɵadvance();
    i0.ɵɵtextInterpolate(ctx_r0.label);
} }
function PoLinkComponent_a_2_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "a", 6);
    i0.ɵɵtext(1);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r1 = i0.ɵɵnextContext();
    i0.ɵɵproperty("href", ctx_r1.url, i0.ɵɵsanitizeUrl)("target", ctx_r1.openNewTab ? "_blank" : "_self");
    i0.ɵɵadvance();
    i0.ɵɵtextInterpolate(ctx_r1.label);
} }
function PoLinkComponent_a_3_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "a", 7);
    i0.ɵɵtext(1);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r2 = i0.ɵɵnextContext();
    i0.ɵɵproperty("routerLink", ctx_r2.url);
    i0.ɵɵadvance();
    i0.ɵɵtextInterpolate(ctx_r2.label);
} }
function PoLinkComponent_a_4_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "a", 8);
    i0.ɵɵtext(1);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r3 = i0.ɵɵnextContext();
    i0.ɵɵproperty("href", ctx_r3.url, i0.ɵɵsanitizeUrl);
    i0.ɵɵadvance();
    i0.ɵɵtextInterpolate(ctx_r3.label);
} }
/**
 * @docsExtends PoLinkBaseComponent
 *
 * @example
 *
 * <example name="po-link-basic" title="PO Link Basic" >
 *  <file name="sample-po-link-basic/sample-po-link-basic.component.html"> </file>
 *  <file name="sample-po-link-basic/sample-po-link-basic.component.ts"> </file>
 * </example>
 *
 * <example name="po-link-labs" title="PO Link Labs" >
 *  <file name="sample-po-link-labs/sample-po-link-labs.component.html"> </file>
 *  <file name="sample-po-link-labs/sample-po-link-labs.component.ts"> </file>
 * </example>
 *
 * <example name="po-link-heroes" title="PO Link Heroes" >
 *  <file name="sample-po-link-heroes/sample-po-link-heroes.component.html"> </file>
 *  <file name="sample-po-link-heroes/sample-po-link-heroes.component.ts"> </file>
 * </example>
 */
export class PoLinkComponent extends PoLinkBaseComponent {
    onClick() {
        if (this.url) {
            return;
        }
        else {
            this.action.emit(null);
        }
    }
    static ɵfac = /*@__PURE__*/ (() => { let ɵPoLinkComponent_BaseFactory; return function PoLinkComponent_Factory(t) { return (ɵPoLinkComponent_BaseFactory || (ɵPoLinkComponent_BaseFactory = i0.ɵɵgetInheritedFactory(PoLinkComponent)))(t || PoLinkComponent); }; })();
    static ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: PoLinkComponent, selectors: [["po-link"]], features: [i0.ɵɵInheritDefinitionFeature], decls: 5, vars: 4, consts: [[3, "ngSwitch"], ["class", "po-link", "type", "button", 3, "click", 4, "ngSwitchCase"], ["class", "po-link", 3, "href", "target", 4, "ngSwitchCase"], ["class", "po-link", 3, "routerLink", 4, "ngSwitchCase"], ["class", "po-link", 3, "href", 4, "ngSwitchDefault"], ["type", "button", 1, "po-link", 3, "click"], [1, "po-link", 3, "href", "target"], [1, "po-link", 3, "routerLink"], [1, "po-link", 3, "href"]], template: function PoLinkComponent_Template(rf, ctx) { if (rf & 1) {
            i0.ɵɵelementContainerStart(0, 0);
            i0.ɵɵtemplate(1, PoLinkComponent_button_1_Template, 2, 1, "button", 1)(2, PoLinkComponent_a_2_Template, 2, 3, "a", 2)(3, PoLinkComponent_a_3_Template, 2, 2, "a", 3)(4, PoLinkComponent_a_4_Template, 2, 2, "a", 4);
            i0.ɵɵelementContainerEnd();
        } if (rf & 2) {
            i0.ɵɵproperty("ngSwitch", ctx.type);
            i0.ɵɵadvance();
            i0.ɵɵproperty("ngSwitchCase", "action");
            i0.ɵɵadvance();
            i0.ɵɵproperty("ngSwitchCase", "externalLink");
            i0.ɵɵadvance();
            i0.ɵɵproperty("ngSwitchCase", "internalLink");
        } }, dependencies: [i1.NgSwitch, i1.NgSwitchCase, i1.NgSwitchDefault, i2.RouterLink], encapsulation: 2 });
}
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(PoLinkComponent, [{
        type: Component,
        args: [{ selector: 'po-link', template: "<ng-container [ngSwitch]=\"type\">\n  <button *ngSwitchCase=\"'action'\" class=\"po-link\" type=\"button\" (click)=\"onClick()\">{{ label }}</button>\n\n  <a *ngSwitchCase=\"'externalLink'\" class=\"po-link\" [href]=\"url\" [target]=\"openNewTab ? '_blank' : '_self'\">{{\n    label\n  }}</a>\n\n  <a *ngSwitchCase=\"'internalLink'\" class=\"po-link\" [routerLink]=\"url\">{{ label }}</a>\n\n  <a *ngSwitchDefault class=\"po-link\" [href]=\"url\">{{ label }}</a>\n</ng-container>\n" }]
    }], null, null); })();
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassDebugInfo(PoLinkComponent, { className: "PoLinkComponent", filePath: "lib/components/po-link/po-link.component.ts", lineNumber: 29 }); })();
