// @po-ui/ng-components/esm2022/lib/components/po-link/po-link.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { PoLinkComponent } from './po-link.component';
import { RouterModule } from '@angular/router';
import * as i0 from "@angular/core";
/**
 * @description
 *
 * Módulo do componente po-link.
 */
export class PoLinkModule {
    static ɵfac = function PoLinkModule_Factory(t) { return new (t || PoLinkModule)(); };
    static ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: PoLinkModule });
    static ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [CommonModule, RouterModule] });
}
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(PoLinkModule, [{
        type: NgModule,
        args: [{
                declarations: [PoLinkComponent],
                imports: [CommonModule, RouterModule],
                exports: [PoLinkComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(PoLinkModule, { declarations: [PoLinkComponent], imports: [CommonModule, RouterModule], exports: [PoLinkComponent] }); })();
