// @totvs/protheus-lib-core/esm2022/services/pro-app-config/interceptors/pro-app-config.inteceptor.mjs
import { Injectable } from '@angular/core';
import { ProAppConfigService } from '../services/pro-app-config.service';
import * as i0 from "@angular/core";
/**
 * @description
 *
 * Interceptor que pega a requisição html e insere o endereço e o header que define o erro padrão do Protheus.
 */
export class ProAppConfigInteceptor {
    constructor(injector) {
        this.injector = injector;
    }
    intercept(original_request, next) {
        let requestResult;
        const proAppConfigService = this.injector.get(ProAppConfigService);
        if (original_request.url.search(/assets/) >= 0) {
            requestResult = this.appendPathNameToUrl(original_request);
        }
        else {
            const isProtheus = proAppConfigService.isProtheusRender;
            if (isProtheus) {
                requestResult = this.appendDomainToUrl(original_request);
            }
            else {
                requestResult = original_request;
            }
        }
        return next.handle(requestResult);
    }
    /**
     * @description Altera a URL da requisição para acesso aos assets do app
     * @param request Requisição HTTP original
     * @returns Requisição HTTP com a alteração do path, somente para assets
     */
    appendPathNameToUrl(request) {
        const location = window.location;
        let appName = sessionStorage['appName'];
        let clone;
        if (appName === undefined) {
            clone = request.clone();
        }
        else {
            if (request.url.substr(0, 1) !== '/') {
                appName += '/';
            }
            const urlWithPathName = `${location.protocol}//${location.host}/app-root/${appName}${request.url}`;
            clone = request.clone({
                url: urlWithPathName
            });
        }
        return clone;
    }
    /**
     * @description Altera da URL da requisição conforme regras
     * @param request Requisição HTTP Original
     * @returns Requisição HTTP, podendo ter sua URL alterada
     */
    appendDomainToUrl(request) {
        const proAppConfigService = this.injector.get(ProAppConfigService);
        const domain = proAppConfigService.serverWithApiUrl;
        let urlWithDomain = '';
        //Caso a URL comece com http ou https, concatenar o domínio vai gerar um URL inválida
        //Essa situação pode ocorrer quando o app consulta serviços de terceiros
        const lowerUrl = request.url.toLowerCase();
        if (lowerUrl.startsWith('http://') || lowerUrl.startsWith('https://')) {
            urlWithDomain = request.url;
        }
        else {
            urlWithDomain = `${domain}${request.url}`;
        }
        return request.clone({
            url: urlWithDomain
        });
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProAppConfigInteceptor, deps: [{ token: i0.Injector }], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProAppConfigInteceptor, providedIn: 'root' }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProAppConfigInteceptor, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root'
                }]
        }], ctorParameters: () => [{ type: i0.Injector }] });
