// @po-ui/ng-components/esm2022/lib/components/po-popup/po-popup.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { PoListBoxModule } from '../po-listbox/po-listbox.module';
import { PoPopupComponent } from './po-popup.component';
import * as i0 from "@angular/core";
/**
 * @description
 *
 * Módulo do componente po-popup.
 */
export class PoPopupModule {
    static ɵfac = function PoPopupModule_Factory(t) { return new (t || PoPopupModule)(); };
    static ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: PoPopupModule });
    static ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [CommonModule, PoListBoxModule] });
}
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(PoPopupModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, PoListBoxModule],
                declarations: [PoPopupComponent],
                exports: [PoPopupComponent],
                providers: [],
                schemas: []
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(PoPopupModule, { declarations: [PoPopupComponent], imports: [CommonModule, PoListBoxModule], exports: [PoPopupComponent] }); })();
