// @po-ui/ng-components/esm2022/lib/components/po-overlay/po-overlay.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { PoOverlayComponent } from './po-overlay.component';
import * as i0 from "@angular/core";
export class PoOverlayModule {
    static ɵfac = function PoOverlayModule_Factory(t) { return new (t || PoOverlayModule)(); };
    static ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: PoOverlayModule });
    static ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [CommonModule] });
}
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(PoOverlayModule, [{
        type: NgModule,
        args: [{
                declarations: [PoOverlayComponent],
                imports: [CommonModule],
                exports: [PoOverlayComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(PoOverlayModule, { declarations: [PoOverlayComponent], imports: [CommonModule], exports: [PoOverlayComponent] }); })();
