// @totvs/protheus-lib-core/esm2022/services/pro-i18nConfig/pro-i18nConfig.module.mjs
import { NgModule } from '@angular/core';
import { PoI18nModule } from '@po-ui/ng-components';
import { literalsEnUS as loginEnUS } from './i18n/pro-login.literals-EnUS';
import { literalsEsES as loginEsES } from './i18n/pro-login.literals-EsES';
import { literalsPtBR as loginPtBR } from './i18n/pro-login.literals-PtBR';
import { literalsRuRU as loginRuRU } from './i18n/pro-login.literals-RuRU';
import { literalsEnUS as sessionEnUS } from './i18n/pro-session-settings.literals-enUS';
import { literalsEsES as sessionEsES } from './i18n/pro-session-settings.literals-esES';
import { literalsPtBR as sessionPtBR } from './i18n/pro-session-settings.literals-ptBR';
import { literalsRuRU as sessionRuRU } from './i18n/pro-session-settings.literals-ruRU';
import { literalsEnUS as profileServiceEnUS } from './i18n/pro-user-profile-EnUS';
import { literalsEsES as profileServiceEsEs } from './i18n/pro-user-profile-EsES';
import { literalsPtBR as profileServicePtBR } from './i18n/pro-user-profile-PtBR';
import { literalsRuRU as profileServiceRuRU } from './i18n/pro-user-profile-RuRU';
import * as i0 from "@angular/core";
import * as i1 from "@po-ui/ng-components";
const i18nConfig = {
    default: {
        language: 'pt',
        context: 'general',
        cache: true
    },
    contexts: {
        login: {
            en: loginEnUS,
            es: loginEsES,
            pt: loginPtBR,
            ru: loginRuRU
        },
        session: {
            en: sessionEnUS,
            es: sessionEsES,
            pt: sessionPtBR,
            ru: sessionRuRU
        },
        profileService: {
            en: profileServiceEnUS,
            es: profileServiceEsEs,
            pt: profileServicePtBR,
            ru: profileServiceRuRU
        },
        general: {}
    }
};
/**
 * @description
 * Módulo para concentrar todas as strings e serviços de tradução
 * que envolvem o PoI18nConfig e o PoI18nModule
 */
export class ProI18nConfigModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProI18nConfigModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "17.3.9", ngImport: i0, type: ProI18nConfigModule, imports: [i1.PoI18nModule] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProI18nConfigModule, imports: [PoI18nModule.config(i18nConfig)] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProI18nConfigModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        PoI18nModule.config(i18nConfig),
                    ]
                }]
        }] });
