// @po-ui/ng-components/esm2022/lib/components/po-tree-view/po-tree-view-item-header/po-tree-view-item-header.component.mjs
import { ChangeDetectionStrategy, Component, EventEmitter, Input, Output, ViewChild } from '@angular/core';
import { uuid } from '../../../utils/util';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
import * as i2 from "@angular/forms";
import * as i3 from "../../po-field/po-checkbox/po-checkbox.component";
import * as i4 from "../../po-field/po-radio/po-radio.component";
import * as i5 from "../../po-icon/po-icon.component";
const _c0 = ["inputCheckbox"];
function PoTreeViewItemHeaderComponent_button_1_Template(rf, ctx) { if (rf & 1) {
    const _r11 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "button", 7);
    i0.ɵɵlistener("click", function PoTreeViewItemHeaderComponent_button_1_Template_button_click_0_listener($event) { i0.ɵɵrestoreView(_r11); const ctx_r10 = i0.ɵɵnextContext(); return i0.ɵɵresetView(ctx_r10.expanded.emit($event)); });
    i0.ɵɵelement(1, "po-icon", 8);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = i0.ɵɵnextContext();
    i0.ɵɵadvance();
    i0.ɵɵclassProp("po-tree-view-item-header-button-icon-transform", ctx_r0.item.expanded);
} }
function PoTreeViewItemHeaderComponent_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainer(0);
} }
function PoTreeViewItemHeaderComponent_ng_template_3_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "span", 9);
    i0.ɵɵtext(1);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r2 = i0.ɵɵnextContext();
    i0.ɵɵclassProp("po-tree-view-item-header-padding", !ctx_r2.hasSubItems);
    i0.ɵɵadvance();
    i0.ɵɵtextInterpolate1(" ", ctx_r2.item.label, " ");
} }
function PoTreeViewItemHeaderComponent_ng_template_5_ng_container_0_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainer(0);
} }
function PoTreeViewItemHeaderComponent_ng_template_5_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵtemplate(0, PoTreeViewItemHeaderComponent_ng_template_5_ng_container_0_Template, 1, 0, "ng-container", 2);
} if (rf & 2) {
    const ctx_r4 = i0.ɵɵnextContext();
    const _r9 = i0.ɵɵreference(10);
    const _r7 = i0.ɵɵreference(8);
    i0.ɵɵproperty("ngIf", ctx_r4.singleSelect)("ngIfThen", _r9)("ngIfElse", _r7);
} }
function PoTreeViewItemHeaderComponent_ng_template_7_Template(rf, ctx) { if (rf & 1) {
    const _r14 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "po-checkbox", 10);
    i0.ɵɵtwoWayListener("ngModelChange", function PoTreeViewItemHeaderComponent_ng_template_7_Template_po_checkbox_ngModelChange_0_listener($event) { i0.ɵɵrestoreView(_r14); const ctx_r13 = i0.ɵɵnextContext(); i0.ɵɵtwoWayBindingSet(ctx_r13.item.selected, $event) || (ctx_r13.item.selected = $event); return i0.ɵɵresetView($event); });
    i0.ɵɵlistener("p-change", function PoTreeViewItemHeaderComponent_ng_template_7_Template_po_checkbox_p_change_0_listener() { i0.ɵɵrestoreView(_r14); const ctx_r15 = i0.ɵɵnextContext(); return i0.ɵɵresetView(ctx_r15.selected.emit(ctx_r15.item)); });
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r6 = i0.ɵɵnextContext();
    i0.ɵɵclassProp("po-tree-view-item-header-padding", !ctx_r6.hasSubItems);
    i0.ɵɵproperty("p-label", ctx_r6.item.label);
    i0.ɵɵtwoWayProperty("ngModel", ctx_r6.item.selected);
    i0.ɵɵproperty("p-disabled", ctx_r6.item.isSelectable === false);
} }
function PoTreeViewItemHeaderComponent_ng_template_9_Template(rf, ctx) { if (rf & 1) {
    const _r18 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "po-radio", 11, 12);
    i0.ɵɵtwoWayListener("ngModelChange", function PoTreeViewItemHeaderComponent_ng_template_9_Template_po_radio_ngModelChange_0_listener($event) { i0.ɵɵrestoreView(_r18); const ctx_r17 = i0.ɵɵnextContext(); i0.ɵɵtwoWayBindingSet(ctx_r17.item.selected, $event) || (ctx_r17.item.selected = $event); return i0.ɵɵresetView($event); });
    i0.ɵɵlistener("p-change-selected", function PoTreeViewItemHeaderComponent_ng_template_9_Template_po_radio_p_change_selected_0_listener() { i0.ɵɵrestoreView(_r18); const ctx_r19 = i0.ɵɵnextContext(); return i0.ɵɵresetView(ctx_r19.selected.emit(ctx_r19.item)); });
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r8 = i0.ɵɵnextContext();
    i0.ɵɵclassProp("po-tree-view-item-header-padding", !ctx_r8.hasSubItems);
    i0.ɵɵproperty("name", ctx_r8.idRadio);
    i0.ɵɵtwoWayProperty("ngModel", ctx_r8.item.selected);
    i0.ɵɵproperty("p-label", ctx_r8.item.label)("p-value", ctx_r8.item.value)("p-checked", ctx_r8.item.value === ctx_r8.selectedValue)("p-disabled", ctx_r8.item.isSelectable === false);
} }
export class PoTreeViewItemHeaderComponent {
    inputCheckbox;
    item;
    selectable = false;
    singleSelect;
    expanded = new EventEmitter();
    selected = new EventEmitter();
    selectedValue;
    idRadio = `po-radio[${uuid()}]`;
    get hasSubItems() {
        return !!(this.item.subItems && this.item.subItems.length);
    }
    static ɵfac = function PoTreeViewItemHeaderComponent_Factory(t) { return new (t || PoTreeViewItemHeaderComponent)(); };
    static ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: PoTreeViewItemHeaderComponent, selectors: [["po-tree-view-item-header"]], viewQuery: function PoTreeViewItemHeaderComponent_Query(rf, ctx) { if (rf & 1) {
            i0.ɵɵviewQuery(_c0, 5);
        } if (rf & 2) {
            let _t;
            i0.ɵɵqueryRefresh(_t = i0.ɵɵloadQuery()) && (ctx.inputCheckbox = _t.first);
        } }, inputs: { item: [i0.ɵɵInputFlags.None, "p-item", "item"], selectable: [i0.ɵɵInputFlags.None, "p-selectable", "selectable"], singleSelect: [i0.ɵɵInputFlags.None, "p-single-select", "singleSelect"], selectedValue: [i0.ɵɵInputFlags.None, "p-selected-value", "selectedValue"] }, outputs: { expanded: "p-expanded", selected: "p-selected" }, decls: 11, vars: 4, consts: [[1, "po-tree-view-item-header"], ["class", "po-tree-view-item-header-button", 3, "click", 4, "ngIf"], [4, "ngIf", "ngIfThen", "ngIfElse"], ["labelTemplate", ""], ["selectionTemplate", ""], ["checkboxTemplate", ""], ["radioTemplate", ""], [1, "po-tree-view-item-header-button", 3, "click"], ["p-icon", "ICON_ARROW_DOWN", 1, "po-tree-view-item-header-button-icon"], [1, "po-tree-view-item-header-label"], [1, "po-tree-view-item-header-checkbox", 3, "p-label", "ngModel", "p-disabled", "ngModelChange", "p-change"], [1, "po-tree-view-item-header-checkbox", 3, "name", "ngModel", "p-label", "p-value", "p-checked", "p-disabled", "ngModelChange", "p-change-selected"], ["inputRadio", ""]], template: function PoTreeViewItemHeaderComponent_Template(rf, ctx) { if (rf & 1) {
            i0.ɵɵelementStart(0, "div", 0);
            i0.ɵɵtemplate(1, PoTreeViewItemHeaderComponent_button_1_Template, 2, 2, "button", 1)(2, PoTreeViewItemHeaderComponent_ng_container_2_Template, 1, 0, "ng-container", 2);
            i0.ɵɵelementEnd();
            i0.ɵɵtemplate(3, PoTreeViewItemHeaderComponent_ng_template_3_Template, 2, 3, "ng-template", null, 3, i0.ɵɵtemplateRefExtractor)(5, PoTreeViewItemHeaderComponent_ng_template_5_Template, 1, 3, "ng-template", null, 4, i0.ɵɵtemplateRefExtractor)(7, PoTreeViewItemHeaderComponent_ng_template_7_Template, 1, 5, "ng-template", null, 5, i0.ɵɵtemplateRefExtractor)(9, PoTreeViewItemHeaderComponent_ng_template_9_Template, 2, 8, "ng-template", null, 6, i0.ɵɵtemplateRefExtractor);
        } if (rf & 2) {
            const _r3 = i0.ɵɵreference(4);
            const _r5 = i0.ɵɵreference(6);
            i0.ɵɵadvance();
            i0.ɵɵproperty("ngIf", ctx.hasSubItems);
            i0.ɵɵadvance();
            i0.ɵɵproperty("ngIf", ctx.selectable)("ngIfThen", _r5)("ngIfElse", _r3);
        } }, dependencies: [i1.NgIf, i2.NgControlStatus, i2.NgModel, i3.PoCheckboxComponent, i4.PoRadioComponent, i5.PoIconComponent], encapsulation: 2, changeDetection: 0 });
}
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(PoTreeViewItemHeaderComponent, [{
        type: Component,
        args: [{ selector: 'po-tree-view-item-header', changeDetection: ChangeDetectionStrategy.OnPush, template: "<div class=\"po-tree-view-item-header\">\n  <button *ngIf=\"hasSubItems\" class=\"po-tree-view-item-header-button\" (click)=\"expanded.emit($event)\">\n    <po-icon\n      p-icon=\"ICON_ARROW_DOWN\"\n      class=\"po-tree-view-item-header-button-icon\"\n      [class.po-tree-view-item-header-button-icon-transform]=\"item.expanded\"\n    >\n    </po-icon>\n  </button>\n\n  <ng-container *ngIf=\"selectable; then selectionTemplate; else labelTemplate\"></ng-container>\n</div>\n\n<ng-template #labelTemplate>\n  <span class=\"po-tree-view-item-header-label\" [class.po-tree-view-item-header-padding]=\"!hasSubItems\">\n    {{ item.label }}\n  </span>\n</ng-template>\n\n<ng-template #selectionTemplate>\n  <ng-container *ngIf=\"singleSelect; then radioTemplate; else checkboxTemplate\"></ng-container>\n</ng-template>\n\n<ng-template #checkboxTemplate>\n  <po-checkbox\n    class=\"po-tree-view-item-header-checkbox\"\n    [class.po-tree-view-item-header-padding]=\"!hasSubItems\"\n    [p-label]=\"item.label\"\n    [(ngModel)]=\"item.selected\"\n    (p-change)=\"selected.emit(item)\"\n    [p-disabled]=\"item.isSelectable === false\"\n  >\n  </po-checkbox>\n</ng-template>\n\n<ng-template #radioTemplate>\n  <po-radio\n    class=\"po-tree-view-item-header-checkbox\"\n    [class.po-tree-view-item-header-padding]=\"!hasSubItems\"\n    #inputRadio\n    [name]=\"idRadio\"\n    [(ngModel)]=\"item.selected\"\n    [p-label]=\"item.label\"\n    [p-value]=\"item.value\"\n    [p-checked]=\"item.value === selectedValue\"\n    (p-change-selected)=\"selected.emit(item)\"\n    [p-disabled]=\"item.isSelectable === false\"\n  >\n  </po-radio>\n</ng-template>\n" }]
    }], null, { inputCheckbox: [{
            type: ViewChild,
            args: ['inputCheckbox']
        }], item: [{
            type: Input,
            args: ['p-item']
        }], selectable: [{
            type: Input,
            args: ['p-selectable']
        }], singleSelect: [{
            type: Input,
            args: ['p-single-select']
        }], expanded: [{
            type: Output,
            args: ['p-expanded']
        }], selected: [{
            type: Output,
            args: ['p-selected']
        }], selectedValue: [{
            type: Input,
            args: ['p-selected-value']
        }] }); })();
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassDebugInfo(PoTreeViewItemHeaderComponent, { className: "PoTreeViewItemHeaderComponent", filePath: "lib/components/po-tree-view/po-tree-view-item-header/po-tree-view-item-header.component.ts", lineNumber: 11 }); })();
