// @po-ui/ng-components/esm2022/lib/components/po-navbar/po-navbar-actions/po-navbar-action/po-navbar-action.component.mjs
import { Component, Input } from '@angular/core';
import { isExternalLink, openExternalLink } from '../../../../utils/util';
import * as i0 from "@angular/core";
import * as i1 from "@angular/router";
import * as i2 from "../../../../directives/po-tooltip/po-tooltip.directive";
import * as i3 from "../../../po-icon/po-icon.component";
export class PoNavbarActionComponent {
    router;
    action;
    icon;
    label;
    link;
    tooltip;
    constructor(viewContainerRef, router) {
        this.router = router;
    }
    click() {
        if (this.action) {
            this.action();
            return;
        }
        if (this.link) {
            return this.openUrl(this.link);
        }
    }
    openUrl(url) {
        if (isExternalLink(url)) {
            return openExternalLink(url);
        }
        if (url) {
            return this.router.navigate([url]);
        }
    }
    static ɵfac = function PoNavbarActionComponent_Factory(t) { return new (t || PoNavbarActionComponent)(i0.ɵɵdirectiveInject(i0.ViewContainerRef), i0.ɵɵdirectiveInject(i1.Router)); };
    static ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: PoNavbarActionComponent, selectors: [["po-navbar-action"]], inputs: { action: [i0.ɵɵInputFlags.None, "p-action", "action"], icon: [i0.ɵɵInputFlags.None, "p-icon", "icon"], label: [i0.ɵɵInputFlags.None, "p-label", "label"], link: [i0.ɵɵInputFlags.None, "p-link", "link"], tooltip: [i0.ɵɵInputFlags.None, "p-tooltip-text", "tooltip"] }, decls: 1, vars: 2, consts: [["tabindex", "0", 1, "po-navbar-action-content", "po-clickable", 3, "p-icon", "p-tooltip", "click"]], template: function PoNavbarActionComponent_Template(rf, ctx) { if (rf & 1) {
            i0.ɵɵelementStart(0, "po-icon", 0);
            i0.ɵɵlistener("click", function PoNavbarActionComponent_Template_po_icon_click_0_listener() { return ctx.click(); });
            i0.ɵɵelementEnd();
        } if (rf & 2) {
            i0.ɵɵproperty("p-icon", ctx.icon)("p-tooltip", ctx.tooltip);
        } }, dependencies: [i2.PoTooltipDirective, i3.PoIconComponent], encapsulation: 2 });
}
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(PoNavbarActionComponent, [{
        type: Component,
        args: [{ selector: 'po-navbar-action', template: "<po-icon\n  tabindex=\"0\"\n  class=\"po-navbar-action-content po-clickable\"\n  [p-icon]=\"icon\"\n  [p-tooltip]=\"tooltip\"\n  (click)=\"click()\"\n></po-icon>\n" }]
    }], () => [{ type: i0.ViewContainerRef }, { type: i1.Router }], { action: [{
            type: Input,
            args: ['p-action']
        }], icon: [{
            type: Input,
            args: ['p-icon']
        }], label: [{
            type: Input,
            args: ['p-label']
        }], link: [{
            type: Input,
            args: ['p-link']
        }], tooltip: [{
            type: Input,
            args: ['p-tooltip-text']
        }] }); })();
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassDebugInfo(PoNavbarActionComponent, { className: "PoNavbarActionComponent", filePath: "lib/components/po-navbar/po-navbar-actions/po-navbar-action/po-navbar-action.component.ts", lineNumber: 10 }); })();
