// @po-ui/ng-components/esm2022/lib/components/po-navbar/po-navbar-actions/po-navbar-actions.component.mjs
import { Component, Input } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
import * as i2 from "./po-navbar-action/po-navbar-action.component";
import * as i3 from "./po-navbar-action-popup/po-navbar-action-popup.component";
function PoNavbarActionsComponent_po_navbar_action_0_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelement(0, "po-navbar-action", 2);
} if (rf & 2) {
    const iconAction_r1 = ctx.$implicit;
    i0.ɵɵproperty("p-action", iconAction_r1.action)("p-icon", iconAction_r1.icon)("p-label", iconAction_r1.label)("p-link", iconAction_r1.link)("p-tooltip-text", iconAction_r1.tooltip);
} }
export class PoNavbarActionsComponent {
    _iconActions;
    set iconActions(actions) {
        this._iconActions = actions.map(action => ({ ...action, separator: true, url: action.link }));
    }
    get iconActions() {
        return this._iconActions;
    }
    static ɵfac = function PoNavbarActionsComponent_Factory(t) { return new (t || PoNavbarActionsComponent)(); };
    static ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: PoNavbarActionsComponent, selectors: [["po-navbar-actions"]], inputs: { iconActions: [i0.ɵɵInputFlags.None, "p-icon-actions", "iconActions"] }, decls: 2, vars: 2, consts: [["class", "po-navbar-action", 3, "p-action", "p-icon", "p-label", "p-link", "p-tooltip-text", 4, "ngFor", "ngForOf"], [1, "po-navbar-action-popup", 3, "p-icon-actions"], [1, "po-navbar-action", 3, "p-action", "p-icon", "p-label", "p-link", "p-tooltip-text"]], template: function PoNavbarActionsComponent_Template(rf, ctx) { if (rf & 1) {
            i0.ɵɵtemplate(0, PoNavbarActionsComponent_po_navbar_action_0_Template, 1, 5, "po-navbar-action", 0);
            i0.ɵɵelement(1, "po-navbar-action-popup", 1);
        } if (rf & 2) {
            i0.ɵɵproperty("ngForOf", ctx.iconActions);
            i0.ɵɵadvance();
            i0.ɵɵproperty("p-icon-actions", ctx.iconActions);
        } }, dependencies: [i1.NgForOf, i2.PoNavbarActionComponent, i3.PoNavbarActionPopupComponent], encapsulation: 2 });
}
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(PoNavbarActionsComponent, [{
        type: Component,
        args: [{ selector: 'po-navbar-actions', template: "<po-navbar-action\n  *ngFor=\"let iconAction of iconActions\"\n  class=\"po-navbar-action\"\n  [p-action]=\"iconAction.action\"\n  [p-icon]=\"iconAction.icon\"\n  [p-label]=\"iconAction.label\"\n  [p-link]=\"iconAction.link\"\n  [p-tooltip-text]=\"iconAction.tooltip\"\n>\n</po-navbar-action>\n\n<po-navbar-action-popup class=\"po-navbar-action-popup\" [p-icon-actions]=\"iconActions\"> </po-navbar-action-popup>\n" }]
    }], null, { iconActions: [{
            type: Input,
            args: ['p-icon-actions']
        }] }); })();
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassDebugInfo(PoNavbarActionsComponent, { className: "PoNavbarActionsComponent", filePath: "lib/components/po-navbar/po-navbar-actions/po-navbar-actions.component.ts", lineNumber: 9 }); })();
