// @totvs/protheus-lib-core/esm2022/components/pro-pages/pro-home/pro-home.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { ProHomeComponent } from './pro-home.component';
import * as i0 from "@angular/core";
export class ProHomeModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProHomeModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "17.3.9", ngImport: i0, type: ProHomeModule, declarations: [ProHomeComponent], imports: [CommonModule], exports: [ProHomeComponent] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProHomeModule, imports: [CommonModule] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProHomeModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule],
                    declarations: [ProHomeComponent],
                    exports: [ProHomeComponent]
                }]
        }] });
