// @po-ui/ng-components/esm2022/lib/components/po-navbar/po-navbar.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { PoMenuModule } from '../po-menu/po-menu.module';
import { PoIconModule } from './../po-icon/po-icon.module';
import { PoLogoModule } from './../po-logo/po-logo.module';
import { PoNavbarActionsModule } from './po-navbar-actions/po-navbar-actions.module';
import { PoNavbarItemNavigationModule } from './po-navbar-item-navigation/po-navbar-item-navigation.module';
import { PoNavbarItemsModule } from './po-navbar-items/po-navbar-items.module';
import { PoNavbarLogoComponent } from './po-navbar-logo/po-navbar-logo.component';
import { PoNavbarComponent } from './po-navbar.component';
import * as i0 from "@angular/core";
/**
 * @description
 *
 * Módulo do componente `po-navbar`.
 *
 * > Para o correto funcionamento do componente `po-navbar`, deve ser importado o módulo `BrowserAnimationsModule` no
 * > módulo principal da sua aplicação.
 *
 * Módulo da aplicação:
 * ```
 * import { BrowserAnimationsModule } from '@angular/platform-browser/animations';
 * import { PoModule } from '@po-ui/ng-components';
 * ...
 *
 * @NgModule({
 *   imports: [
 *     BrowserModule,
 *     BrowserAnimationsModule,
 *     ...
 *     PoModule
 *   ],
 *   declarations: [
 *     AppComponent,
 *     ...
 *   ],
 *   providers: [],
 *   bootstrap: [AppComponent]
 * })
 * export class AppModule { }
 * ```
 */
export class PoNavbarModule {
    static ɵfac = function PoNavbarModule_Factory(t) { return new (t || PoNavbarModule)(); };
    static ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: PoNavbarModule });
    static ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [CommonModule,
            PoIconModule,
            PoLogoModule,
            PoMenuModule,
            PoNavbarActionsModule,
            PoNavbarItemNavigationModule,
            PoNavbarItemsModule] });
}
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(PoNavbarModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    PoIconModule,
                    PoLogoModule,
                    PoMenuModule,
                    PoNavbarActionsModule,
                    PoNavbarItemNavigationModule,
                    PoNavbarItemsModule
                ],
                declarations: [PoNavbarComponent, PoNavbarLogoComponent],
                exports: [PoNavbarComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(PoNavbarModule, { declarations: [PoNavbarComponent, PoNavbarLogoComponent], imports: [CommonModule,
        PoIconModule,
        PoLogoModule,
        PoMenuModule,
        PoNavbarActionsModule,
        PoNavbarItemNavigationModule,
        PoNavbarItemsModule], exports: [PoNavbarComponent] }); })();
