// @po-ui/ng-components/esm2022/lib/components/po-tabs/po-tabs.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { PoPopoverModule } from '../po-popover/po-popover.module';
import { PoTabButtonComponent } from './po-tab-button/po-tab-button.component';
import { PoTabComponent } from './po-tab/po-tab.component';
import { PoTabDropdownComponent } from './po-tab-dropdown/po-tab-dropdown.component';
import { PoTabsComponent } from './po-tabs.component';
import { PoDropdownModule } from '../po-dropdown/po-dropdown.module';
import { PoButtonModule } from '../po-button/po-button.module';
import { PoListBoxModule } from '../po-listbox/po-listbox.module';
import { PoTooltipModule } from '../../directives/po-tooltip/po-tooltip.module';
import { PoTabsService } from './po-tabs.service';
import * as i0 from "@angular/core";
/**
 * @description
 * Módulo do componente po-tabs
 */
export class PoTabsModule {
    static ɵfac = function PoTabsModule_Factory(t) { return new (t || PoTabsModule)(); };
    static ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: PoTabsModule });
    static ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ providers: [PoTabsService], imports: [CommonModule, PoPopoverModule, PoDropdownModule, PoTooltipModule, PoButtonModule, PoListBoxModule] });
}
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(PoTabsModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, PoPopoverModule, PoDropdownModule, PoTooltipModule, PoButtonModule, PoListBoxModule],
                declarations: [PoTabButtonComponent, PoTabComponent, PoTabDropdownComponent, PoTabsComponent],
                exports: [PoTabComponent, PoTabsComponent],
                providers: [PoTabsService]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(PoTabsModule, { declarations: [PoTabButtonComponent, PoTabComponent, PoTabDropdownComponent, PoTabsComponent], imports: [CommonModule, PoPopoverModule, PoDropdownModule, PoTooltipModule, PoButtonModule, PoListBoxModule], exports: [PoTabComponent, PoTabsComponent] }); })();
