// @po-ui/ng-templates/esm2022/lib/components/po-page-background/po-page-background.component.mjs
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { poLanguageDefault } from '@po-ui/ng-components';
import { convertToBoolean } from './../../utils/util';
import * as i0 from "@angular/core";
import * as i1 from "@po-ui/ng-components";
import * as i2 from "@angular/common";
import * as i3 from "@angular/forms";
function PoPageBackgroundComponent_po_logo_3_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelement(0, "po-logo", 13);
} if (rf & 2) {
    const ctx_r0 = i0.ɵɵnextContext();
    i0.ɵɵproperty("p-link", false)("p-logo", ctx_r0.logo)("p-logo-alt", ctx_r0.logoAlt);
} }
function PoPageBackgroundComponent_div_10_Template(rf, ctx) { if (rf & 1) {
    const _r5 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "div", 14)(1, "po-select", 15);
    i0.ɵɵtwoWayListener("ngModelChange", function PoPageBackgroundComponent_div_10_Template_po_select_ngModelChange_1_listener($event) { i0.ɵɵrestoreView(_r5); const ctx_r4 = i0.ɵɵnextContext(); i0.ɵɵtwoWayBindingSet(ctx_r4.selectedLanguageOption, $event) || (ctx_r4.selectedLanguageOption = $event); return i0.ɵɵresetView($event); });
    i0.ɵɵlistener("p-change", function PoPageBackgroundComponent_div_10_Template_po_select_p_change_1_listener() { i0.ɵɵrestoreView(_r5); const ctx_r6 = i0.ɵɵnextContext(); return i0.ɵɵresetView(ctx_r6.onChangeLanguage()); });
    i0.ɵɵelementEnd()();
} if (rf & 2) {
    const ctx_r1 = i0.ɵɵnextContext();
    i0.ɵɵadvance();
    i0.ɵɵtwoWayProperty("ngModel", ctx_r1.selectedLanguageOption);
    i0.ɵɵproperty("p-options", ctx_r1.selectLanguageOptions);
} }
function PoPageBackgroundComponent_po_logo_12_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelement(0, "po-logo", 16);
} if (rf & 2) {
    const ctx_r2 = i0.ɵɵnextContext();
    i0.ɵɵproperty("p-link", false)("p-logo", ctx_r2.secondaryLogo);
} }
function PoPageBackgroundComponent_div_13_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "div", 17)(1, "div", 18)(2, "div", 19);
    i0.ɵɵtext(3);
    i0.ɵɵelementEnd()()();
} if (rf & 2) {
    const ctx_r3 = i0.ɵɵnextContext();
    i0.ɵɵstyleProp("background-image", "url(" + ctx_r3.background + ")");
    i0.ɵɵproperty("ngClass", ctx_r3.background ? "po-page-login-highlight-image" : "po-page-login-highlight-image-off");
    i0.ɵɵadvance(3);
    i0.ɵɵtextInterpolate(ctx_r3.highlightInfo);
} }
const _c0 = ["*"];
/**
 * @docsPrivate
 *
 * @description
 *
 * Componente para definição de cor de fundo e dos logotipos primário e secundário para os templates
 * de `po-page-login` e demais templates de login.
 */
export class PoPageBackgroundComponent {
    poLanguageService;
    /** Insere uma imagem de destaque ao lado direito do container. */
    background;
    /** Idioma inicial selecionado no combo */
    initialSelectLanguage;
    /** Designa se o logotipo deve desaparecer em resoluções menores. */
    hideLogo;
    /** Texto de destaque sobreposto à imagem de destaque. Essa opção é utilizada em conjunto com o atributo `p-background`. */
    highlightInfo;
    /**
     * @optional
     *
     * @description
     *
     * Evento disparado ao selecionar alguma opção no seletor de idiomas.
     * Para este evento será passado como parâmetro o valor de idioma selecionado.
     */
    selectedLanguage = new EventEmitter();
    selectedLanguageOption;
    _showSelectLanguage = false;
    _languagesList;
    _selectLanguageOptions;
    /** Lista de idiomas para o combo box */
    set languagesList(value) {
        this._languagesList = value;
        this.setLanguageOptions();
    }
    get languagesList() {
        if (this._languagesList?.length) {
            return this._languagesList;
        }
        return poLanguageDefault;
    }
    /** Caminho para a logomarca localizada na parte superior. */
    logo;
    /** Texto alternativo para a logomarca. */
    logoAlt;
    /**
     * @optional
     *
     * @description
     *
     * Caminho para a logomarca localizada no rodapé.
     */
    secondaryLogo;
    /** Define se o seletor de idiomas deve ser exibido. */
    set showSelectLanguage(showSelectLanguage) {
        this._showSelectLanguage = convertToBoolean(showSelectLanguage);
    }
    get showSelectLanguage() {
        return this._showSelectLanguage;
    }
    constructor(poLanguageService) {
        this.poLanguageService = poLanguageService;
    }
    ngOnInit() {
        this.selectedLanguageOption = this.initialSelectLanguage || this.poLanguageService.getShortLanguage();
    }
    onChangeLanguage() {
        this.selectedLanguage.emit(this.selectedLanguageOption);
    }
    get selectLanguageOptions() {
        return this._selectLanguageOptions;
    }
    setLanguageOptions() {
        this._selectLanguageOptions = this.languagesList.map(language => ({
            label: language.description,
            value: language.language
        }));
    }
    static ɵfac = function PoPageBackgroundComponent_Factory(t) { return new (t || PoPageBackgroundComponent)(i0.ɵɵdirectiveInject(i1.PoLanguageService)); };
    static ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: PoPageBackgroundComponent, selectors: [["po-page-background"]], inputs: { background: [i0.ɵɵInputFlags.None, "p-background", "background"], initialSelectLanguage: [i0.ɵɵInputFlags.None, "p-initial-language", "initialSelectLanguage"], hideLogo: [i0.ɵɵInputFlags.None, "p-hide-logo", "hideLogo"], highlightInfo: [i0.ɵɵInputFlags.None, "p-highlight-info", "highlightInfo"], languagesList: [i0.ɵɵInputFlags.None, "p-languages", "languagesList"], logo: [i0.ɵɵInputFlags.None, "p-logo", "logo"], logoAlt: [i0.ɵɵInputFlags.None, "p-logo-alt", "logoAlt"], secondaryLogo: [i0.ɵɵInputFlags.None, "p-secondary-logo", "secondaryLogo"], showSelectLanguage: [i0.ɵɵInputFlags.None, "p-show-select-language", "showSelectLanguage"] }, outputs: { selectedLanguage: "p-selected-language" }, ngContentSelectors: _c0, decls: 14, vars: 7, consts: [[1, "po-page-login-container"], [1, "po-page-login-panel"], [1, "po-page-background-main-logo"], ["p-class", "po-page-background-main-logo-image", 3, "p-link", "p-logo", "p-logo-alt", 4, "ngIf"], [1, "po-page-login-body"], [1, "po-page-login-panel-content"], [1, "po-page-background-footer", "po-sm-12"], [1, "po-page-background-footer-mobile-only"], [1, "po-page-background-footer-content"], ["class", "po-page-background-footer-select", 4, "ngIf"], [1, "po-page-background-secondary-logo", 3, "ngClass"], ["p-class", "po-page-background-secondary-logo-image", "p-logo-alt", "secondary-logomarca", 3, "p-link", "p-logo", 4, "ngIf"], [3, "ngClass", "background-image", 4, "ngIf"], ["p-class", "po-page-background-main-logo-image", 3, "p-link", "p-logo", "p-logo-alt"], [1, "po-page-background-footer-select"], ["name", "selectedLanguageOption", 3, "ngModel", "p-options", "ngModelChange", "p-change"], ["p-class", "po-page-background-secondary-logo-image", "p-logo-alt", "secondary-logomarca", 3, "p-link", "p-logo"], [3, "ngClass"], [1, "po-page-login-highlight-text"], [1, "po-font-display"]], template: function PoPageBackgroundComponent_Template(rf, ctx) { if (rf & 1) {
            i0.ɵɵprojectionDef();
            i0.ɵɵelementStart(0, "div", 0)(1, "div", 1)(2, "div", 2);
            i0.ɵɵtemplate(3, PoPageBackgroundComponent_po_logo_3_Template, 1, 3, "po-logo", 3);
            i0.ɵɵelementEnd();
            i0.ɵɵelementStart(4, "div", 4)(5, "div", 5);
            i0.ɵɵprojection(6);
            i0.ɵɵelementEnd()();
            i0.ɵɵelementStart(7, "div", 6);
            i0.ɵɵelement(8, "po-divider", 7);
            i0.ɵɵelementStart(9, "div", 8);
            i0.ɵɵtemplate(10, PoPageBackgroundComponent_div_10_Template, 2, 2, "div", 9);
            i0.ɵɵelementStart(11, "div", 10);
            i0.ɵɵtemplate(12, PoPageBackgroundComponent_po_logo_12_Template, 1, 2, "po-logo", 11);
            i0.ɵɵelementEnd()()()();
            i0.ɵɵtemplate(13, PoPageBackgroundComponent_div_13_Template, 4, 4, "div", 12);
            i0.ɵɵelementEnd();
        } if (rf & 2) {
            i0.ɵɵadvance(2);
            i0.ɵɵclassProp("po-page-background-hide-logo", ctx.hideLogo);
            i0.ɵɵadvance();
            i0.ɵɵproperty("ngIf", ctx.logo);
            i0.ɵɵadvance(7);
            i0.ɵɵproperty("ngIf", ctx.showSelectLanguage);
            i0.ɵɵadvance();
            i0.ɵɵproperty("ngClass", ctx.showSelectLanguage ? "po-page-background-secondary-logo-right" : "po-page-background-secondary-logo-centered");
            i0.ɵɵadvance();
            i0.ɵɵproperty("ngIf", ctx.secondaryLogo);
            i0.ɵɵadvance();
            i0.ɵɵproperty("ngIf", ctx.background);
        } }, dependencies: [i2.NgClass, i2.NgIf, i3.NgControlStatus, i3.NgModel, i1.PoDividerComponent, i1.PoSelectComponent, i1.PoLogoComponent], encapsulation: 2 });
}
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(PoPageBackgroundComponent, [{
        type: Component,
        args: [{ selector: 'po-page-background', template: "<div class=\"po-page-login-container\">\n  <div class=\"po-page-login-panel\">\n    <div class=\"po-page-background-main-logo\" [class.po-page-background-hide-logo]=\"hideLogo\">\n      <po-logo\n        *ngIf=\"logo\"\n        p-class=\"po-page-background-main-logo-image\"\n        [p-link]=\"false\"\n        [p-logo]=\"logo\"\n        [p-logo-alt]=\"logoAlt\"\n      ></po-logo>\n    </div>\n\n    <div class=\"po-page-login-body\">\n      <div class=\"po-page-login-panel-content\">\n        <ng-content></ng-content>\n      </div>\n    </div>\n    <div class=\"po-page-background-footer po-sm-12\">\n      <po-divider class=\"po-page-background-footer-mobile-only\"></po-divider>\n\n      <div class=\"po-page-background-footer-content\">\n        <div *ngIf=\"showSelectLanguage\" class=\"po-page-background-footer-select\">\n          <po-select\n            name=\"selectedLanguageOption\"\n            [(ngModel)]=\"selectedLanguageOption\"\n            [p-options]=\"selectLanguageOptions\"\n            (p-change)=\"onChangeLanguage()\"\n          ></po-select>\n        </div>\n\n        <div\n          class=\"po-page-background-secondary-logo\"\n          [ngClass]=\"\n            showSelectLanguage\n              ? 'po-page-background-secondary-logo-right'\n              : 'po-page-background-secondary-logo-centered'\n          \"\n        >\n          <po-logo\n            *ngIf=\"secondaryLogo\"\n            p-class=\"po-page-background-secondary-logo-image\"\n            [p-link]=\"false\"\n            [p-logo]=\"secondaryLogo\"\n            p-logo-alt=\"secondary-logomarca\"\n          ></po-logo>\n        </div>\n      </div>\n    </div>\n  </div>\n\n  <div\n    *ngIf=\"background\"\n    [ngClass]=\"background ? 'po-page-login-highlight-image' : 'po-page-login-highlight-image-off'\"\n    [style.background-image]=\"'url(' + background + ')'\"\n  >\n    <div class=\"po-page-login-highlight-text\">\n      <div class=\"po-font-display\">{{ highlightInfo }}</div>\n    </div>\n  </div>\n</div>\n" }]
    }], () => [{ type: i1.PoLanguageService }], { background: [{
            type: Input,
            args: ['p-background']
        }], initialSelectLanguage: [{
            type: Input,
            args: ['p-initial-language']
        }], hideLogo: [{
            type: Input,
            args: ['p-hide-logo']
        }], highlightInfo: [{
            type: Input,
            args: ['p-highlight-info']
        }], selectedLanguage: [{
            type: Output,
            args: ['p-selected-language']
        }], languagesList: [{
            type: Input,
            args: ['p-languages']
        }], logo: [{
            type: Input,
            args: ['p-logo']
        }], logoAlt: [{
            type: Input,
            args: ['p-logo-alt']
        }], secondaryLogo: [{
            type: Input,
            args: ['p-secondary-logo']
        }], showSelectLanguage: [{
            type: Input,
            args: ['p-show-select-language']
        }] }); })();
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassDebugInfo(PoPageBackgroundComponent, { className: "PoPageBackgroundComponent", filePath: "lib/components/po-page-background/po-page-background.component.ts", lineNumber: 20 }); })();
