// @po-ui/ng-templates/esm2022/lib/components/po-page-background/po-page-background.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { RouterModule } from '@angular/router';
import { PoDividerModule, PoFieldModule, PoLogoModule } from '@po-ui/ng-components';
import { PoPageBackgroundComponent } from './po-page-background.component';
import * as i0 from "@angular/core";
/**
 * @description
 *
 * Módulo do template do po-page-background.
 */
export class PoPageBackgroundModule {
    static ɵfac = function PoPageBackgroundModule_Factory(t) { return new (t || PoPageBackgroundModule)(); };
    static ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: PoPageBackgroundModule });
    static ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [CommonModule, FormsModule, RouterModule, PoDividerModule, PoFieldModule, PoLogoModule] });
}
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(PoPageBackgroundModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, FormsModule, RouterModule, PoDividerModule, PoFieldModule, PoLogoModule],
                declarations: [PoPageBackgroundComponent],
                exports: [PoPageBackgroundComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(PoPageBackgroundModule, { declarations: [PoPageBackgroundComponent], imports: [CommonModule, FormsModule, RouterModule, PoDividerModule, PoFieldModule, PoLogoModule], exports: [PoPageBackgroundComponent] }); })();
