// @po-ui/ng-components/esm2022/lib/components/po-divider/po-divider.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { PoDividerComponent } from './po-divider.component';
import * as i0 from "@angular/core";
/**
 * @description
 *
 * Módulo do componente po-divider.
 */
export class PoDividerModule {
    static ɵfac = function PoDividerModule_Factory(t) { return new (t || PoDividerModule)(); };
    static ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: PoDividerModule });
    static ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [CommonModule] });
}
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(PoDividerModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule],
                declarations: [PoDividerComponent],
                exports: [PoDividerComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(PoDividerModule, { declarations: [PoDividerComponent], imports: [CommonModule], exports: [PoDividerComponent] }); })();
