// @po-ui/ng-components/esm2022/lib/components/po-divider/po-divider-base.component.mjs
import { Input, Directive } from '@angular/core';
import { PoDividerSize } from './po-divider-size.enum';
import * as i0 from "@angular/core";
/**
 * @description
 *
 * Este componente apresenta uma linha demarcadora de blocos e pode conter um *label*. Seu uso é indicado para definição
 * e organização de informações em uma tela e sua característica é semelhante à tag `<hr>`.
 *
 * #### Tokens customizáveis
 *
 * É possível alterar o estilo do componente usando os seguintes tokens (CSS):
 *
 * > Para maiores informações, acesse o guia [Personalizando o Tema Padrão com Tokens CSS](https://po-ui.io/guides/theme-customization).
 *
 * | Propriedade                            | Descrição                                             | Valor Padrão                                    |
 * |----------------------------------------|-------------------------------------------------------|-------------------------------------------------|
 * | **Default Values**                     |                                                       |                                                 |
 * | `--color`                              | Cor principla do divider&nbsp;                        | `var(--color-neutral-mid-40)`                   |
 * | `--stroke-linecap`                     | Extremidade da linha&nbsp;                            | `round`                                         |
 *
 */
export class PoDividerBaseComponent {
    coordinateX1;
    coordinateX2;
    _borderWidth = PoDividerSize.small;
    /** Valor do rótulo a ser exibido. */
    label;
    /**
     * @optional
     *
     * @description
     *
     * Define a espessura da linha.
     *
     * Valores válidos:
     * - small
     * - medium
     * - large
     *
     * @default `small`
     */
    set borderWidth(value) {
        this._borderWidth = PoDividerSize[value] ? PoDividerSize[value] : PoDividerSize.small;
        this.getCoordinates();
    }
    get borderWidth() {
        return this._borderWidth;
    }
    ngOnInit() {
        this.getCoordinates();
    }
    getCoordinates() {
        if (this.borderWidth === PoDividerSize.small) {
            this.coordinateX1 = '0.1%';
            this.coordinateX2 = '99.9%';
        }
        else if (this.borderWidth === PoDividerSize.medium) {
            this.coordinateX1 = '0.2%';
            this.coordinateX2 = '99.8%';
        }
        else {
            this.coordinateX1 = '0.3%';
            this.coordinateX2 = '99.7%';
        }
    }
    static ɵfac = function PoDividerBaseComponent_Factory(t) { return new (t || PoDividerBaseComponent)(); };
    static ɵdir = /*@__PURE__*/ i0.ɵɵdefineDirective({ type: PoDividerBaseComponent, inputs: { label: [i0.ɵɵInputFlags.None, "p-label", "label"], borderWidth: [i0.ɵɵInputFlags.None, "p-border-width", "borderWidth"] } });
}
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(PoDividerBaseComponent, [{
        type: Directive
    }], null, { label: [{
            type: Input,
            args: ['p-label']
        }], borderWidth: [{
            type: Input,
            args: ['p-border-width']
        }] }); })();
