// @po-ui/ng-components/esm2022/lib/components/po-divider/po-divider.component.mjs
import { ChangeDetectionStrategy, Component } from '@angular/core';
import { PoDividerBaseComponent } from './po-divider-base.component';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
function PoDividerComponent_div_1_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "div", 4);
    i0.ɵɵtext(1);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = i0.ɵɵnextContext();
    i0.ɵɵadvance();
    i0.ɵɵtextInterpolate(ctx_r0.label);
} }
/**
 * @docsExtends PoDividerBaseComponent
 *
 * @example
 *
 * <example name="po-divider-basic" title="PO Divider Basic" >
 *  <file name="sample-po-divider-basic/sample-po-divider-basic.component.html"> </file>
 *  <file name="sample-po-divider-basic/sample-po-divider-basic.component.ts"> </file>
 * </example>
 *
 * <example name="po-divider-labs" title="PO Divider Labs" >
 *  <file name="sample-po-divider-labs/sample-po-divider-labs.component.html"> </file>
 *  <file name="sample-po-divider-labs/sample-po-divider-labs.component.ts"> </file>
 * </example>
 *
 * <example name="po-divider-user-detail" title="PO Divider - User Detail" >
 *  <file name="sample-po-divider-user-detail/sample-po-divider-user-detail.component.html"> </file>
 *  <file name="sample-po-divider-user-detail/sample-po-divider-user-detail.component.ts"> </file>
 * </example>
 */
export class PoDividerComponent extends PoDividerBaseComponent {
    static ɵfac = /*@__PURE__*/ (() => { let ɵPoDividerComponent_BaseFactory; return function PoDividerComponent_Factory(t) { return (ɵPoDividerComponent_BaseFactory || (ɵPoDividerComponent_BaseFactory = i0.ɵɵgetInheritedFactory(PoDividerComponent)))(t || PoDividerComponent); }; })();
    static ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: PoDividerComponent, selectors: [["po-divider"]], features: [i0.ɵɵInheritDefinitionFeature], decls: 4, vars: 4, consts: [[1, "po-divider"], ["class", "po-divider-label", 4, "ngIf"], ["xmlns", "http://www.w3.org/2000/svg"], ["y1", "50%", "y2", "50%"], [1, "po-divider-label"]], template: function PoDividerComponent_Template(rf, ctx) { if (rf & 1) {
            i0.ɵɵelementStart(0, "div", 0);
            i0.ɵɵtemplate(1, PoDividerComponent_div_1_Template, 2, 1, "div", 1);
            i0.ɵɵnamespaceSVG();
            i0.ɵɵelementStart(2, "svg", 2);
            i0.ɵɵelement(3, "line", 3);
            i0.ɵɵelementEnd()();
        } if (rf & 2) {
            i0.ɵɵattribute("p-size", ctx.borderWidth);
            i0.ɵɵadvance();
            i0.ɵɵproperty("ngIf", ctx.label);
            i0.ɵɵadvance(2);
            i0.ɵɵattribute("x1", ctx.coordinateX1)("x2", ctx.coordinateX2);
        } }, dependencies: [i1.NgIf], encapsulation: 2, changeDetection: 0 });
}
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(PoDividerComponent, [{
        type: Component,
        args: [{ selector: 'po-divider', changeDetection: ChangeDetectionStrategy.OnPush, template: "<div class=\"po-divider\" [attr.p-size]=\"borderWidth\">\n  <div *ngIf=\"label\" class=\"po-divider-label\">{{ label }}</div>\n  <svg xmlns=\"http://www.w3.org/2000/svg\">\n    <line [attr.x1]=\"coordinateX1\" y1=\"50%\" [attr.x2]=\"coordinateX2\" y2=\"50%\" />\n  </svg>\n</div>\n" }]
    }], null, null); })();
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassDebugInfo(PoDividerComponent, { className: "PoDividerComponent", filePath: "lib/components/po-divider/po-divider.component.ts", lineNumber: 30 }); })();
