// @totvs/protheus-lib-core/esm2022/components/pro-pages/pro-session-settings/pro-session-settings.component.mjs
import { Component, Input, ViewChild } from '@angular/core';
import { Validators } from '@angular/forms';
import { of } from 'rxjs';
import { delay, retry, tap } from 'rxjs/operators';
import { SubSink } from 'subsink';
import { environment } from '../../../environments/environment';
import { isExternalLink } from '../../../util';
import { ProBranchLookupComponent } from '../../pro-fields/pro-branch-lookup/pro-branch-lookup.component';
import { ProCompanyLookupComponent } from '../../pro-fields/pro-company-lookup/pro-company-lookup.component';
import * as i0 from "@angular/core";
import * as i1 from "@angular/forms";
import * as i2 from "@angular/router";
import * as i3 from "./../../../services/pro-messages/services/pro-message.service";
import * as i4 from "./services/pro-session-settings.service";
import * as i5 from "./../../../services/pro-tenant/pro-company/services/pro-company.service";
import * as i6 from "./../../../services/pro-role/pro-role-modules/services/pro-role.service";
import * as i7 from "./../../../services/pro-tenant/pro-branch/services/pro-branch.service";
import * as i8 from "@po-ui/ng-components";
import * as i9 from "../../../services/pro-translate-string/services/pro-translate-string.service";
import * as i10 from "../../../services/pro-brand/services/pro-brand.service";
import * as i11 from "./../../../services/pro-system-info/pro-system-modules/services/pro-system-modules.service";
import * as i12 from "@angular/common";
import * as i13 from "../pro-page-background/pro-page-background.component";
import * as i14 from "../../pro-fields/pro-company-lookup/pro-company-lookup.component";
import * as i15 from "../../pro-fields/pro-branch-lookup/pro-branch-lookup.component";
import * as i16 from "../../pro-fields/pro-role-lookup/pro-role-lookup.component";
import * as i17 from "../../pro-fields/pro-system-module-lookup/pro-system-module-lookup.component";
export class ProSessionSettingsComponent {
    /**
     * @description metodo chamado na mudança do combox de idioma da tela de parametros da sessão
     * @param language opção de idioma escolhido pelo usuário
     * @returns void
     */
    onSelectedLanguage(language) {
        this.poI18nService.setLanguage(language, false);
        this.poI18nService
            .getLiterals({ language, context: 'session' })
            .subscribe(literals => {
            this.updLiterals(literals);
            this.updateProductName();
        });
        this.messageService.changeLanguage(language).subscribe({
            next: () => { this.updateModuleDescription(); }
        });
    }
    /**
     * @description Atualiza as strings traduzíveis da interface
     * @param literals Objeto de literais, contendo as traduções da interface
     */
    updLiterals(literals) {
        this.literals = literals;
        this.companyColumns[0].label = this.literals.str0020; // 'Código'
        this.companyColumns[1].label = this.literals.str0021; // 'Descrição'
        this.moduleColumns[0].label = this.literals.str0020; // 'Código'
        this.moduleColumns[1].label = this.literals.str0022; // 'Nome'
        this.moduleColumns[2].label = this.literals.str0021; // 'Descrição'
        this.roleColumns[0].label = this.literals.str0020; // 'Código'
        this.roleColumns[1].label = this.literals.str0021; // 'Descrição'
        this.updateBranchLiterals();
    }
    /**
     * @description Atualiza as strings de tradução as colunas do Lookup de filial
     */
    updateBranchLiterals() {
        this.branchColumns[0].label = this.literals.str0020; // 'Código'
        this.branchColumns[1].label = this.literals.str0021; // 'Descrição'
        this.branchColumns[2].label = this.literals.str0026; // 'CNPJ'
        this.branchColumns[3].label = this.literals.str0025; // 'Nome comercial'
        this.updateCnpjTitleColumn();
    }
    activateSupport(url = this.getSupportLink()) {
        this.setUrlRedirect(url);
    }
    /**
     * @description Método para obter o link de suporte
     * @returns Retorna uma string contendo o link de suporte do ERP
     */
    getSupportLink() {
        const language = this.poI18nService.getShortLanguage();
        let supportLink = 'http://suporte.totvs.com/';
        if (this.brand == 'MA3') {
            supportLink = language === 'en' ? 'https://www.national-platform.com/' : 'https://www.national-platform.ru/';
        }
        else {
            if (language === 'es') {
                supportLink = 'https://totvscst.zendesk.com/hc/es#home';
            }
            else if (language === 'en') {
                supportLink = 'https://totvscst.zendesk.com/hc/en-us#home';
            }
        }
        return supportLink;
    }
    /**
     * @description Efetua o redirect conforme a URL recebida
     * @param url URL que será verificada, podendo abrir uma janela ou mudar a rota
     */
    setUrlRedirect(url) {
        isExternalLink(url)
            ? window.open(url, '_blank')
            : this.router.navigate([url]);
    }
    /**
     * @description Construtor da classe
     * @param formBuilder Objeto FormBuilder para controle do formulário
     * @param router Objeto Router para controle da rota
     * @param messageService Serviço ProMessageService para exibição de notificações
     * @param route Objeto ActivatedRoute para controle da rota atual
     * @param proSessionSettingsService Serviço ProSessionSettingsService para obter os dados de login salvos
     * @param proCompanyService Serviço ProCompanyService para obter dados do grupo de empresas
     * @param proRoleService Serviço ProRoleService para obter dados de papel de trabalho
     * @param proBranchService Serviço ProBranchService para obter dados da filial
     * @param poI18nService Serviço PoI18nService para tradução da interface via dados locais
     * @param proTranslateStringService Serviço ProTranslateStringService para tradução da interface via API do Protheus
     * @param proBrandService Serviço para identificação da marca, MA3 na russia
     */
    constructor(formBuilder, router, messageService, route, proSessionSettingsService, proCompanyService, proRoleService, proBranchService, poI18nService, proTranslateStringService, proBrandService, proSystemModuleService) {
        this.formBuilder = formBuilder;
        this.router = router;
        this.messageService = messageService;
        this.route = route;
        this.proSessionSettingsService = proSessionSettingsService;
        this.proCompanyService = proCompanyService;
        this.proRoleService = proRoleService;
        this.proBranchService = proBranchService;
        this.poI18nService = poI18nService;
        this.proTranslateStringService = proTranslateStringService;
        this.proBrandService = proBrandService;
        this.proSystemModuleService = proSystemModuleService;
        this.useHTTP = environment.useHTTP;
        this.subs = new SubSink();
        this.companyColumns = [
            { property: 'Code', label: '' },
            { property: 'CorporateName', label: '' }
        ];
        this.branchColumns = this.getDefaultBranchColumns();
        this.moduleColumns = [
            { property: 'id', label: '' },
            { property: 'name', label: '' },
            { property: 'description', label: '' }
        ];
        this.roleColumns = [
            { property: 'Code', label: '' },
            { property: 'Description', label: '' }
        ];
        const language = poI18nService.getShortLanguage();
        poI18nService
            .getLiterals({ language, context: 'session' })
            .subscribe(literals => {
            this.updLiterals(literals);
            this.updateBrand();
        });
        this.messageService.changeLanguage(language);
        this.changeServicesChannel();
    }
    /**
     * @description Lista de colunas do Lookup de filiais
     * @returns Retorna uma lista de PoLookupColumn, contendo as colunas padrões do Lookup de filiais
     */
    getDefaultBranchColumns() {
        return [
            { property: 'Code', label: '', width: '140px' },
            { property: 'Description', label: '', width: '200px' },
            { property: 'Cgc', label: '', width: '180px' },
            { property: 'CommercialName', label: '', width: '200px' },
        ];
    }
    /**
     * @description metodo disparado a partir de um evento no lookup de company
     * utilizado para forçar o foco no lookup de branch
     */
    branchSettedEvent() {
        this.proBranchLookupComponent.setBranchFocus();
    }
    /**
     * @description Método disparado via o evento de troca de grupo de empresas
     * @param event Evento do formulário
     */
    changeBranchColumns(event) {
        this.updateCnpjTitle();
        const layout = event.layout || this.settingForm.value.companyLayout;
        if (layout) {
            this.branchColumns = this.getDefaultBranchColumns();
            this.updateBranchLiterals();
            const company = layout.includes('E');
            const unit = layout.includes('U');
            if (company) {
                this.branchColumns.push({ property: 'CompanyCode', label: this.literals.str0027, width: '100px' });
            }
            if (unit) {
                this.branchColumns.push({ property: 'UnitOfBusiness', label: this.literals.str0028, width: '100px' });
            }
            if ((unit || company) && layout.includes('F')) {
                this.branchColumns.push({ property: 'ParentCode', label: this.literals.str0012, width: '100px' });
            }
        }
    }
    /**
     * @description Método do ciclo de vida do Angular, esse método é chamado após o construtor da classe
     */
    ngOnInit() {
        this.showGoEmpFil = false;
        this.showMDIMenuInfo = false;
        this.isLoading = false;
        this.settingForm = this.buildForm();
        this.subs.sink = this.subscribeToResolver();
        this.logo = './assets/images/totvs/totvs.svg';
        this.secondaryLogo = './assets/images/totvs/totvs.svg';
        this.updateCnpjTitle();
    }
    /**
     * @description Responda depois que o Angular inicializar as visualizações do componente e as visualizações filhas ou a visualização que contém a diretiva.
     * utilizado aqui para setar o foco no componente de grupo de empresa na abertura da tela
     * e definir a variavel started para true do componente company
     */
    ngAfterViewInit() {
        setTimeout(() => {
            this.proCompanyLookupComponent.setCompanyFocus();
            this.proCompanyLookupComponent.setStarted(true);
        }, 500);
    }
    /**
     * @description Atualiza o título da coluna de CNPJ da consulta de filiais conforme API
     */
    updateCnpjTitle() {
        let unsubscribe;
        unsubscribe = this.proTranslateStringService.getStrList('fwfilial').subscribe({
            next: (strs) => {
                if (strs && strs.str0010) {
                    this.cnpjTitle = strs.str0010;
                    if (this.literals) {
                        this.updateCnpjTitleColumn();
                    }
                }
            },
            complete: () => unsubscribe ? unsubscribe.unsubscribe() : null
        });
    }
    /**
     * @description Atualiza o título da coluna de CNPJ no objeto da coluna da interface já criada
     */
    updateCnpjTitleColumn() {
        //A prioridade é o valor da API, caso contrário, utiliza o valor de tradução local
        this.cnpjTitle = this.cnpjTitle || this.literals.str0026; // 'CNPJ'
        this.branchColumns[2].label = this.cnpjTitle;
    }
    /**
     * @description Atualiza a descrição do modulo selecionado
     */
    updateModuleDescription() {
        this.proSystemModuleService.getSystemModule(this.settingForm.get('environment_code').value).subscribe((systemModule) => {
            this.settingForm.get('environment_description').patchValue(systemModule.description);
        });
    }
    /**
     * @description Cria e retorna o dados de controle de formulário das configurações de login
     * @returns Retorna o FormGroup do formulário de dados de login
     */
    buildForm() {
        return this.formBuilder.group({
            base_date: [null, [Validators.required]],
            company_code: [null, [Validators.required]],
            company_description: [null, [Validators.required]],
            branch_code: [null, [Validators.required]],
            branch_description: [null, [Validators.required]],
            environment_code: [
                null,
                [this.moduleValidator.bind(this), Validators.required]
            ],
            environment_description: [null, [Validators.required]],
            role_code: [null, []],
            role_description: [null, []],
            mdi_menu_info: [false, []],
            go_emp_fil: [false, []],
            companyLayout: [null, []],
        });
    }
    moduleValidator(control) {
        if (!this.fixedModule && parseInt(control.value, 10) === 99) {
            return {
                configurador: this.literals.str0019
            };
        }
        else {
            return Validators.required(control);
        }
    }
    subscribeToResolver() {
        return this.route.data.subscribe(resultOfResolver => {
            this.validateResolverData(resultOfResolver);
        });
    }
    validateResolverData(resolveData) {
        if (resolveData.defaultsSettings) {
            this.sessionSettingsDefaults = resolveData.defaultsSettings;
            this.fixedModule =
                typeof resolveData.defaultsSettings.environment_when === 'undefined'
                    ? false
                    : !resolveData.defaultsSettings.environment_when;
            this.fixedRole = !resolveData.defaultsSettings.role_when;
            this.disableBackButton = resolveData.defaultsSettings.disable_back_button;
            this.showMDIMenuInfo = resolveData.defaultsSettings.show_mdi_menu_info;
            this.showGoEmpFil = resolveData.defaultsSettings.show_go_emp_fil;
            this.loadSettingForm();
        }
        if (resolveData.userInfo) {
            this.displayName = resolveData.userInfo.displayName;
        }
    }
    set sessionSettingsDefaults(defaults) {
        if (defaults.base_date) {
            defaults.base_date = this.convertDateforPoDatePicker(defaults.base_date);
        }
        else {
            defaults.base_date = new Date();
        }
        this._ProSessionSettingsDefaults = defaults;
    }
    get sessionSettingsDefaults() {
        return this._ProSessionSettingsDefaults;
    }
    convertDateforPoDatePicker(base_date) {
        let convertedDate = base_date;
        if (base_date && !(base_date instanceof Date)) {
            convertedDate = `${base_date}T00:00:00-02:00`;
        }
        return convertedDate;
    }
    loadSettingForm() {
        // Carrega o campo de empresa primeiro, pois ele é utilizado no campo de filial.
        this.settingForm
            .get('company_code')
            .patchValue(this.sessionSettingsDefaults.company_code);
        this.settingForm.patchValue(this.sessionSettingsDefaults);
        this.proCompanyService.company = {
            Code: this.sessionSettingsDefaults.company_code
        };
        this.proBranchService.branch = {
            Code: this.sessionSettingsDefaults.branch_code
        };
        this.proRoleService.role = {
            Code: this.sessionSettingsDefaults.role_code,
            Description: this.sessionSettingsDefaults.role_description
        };
    }
    onSessionSettingBack() {
        this.proSessionSettingsService.clearSettingsDefaults();
        this.router.navigate(['/login']);
    }
    onSessionSettingSubmit() {
        this.isLoading = true;
        of(true)
            .pipe(delay(100), tap(() => {
            if (this.settingForm.pending) {
                throw new Error();
            }
        }), retry(25), tap(() => {
            this.submitForm();
        }))
            .subscribe({
            error: () => {
                this.isLoading = false;
                this.onErrorSubmit(this.literals.str0024);
            }
        });
    }
    submitForm() {
        if (this.settingForm.valid) {
            this.subs.sink = this.saveSettings(this.settingForm).subscribe({
                next: () => {
                    this.messageService.changeLanguage(this.poI18nService.getShortLanguage());
                    this.router.navigate(['/home']);
                },
                error: () => {
                    this.onErrorSubmit();
                }
            });
        }
        else {
            this.onErrorSubmit(this.literals.str0019); // 'Formulário inválido'
        }
    }
    ngOnDestroy() {
        this.subs.unsubscribe();
    }
    onErrorSubmit(message) {
        this.isLoading = false;
        if (message) {
            this.messageService.showMessage(message, 2);
        }
        this.router.navigate(['/session-settings']);
    }
    saveSettings(form) {
        const newSession = {
            base_date: form.get('base_date').value,
            company_code: form.get('company_code').value,
            branch_code: form.get('branch_code').value,
            environment_code: form.get('environment_code').value,
            role_code: form.get('role_code').value,
            mdi_menu_info: form.get('mdi_menu_info').value,
            go_emp_fil: form.get('go_emp_fil').value
        };
        return this.proSessionSettingsService.saveSettingsDefaults(newSession);
    }
    setChannelAsHTTP(value) {
        this.useHTTP = value;
        this.changeServicesChannel();
    }
    isChannelHTTP() {
        return this.useHTTP;
    }
    /**
     * @description Atualiza o Canal de comunicação
     * dos serviços utilizados pelo componente
     * @returns void
     */
    changeServicesChannel() {
        this.proCompanyService.setChannelAsHTTP(this.useHTTP);
        this.proBranchService.setChannelAsHTTP(this.useHTTP);
        this.proTranslateStringService.setChannelAsHTTP(this.useHTTP);
        this.proSystemModuleService.setChannelAsHTTP(this.useHTTP);
    }
    /**
     * @description Atualiza as variaveis a partir da marca
     * ma3 na russia e protheus nas demais
     * @returns void
     */
    updateResourceByBrand() {
        if (this.brand == "MA3") {
            this.logo = './assets/images/nationalplatform/nationalplatform.svg';
            this.secondaryLogo = './assets/images/nationalplatform/nationalplatform.svg';
            this.productName = 'Ma-3 Line';
        }
        else {
            this.logo = './assets/images/totvs/totvs.svg';
            this.secondaryLogo = './assets/images/totvs/totvs.svg';
            this.productName = this.literals.str0001;
        }
    }
    /**
     * @description atualiza a variavel productName.
     * quando não for MA3 precisa atualziar com a literals cada vez que muda o idioma
     * @returns void
     */
    updateProductName() {
        if (this.brand != "MA3") {
            this.productName = this.literals.str0001;
        }
    }
    /**
     * @description atualiza a variavel de marca
     * @returns void
     */
    updateBrand() {
        this.proBrandService.setChannelAsHTTP(environment.useHTTP);
        this.proBrandService.getERPBrand().subscribe({
            next: (brand) => {
                this.brand = brand;
                this.updateResourceByBrand();
            },
            error: (error) => {
                this.brand = "TOTVS";
                this.updateResourceByBrand();
            }
        });
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProSessionSettingsComponent, deps: [{ token: i1.UntypedFormBuilder }, { token: i2.Router }, { token: i3.ProMessageService }, { token: i2.ActivatedRoute }, { token: i4.ProSessionSettingsService }, { token: i5.ProCompanyService }, { token: i6.ProRoleService }, { token: i7.ProBranchService }, { token: i8.PoI18nService }, { token: i9.ProTranslateStringService }, { token: i10.ProBrandService }, { token: i11.ProSystemModuleService }], target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "17.3.9", type: ProSessionSettingsComponent, selector: "pro-session-settings", inputs: { environment: ["p-environment", "environment"], background: ["p-background", "background"], hideLogo: ["p-hide-logo", "hideLogo"], highlightInfo: ["p-highlight-info", "highlightInfo"] }, viewQueries: [{ propertyName: "proBranchLookupComponent", first: true, predicate: ProBranchLookupComponent, descendants: true }, { propertyName: "proCompanyLookupComponent", first: true, predicate: ProCompanyLookupComponent, descendants: true }], ngImport: i0, template: "<button\r\n  class=\"po-page-login-support\"\r\n  (click)=\"activateSupport()\"\r\n  [hidden]=\"!literals?.str0004\"\r\n>\r\n  <span class=\"po-icon po-icon-help\"></span>\r\n  {{ literals?.str0004 }}\r\n  <!-- Suporte -->\r\n</button>\r\n\r\n<pro-page-background\r\n  p-show-select-language\r\n  [p-background]=\"background\"\r\n  [p-highlight-info]=\"highlightInfo\"\r\n  [p-logo]=\"logo\"\r\n  [p-secondary-logo]=\"secondaryLogo\"\r\n  (p-selected-language)=\"onSelectedLanguage($event)\"\r\n>\r\n  <header class=\"po-page-login-header\">\r\n    <div class=\"session-display-top session-display-bot\">\r\n      <div class=\"po-page-login-header-product-name\">\r\n        <h1>{{ productName }}</h1>\r\n        <!-- Linha Protheus -->\r\n      </div>\r\n      <po-tag *ngIf=\"environment\" p-type=\"warning\" [p-value]=\"environment\">\r\n      </po-tag>\r\n    </div>\r\n    <div class=\"po-page-login-header-welcome session-display-bot\">\r\n      {{\r\n        displayName\r\n          ? (literals?.str0009 | poI18n: [displayName])\r\n          : literals?.str0006\r\n      }}<!-- Boas-vindas, {0}. --><!-- Carregando... -->\r\n    </div>\r\n  </header>\r\n  <form class=\"session-settings-form\" [formGroup]=\"settingForm\">\r\n    <div class=\"po-row\">\r\n      <div class=\"po-md-12 po-lg-12 po-xl-12 po-page-login-info-container session-date-container\">\r\n        <po-datepicker\r\n          class=\"po-md-7 po-lg-7 po-xl-7\"\r\n          name=\"base_date\"\r\n          formControlName=\"base_date\"\r\n          p-clean\r\n          p-format=\"dd/mm/yyyy\"\r\n          [p-label]=\"literals?.str0018\"\r\n          p-required\r\n        ></po-datepicker\r\n        ><!-- Data base -->\r\n        <div class=\"po-page-login-info-icon-container session-hint-div\">\r\n          <span\r\n            class=\"po-icon po-field-icon po-icon-info\"\r\n            [p-tooltip]=\"literals?.str0002\"\r\n            p-tooltip-position=\"right\"\r\n            ><!-- Escolha a data base do sistema. -->\r\n          </span>\r\n        </div>\r\n      </div>\r\n      <pro-company-lookup\r\n        class=\"po-md-12\"\r\n        [parent]=\"settingForm\"\r\n        [p-label]=\"literals?.str0010\"\r\n        [p-hint]=\"literals?.str0011\"\r\n        [pro-columns]=\"companyColumns\"\r\n        (companySetted)=\"changeBranchColumns($event)\"\r\n        (branchSetted)=\"branchSettedEvent()\"\r\n      ></pro-company-lookup\r\n      ><!-- Grupo --><!-- Selecione o Grupo de Empresas. -->\r\n      <pro-branch-lookup\r\n        [p-label]=\"literals?.str0012\"\r\n        [p-hint]=\"literals?.str0013\"\r\n        [pro-columns]=\"branchColumns\"\r\n        class=\"po-md-12\"\r\n        [parent]=\"settingForm\"\r\n        ><!-- Filial --><!-- Selecione a Filial. -->\r\n      </pro-branch-lookup>\r\n      <pro-system-module-lookup\r\n        [p-label]=\"literals?.str0014\"\r\n        [p-disabled]=\"fixedModule\"\r\n        [p-hint]=\"literals?.str0015\"\r\n        [pro-columns]=\"moduleColumns\"\r\n        class=\"po-md-12\"\r\n        [parent]=\"settingForm\"\r\n        ><!-- Ambiente --><!-- Selecione o m\u00F3dulo do sistema. -->\r\n      </pro-system-module-lookup>\r\n      <pro-role-lookup\r\n        [p-label]=\"literals?.str0016\"\r\n        [p-disabled]=\"fixedRole\"\r\n        [p-hint]=\"literals?.str0017\"\r\n        class=\"po-md-12\"\r\n        [parent]=\"settingForm\"\r\n      ></pro-role-lookup\r\n      ><!-- Papel de trabalho --><!-- Selecione um Papel de trabalho. -->\r\n      <div *ngIf=\"showGoEmpFil\" class=\"session-reminder-container\">\r\n        <po-switch\r\n          class=\"session-remider-switch\"\r\n          name=\"mdi_menu_info\"\r\n          formControlName=\"mdi_menu_info\"\r\n          p-label=\" \"\r\n          [p-label-off]=\"literals?.str0005\"\r\n          [p-label-on]=\"literals?.str0005\"\r\n        ></po-switch\r\n        ><!-- Usar as informa\u00E7\u00F5es acima em todas as sess\u00F5es. -->\r\n      </div>\r\n      <div *ngIf=\"showMDIMenuInfo\" class=\"session-reminder-container\">\r\n        <po-switch\r\n          class=\"session-remider-switch\"\r\n          name=\"go_emp_fil\"\r\n          formControlName=\"go_emp_fil\"\r\n          p-label=\" \"\r\n          [p-label-off]=\"literals?.str0023\"\r\n          [p-label-on]=\"literals?.str0023\"\r\n        ></po-switch\r\n        ><!-- Iniciar com as informa\u00E7\u00F5es da \u00FAltima sess\u00E3o. -->\r\n        <span\r\n          class=\"po-icon po-field-icon po-icon-info session-reminder-tooltip\"\r\n          [p-tooltip]=\"literals?.str0003\"\r\n          p-tooltip-position=\"right\"\r\n          ><!-- Voc\u00EA pode desabilitar essa op\u00E7\u00E3o no menu do sistema. -->\r\n        </span>\r\n      </div>\r\n      <div\r\n        class=\"po-md-12 po-page-login-info-container session-settings-buttons\"\r\n      >\r\n        <po-button\r\n          class=\"po-md-6\"\r\n          name=\"back\"\r\n          [p-disabled]=\"isLoading || disableBackButton\"\r\n          [p-label]=\"literals?.str0008\"\r\n          (p-click)=\"onSessionSettingBack()\"\r\n          ><!-- Voltar -->\r\n        </po-button>\r\n        <po-button\r\n          class=\"po-md-6 session-settings-button-enter\"\r\n          name=\"submmit\"\r\n          p-kind=\"primary\"\r\n          [p-disabled]=\"\r\n            !(\r\n              this.settingForm.get('company_code').value &&\r\n              this.settingForm.get('branch_code').value &&\r\n              this.settingForm.get('environment_code').value\r\n            )\r\n          \"\r\n          [p-label]=\"isLoading ? literals?.str0006 : literals?.str0007\"\r\n          [p-loading]=\"isLoading\"\r\n          (p-click)=\"onSessionSettingSubmit()\"\r\n          ><!-- Carregando... --><!-- Entrar -->\r\n        </po-button>\r\n        <div class=\"po-page-login-info-icon-container session-empty-div\"></div>\r\n      </div>\r\n    </div>\r\n  </form>\r\n</pro-page-background>\r\n", styles: [".session-hint-div{padding:37px 0 0}.session-empty-div{width:30px}.session-info-diff{width:191px}.session-date-container{display:inline-block}.session-reminder-container{padding:0 16px;display:flex;margin-top:0}.session-reminder-tooltip{padding:27px 0 0 20px}.session-display-top{padding-top:3%}.session-display-bot{padding-bottom:2.2%}.session-settings-buttons{padding-top:10px}@media (min-width: 1367px){.session-empty-div{width:33px}.session-info-diff{width:187px}}@media (min-width: 768px){.session-settings-form{padding-top:0}}@media (max-width: 480px){.session-settings-button-enter{padding-top:15px}}\n"], dependencies: [{ kind: "directive", type: i12.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }, { kind: "component", type: i8.PoButtonComponent, selector: "po-button" }, { kind: "component", type: i8.PoDatepickerComponent, selector: "po-datepicker", inputs: ["p-label", "p-help"] }, { kind: "component", type: i8.PoSwitchComponent, selector: "po-switch", inputs: ["p-value", "p-format-model", "p-label-position", "p-label-off", "p-label-on"] }, { kind: "component", type: i8.PoTagComponent, selector: "po-tag" }, { kind: "component", type: i13.ProPageBackgroundComponent, selector: "pro-page-background", inputs: ["p-background", "p-hide-logo", "p-highlight-info", "p-logo", "p-secondary-logo", "p-show-select-language"], outputs: ["p-selected-language"] }, { kind: "directive", type: i8.PoTooltipDirective, selector: "[p-tooltip]" }, { kind: "component", type: i14.ProCompanyLookupComponent, selector: "pro-company-lookup", inputs: ["parent", "p-label", "p-hint", "pro-columns"], outputs: ["companySetted", "branchSetted"] }, { kind: "component", type: i15.ProBranchLookupComponent, selector: "pro-branch-lookup", inputs: ["parent", "p-label", "p-hint", "pro-columns"], outputs: ["branchSetted"] }, { kind: "component", type: i16.ProRoleLookupComponent, selector: "pro-role-lookup", inputs: ["parent", "p-label", "p-hint", "pro-columns", "p-disabled"], outputs: ["roleSetted"] }, { kind: "component", type: i17.ProSystemModuleLookupComponent, selector: "pro-system-module-lookup", inputs: ["parent", "p-label", "p-disabled", "p-hint", "pro-columns"], outputs: ["systeModuleSetted"] }, { kind: "directive", type: i1.ɵNgNoValidate, selector: "form:not([ngNoForm]):not([ngNativeValidate])" }, { kind: "directive", type: i1.NgControlStatus, selector: "[formControlName],[ngModel],[formControl]" }, { kind: "directive", type: i1.NgControlStatusGroup, selector: "[formGroupName],[formArrayName],[ngModelGroup],[formGroup],form:not([ngNoForm]),[ngForm]" }, { kind: "directive", type: i1.FormGroupDirective, selector: "[formGroup]", inputs: ["formGroup"], outputs: ["ngSubmit"], exportAs: ["ngForm"] }, { kind: "directive", type: i1.FormControlName, selector: "[formControlName]", inputs: ["formControlName", "disabled", "ngModel"], outputs: ["ngModelChange"] }, { kind: "pipe", type: i8.PoI18nPipe, name: "poI18n" }] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProSessionSettingsComponent, decorators: [{
            type: Component,
            args: [{ selector: 'pro-session-settings', template: "<button\r\n  class=\"po-page-login-support\"\r\n  (click)=\"activateSupport()\"\r\n  [hidden]=\"!literals?.str0004\"\r\n>\r\n  <span class=\"po-icon po-icon-help\"></span>\r\n  {{ literals?.str0004 }}\r\n  <!-- Suporte -->\r\n</button>\r\n\r\n<pro-page-background\r\n  p-show-select-language\r\n  [p-background]=\"background\"\r\n  [p-highlight-info]=\"highlightInfo\"\r\n  [p-logo]=\"logo\"\r\n  [p-secondary-logo]=\"secondaryLogo\"\r\n  (p-selected-language)=\"onSelectedLanguage($event)\"\r\n>\r\n  <header class=\"po-page-login-header\">\r\n    <div class=\"session-display-top session-display-bot\">\r\n      <div class=\"po-page-login-header-product-name\">\r\n        <h1>{{ productName }}</h1>\r\n        <!-- Linha Protheus -->\r\n      </div>\r\n      <po-tag *ngIf=\"environment\" p-type=\"warning\" [p-value]=\"environment\">\r\n      </po-tag>\r\n    </div>\r\n    <div class=\"po-page-login-header-welcome session-display-bot\">\r\n      {{\r\n        displayName\r\n          ? (literals?.str0009 | poI18n: [displayName])\r\n          : literals?.str0006\r\n      }}<!-- Boas-vindas, {0}. --><!-- Carregando... -->\r\n    </div>\r\n  </header>\r\n  <form class=\"session-settings-form\" [formGroup]=\"settingForm\">\r\n    <div class=\"po-row\">\r\n      <div class=\"po-md-12 po-lg-12 po-xl-12 po-page-login-info-container session-date-container\">\r\n        <po-datepicker\r\n          class=\"po-md-7 po-lg-7 po-xl-7\"\r\n          name=\"base_date\"\r\n          formControlName=\"base_date\"\r\n          p-clean\r\n          p-format=\"dd/mm/yyyy\"\r\n          [p-label]=\"literals?.str0018\"\r\n          p-required\r\n        ></po-datepicker\r\n        ><!-- Data base -->\r\n        <div class=\"po-page-login-info-icon-container session-hint-div\">\r\n          <span\r\n            class=\"po-icon po-field-icon po-icon-info\"\r\n            [p-tooltip]=\"literals?.str0002\"\r\n            p-tooltip-position=\"right\"\r\n            ><!-- Escolha a data base do sistema. -->\r\n          </span>\r\n        </div>\r\n      </div>\r\n      <pro-company-lookup\r\n        class=\"po-md-12\"\r\n        [parent]=\"settingForm\"\r\n        [p-label]=\"literals?.str0010\"\r\n        [p-hint]=\"literals?.str0011\"\r\n        [pro-columns]=\"companyColumns\"\r\n        (companySetted)=\"changeBranchColumns($event)\"\r\n        (branchSetted)=\"branchSettedEvent()\"\r\n      ></pro-company-lookup\r\n      ><!-- Grupo --><!-- Selecione o Grupo de Empresas. -->\r\n      <pro-branch-lookup\r\n        [p-label]=\"literals?.str0012\"\r\n        [p-hint]=\"literals?.str0013\"\r\n        [pro-columns]=\"branchColumns\"\r\n        class=\"po-md-12\"\r\n        [parent]=\"settingForm\"\r\n        ><!-- Filial --><!-- Selecione a Filial. -->\r\n      </pro-branch-lookup>\r\n      <pro-system-module-lookup\r\n        [p-label]=\"literals?.str0014\"\r\n        [p-disabled]=\"fixedModule\"\r\n        [p-hint]=\"literals?.str0015\"\r\n        [pro-columns]=\"moduleColumns\"\r\n        class=\"po-md-12\"\r\n        [parent]=\"settingForm\"\r\n        ><!-- Ambiente --><!-- Selecione o m\u00F3dulo do sistema. -->\r\n      </pro-system-module-lookup>\r\n      <pro-role-lookup\r\n        [p-label]=\"literals?.str0016\"\r\n        [p-disabled]=\"fixedRole\"\r\n        [p-hint]=\"literals?.str0017\"\r\n        class=\"po-md-12\"\r\n        [parent]=\"settingForm\"\r\n      ></pro-role-lookup\r\n      ><!-- Papel de trabalho --><!-- Selecione um Papel de trabalho. -->\r\n      <div *ngIf=\"showGoEmpFil\" class=\"session-reminder-container\">\r\n        <po-switch\r\n          class=\"session-remider-switch\"\r\n          name=\"mdi_menu_info\"\r\n          formControlName=\"mdi_menu_info\"\r\n          p-label=\" \"\r\n          [p-label-off]=\"literals?.str0005\"\r\n          [p-label-on]=\"literals?.str0005\"\r\n        ></po-switch\r\n        ><!-- Usar as informa\u00E7\u00F5es acima em todas as sess\u00F5es. -->\r\n      </div>\r\n      <div *ngIf=\"showMDIMenuInfo\" class=\"session-reminder-container\">\r\n        <po-switch\r\n          class=\"session-remider-switch\"\r\n          name=\"go_emp_fil\"\r\n          formControlName=\"go_emp_fil\"\r\n          p-label=\" \"\r\n          [p-label-off]=\"literals?.str0023\"\r\n          [p-label-on]=\"literals?.str0023\"\r\n        ></po-switch\r\n        ><!-- Iniciar com as informa\u00E7\u00F5es da \u00FAltima sess\u00E3o. -->\r\n        <span\r\n          class=\"po-icon po-field-icon po-icon-info session-reminder-tooltip\"\r\n          [p-tooltip]=\"literals?.str0003\"\r\n          p-tooltip-position=\"right\"\r\n          ><!-- Voc\u00EA pode desabilitar essa op\u00E7\u00E3o no menu do sistema. -->\r\n        </span>\r\n      </div>\r\n      <div\r\n        class=\"po-md-12 po-page-login-info-container session-settings-buttons\"\r\n      >\r\n        <po-button\r\n          class=\"po-md-6\"\r\n          name=\"back\"\r\n          [p-disabled]=\"isLoading || disableBackButton\"\r\n          [p-label]=\"literals?.str0008\"\r\n          (p-click)=\"onSessionSettingBack()\"\r\n          ><!-- Voltar -->\r\n        </po-button>\r\n        <po-button\r\n          class=\"po-md-6 session-settings-button-enter\"\r\n          name=\"submmit\"\r\n          p-kind=\"primary\"\r\n          [p-disabled]=\"\r\n            !(\r\n              this.settingForm.get('company_code').value &&\r\n              this.settingForm.get('branch_code').value &&\r\n              this.settingForm.get('environment_code').value\r\n            )\r\n          \"\r\n          [p-label]=\"isLoading ? literals?.str0006 : literals?.str0007\"\r\n          [p-loading]=\"isLoading\"\r\n          (p-click)=\"onSessionSettingSubmit()\"\r\n          ><!-- Carregando... --><!-- Entrar -->\r\n        </po-button>\r\n        <div class=\"po-page-login-info-icon-container session-empty-div\"></div>\r\n      </div>\r\n    </div>\r\n  </form>\r\n</pro-page-background>\r\n", styles: [".session-hint-div{padding:37px 0 0}.session-empty-div{width:30px}.session-info-diff{width:191px}.session-date-container{display:inline-block}.session-reminder-container{padding:0 16px;display:flex;margin-top:0}.session-reminder-tooltip{padding:27px 0 0 20px}.session-display-top{padding-top:3%}.session-display-bot{padding-bottom:2.2%}.session-settings-buttons{padding-top:10px}@media (min-width: 1367px){.session-empty-div{width:33px}.session-info-diff{width:187px}}@media (min-width: 768px){.session-settings-form{padding-top:0}}@media (max-width: 480px){.session-settings-button-enter{padding-top:15px}}\n"] }]
        }], ctorParameters: () => [{ type: i1.UntypedFormBuilder }, { type: i2.Router }, { type: i3.ProMessageService }, { type: i2.ActivatedRoute }, { type: i4.ProSessionSettingsService }, { type: i5.ProCompanyService }, { type: i6.ProRoleService }, { type: i7.ProBranchService }, { type: i8.PoI18nService }, { type: i9.ProTranslateStringService }, { type: i10.ProBrandService }, { type: i11.ProSystemModuleService }], propDecorators: { environment: [{
                type: Input,
                args: ['p-environment']
            }], background: [{
                type: Input,
                args: ['p-background']
            }], hideLogo: [{
                type: Input,
                args: ['p-hide-logo']
            }], highlightInfo: [{
                type: Input,
                args: ['p-highlight-info']
            }], proBranchLookupComponent: [{
                type: ViewChild,
                args: [ProBranchLookupComponent]
            }], proCompanyLookupComponent: [{
                type: ViewChild,
                args: [ProCompanyLookupComponent]
            }] } });
