// @totvs/protheus-lib-core/esm2022/components/pro-pages/pro-session-settings/pro-session-settings.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { ReactiveFormsModule } from '@angular/forms';
import { PoButtonModule, PoFieldModule, PoModule, PoTagModule, PoTooltipModule } from '@po-ui/ng-components';
import { ProPageBackgroundModule } from '../pro-page-background';
import { ProFieldsModule } from './../../pro-fields/pro-fields.module';
import { ProSessionSettingsComponent } from './pro-session-settings.component';
import * as i0 from "@angular/core";
export class ProSessionSettingsModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProSessionSettingsModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "17.3.9", ngImport: i0, type: ProSessionSettingsModule, declarations: [ProSessionSettingsComponent], imports: [CommonModule,
            PoButtonModule,
            PoFieldModule,
            ProPageBackgroundModule,
            PoModule,
            PoTagModule,
            PoTooltipModule,
            ProFieldsModule,
            ReactiveFormsModule], exports: [ProSessionSettingsComponent] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProSessionSettingsModule, imports: [CommonModule,
            PoButtonModule,
            PoFieldModule,
            ProPageBackgroundModule,
            PoModule,
            PoTagModule,
            PoTooltipModule,
            ProFieldsModule,
            ReactiveFormsModule] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProSessionSettingsModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        PoButtonModule,
                        PoFieldModule,
                        ProPageBackgroundModule,
                        PoModule,
                        PoTagModule,
                        PoTooltipModule,
                        ProFieldsModule,
                        ReactiveFormsModule
                    ],
                    declarations: [ProSessionSettingsComponent],
                    exports: [ProSessionSettingsComponent]
                }]
        }] });
