// @po-ui/ng-components/esm2022/lib/directives/directives.module.mjs
import { NgModule } from '@angular/core';
import { PoTooltipModule } from './po-tooltip/po-tooltip.module';
import * as i0 from "@angular/core";
export class PoDirectivesModule {
    static ɵfac = function PoDirectivesModule_Factory(t) { return new (t || PoDirectivesModule)(); };
    static ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: PoDirectivesModule });
    static ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [PoTooltipModule, PoTooltipModule] });
}
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(PoDirectivesModule, [{
        type: NgModule,
        args: [{
                declarations: [],
                imports: [PoTooltipModule],
                exports: [PoTooltipModule],
                providers: [],
                bootstrap: []
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(PoDirectivesModule, { imports: [PoTooltipModule], exports: [PoTooltipModule] }); })();
