// @totvs/protheus-lib-core/esm2022/components/pro-pages/pro-session-settings/services/pro-session-settings.service.mjs
import { HttpHeaders } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { valueIsNull } from '../../../../util';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common/http";
import * as i2 from "./../../../../services/pro-js-to-advpl/services/pro-js-to-advpl.service";
const ENDPOINT = '/api/protheus/session/v1/defaults';
export class ProSessionSettingsService {
    constructor(http, advplService) {
        this.http = http;
        this.advplService = advplService;
    }
    requestSettingsDefaults() {
        if (!valueIsNull(this.advplService.getWebChannel())) {
            return this.advplService.buildObservable(({ protheusResponse, subscriber }) => {
                if (protheusResponse.length === 0) {
                    subscriber.next({
                        role_code: '',
                        role_description: '',
                        role_when: false,
                        show_go_emp_fil: false,
                        show_mdi_menu_info: false,
                        environment_code: '',
                        environment_description: '',
                        environment_when: true,
                        disable_back_button: false
                    });
                }
                else {
                    const startKeys = JSON.parse(protheusResponse);
                    subscriber.next(startKeys);
                }
                subscriber.complete();
            }, {
                autoDestruct: true,
                receiveId: 'setSessionStart',
                sendInfo: {
                    type: 'getSessionStart'
                }
            });
        }
        else {
            const headers = new HttpHeaders().append('Accept', 'application/json; charset=utf-8');
            return this.http.get(ENDPOINT, {
                headers
            });
        }
    }
    clearSettingsDefaults() {
        if (this.advplService.protheusConnected()) {
            this.advplService.jsToAdvpl('backButton', '');
        }
        const erpAppConfig = sessionStorage.getItem('ERPAPPCONFIG');
        sessionStorage.clear();
        sessionStorage.setItem('ERPAPPCONFIG', erpAppConfig);
    }
    saveSettingsDefaults(session) {
        if (this.advplService.protheusConnected()) {
            return this.advplService.buildObservable(({ protheusResponse, subscriber }) => {
                if (protheusResponse === 'true') {
                    subscriber.next({});
                }
                else {
                    subscriber.error({});
                }
                subscriber.complete();
            }, {
                autoDestruct: true,
                receiveId: 'submitButtonResponse',
                sendInfo: {
                    type: 'submitButton',
                    content: JSON.stringify(session)
                }
            });
        }
        else {
            const headers = new HttpHeaders().append('Content-Type', 'application/json; charset=utf-8');
            return this.http.post(ENDPOINT, session, { headers });
        }
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProSessionSettingsService, deps: [{ token: i1.HttpClient }, { token: i2.ProJsToAdvplService }], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProSessionSettingsService, providedIn: 'root' }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProSessionSettingsService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root'
                }]
        }], ctorParameters: () => [{ type: i1.HttpClient }, { type: i2.ProJsToAdvplService }] });
