// @po-ui/ng-components/esm2022/lib/components/po-dynamic/shared/po-dynamic-shared-base.mjs
import { Directive } from '@angular/core';
import * as i0 from "@angular/core";
/**
 * @usedBy PoDynamicViewComponent, PoDynamicFormComponent
 * @description
 * Componente para listar dados compartilhados entre o dynamic-view e o dynamic-form
 */
export class PoDynamicSharedBase {
    visibleFields = [];
    containerFields = [];
    hasContainers = false;
    ensureFieldHasContainer(fields) {
        if (fields && fields.some(field => field?.container)) {
            const firstFieldIndex = fields.findIndex(field => field.order === 1);
            if (firstFieldIndex !== -1 && !fields[firstFieldIndex].container) {
                fields[firstFieldIndex].container = '';
            }
        }
    }
    setContainerFields() {
        this.hasContainers = this.visibleFields && this.visibleFields.some(field => field.container);
        if (this.hasContainers) {
            this.containerFields = this.groupFieldsIntoContainers(this.visibleFields);
        }
    }
    groupFieldsIntoContainers(fields) {
        const groups = [[]];
        return fields.reduce((acc, field) => {
            if (field.container && acc[acc.length - 1].length > 0)
                acc.push([]);
            acc[acc.length - 1].push(field);
            return acc;
        }, groups);
    }
    static ɵfac = function PoDynamicSharedBase_Factory(t) { return new (t || PoDynamicSharedBase)(); };
    static ɵdir = /*@__PURE__*/ i0.ɵɵdefineDirective({ type: PoDynamicSharedBase });
}
(() => { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(PoDynamicSharedBase, [{
        type: Directive
    }], null, null); })();
