// @totvs/protheus-lib-core/esm2022/services/pro-messages/services/pro-message.service.mjs
import { Injectable } from '@angular/core';
import { PoToasterOrientation } from '@po-ui/ng-components';
import * as i0 from "@angular/core";
import * as i1 from "@po-ui/ng-components";
import * as i2 from "./../../pro-js-to-advpl/services/pro-js-to-advpl.service";
export class ProMessageService {
    constructor(poNotificationService, advplService) {
        this.poNotificationService = poNotificationService;
        this.advplService = advplService;
        if (this.advplService.protheusConnected()) {
            this.advplService.buildListener('showMessage', ({ protheusResponse }) => {
                if (protheusResponse.length === 0) {
                    return;
                }
                else {
                    const messageObject = JSON.parse(protheusResponse);
                    this.showMessage(messageObject.message, messageObject.type);
                }
            });
        }
    }
    // Método criado para evitar que sejam exibidas duas mensagens para o usuário (uma do Angular e outra do Protheus)
    showOneMessage(message, type = 2) {
        if (!this.advplService.protheusConnected()) {
            this.showMessage(message, type);
        }
    }
    /**
     * @description Exibe uma mensagem via notificação
     * @param message Mensagem
     * @param type Tipo da mensage, como erro, informação etc
     */
    showMessage(message, type = 2) {
        const notification = {
            message,
            orientation: PoToasterOrientation.Top
        };
        if (type === 1) {
            this.poNotificationService.warning(notification);
        }
        if (type === 2) {
            this.poNotificationService.error(notification);
        }
        if (type === 3) {
            this.poNotificationService.information(notification);
        }
    }
    /**
     * @description Executa a mudança do idioma no Protheus via Observable do advplService
     * @param language idioma escolhido via po-select
     * @returns Observable de boolean da confirmação da troca do idioma no BackEnd
     */
    changeLanguage(language) {
        if (this.advplService.protheusConnected()) {
            return this.advplService.buildObservable(({ protheusResponse, subscriber }) => {
                if (protheusResponse !== 'true') {
                    subscriber.error({
                        status: 400,
                        description: 'language could not be changed'
                    });
                }
                else {
                    subscriber.next();
                }
                subscriber.complete();
            }, {
                sendInfo: {
                    type: 'changeLanguage',
                    content: language
                },
                autoDestruct: true,
                receiveId: 'changeLanguage'
            });
        }
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProMessageService, deps: [{ token: i1.PoNotificationService }, { token: i2.ProJsToAdvplService }], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProMessageService, providedIn: 'root' }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.9", ngImport: i0, type: ProMessageService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root'
                }]
        }], ctorParameters: () => [{ type: i1.PoNotificationService }, { type: i2.ProJsToAdvplService }] });
